{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Expectation-Maximization algorithm** (hidden model)\n",
    "### Author: Mart√≠ Cortada Garcia\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import numpy as np\n",
    "from Bio import SeqIO, Phylo\n",
    "from io import StringIO\n",
    "from collections import Counter, OrderedDict\n",
    "import itertools\n",
    "import pandas as pd\n",
    "import math\n",
    "import copy\n",
    "\n",
    "# Import required functions from modules we have programmed\n",
    "from data_simulation import simulate"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Input:**\n",
    "We have a tree $T$ (where the root corresponds to the internal node codified as `Int_0`) and the alignments of the leaves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAGwCAYAAACHJU4LAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy88F64QAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAwJUlEQVR4nO3df1RVZb7H8c8RBPmt+AtBlFJEM0C0MKzM0lKzkuauLJf5o5qYZoFJxmTebMwZU+9NTaa6ldWFppuS2aXu1UmGLLQwC1En/HG1sDqlKI0lCI0HPez7R9dzI1E5CJwHz/u1Fmu593723t/9rL32+fjsffaxWZZlCQAAwEAdPF0AAADA2RBUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACM5evpAi5EfX29Dh06pJCQENlsNk+XAwAAmsCyLB0/flyRkZHq0OHcYybtOqgcOnRI0dHRni4DAAA0wzfffKPevXufs027DiohISGSfjrQ0NBQD1cDAACaorq6WtHR0a7P8XNp10Hl9O2e0NBQggoAAO1MUx7b4GFaAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAPMayLKWlpSk8PFw2m007d+70dEkADENQAdCooqIi2Ww2HTt2rNX2sWHDBuXm5mrdunWqqKjQ5Zdffs72ixcv1pVXXqmQkBD16NFDqamp2rdvX6vVB8DzCCoAPKa8vFy9evXSiBEjFBERIV9f33O237Rpk9LT07V161YVFhbq5MmTuummm1RbW9tGFQNoax4NKps3b9att96qyMhI2Ww2vf32254sB4AbHA6HsrKyFBUVpaCgIA0fPlxFRUWu5UePHtXkyZMVFRWlwMBAxcfHa/Xq1a7lM2bM0MyZM2W322Wz2RQTE3PefW7YsEEzZszQ4MGDlZiYqNzcXNntdpWWlrbCEQIwwbn/+9LKamtrlZiYqHvvvVe/+tWvPFmKqx7AGwUFBbm9TkZGhvbs2aO8vDxFRkYqPz9f48aNU1lZmWJjY3XixAkNGzZMc+bMUWhoqNavX6+pU6eqX79+Sk5OVnZ2tvr166eVK1eqpKREPj4+btdQVVUlSQoPD3d7XQDtg82yLMvTRUiSzWZTfn6+UlNTz9rG4XDI4XC4pqurqxUdHa2qqiqFhoa2SA2AN2rsMlBUVKTrr79eP/zwgzp37txgmd1u16WXXiq73a7IyEjX/DFjxig5OVmLFi1qdD+33HKLBg4cqKVLl0qSVqxYoRUrVuirr75yu+b6+nrddtttOnbsmD766CO31wfgOdXV1QoLC2vS57dHR1TctXjxYi1YsKDFt1tbW6vg4OAW3y5wsSorK5PT6dSAAQMazHc4HOrataskyel0atGiRVqzZo0OHjyouro6ORwOBQYGtkgN6enp2rVrFyEFuMi1q6Ayd+5czZ492zV9ekSlJR05cqRZw+CAN6mpqZGPj49KS0vPuGVzOvQ/9dRTys7O1ooVKxQfH6+goCBlZmaqrq7ugvefkZGhdevWafPmzerdu/cFbw+AudpVUPH395e/v3+r7iMoKIigApxHUlKSnE6nKisrde211zbapri4WBMnTtTdd98t6adbNfv379dll13W7P1alqWZM2cqPz9fRUVFuuSSS5q9LQDtQ7sKKgDaXllZmUJCQlzTNptNiYmJmjJliqZNm6Zly5YpKSlJ3333nTZu3KiEhARNmDBBsbGxWrt2rbZs2aIuXbpo+fLlOnLkyAUFlfT0dK1atUrvvPOOQkJCdPjwYUlSWFiYAgICLvhYAZiHoALgnEaOHNlg2sfHR6dOnVJOTo4WLlyohx9+WAcPHlS3bt101VVX6ZZbbpEkzZs3TwcOHNDYsWMVGBiotLQ0paamur6p0xzPP/+8JGnUqFEN5ufk5GjGjBnN3i4Ac3n0Wz81NTX64osvJP00lLx8+XJdf/31Cg8PV58+fc67vjtPDZ/Lzx+mramp4dYPAACtqN1862fbtm26/vrrXdOnH5SdPn26cnNzPVQVAAAwhUeDyqhRoxp9fwMA72O328/5/MqePXuaNNIK4OLCMyoAjBAZGXnOX0/++YvlAHgPggoAI/j6+qp///6eLgOAYfj1ZAAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAj7EsS2lpaQoPD5fNZtPOnTs9XRIAwxBUADSqqKhINptNx44da7V9bNiwQbm5uVq3bp0qKip0+eWXn7P9888/r4SEBIWGhio0NFQpKSl69913W60+AJ7n0aCyePFiXXnllQoJCVGPHj2Umpqqffv2ebIkAG2ovLxcvXr10ogRIxQRESFfX99ztu/du7eWLFmi0tJSbdu2TTfccIMmTpyo3bt3t1HFANqaR4PKpk2blJ6erq1bt6qwsFAnT57UTTfdpNraWk+WBaAJHA6HsrKyFBUVpaCgIA0fPlxFRUWu5UePHtXkyZMVFRWlwMBAxcfHa/Xq1a7lM2bM0MyZM2W322Wz2RQTE3Pefd566626+eabFRsbqwEDBujJJ59UcHCwtm7d2gpHCMAE5/7vSyvbsGFDg+nc3Fz16NFDpaWlGjlypEdqIiTBGwUFBbm9TkZGhvbs2aO8vDxFRkYqPz9f48aNU1lZmWJjY3XixAkNGzZMc+bMUWhoqNavX6+pU6eqX79+Sk5OVnZ2tvr166eVK1eqpKREPj4+bu3f6XTqzTffVG1trVJSUtyuH0D74NGg8ktVVVWSpPDw8EaXOxwOORwO13R1dXWL19CzZ88W3yZgOsuy3Gpvt9uVk5Mju92uyMhISVJWVpY2bNignJwcLVq0SFFRUcrKynKtM3PmTBUUFGjNmjVKTk5WWFiYQkJC5OPjo4iIiCbvu6ysTCkpKTpx4oSCg4OVn5+vyy67zK36AbQfxgSV+vp6ZWZm6uqrrz7rA3WLFy/WggUL2rgyAL9UVlYmp9OpAQMGNJjvcDjUtWtXST+NeCxatEhr1qzRwYMHVVdXJ4fDocDAwAvad1xcnHbu3KmqqiqtXbtW06dP16ZNmwgrwEXKmKCSnp6uXbt26aOPPjprm7lz52r27Nmu6erqakVHR7doHUeOHGnWMDjgTWpqauTj46PS0tIzbtkEBwdLkp566illZ2drxYoVio+PV1BQkDIzM1VXV3dB+/bz81P//v0lScOGDVNJSYmys7P14osvXtB2AZjJiKCSkZGhdevWafPmzerdu/dZ2/n7+8vf379VawkKCiKoAOeRlJQkp9OpyspKXXvttY22KS4u1sSJE3X33XdL+mnUdP/+/S0+8lFfX9/gljCAi4tHg4plWZo5c6by8/NVVFSkSy65xJPlAGhEWVmZQkJCXNM2m02JiYmaMmWKpk2bpmXLlikpKUnfffedNm7cqISEBE2YMEGxsbFau3attmzZoi5dumj58uU6cuTIBQWVuXPnavz48erTp4+OHz+uVatWqaioSAUFBS1xqAAM5NGgkp6erlWrVumdd95RSEiIDh8+LEkKCwtTQECAJ0sD8H9++Q08Hx8fnTp1Sjk5OVq4cKEefvhhHTx4UN26ddNVV12lW265RZI0b948HThwQGPHjlVgYKDS0tKUmprqemi+OSorKzVt2jRVVFQoLCxMCQkJKigo0I033nhBxwjAXDbL3cf9W3LnNluj83NycjRjxozzrl9dXa2wsDBVVVUpNDS02XXU1ta67qvX1NRw6wcAgFbkzue3x2/9AAAAnA2/9QPACHa7XcHBwWf9s9vtni4RgAcY8a0fAIiMjDznryeffrEcAO9CUAFgBF9fX9f7UQDgNG79AAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkHlPIqKimSz2XTs2LEmrzNjxgylpqa6tZ/vv/9eU6ZMUWhoqDp37qz77rtPNTU17hULAMBFhqBiiClTpmj37t0qLCx0/ZJ0Wlqap8sCAMCjCCpuys3NVefOnVVQUKBBgwYpODhY48aNU0VFhSTpiSee0Kuvvqp33nlHNptNNptNRUVF59zm3r17tWHDBr388ssaPny4rrnmGj3zzDPKy8vToUOH2uCoAAAwE0GlGX788UctXbpUr732mjZv3iy73a6srCxJUlZWliZNmuQKLxUVFRoxYsQ5t/fxxx+rc+fOuuKKK1zzxowZow4dOuiTTz5p1WMBAMBkvJm2GU6ePKkXXnhB/fr1kyRlZGToD3/4gyQpODhYAQEBcjgcioiIaNL2Dh8+rB49ejSY5+vrq/DwcB0+fLhliwcAoB1hRKUZAgMDXSFFknr16qXKykoPVgQAwMWJoNIMHTt2bDBts9lkWVaztxcREXFG0Dl16pS+//77Jo/KAABwMSKotAI/Pz85nc4mt09JSdGxY8dUWlrqmvf++++rvr5ew4cPb40SAQBoFwgqrSAmJkafffaZ9u3bp7///e86efLkOdsPGjRI48aN0/33369PP/1UxcXFysjI0F133cVP2wMAvBpBpRXcf//9iouL0xVXXKHu3buruLj4vOu8/vrrGjhwoEaPHq2bb75Z11xzjVauXNkG1QIAYC6bdSEPV3hYdXW1wsLCVFVVpdDQ0GZvp7a2VsHBwZKkmpoaBQUFtVSJAADgF9z5/GZEBQAAGIug0gYWLVqk4ODgRv/Gjx/v6fIAADBWs174tm3bNq1Zs0Z2u111dXUNlv3nf/5nixR2MXnggQc0adKkRpcFBAS0cTUAALQfbgeVvLw8TZs2TWPHjtVf//pX3XTTTdq/f7+OHDmi22+/vTVqbPfCw8MVHh7u6TIAAGh33L71s2jRIj399NP67//+b/n5+Sk7O1v/8z//o0mTJqlPnz6tUSMAAPBSbgeV8vJyTZgwQdJPLzarra2VzWbTQw89xNdpAQBAi3I7qHTp0kXHjx+XJEVFRWnXrl2SpGPHjunHH39s2eoAAIBXc/sZlZEjR6qwsFDx8fG64447NGvWLL3//vsqLCzU6NGjW6NGAADgpdwOKs8++6xOnDghSXrsscfUsWNHbdmyRf/0T/+kefPmtXiBAADAe7kdVH7+7ZUOHTro0UcfbdGCAAAATnP7GZXc3NxG5586dUpz58690HoAAABc3A4qDz74oO644w798MMPrnn79u3T8OHDtXr16hYtDgAAeDe3g8qOHTv07bffKj4+XoWFhXruuec0dOhQDRw4UH/7299ao0YAAOCl3H5GpV+/fiouLlZmZqbGjRsnHx8fvfrqq5o8eXJr1AcAALxYs36UcP369crLy1NKSoo6d+6sV155RYcOHWrp2gAAgJdzO6j85je/0R133KE5c+boww8/1GeffSY/Pz/Fx8drzZo1rVEjAADwUm7f+ikuLtYnn3yixMRESVJERIT+8pe/6LnnntO999571l8JBgAAcJfbQaW0tFT+/v5nzE9PT9eYMWNapCgAAACpGbd+Ggspp8XFxV1QMQAAAD/n9oiKJK1du1Zr1qyR3W5XXV1dg2Xbt29vkcIAAADcHlH505/+pHvuuUc9e/bUjh07lJycrK5du+rAgQMaP358a9QIAAC8lNtB5d/+7d+0cuVKPfPMM/Lz89MjjzyiwsJCPfjgg6qqqmqNGgEAgJdyO6jY7XaNGDFCkhQQEKDjx49LkqZOncor9AEAQItyO6hERETo+++/lyT16dNHW7dulSR9+eWXsiyrZasDAABeze2gcsMNN+i//uu/JEn33HOPHnroId1444268847dfvtt7d4gQAAwHu5/a2fxx57TFFRUZJ+endK165dtWXLFt12220aN25cixcIAAC8l81y836Nj4+PKioq1KNHjwbzjx49qh49esjpdLZogedSXV2tsLAwVVVVKTQ0tNnbqa2tVXBwsCSppqZGQUFBLVUiAAD4BXc+v92+9XO2XFNTU6NOnTq5uzkAAICzavKtn9mzZ0uSbDabfv/73yswMNC1zOl06pNPPtGQIUNavEAAAOC9mhxUduzYIemnEZWysjL5+fm5lvn5+SkxMVFZWVktXyEAAPBaTQ4qH3zwgaSfvumTnZ19Qc+EAAAANIXb3/rJyclpjToAAADO4PbDtAAAAG2FoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAIA2ZlmW0tLSFB4eLpvNpp07d3q6JGMRVAAA+JmioiLZbDYdO3as1faxYcMG5ebmat26daqoqNDll1/e5HWXLFkim82mzMzMVqvPJB4NKs8//7wSEhIUGhqq0NBQpaSk6N133/VkSQAAtLry8nL16tVLI0aMUEREhHx9fZu0XklJiV588UUlJCS0coXm8GhQ6d27t5YsWaLS0lJt27ZNN9xwgyZOnKjdu3d7siwAAM7K4XAoKytLUVFRCgoK0vDhw1VUVORafvToUU2ePFlRUVEKDAxUfHy8Vq9e7Vo+Y8YMzZw5U3a7XTabTTExMU3ab01NjaZMmaKXXnpJXbp0aeGjMlfTIlwrufXWWxtMP/nkk3r++ee1detWDR482ENVAd6ntrbW0yUAHhEUFOT2OhkZGdqzZ4/y8vIUGRmp/Px8jRs3TmVlZYqNjdWJEyc0bNgwzZkzR6GhoVq/fr2mTp2qfv36KTk5WdnZ2erXr59WrlypkpIS+fj4NGm/6enpmjBhgsaMGaOFCxe6XXd75dGg8nNOp1NvvvmmamtrlZKS0mgbh8Mhh8Phmq6urm6r8oCLWnBwsKdLADzCsiy32tvtduXk5MhutysyMlKSlJWVpQ0bNignJ0eLFi1SVFSUsrKyXOvMnDlTBQUFWrNmjZKTkxUWFqaQkBD5+PgoIiKiSfvNy8vT9u3bVVJS4la9FwOPB5WysjKlpKToxIkTCg4OVn5+vi677LJG2y5evFgLFixo4wqBi1dtbS0hBXBDWVmZnE6nBgwY0GC+w+FQ165dJf30H+9FixZpzZo1OnjwoOrq6uRwOBQYGNisfX7zzTeaNWuWCgsL1alTpws+hvbG40ElLi5OO3fuVFVVldauXavp06dr06ZNjYaVuXPnavbs2a7p6upqRUdHt2W5wEXryJEjzRoGB7xJTU2NfHx8VFpaesYtm9Oh/6mnnlJ2drZWrFih+Ph4BQUFKTMzU3V1dc3aZ2lpqSorKzV06FDXPKfTqc2bN+vZZ5+Vw+Fo8u2j9sjjQcXPz0/9+/eXJA0bNkwlJSXKzs7Wiy++eEZbf39/+fv7t3WJgFcICgoiqADnkZSUJKfTqcrKSl177bWNtikuLtbEiRN19913S5Lq6+u1f//+s94tOJ/Ro0errKyswbx77rlHAwcO1Jw5cy7qkCIZEFR+qb6+vsFzKAAAeEJZWZlCQkJc0zabTYmJiZoyZYqmTZumZcuWKSkpSd999502btyohIQETZgwQbGxsVq7dq22bNmiLl26aPny5Tpy5Eizg0pISMgZ71kJCgpS165d3Xr/Snvl0aAyd+5cjR8/Xn369NHx48e1atUqFRUVqaCgwJNlAQCgkSNHNpj28fHRqVOnlJOTo4ULF+rhhx/WwYMH1a1bN1111VW65ZZbJEnz5s3TgQMHNHbsWAUGBiotLU2pqamqqqryxGG0ezbL3UeeW9B9992njRs3qqKiQmFhYUpISNCcOXN04403Nmn96upqhYWFqaqqSqGhoc2u4+cPFNbU1DD8Da/BuQ/AE9z5/PboiMorr7ziyd0DAADD8Vs/AAB4kN1uV3Bw8Fn/7Ha7p0v0KOMepgUAwJtERkae89eTT79YzlsRVAAA8CBfX1/XazpwJm79AAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUADSqqKhINptNx44da/I6M2bMUGpqqlv7efLJJzVixAgFBgaqc+fObq0L4OJHUAHgUXV1dbrjjjv029/+1tOlADAQQQVAk+Tm5qpz584qKCjQoEGDFBwcrHHjxqmiokKS9MQTT+jVV1/VO++8I5vNJpvNpqKiovNud8GCBXrooYcUHx/fykcAoD3y9XQBpqmtrfV0CUCbcfd8//HHH7V06VK99tpr6tChg+6++25lZWXp9ddfV1ZWlvbu3avq6mrl5ORIksLDw1ujbABehKDyCz179vR0CYCxTp48qRdeeEH9+vWTJGVkZOgPf/iDJCk4OFgBAQFyOByKiIjwZJkALiLc+pEUFBQky7I8XQZgvMDAQFdIkaRevXqpsrLSgxUBuNgxovIzNTU1ni4BaFO1tbVujSJ27NixwbTNZiPkA2hVBJWfCQoK8nQJQLvm5+cnp9Pp6TIAXEQIKgBaTExMjAoKCrRv3z517dpVYWFhZ4zC/JLdbtf3338vu90up9OpnTt3SpL69++v4ODgNqgagMl4RgVAi7n//vsVFxenK664Qt27d1dxcfF51/n973+vpKQkzZ8/XzU1NUpKSlJSUpK2bdvWBhUDMJ3Nasc3mKurqxUWFqaqqiqFhoZ6uhyg3amtrXWNWtTU1HD7E0CbcOfzmxEVAABgLIIKgFazaNEiBQcHN/o3fvx4T5cHoB3gYVoAreaBBx7QpEmTGl0WEBDQxtUAaI8IKgBaTXh4OK/RB3BBuPUDAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAEAbsyxLaWlpCg8Pl81m086dOz1dkrEIKgAA/ExRUZFsNpuOHTvWavvYsGGDcnNztW7dOlVUVOjyyy8/Z/snnnhCNputwd/AgQNbrT6T+Hq6AAAAvE15ebl69eqlESNGNHmdwYMH67333nNN+/p6x0e4MSMqS5Yskc1mU2ZmpqdLAQDgrBwOh7KyshQVFaWgoCANHz5cRUVFruVHjx7V5MmTFRUVpcDAQMXHx2v16tWu5TNmzNDMmTNlt9tls9kUExPTpP36+voqIiLC9detW7cWPjIzGRHHSkpK9OKLLyohIcHTpQBeq7a21tMlAG0uKCjI7XUyMjK0Z88e5eXlKTIyUvn5+Ro3bpzKysoUGxurEydOaNiwYZozZ45CQ0O1fv16TZ06Vf369VNycrKys7PVr18/rVy5UiUlJfLx8WnSfj///HNFRkaqU6dOSklJ0eLFi9WnTx+3629vPB5UampqNGXKFL300ktauHDhOds6HA45HA7XdHV1dWuXB3iNnj17eroEoM1ZluVWe7vdrpycHNntdkVGRkqSsrKytGHDBuXk5GjRokWKiopSVlaWa52ZM2eqoKBAa9asUXJyssLCwhQSEiIfHx9FREQ0ab/Dhw9Xbm6u4uLiVFFRoQULFujaa6/Vrl27FBIS4tYxtDceDyrp6emaMGGCxowZc96gsnjxYi1YsKCNKgMufkFBQbIsSzabzdOlAO1CWVmZnE6nBgwY0GC+w+FQ165dJUlOp1OLFi3SmjVrdPDgQdXV1cnhcCgwMLDZ+x0/frzr3wkJCRo+fLj69u2rNWvW6L777mv2dtsDjwaVvLw8bd++XSUlJU1qP3fuXM2ePds1XV1drejo6NYqD/AaNTU1ni4BaBdqamrk4+Oj0tLSM27ZBAcHS5KeeuopZWdna8WKFYqPj1dQUJAyMzNVV1fXYnV07txZAwYM0BdffNFi2zSVx4LKN998o1mzZqmwsFCdOnVq0jr+/v7y9/dv5coA79Oc+/SAN0pKSpLT6VRlZaWuvfbaRtsUFxdr4sSJuvvuuyVJ9fX12r9/vy677LIWq6Ompkbl5eWaOnVqi23TVB4LKqWlpaqsrNTQoUNd85xOpzZv3qxnn31WDoejyQ8YAQDQ0srKyho8/2Gz2ZSYmKgpU6Zo2rRpWrZsmZKSkvTdd99p48aNSkhI0IQJExQbG6u1a9dqy5Yt6tKli5YvX64jR45cUFDJysrSrbfeqr59++rQoUOaP3++fHx8NHny5JY4VKN5LKiMHj1aZWVlDebdc889GjhwoObMmUNIAQB41MiRIxtM+/j46NSpU8rJydHChQv18MMP6+DBg+rWrZuuuuoq3XLLLZKkefPm6cCBAxo7dqwCAwOVlpam1NRUVVVVNbuWb7/9VpMnT9bRo0fVvXt3XXPNNdq6dau6d+9+QcfYHtgsdx95bkWjRo3SkCFDtGLFiia1r66uVlhYmKqqqhQaGtq6xQEAgBbhzue3MS98AwAA+CWPfz35537+Zj8AALyB3W4/5/Mre/bs8YoXu52NUUEFAABvExkZec5fTz79YjlvRVABAMCDfH191b9/f0+XYSyeUQEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMJavpwu4EJZlSZKqq6s9XAkAAGiq05/bpz/Hz6VdB5Xjx49LkqKjoz1cCQAAcNfx48cVFhZ2zjY2qylxxlD19fU6dOiQQkJCZLPZLmhb1dXVio6O1jfffKPQ0NAWqvDiRp+5h/5yD/3lHvrLffSZe1qyvyzL0vHjxxUZGakOHc79FEq7HlHp0KGDevfu3aLbDA0N5YR1E33mHvrLPfSXe+gv99Fn7mmp/jrfSMppPEwLAACMRVABAADGIqj8H39/f82fP1/+/v6eLqXdoM/cQ3+5h/5yD/3lPvrMPZ7qr3b9MC0AALi4MaICAACMRVABAADGIqgAAABjEVQAAICxvCqoPPfcc4qJiVGnTp00fPhwffrpp+ds/+abb2rgwIHq1KmT4uPj9Ze//KWNKjWDO/2Vm5srm83W4K9Tp05tWK1nbd68WbfeeqsiIyNls9n09ttvn3edoqIiDR06VP7+/urfv79yc3NbvU6TuNtnRUVFZ5xjNptNhw8fbpuCPWzx4sW68sorFRISoh49eig1NVX79u0773reeh1rTn9583Xs+eefV0JCgutlbikpKXr33XfPuU5bnVteE1TeeOMNzZ49W/Pnz9f27duVmJiosWPHqrKystH2W7Zs0eTJk3Xfffdpx44dSk1NVWpqqnbt2tXGlXuGu/0l/fS2woqKCtff119/3YYVe1Ztba0SExP13HPPNan9l19+qQkTJuj666/Xzp07lZmZqV//+tcqKCho5UrN4W6fnbZv374G51mPHj1aqUKzbNq0Senp6dq6dasKCwt18uRJ3XTTTaqtrT3rOt58HWtOf0neex3r3bu3lixZotLSUm3btk033HCDJk6cqN27dzfavk3PLctLJCcnW+np6a5pp9NpRUZGWosXL260/aRJk6wJEyY0mDd8+HDrN7/5TavWaQp3+ysnJ8cKCwtro+rMJsnKz88/Z5tHHnnEGjx4cIN5d955pzV27NhWrMxcTemzDz74wJJk/fDDD21Sk+kqKystSdamTZvO2sbbr2M/15T+4jrWUJcuXayXX3650WVteW55xYhKXV2dSktLNWbMGNe8Dh06aMyYMfr4448bXefjjz9u0F6Sxo4de9b2F5Pm9Jck1dTUqG/fvoqOjj5nEod3n18XasiQIerVq5duvPFGFRcXe7ocj6mqqpIkhYeHn7UN59n/a0p/SVzHJMnpdCovL0+1tbVKSUlptE1bnlteEVT+/ve/y+l0qmfPng3m9+zZ86z3tw8fPuxW+4tJc/orLi5O//7v/6533nlH//Ef/6H6+nqNGDFC3377bVuU3O6c7fyqrq7WP/7xDw9VZbZevXrphRde0FtvvaW33npL0dHRGjVqlLZv3+7p0tpcfX29MjMzdfXVV+vyyy8/aztvvo79XFP7y9uvY2VlZQoODpa/v78eeOAB5efn67LLLmu0bVueW+3615NhjpSUlAbJe8SIERo0aJBefPFF/fGPf/RgZbhYxMXFKS4uzjU9YsQIlZeX6+mnn9Zrr73mwcraXnp6unbt2qWPPvrI06W0C03tL2+/jsXFxWnnzp2qqqrS2rVrNX36dG3atOmsYaWteMWISrdu3eTj46MjR440mH/kyBFFREQ0uk5ERIRb7S8mzemvX+rYsaOSkpL0xRdftEaJ7d7Zzq/Q0FAFBAR4qKr2Jzk52evOsYyMDK1bt04ffPCBevfufc623nwdO82d/volb7uO+fn5qX///ho2bJgWL16sxMREZWdnN9q2Lc8trwgqfn5+GjZsmDZu3OiaV19fr40bN571/ltKSkqD9pJUWFh41vYXk+b01y85nU6VlZWpV69erVVmu+bN51dL2rlzp9ecY5ZlKSMjQ/n5+Xr//fd1ySWXnHcdbz7PmtNfv+Tt17H6+no5HI5Gl7XpudXij+caKi8vz/L397dyc3OtPXv2WGlpaVbnzp2tw4cPW5ZlWVOnTrUeffRRV/vi4mLL19fXWrp0qbV3715r/vz5VseOHa2ysjJPHUKbcre/FixYYBUUFFjl5eVWaWmpddddd1mdOnWydu/e7alDaFPHjx+3duzYYe3YscOSZC1fvtzasWOH9fXXX1uWZVmPPvqoNXXqVFf7AwcOWIGBgdbvfvc7a+/evdZzzz1n+fj4WBs2bPDUIbQ5d/vs6aeftt5++23r888/t8rKyqxZs2ZZHTp0sN577z1PHUKb+u1vf2uFhYVZRUVFVkVFhevvxx9/dLXhOvb/mtNf3nwde/TRR61NmzZZX375pfXZZ59Zjz76qGWz2ay//vWvlmV59tzymqBiWZb1zDPPWH369LH8/Pys5ORka+vWra5l1113nTV9+vQG7desWWMNGDDA8vPzswYPHmytX7++jSv2LHf6KzMz09W2Z8+e1s0332xt377dA1V7xumvzv7y73QfTZ8+3bruuuvOWGfIkCGWn5+fdemll1o5OTltXrcnudtn//Iv/2L169fP6tSpkxUeHm6NGjXKev/99z1TvAc01leSGpw3XMf+X3P6y5uvY/fee6/Vt29fy8/Pz+revbs1evRoV0ixLM+eWzbLsqyWH6cBAAC4cF7xjAoAAGifCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVICL3KhRo5SZmenpMhr11VdfyWazaefOnW6tZ7PZ9Pbbb7dKTe564oknNGTIEE+XAVy0CCoA0EQmBSTAWxBUAJyhrq7O0yUAgCSCCuAVTp06pYyMDIWFhalbt256/PHH9fOf+YqJidEf//hHTZs2TaGhoUpLS5MkzZkzRwMGDFBgYKAuvfRSPf744zp58qRrvdO3PV577TXFxMQoLCxMd911l44fP+5qU19fr3/9139V//795e/vrz59+ujJJ59sUN+BAwd0/fXXKzAwUImJifr444/dOr5vvvlGkyZNUufOnRUeHq6JEyfqq6++ci2fMWOGUlNTtXTpUvXq1Utdu3ZVenp6g2OpqKjQhAkTFBAQoEsuuUSrVq1STEyMVqxY4eojSbr99ttls9lc06edqw8ANB9BBfACr776qnx9ffXpp58qOztby5cv18svv9ygzdKlS5WYmKgdO3bo8ccflySFhIQoNzdXe/bsUXZ2tl566SU9/fTTDdYrLy/X22+/rXXr1mndunXatGmTlixZ4lo+d+5cLVmyRI8//rj27NmjVatWqWfPng228dhjjykrK0s7d+7UgAEDNHnyZJ06dapJx3by5EmNHTtWISEh+vDDD1VcXKzg4GCNGzeuwcjQBx98oPLycn3wwQd69dVXlZubq9zcXNfyadOm6dChQyoqKtJbb72llStXqrKy0rW8pKREkpSTk6OKigrXdFP6AMAFaJXfZAZgjOuuu84aNGiQVV9f75o3Z84ca9CgQa7pvn37Wqmpqefd1lNPPWUNGzbMNT1//nwrMDDQqq6uds373e9+Zw0fPtyyLMuqrq62/P39rZdeeqnR7X355ZeWJOvll192zdu9e7clydq7d+9Z65Bk5efnW5ZlWa+99poVFxfX4PgcDocVEBBgFRQUWJZlWdOnT7f69u1rnTp1ytXmjjvusO68807Lsixr7969liSrpKTEtfzzzz+3JFlPP/10o/ttah8AuDCMqABe4KqrrpLNZnNNp6Sk6PPPP5fT6XTNu+KKK85Y74033tDVV1+tiIgIBQcHa968ebLb7Q3axMTEKCQkxDXdq1cv10jE3r175XA4NHr06HPWl5CQ0GB9SQ1GM87lb3/7m7744guFhIQoODhYwcHBCg8P14kTJ1ReXu5qN3jwYPn4+DRa5759++Tr66uhQ4e6lvfv319dunRpUg3n6gMAF8bX0wUAMENQUFCD6Y8//lhTpkzRggULNHbsWIWFhSkvL0/Lli1r0K5jx44Npm02m+rr6yVJAQEBTdr3z7dxOlCd3sb51NTUaNiwYXr99dfPWNa9e/cm1XmhWnPbgLcjqABe4JNPPmkwvXXrVsXGxjYYYfilLVu2qG/fvnrsscdc877++mu39hsbG6uAgABt3LhRv/71r90ruomGDh2qN954Qz169FBoaGizthEXF6dTp05px44dGjZsmCTpiy++0A8//NCgXceOHRuMQgFofdz6AbyA3W7X7NmztW/fPq1evVrPPPOMZs2adc51YmNjZbfblZeXp/Lycv3pT39Sfn6+W/vt1KmT5syZo0ceeUR//vOfVV5erq1bt+qVV165kMNpYMqUKerWrZsmTpyoDz/8UF9++aWKior04IMP6ttvv23SNgYOHKgxY8YoLS1Nn376qXbs2KG0tDQFBAQ0uGUWExOjjRs36vDhw2eEGACtg6ACeIFp06bpH//4h5KTk5Wenq5Zs2a5voJ8NrfddpseeughZWRkaMiQIdqyZYvr20DuePzxx/Xwww/r97//vQYNGqQ777yzRZ/fCAwM1ObNm9WnTx/96le/0qBBg3TffffpxIkTbo2w/PnPf1bPnj01cuRI3X777br//vsVEhKiTp06udosW7ZMhYWFio6OVlJSUosdA4Czs1nWz16mAACQJH377beKjo7We++9d96HgQG0HoIKAEh6//33VVNTo/j4eFVUVOiRRx7RwYMHtX///jMelgXQdniYFgD004vj/vmf/1kHDhxQSEiIRowYoddff52QAngYIyoAAMBYPEwLAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABjrfwGND7zx0u8eAQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "newick_tree = open(\"./newick.txt\", \"r\")\n",
    "newick = newick_tree.read()\n",
    "newick_tree.close()\n",
    "tree = Phylo.read(StringIO(newick), \"newick\")\n",
    "\n",
    "# Change nodes names\n",
    "for idx, clade in enumerate(tree.get_nonterminals()):\n",
    "    clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "\n",
    "# Change leaves names\n",
    "for idx, clade in enumerate(tree.get_terminals()):\n",
    "    clade.name = \"Leaf_\" + clade.name \n",
    "\n",
    "Phylo.draw(tree)\n",
    "\n",
    "net = Phylo.to_networkx(tree) # to graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "simulate() missing 1 required positional argument: 'directory'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m edges, node_distr, filename \u001b[39m=\u001b[39m simulate(net, \u001b[39m500\u001b[39;49m) \u001b[39m# simulate data calling the module data_simulation.py \u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[39m# Preprocess leave alignments\u001b[39;00m\n\u001b[1;32m      4\u001b[0m leave_alignments \u001b[39m=\u001b[39m [i \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m SeqIO\u001b[39m.\u001b[39mparse(filename, \u001b[39m\"\u001b[39m\u001b[39mfasta\u001b[39m\u001b[39m\"\u001b[39m)]\n",
      "\u001b[0;31mTypeError\u001b[0m: simulate() missing 1 required positional argument: 'directory'"
     ]
    }
   ],
   "source": [
    "edges, node_distr, filename = simulate(net, 500) # simulate data calling the module data_simulation.py \n",
    "\n",
    "# Preprocess leave alignments\n",
    "leave_alignments = [i for i in SeqIO.parse(filename, \"fasta\")]\n",
    "for i in leave_alignments:\n",
    "    print(\"Seq>\", i.seq)\n",
    "# Codify the alphabet {A,G,C,T} -> {0,1,2,3}\n",
    "leave_sequences = []\n",
    "for i in leave_alignments:\n",
    "    seq = \"\"\n",
    "    for j in i.seq:\n",
    "        if j == \"A\":\n",
    "            seq += \"0\"\n",
    "        elif j == \"G\":\n",
    "            seq += \"1\"\n",
    "        elif j == \"C\":\n",
    "            seq += \"2\"\n",
    "        elif j == \"T\":\n",
    "            seq += \"3\"\n",
    "        else:\n",
    "            raise ValueError(\"Invalid nucleotide\")\n",
    "    leave_sequences.append(seq)\n",
    "#leave_sequences\n",
    "edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## **EM algorithm**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from evaluate import init_EM\n",
    "\n",
    "Results = init_EM(newick, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Source: Int_0\n",
      "Target: Leaf_2\n",
      "Matrix:\n",
      " [[0.80529814 0.06095358 0.09251042 0.04123786]\n",
      " [0.05611119 0.71483627 0.11102639 0.11802615]\n",
      " [0.0952762  0.07415636 0.731184   0.09938344]\n",
      " [0.02196288 0.0820518  0.08234946 0.81363586]]\n",
      "Source: Int_0\n",
      "Target: Leaf_3\n",
      "Matrix:\n",
      " [[0.63632807 0.06782976 0.1446222  0.15121997]\n",
      " [0.12810897 0.67208297 0.04706075 0.15274731]\n",
      " [0.10204507 0.09021686 0.74449869 0.06323937]\n",
      " [0.06570111 0.08178298 0.05155947 0.80095644]]\n",
      "Source: Int_0\n",
      "Target: Int_1\n",
      "Matrix:\n",
      " [[0.72577892 0.07421689 0.0831795  0.11682468]\n",
      " [0.08700214 0.7261134  0.06686419 0.12002027]\n",
      " [0.09519544 0.182137   0.68012648 0.04254107]\n",
      " [0.06761601 0.06509887 0.08961157 0.77767356]]\n",
      "Source: Int_1\n",
      "Target: Leaf_4\n",
      "Matrix:\n",
      " [[0.85639817 0.07234946 0.04036045 0.03089192]\n",
      " [0.0298384  0.73128255 0.0944944  0.14438465]\n",
      " [0.03533407 0.10420379 0.77836777 0.08209437]\n",
      " [0.06403097 0.03195277 0.06966545 0.8343508 ]]\n",
      "Source: Int_1\n",
      "Target: Leaf_5\n",
      "Matrix:\n",
      " [[0.80075352 0.06438893 0.11198018 0.02287738]\n",
      " [0.07400816 0.72459648 0.10712499 0.09427037]\n",
      " [0.14358855 0.08350079 0.74500083 0.02790983]\n",
      " [0.01210798 0.14166184 0.06442053 0.78180965]]\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(Results)):\n",
    "    print(\"Source:\", Results[i].source)\n",
    "    print(\"Target:\", Results[i].target)\n",
    "    print(\"Matrix:\\n\", Results[i].matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "l -73482.53888183684\n",
      "73482.53888183684\n",
      "--------\n",
      "2803.2235244028125\n",
      "--------\n",
      "403.42311980716477\n",
      "--------\n",
      "147.34679458629398\n",
      "--------\n",
      "65.81431201811938\n",
      "--------\n",
      "32.62458047835389\n",
      "--------\n",
      "17.4679848113592\n",
      "--------\n",
      "9.96598948755127\n",
      "--------\n",
      "5.979527266841615\n",
      "--------\n",
      "3.724567122422741\n",
      "--------\n",
      "2.382098244605004\n",
      "--------\n",
      "1.5508962869207608\n",
      "--------\n",
      "1.021128828782821\n",
      "--------\n",
      "0.6763189632765716\n",
      "--------\n",
      "0.44849936086393427\n",
      "--------\n",
      "0.29640033769828733\n",
      "--------\n",
      "0.19417478849936742\n",
      "--------\n",
      "0.12524008554464672\n",
      "--------\n",
      "0.07875408534891903\n",
      "--------\n",
      "0.047520216307020746\n",
      "--------\n",
      "0.02670328039675951\n",
      "--------\n",
      "0.01302253625181038\n",
      "--------\n",
      "0.004234299674862996\n",
      "--------\n",
      "0.0012054162216372788\n",
      "--------\n",
      "0.004364024440292269\n",
      "--------\n",
      "0.005981738388072699\n",
      "--------\n",
      "0.006572629208676517\n",
      "--------\n",
      "0.006493631357443519\n",
      "--------\n",
      "0.005992020829580724\n",
      "--------\n",
      "0.005238327459665015\n",
      "--------\n",
      "0.00434925795707386\n",
      "--------\n",
      "0.003403757931664586\n",
      "--------\n",
      "0.002454297398799099\n",
      "--------\n",
      "0.0015348267188528553\n",
      "--------\n",
      "Done!\n",
      "Number of iterations:  34\n"
     ]
    }
   ],
   "source": [
    "# Auxiliary functions\n",
    "\n",
    "def init_root_distribution():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the root distribution (all probabilities between 0.2 and 0.3)\n",
    "    \"\"\"\n",
    "    estimated_root_distribution = []\n",
    "    while True:\n",
    "        estimated_root_distribution = np.random.dirichlet([1,1,1,1])\n",
    "        i = 0\n",
    "        while i<4:\n",
    "            if estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3:\n",
    "                i += 1\n",
    "            else:\n",
    "                break\n",
    "        if i == 4:\n",
    "            assert(np.sum(estimated_root_distribution) < 1.0000001 and np.sum(estimated_root_distribution) > 0.9999999)\n",
    "            assert(estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3 for i in range(4))\n",
    "            break\n",
    "    return estimated_root_distribution\n",
    "\n",
    "def init_transition_matrix():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the transition matrices (which we estimate as LARGE-DIAGONAL-VALUE matrices)\n",
    "    \"\"\"\n",
    "    M = np.zeros((4,4))\n",
    "    i=0\n",
    "    while i<4:\n",
    "        R = np.random.dirichlet([1,1,1,1])\n",
    "        if R[i] > 0.6 and R[i] < 0.8:\n",
    "            M[i,:] = R \n",
    "            i = i + 1\n",
    "    assert (M[i,i] > 0.6 and M[i,i] < 0.8 for i in range(4))\n",
    "    assert (np.sum(M[i,:]) < 1.0000001 and np.sum(M[i,:]) > 0.9999999 for i in range(4))\n",
    "    return M\n",
    "\n",
    "def log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "    \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "    logL = 0\n",
    "    for obs in states:\n",
    "        p_i = 0 \n",
    "        observed_data = obs[n_int:]\n",
    "        if observed_data in u_i:\n",
    "            for state in states:\n",
    "                if state[n_int:] == observed_data:\n",
    "                    pi = root_distribution[int(state[0])]\n",
    "                    for p in params.items():\n",
    "                        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                        pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                    p_i += pi\n",
    "        logL += (u_i[observed_data] * math.log(p_i))\n",
    "    return logL\n",
    "\n",
    "############################################################################################################\n",
    "########################################## FELSENSTEIN ALGORITHM ###########################################\n",
    "############################################################################################################\n",
    "# # Felsentstein algorithm\n",
    "# # \"((4,5),(2,3))\"\n",
    "# newick_tree = \"(((4:1,5:1):1,(6:1,7:1):1):1,(8:1,9:1):1);\"\n",
    "# L_tree = Phylo.read(StringIO(newick_tree), \"newick\")\n",
    "# # Change nodes names\n",
    "# for idx, clade in enumerate(L_tree.get_nonterminals()):\n",
    "#     id = 0\n",
    "#     if idx == 2:\n",
    "#         clade.name = \"Int_Id_\" + str(id)\n",
    "#         id += 1\n",
    "#     else:\n",
    "#         clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "# # Change leaves names\n",
    "# for idx, clade in enumerate(L_tree.get_terminals()):\n",
    "#     clade.name = \"Leaf_\" + clade.name \n",
    "# L_net = Phylo.to_networkx(L_tree) # to graph\n",
    "# L_edges = []\n",
    "# for edge in L_net.edges():\n",
    "#     if edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     elif edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] != \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     elif edge[0].name.split(\"_\")[1] != \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     else:\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     L_edges.append((u,v))\n",
    "# # Create a dictionary to store the father and son information\n",
    "# family_tree = {}\n",
    "# # Loop through the edges and update the dictionary\n",
    "# for father, son in L_edges:\n",
    "#     if father not in family_tree:\n",
    "#         family_tree[father] = []\n",
    "#     family_tree[father].append(son)\n",
    "# # Create a DataFrame from the dictionary\n",
    "# tree_df = pd.DataFrame.from_dict(family_tree, orient='index')\n",
    "# # Fill NaN values with empty strings\n",
    "# tree_df = tree_df.fillna('')\n",
    "# # Change column names\n",
    "# tree_df.columns = ['child_1', 'child_2']\n",
    "\n",
    "# class TreeInference():\n",
    "#     def __init__(self, tree_data, leaf_values, params, root_distr):\n",
    "#         # Set parameters\n",
    "#         self.tree_data = tree_data.copy()\n",
    "#         self.leaf_values = leaf_values\n",
    "#         self.params = params\n",
    "#         self.root_distr = root_distr\n",
    "\n",
    "#         # Find IDs of internal nodes, all nodes and leaf nodes\n",
    "#         self.internal_node_ids = np.array(tree_data.index)\n",
    "#         self.all_node_ids = np.unique(np.concatenate((np.array(self.tree_data.index),\n",
    "#                                                       np.array(self.tree_data[\"child_1\"]),\n",
    "#                                                       np.array(self.tree_data[\"child_2\"]))))\n",
    "#         self.leaf_node_ids = list(set(self.all_node_ids) - set(self.internal_node_ids))\n",
    "        \n",
    "#         # Count internal nodes and leaves for each row in tree_data and set root node\n",
    "#         self.treeNodeCounts()\n",
    "#         self.root_node = 0\n",
    "        \n",
    "#         # Get indexes for matricies for states, e.g. A is row 0, G is row 2...\n",
    "#         states = [\"A\", \"G\", \"C\", \"T\"]\n",
    "#         self.matrix_idx = dict(zip(states, range(len(states))))\n",
    "        \n",
    "#         # Empty matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "#         # Log likelihood of whole tree\n",
    "#         self.L_root = None\n",
    "        \n",
    "\n",
    "#     \"\"\"\n",
    "#     Calculates the number of internal nodes and leaves below each internal node.\n",
    "#     This works by recursively iterating down the child nodes from the original internal node until a\n",
    "#     leaf node is found, and then returning upwards counting the leaves and nodes.\n",
    "#     The new derived information is then added as two new columns at the end of the data.\n",
    "#     \"\"\"\n",
    "#     def treeNodeCounts(self):\n",
    "#         n_nodes = np.array([])\n",
    "#         n_leaves = np.array([])\n",
    "\n",
    "#         # Set the number of internal nodes and leaves for each internal node row\n",
    "#         for n in self.internal_node_ids:\n",
    "#             num_nodes, num_leaves = self.checkNode(self.tree_data, n, self.internal_node_ids)\n",
    "#             n_nodes = np.append(n_nodes, num_nodes)\n",
    "#             n_leaves = np.append(n_leaves, num_leaves)\n",
    "\n",
    "#         # Add new information to dataframe\n",
    "#         self.tree_data[\"N_Internal\"] = n_nodes\n",
    "#         self.tree_data[\"N_Leaves\"] = n_leaves\n",
    "                \n",
    "#     # Recursively find the number of internal nodes and leaves given a node ID\n",
    "#     def checkNode(self, tree_data, node_id, node_array):\n",
    "#         # Record the number of internal nodes and leaves below\n",
    "#         num_nodes = 0\n",
    "#         num_leaves = 0\n",
    "\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[node_id][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[node_id][\"child_2\"]\n",
    "\n",
    "#         # Check if found an internal node to the left\n",
    "#         if child_1 in node_array:\n",
    "#             # Recursively check nodes below\n",
    "#             d1_nodes, d1_leaves = self.checkNode(self.tree_data, child_1, node_array)\n",
    "#             num_nodes += d1_nodes + 1\n",
    "#             num_leaves += d1_leaves\n",
    "\n",
    "#         # Check if found an internal node to the right\n",
    "#         if child_2 in node_array:\n",
    "#             d2_nodes, d2_leaves = self.checkNode(self.tree_data, child_2, node_array)\n",
    "#             num_nodes += d2_nodes + 1\n",
    "#             num_leaves += d2_leaves\n",
    "\n",
    "#         return num_nodes, num_leaves\n",
    "    \n",
    "    \n",
    "#     # Calculate likelihood for an internal node\n",
    "#     def nodeLikelihood(self, node_n, params):\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[str(node_n)][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[str(node_n)][\"child_2\"]\n",
    "        \n",
    "#         # Check if likelihood of left child node set\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(child_1, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_1), 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(int(child_1), params)\n",
    "#         # Check if likelihood of right child node set\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 self.nodeLikelihood(child_2, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_2), 0] is None):\n",
    "#                 self.nodeLikelihood(int(child_2), params)\n",
    "        \n",
    "#         M1, M2 = np.full((4,4),None), np.full((4,4),None)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             # Create transition matrix for left and right child nodes\n",
    "#             matrix_name_child1 = \"M_0_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_0_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_1 == \"Id_0\":\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = np.identity(4)\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_2 == \"Id_0\":\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = np.identity(4)\n",
    "#         else:\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "\n",
    "#         # Get conditional likelihood of child nodes\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             d1_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d1_likelihood = self.L[int(child_1)]\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             d2_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d2_likelihood = self.L[int(child_2)]\n",
    "        \n",
    "#         # Calculate the likelihood of each state for the node\n",
    "#         left_prob = np.sum(M1 * d1_likelihood, axis = 1)\n",
    "#         right_prob = np.sum(M2 * d2_likelihood, axis = 1)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             self.L[-1,:] = np.multiply(left_prob, right_prob)\n",
    "#         else:\n",
    "#             self.L[int(node_n),:] = np.multiply(left_prob, right_prob)\n",
    "    \n",
    "#     # Run Felsenstein‚Äôs algorithm to calculate the log likelihood of the tree\n",
    "#     def calculateLikelihood(self, params, Estep):\n",
    "#         # Clear matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "        \n",
    "#         # Set likelihood of leaf states\n",
    "#         for leaf_n in self.leaf_node_ids:\n",
    "#             if leaf_n == \"Id_0\":\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[-1, :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[-1, self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "#             else:\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[int(leaf_n), :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[int(leaf_n), self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "        \n",
    "#         for node_n in self.internal_node_ids:\n",
    "#             # Check likelihood has not been set\n",
    "#             if node_n == \"Id_0\":\n",
    "#                 if self.L[-1, 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "#             else:\n",
    "#                 if self.L[int(node_n), 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "        \n",
    "#         if Estep == True:\n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = np.sum(np.multiply(self.root_distr, self.L[self.root_node]))\n",
    "#         else: \n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = math.log(np.sum(np.multiply(self.root_distr, self.L[self.root_node])))\n",
    "        \n",
    "#         return self.L_root\n",
    "\n",
    "# def Felsenstein_log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "#     \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "#     logL = 0\n",
    "#     for obs in states:\n",
    "#         observed_data = obs[n_int:]\n",
    "#         tree_leaves = dict()\n",
    "#         for i in range(len(observed_data)):\n",
    "#             if observed_data[i] == str(0):\n",
    "#                 tree_leaves[n_int+i] = \"A\"\n",
    "#             elif observed_data[i] == str(1):\n",
    "#                 tree_leaves[n_int+i] = \"G\"\n",
    "#             elif observed_data[i] == str(2):\n",
    "#                 tree_leaves[n_int+i] = \"C\"\n",
    "#             elif observed_data[i] == str(3):\n",
    "#                 tree_leaves[n_int+i] = \"T\"\n",
    "#         tree_inference = TreeInference(tree_df, tree_leaves, params, root_distribution)\n",
    "#         tree_inference.calculateLikelihood(params, False)\n",
    "#         logL += (u_i[observed_data] * tree_inference.L_root)\n",
    "#         del tree_inference\n",
    "#     return logL\n",
    "\n",
    "\n",
    "############################################################################################################\n",
    "\n",
    "class Param:\n",
    "    \"\"\"\n",
    "    Class to store the parameters of the tree\n",
    "    \"\"\"\n",
    "    def __init__(self, edge, transition_matrix, alignment=None):\n",
    "        self.edge = edge\n",
    "        self.transition_matrix = transition_matrix\n",
    "        self.alignment = alignment # it will only be placed in the leaves\n",
    "\n",
    "# Obtain number of leaves and internal nodes\n",
    "n_leaves, n_int = 0, 0 \n",
    "for node in net.nodes():\n",
    "    if node.name.startswith(\"L\"):\n",
    "        n_leaves += 1\n",
    "    else:\n",
    "        n_int += 1\n",
    "\n",
    "# Count all the leaves ocurrences \n",
    "sequence_length = len(leave_sequences[0])\n",
    "number_of_sequences = len(leave_sequences)\n",
    "occurrences = []\n",
    "for i in range(sequence_length):\n",
    "    _ = ''\n",
    "    for j in range(number_of_sequences):\n",
    "        _ += leave_sequences[j][i]\n",
    "    occurrences.append(_)\n",
    "c = Counter(occurrences)\n",
    "u_i = OrderedDict(sorted(c.items()))\n",
    "\n",
    "# Define all possible (hidden) states of the internal nodes \n",
    "hidden_combinations = itertools.product([0,1,2,3], repeat=n_int)\n",
    "sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "for i in range(len(sorted_hidden_combinations)):\n",
    "    sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    " \n",
    "# Define all possible fully observed states (i.e. combining u_i + internal nodes)\n",
    "# '000000' would mean to have an 'A' in all nodes: Int_0, Int_1, Leaf_2, Leaf_3, Leaf_4, Leaf_5 \n",
    "states = list(itertools.product(list(sorted_hidden_combinations), list(u_i.keys())))\n",
    "states = [i[0]+i[1] for i in states]\n",
    "\n",
    "############################################################################################################\n",
    "###################################### STEP 0: Initialise parameters #######################################\n",
    "############################################################################################################\n",
    "estimated_root_distribution = init_root_distribution() # First estimation for the root distribution\n",
    "\n",
    "# PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "# for edge in net.edges():\n",
    "#     # If we are in a leaf, we need to specify the alignment\n",
    "#     iter_leaves = 0\n",
    "#     if edge[1].name.startswith(\"L\"):\n",
    "#         new_edge = Param(edge, init_transition_matrix(), leave_alignments[iter_leaves].seq)\n",
    "#         iter_leaves += 1\n",
    "#     # Otherwise, are in internal nodes\n",
    "#     else:\n",
    "#         new_edge = Param(edge, init_transition_matrix())\n",
    "#     u = str(edge[0].name.split(\"_\")[1])\n",
    "#     v = str(edge[1].name.split(\"_\")[1])\n",
    "#     name = \"M_\" + u + \"_to_\" + v\n",
    "#     PARAMS[name] = new_edge\n",
    "\n",
    "# ADRIA CODE\n",
    "i=0\n",
    "PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "for edge in net.edges():\n",
    "    # If we are in a leaf, we need to specify the alignment\n",
    "    iter_leaves = 0\n",
    "    if edge[1].name.startswith(\"L\"):\n",
    "        new_edge = Param(edge, Results[i].matrix, leave_alignments[iter_leaves].seq)\n",
    "        iter_leaves += 1\n",
    "    # Otherwise, are in internal nodes\n",
    "    else:\n",
    "        new_edge = Param(edge, Results[i].matrix)\n",
    "    u = str(edge[0].name.split(\"_\")[1])\n",
    "    v = str(edge[1].name.split(\"_\")[1])\n",
    "    name = \"M_\" + u + \"_to_\" + v\n",
    "    PARAMS[name] = new_edge\n",
    "    i += 1\n",
    "\n",
    "eps = 10**(-3)                  # epsilon for convergence\n",
    "iter = 0                        # iteration counter \n",
    "params = copy.copy(PARAMS)      # copy of the parameters (will be updated at each iteration)\n",
    "# logF = Felsenstein_log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "logL = log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "# print(logF)\n",
    "print(\"l\", logL)\n",
    "logL_ = 0\n",
    "\n",
    "while np.abs(logL_ - logL) > eps:\n",
    "    print(np.abs(logL_ - logL))\n",
    "    # print(np.abs(logF_ - logF))\n",
    "    print(\"--------\")\n",
    "    if iter > 0:\n",
    "        logL = logL_\n",
    "        # logF = logF_\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## E-step ##################################################\n",
    "    ############################################################################################################\n",
    "    # Define the expected hidden data matrix\n",
    "    U = dict()\n",
    "    root_distr = estimated_root_distribution ########\n",
    "\n",
    "    for obs in states:\n",
    "        U[obs] = u_i[obs[n_int:]] # initialise by computing u_i of the expected hidden data matrix formula\n",
    "        f_ij = estimated_root_distribution[int(obs[0])] \n",
    "        for p in params.items():\n",
    "            u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "            f_ij *= p[1].transition_matrix[int(obs[u]), int(obs[v])]\n",
    "        # Now f_ij := numerator of the expected hidden data matrix formula \n",
    "\n",
    "        # We need to marginalise to obtain f_i := denominator of the expected hidden data matrix formula\n",
    "        f_i = 0 # we initialise the partial data model\n",
    "        observed_data = obs[n_int:]\n",
    "        for state in states:\n",
    "            if state[n_int:] == observed_data:\n",
    "                pi = estimated_root_distribution[int(state[0])]\n",
    "                for p in params.items():\n",
    "                    u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                    pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                f_i += pi\n",
    "\n",
    "        # observed_data = obs[n_int:]\n",
    "        # tree_leaves = dict()\n",
    "        # for i in range(len(observed_data)):\n",
    "        #     if observed_data[i] == str(0):\n",
    "        #         tree_leaves[n_int+i] = \"A\"\n",
    "        #     elif observed_data[i] == str(1):\n",
    "        #         tree_leaves[n_int+i] = \"G\"\n",
    "        #     elif observed_data[i] == str(2):\n",
    "        #         tree_leaves[n_int+i] = \"C\"\n",
    "        #     elif observed_data[i] == str(3):\n",
    "        #         tree_leaves[n_int+i] = \"T\"\n",
    "        # tree_inference = TreeInference(tree_df, tree_leaves, params, root_distr)\n",
    "        # tree_inference.calculateLikelihood(params, True)\n",
    "\n",
    "        U[obs] *= f_ij/f_i\n",
    "        # U[obs] *= f_ij/tree_inference.L_root\n",
    "        # del tree_inference\n",
    "        # del tree_leaves\n",
    "\n",
    "\n",
    "    assert(len(U) <= 4**(n_int+n_leaves))\n",
    "    sum = 0\n",
    "    sum2 = 0\n",
    "    for i in U.items():\n",
    "        sum += i[1]\n",
    "    for i in u_i.items():\n",
    "        sum2 += i[1]\n",
    "    assert(sum < sequence_length+0.001 and sum > sequence_length-0.001)\n",
    "    assert(sum2 < sequence_length+0.001 and sum2 > sequence_length-0.001)\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## M-step ##################################################\n",
    "    ############################################################################################################\n",
    "    estimated_parameters = dict()\n",
    "    root_distr = np.zeros(4)\n",
    "    iter_root = 0 # counter for the root distribution, we only want to count once\n",
    "\n",
    "    for p in params.items():\n",
    "        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "        name = \"M_\" + str(u) + \"_to_\" + str(v)\n",
    "        estimated_parameters[name] = np.zeros((4,4))\n",
    "        for i in range(4):\n",
    "            for j in range(4):\n",
    "                ui = 0\n",
    "                for obs in U.items():\n",
    "                    # For the root distribution, we only need to consider the first internal node\n",
    "                    if iter_root == 0 and obs[0][0] == str(i):\n",
    "                        root_distr[i] += obs[1]\n",
    "                    # For the transition matrices:\n",
    "                    if obs[0][u] == str(i) and obs[0][v] == str(j):\n",
    "                        estimated_parameters[name][i][j] += obs[1]\n",
    "                    if obs[0][u] == str(i):\n",
    "                        ui += obs[1]\n",
    "                # Update root distribution\n",
    "                if iter_root == 0:\n",
    "                    root_distr[i] /= sequence_length\n",
    "                iter_root = 1\n",
    "                estimated_parameters[name][i][j] /= ui\n",
    "            iter_root = 0\n",
    "            \n",
    "        # Update parameters\n",
    "        params[name].transition_matrix = estimated_parameters[name]\n",
    "\n",
    "    ##########################################################################################################\n",
    "    ########################### Compute log-likelihood of the estimated parameters ###########################\n",
    "    ##########################################################################################################\n",
    "    # logF_ = Felsenstein_log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "    logL_ = log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "\n",
    "    iter += 1\n",
    "\n",
    "print(\"Done!\")\n",
    "print(\"Number of iterations: \", iter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################### DEBUGGING ############################\n",
    "\n",
    "# # All possible (theoretical) occurrences of nucleotides in the leaves\n",
    "# alphabet = [0, 1, 2, 3]\n",
    "# combinations = itertools.product(alphabet, repeat=n_leaves)\n",
    "# sorted_combinations = sorted(combinations)\n",
    "# for i in range(len(sorted_combinations)):\n",
    "#     sorted_combinations[i] = ''.join([str(s) for s in sorted_combinations[i]])\n",
    "\n",
    "# # All possible (hidden) states of the internal nodes (i.e. indexes of columns of U)\n",
    "# hidden_combinations = itertools.product(alphabet, repeat=n_int)\n",
    "# sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "# for i in range(len(sorted_hidden_combinations)):\n",
    "#     sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    "\n",
    "# # To check if an element is within a dictionary or not \n",
    "# if '0101' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')\n",
    "# if '0000' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix: [[0.64642183 0.02803076 0.27483013 0.05071728]\n",
      " [0.03848122 0.7757021  0.10189982 0.08391686]\n",
      " [0.03184807 0.07934462 0.82109495 0.06771236]\n",
      " [0.04702478 0.01879095 0.04076404 0.89342023]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix: [[0.55850753 0.13612552 0.2499751  0.05539185]\n",
      " [0.01791204 0.82605787 0.05907209 0.096958  ]\n",
      " [0.04534628 0.008074   0.85149918 0.09508054]\n",
      " [0.00646156 0.06355415 0.02348564 0.90649866]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix: [[0.75152183 0.09827146 0.07355421 0.0766525 ]\n",
      " [0.04888819 0.86735324 0.02989229 0.05386628]\n",
      " [0.05446854 0.02450495 0.75018255 0.17084396]\n",
      " [0.12039713 0.03979299 0.05715314 0.78265675]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix: [[0.82835309 0.03976936 0.05146899 0.08040856]\n",
      " [0.02327998 0.80873654 0.06043788 0.10754561]\n",
      " [0.11544022 0.13895831 0.71027209 0.03532938]\n",
      " [0.09988558 0.08065349 0.01737091 0.80209001]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix: [[0.78296278 0.06942125 0.08171685 0.06589912]\n",
      " [0.02166725 0.83743552 0.01932756 0.12156967]\n",
      " [0.11182788 0.03378373 0.78484422 0.06954417]\n",
      " [0.11063295 0.0881492  0.0495941  0.75162375]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23195748 0.24737645 0.23524372 0.28542235]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.6464218290229772"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the theoretical parameters:\n",
    "theoretical_matrices = dict()\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    u = edges[i].edge[0].name.split(\"_\")[1]\n",
    "    v = edges[i].edge[1].name.split(\"_\")[1]\n",
    "    matrix_name = \"M_\" + u + \"_to_\" + v\n",
    "    print(\"Transition matrix:\", edges[i].transition_matrix)\n",
    "    theoretical_matrices[matrix_name] = np.array(edges[i].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", node_distr['Int_0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'M_0_to_2': array([[0.64642183, 0.02803076, 0.27483013, 0.05071728],\n",
      "       [0.03848122, 0.7757021 , 0.10189982, 0.08391686],\n",
      "       [0.03184807, 0.07934462, 0.82109495, 0.06771236],\n",
      "       [0.04702478, 0.01879095, 0.04076404, 0.89342023]]), 'M_0_to_3': array([[0.55850753, 0.13612552, 0.2499751 , 0.05539185],\n",
      "       [0.01791204, 0.82605787, 0.05907209, 0.096958  ],\n",
      "       [0.04534628, 0.008074  , 0.85149918, 0.09508054],\n",
      "       [0.00646156, 0.06355415, 0.02348564, 0.90649866]]), 'M_0_to_1': array([[0.75152183, 0.09827146, 0.07355421, 0.0766525 ],\n",
      "       [0.04888819, 0.86735324, 0.02989229, 0.05386628],\n",
      "       [0.05446854, 0.02450495, 0.75018255, 0.17084396],\n",
      "       [0.12039713, 0.03979299, 0.05715314, 0.78265675]]), 'M_1_to_4': array([[0.82835309, 0.03976936, 0.05146899, 0.08040856],\n",
      "       [0.02327998, 0.80873654, 0.06043788, 0.10754561],\n",
      "       [0.11544022, 0.13895831, 0.71027209, 0.03532938],\n",
      "       [0.09988558, 0.08065349, 0.01737091, 0.80209001]]), 'M_1_to_5': array([[0.78296278, 0.06942125, 0.08171685, 0.06589912],\n",
      "       [0.02166725, 0.83743552, 0.01932756, 0.12156967],\n",
      "       [0.11182788, 0.03378373, 0.78484422, 0.06954417],\n",
      "       [0.11063295, 0.0881492 , 0.0495941 , 0.75162375]])}\n"
     ]
    }
   ],
   "source": [
    "# # save the dictionary to a file\n",
    "# np.save('data.npy', theoretical_matrices)\n",
    "# # load the dictionary from the file\n",
    "# loaded_dict = np.load('data.npy', allow_pickle=True).item()\n",
    "# # print the loaded dictionary\n",
    "# print(loaded_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M_0_to_2\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix:\n",
      " [[0.86753892 0.05213542 0.06946348 0.01086218]\n",
      " [0.08065343 0.85671968 0.04982893 0.01279796]\n",
      " [0.11219846 0.17464684 0.65461785 0.05853685]\n",
      " [0.02775028 0.08494961 0.04088151 0.8464186 ]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_3\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix:\n",
      " [[0.75527498 0.04249873 0.00019955 0.20202674]\n",
      " [0.10446698 0.73349645 0.10777451 0.05426206]\n",
      " [0.009178   0.09167603 0.88105832 0.01808765]\n",
      " [0.10672648 0.08298123 0.01949473 0.79079756]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_1\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix:\n",
      " [[0.69009493 0.15321495 0.12518281 0.03150731]\n",
      " [0.03390489 0.89612499 0.03748613 0.032484  ]\n",
      " [0.13612944 0.00355591 0.86031238 0.00000227]\n",
      " [0.13829849 0.07169377 0.04358193 0.74642581]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_4\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix:\n",
      " [[0.79787321 0.09123409 0.0748348  0.0360579 ]\n",
      " [0.1525167  0.62670853 0.12700414 0.09377063]\n",
      " [0.06182208 0.09423139 0.82703856 0.01690797]\n",
      " [0.09434814 0.05487633 0.01727208 0.83350345]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_5\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix:\n",
      " [[0.77670263 0.03884329 0.11981174 0.06464234]\n",
      " [0.07306013 0.6964252  0.14381485 0.08669983]\n",
      " [0.0001472  0.02174493 0.93060026 0.04750762]\n",
      " [0.01290462 0.10677068 0.06826479 0.81205991]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23508015 0.26506847 0.26101933 0.23983305]\n"
     ]
    }
   ],
   "source": [
    "# Check the estimated parameters:\n",
    "for i in params.items():\n",
    "    print(i[0])\n",
    "    print(\"Edge:\", i[1].edge)\n",
    "    print(\"Transition matrix:\\n\", i[1].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", root_distr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix difference:\n",
      " [[0.09960865 0.07093256 0.00498621 0.03366231]\n",
      " [0.01576357 0.02703813 0.00892483 0.03387688]\n",
      " [0.00533819 0.03058066 0.01301418 0.01222828]\n",
      " [0.00504287 0.04282721 0.00457214 0.05244223]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix difference:\n",
      " [[0.00386991 0.01431933 0.0324696  0.04291902]\n",
      " [0.03240956 0.00438465 0.00578344 0.03101077]\n",
      " [0.00129948 0.00595081 0.00836875 0.00111846]\n",
      " [0.00368219 0.0201782  0.01427531 0.03077132]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix difference:\n",
      " [[0.00043271 0.04586535 0.01971483 0.02571781]\n",
      " [0.04628465 0.03270525 0.01225697 0.00132243]\n",
      " [0.00405161 0.02596747 0.04979691 0.01977783]\n",
      " [0.05635943 0.0032355  0.00691523 0.06003916]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix difference:\n",
      " [[0.02206907 0.00438435 0.02045318 0.00276846]\n",
      " [0.04114356 0.07985678 0.0152662  0.02344702]\n",
      " [0.01766089 0.02629588 0.04017144 0.00378534]\n",
      " [0.00126844 0.00706299 0.01874386 0.02707528]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix difference:\n",
      " [[0.04539034 0.01368587 0.03621623 0.00451176]\n",
      " [0.00302602 0.00613968 0.01219578 0.00908212]\n",
      " [0.00442834 0.01538392 0.00158094 0.0213932 ]\n",
      " [0.02699704 0.00888034 0.00551462 0.03036276]]\n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Compute the matrix difference between the theoretical and estimated parameters\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    print(\"Transition matrix difference:\\n\", np.abs(edges[i].transition_matrix - params['M_' + str(edges[i].edge[0].name.split(\"_\")[1]) + '_to_' + str(edges[i].edge[1].name.split(\"_\")[1])].transition_matrix))\n",
    "    print(\"---\"*30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Real pi\n",
      "[0.29373344 0.24294165 0.24128653 0.22203839]\n",
      "Real M\n",
      "M_0_to_2 [[0.7392685  0.04826224 0.07275359 0.13971567]\n",
      " [0.09209052 0.74985428 0.13775508 0.02030012]\n",
      " [0.06351693 0.19775401 0.65315908 0.08556998]\n",
      " [0.06980183 0.02372324 0.03643587 0.87003906]]\n",
      "M_0_to_3 [[0.68678405 0.1520296  0.0296847  0.13150165]\n",
      " [0.0339812  0.72700161 0.12923714 0.10978006]\n",
      " [0.10363487 0.08120426 0.71519849 0.09996238]\n",
      " [0.01039317 0.08601443 0.05435231 0.84924008]]\n",
      "M_0_to_1 [[0.71374564 0.05670601 0.05578735 0.173761  ]\n",
      " [0.10408603 0.73417494 0.13668183 0.02505721]\n",
      " [0.06384664 0.03016598 0.72032436 0.18566302]\n",
      " [0.05114861 0.10359904 0.04644648 0.79880587]]\n",
      "M_1_to_4 [[0.80204429 0.02898907 0.09814293 0.07082372]\n",
      " [0.08131525 0.75900631 0.09636563 0.06331282]\n",
      " [0.1968642  0.05870861 0.64476495 0.09966224]\n",
      " [0.08388755 0.12295254 0.01998895 0.77317096]]\n",
      "M_1_to_5 [[0.84109377 0.10952739 0.03332241 0.01605643]\n",
      " [0.20055762 0.56194241 0.12236898 0.11513099]\n",
      " [0.01217838 0.05559775 0.8788547  0.05336917]\n",
      " [0.09520559 0.05418108 0.11884368 0.73176965]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Estimated M\n",
      "M_0_to_2 [[0.76067273 0.05165304 0.06661134 0.12106289]\n",
      " [0.1068544  0.76480921 0.11737767 0.01095872]\n",
      " [0.06581898 0.23138976 0.6155466  0.08724466]\n",
      " [0.09053729 0.03391333 0.03843486 0.83711452]]\n",
      "M_0_to_3 [[0.71223705 0.14425068 0.02529606 0.11821621]\n",
      " [0.02512368 0.76594502 0.10382435 0.10510694]\n",
      " [0.10143799 0.11985421 0.68026361 0.09844419]\n",
      " [0.03584451 0.09145168 0.05242664 0.82027717]]\n",
      "M_0_to_1 [[7.42896384e-01 4.86687009e-02 3.75522113e-02 1.70882704e-01]\n",
      " [1.11060470e-01 7.79249843e-01 1.09674754e-01 1.49337917e-05]\n",
      " [7.54054063e-02 3.05811746e-02 7.04708979e-01 1.89304440e-01]\n",
      " [8.21426881e-02 8.78562240e-02 4.71418868e-02 7.82859201e-01]]\n",
      "M_1_to_4 [[0.82817041 0.02379823 0.09586435 0.052167  ]\n",
      " [0.08760286 0.78175047 0.07295016 0.05769651]\n",
      " [0.19274362 0.05718948 0.65516482 0.09490207]\n",
      " [0.08465503 0.12762195 0.02154252 0.7661805 ]]\n",
      "M_1_to_5 [[0.82783764 0.1138883  0.03931322 0.01896084]\n",
      " [0.20495855 0.57536498 0.12081408 0.09886238]\n",
      " [0.02270762 0.06139096 0.86083036 0.05507105]\n",
      " [0.11411889 0.05724385 0.11450632 0.71413094]]\n",
      "Estimated l\n",
      "M_0_to_2 1.3053349672714898\n",
      "M_0_to_3 1.2516717783483513\n",
      "M_0_to_1 1.1893810273955734\n",
      "M_1_to_4 1.1741069375209396\n",
      "M_1_to_5 1.29393766372579\n",
      "Estimated pi\n",
      "[0.26749267 0.21444507 0.27813979 0.24002248]\n",
      "Niter\n",
      "100\n",
      "TExec\n",
      "255.90280890464783\n"
     ]
    }
   ],
   "source": [
    "print(\"Real pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a125_b125_10000_random/real_root_distr.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Real M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a125_b125_10000_random/real_matrices.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"---\"*30)\n",
    "print(\"Estimated M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a125_b125_10000_random/RESULTS/repetition_99/M_estimation.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated l\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a125_b125_10000_random/RESULTS/repetition_99/estimated_branch_lengths.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a125_b125_10000_random/RESULTS/repetition_99/root_estimation.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Niter\")\n",
    "file = open(\"./OUTPUTS/EM_a125_b125_10000_random/RESULTS/repetition_99/Niter.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)\n",
    "print(\"TExec\")\n",
    "file = open(\"./OUTPUTS/EM_a125_b125_10000_random/RESULTS/repetition_99/TExec.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "debugger",
   "language": "python",
   "name": "debugger"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "03f790eb89590da6bbc9dab944ceea1669c4a69ff2bd32047ab9c0f81f755bd1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
