{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Expectation-Maximization algorithm** (hidden model)\n",
    "### Author: Martí Cortada Garcia\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import numpy as np\n",
    "from Bio import SeqIO, Phylo\n",
    "from io import StringIO\n",
    "from collections import Counter, OrderedDict\n",
    "import itertools\n",
    "import pandas as pd\n",
    "import math\n",
    "import copy\n",
    "\n",
    "# Import required functions from modules we have programmed\n",
    "from data_simulation import simulate"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Input:**\n",
    "We have a tree $T$ (where the root corresponds to the internal node codified as `Int_0`) and the alignments of the leaves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAGwCAYAAACHJU4LAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy88F64QAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAxPUlEQVR4nO3de1SVdaLG8WcLglz2RvGGIGrhNQPzMhCWXS01M23OysYxL03FdAbNS0zkqGOWoafUZMqptA44LZUhT9QcbWDMIkuzjGSG0tG8NLsUtSy5NW4E3vNHx70iUdnIZv+Q72ct1uq9P+9vEfvxfd+9t82yLEsAAAAGauXrAAAAAOdCUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMJa/rwNcjJqaGh05ckR2u102m83XcQAAQD1YlqWysjJFRkaqVavzXzNp1kXlyJEjio6O9nUMAADQAF9++aW6du163nWadVGx2+2SfjhRh8Ph4zQAAKA+SktLFR0d7X4dP59mXVTO3O5xOBwUFQAAmpn6PLbBw7QAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAD4jGVZSkpKUnh4uGw2mwoLC30dCYBhKCoA6pSfny+bzaaTJ0967Ri5ubnKzMzUxo0bVVxcrCuvvPK86y9evFg/+9nPZLfb1alTJ40bN0579+71Wj4AvkdRAeAzBw4cUJcuXTR06FBFRETI39//vOu/++67Sk5O1o4dO7R582adPn1at956qyoqKpooMYCm5tOisnXrVo0ZM0aRkZGy2Wx6/fXXfRkHgAdcLpdSUlIUFRWlkJAQJSQkKD8/3738xIkTmjBhgqKiohQcHKzY2FitX7/evXzq1KmaPn26nE6nbDabevToccFj5ubmaurUqerfv78GDBigzMxMOZ1OFRQUeOEMAZjg/P988bKKigoNGDBAv/rVr/Tzn//cl1HceYCWKCQkxONtpk2bpt27dysrK0uRkZHKycnRyJEjVVRUpF69eunUqVMaPHiwUlNT5XA4tGnTJk2aNEkxMTGKj49Xenq6YmJitGrVKu3cuVN+fn4eZygpKZEkhYeHe7wtgObBZlmW5esQkmSz2ZSTk6Nx48adcx2XyyWXy+WeLi0tVXR0tEpKSuRwOBolA9AS1fVnID8/XzfeeKO+++47tW3bttYyp9Opyy+/XE6nU5GRke75w4cPV3x8vNLS0uo8zu23366+fftq6dKlkqQVK1ZoxYoV+uKLLzzOXFNTozvuuEMnT57U+++/7/H2AHyntLRUYWFh9Xr99ukVFU8tXrxYCxcubPT9VlRUKDQ0tNH3C1yqioqKVF1drd69e9ea73K51L59e0lSdXW10tLSlJ2drcOHD6uyslIul0vBwcGNkiE5OVmffvopJQW4xDWrojJnzhzNnj3bPX3mikpjOnbsWIMugwMtSXl5ufz8/FRQUHDWLZszpf/pp59Wenq6VqxYodjYWIWEhGjmzJmqrKy86ONPmzZNGzdu1NatW9W1a9eL3h8AczWrohIYGKjAwECvHiMkJISiAlzAwIEDVV1drePHj2vYsGF1rrNt2zaNHTtW99xzj6QfbtXs27dPV1xxRYOPa1mWpk+frpycHOXn5+uyyy5r8L4ANA/NqqgAaHpFRUWy2+3uaZvNpgEDBmjixImaPHmyli1bpoEDB+rrr7/Wli1bFBcXp9GjR6tXr17asGGDtm/frnbt2mn58uU6duzYRRWV5ORkrVu3Tm+88YbsdruOHj0qSQoLC1NQUNBFnysA81BUAJzXddddV2vaz89PVVVVysjI0KJFi/Twww/r8OHD6tChg66++mrdfvvtkqR58+bp4MGDGjFihIKDg5WUlKRx48a536nTEM8//7wk6YYbbqg1PyMjQ1OnTm3wfgGYy6fv+ikvL9f+/fsl/XApefny5brxxhsVHh6ubt26XXB7T54aPp8fP0xbXl7OrR8AALyo2bzr5+OPP9aNN97onj7zoOyUKVOUmZnpo1QAAMAUPi0qN9xwQ52f3wCg5XE6ned9fmX37t31utIK4NLCMyoAjBAZGXneb0/+8QfLAWg5KCoAjODv76+ePXv6OgYAw/DtyQAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAPmNZlpKSkhQeHi6bzabCwkJfRwJgGIoKgDrl5+fLZrPp5MmTXjtGbm6uMjMztXHjRhUXF+vKK6887/rPP/+84uLi5HA45HA4lJiYqL/+9a9eywfA93xaVBYvXqyf/exnstvt6tSpk8aNG6e9e/f6MhKAJnTgwAF16dJFQ4cOVUREhPz9/c+7fteuXbVkyRIVFBTo448/1k033aSxY8fqs88+a6LEAJqaT4vKu+++q+TkZO3YsUObN2/W6dOndeutt6qiosKXsQDUg8vlUkpKiqKiohQSEqKEhATl5+e7l584cUITJkxQVFSUgoODFRsbq/Xr17uXT506VdOnT5fT6ZTNZlOPHj0ueMwxY8botttuU69evdS7d289+eSTCg0N1Y4dO7xwhgBMcP5/vnhZbm5urenMzEx16tRJBQUFuu6663ySiZKEligkJMTjbaZNm6bdu3crKytLkZGRysnJ0ciRI1VUVKRevXrp1KlTGjx4sFJTU+VwOLRp0yZNmjRJMTExio+PV3p6umJiYrRq1Srt3LlTfn5+Hh2/urpar776qioqKpSYmOhxfgDNg0+Lyk+VlJRIksLDw+tc7nK55HK53NOlpaWNnqFz586Nvk/AdJZlebS+0+lURkaGnE6nIiMjJUkpKSnKzc1VRkaG0tLSFBUVpZSUFPc206dPV15enrKzsxUfH6+wsDDZ7Xb5+fkpIiKi3scuKipSYmKiTp06pdDQUOXk5OiKK67wKD+A5sOYolJTU6OZM2fqmmuuOecDdYsXL9bChQubOBmAnyoqKlJ1dbV69+5da77L5VL79u0l/XDFIy0tTdnZ2Tp8+LAqKyvlcrkUHBx8Ucfu06ePCgsLVVJSog0bNmjKlCl69913KSvAJcqYopKcnKxPP/1U77///jnXmTNnjmbPnu2eLi0tVXR0dKPmOHbsWIMugwMtSXl5ufz8/FRQUHDWLZvQ0FBJ0tNPP6309HStWLFCsbGxCgkJ0cyZM1VZWXlRxw4ICFDPnj0lSYMHD9bOnTuVnp6uF1988aL2C8BMRhSVadOmaePGjdq6dau6du16zvUCAwMVGBjo1SwhISEUFeACBg4cqOrqah0/flzDhg2rc51t27Zp7NixuueeeyT9cNV03759jX7lo6amptYtYQCXFp8WFcuyNH36dOXk5Cg/P1+XXXaZL+MAqENRUZHsdrt72mazacCAAZo4caImT56sZcuWaeDAgfr666+1ZcsWxcXFafTo0erVq5c2bNig7du3q127dlq+fLmOHTt2UUVlzpw5GjVqlLp166aysjKtW7dO+fn5ysvLa4xTBWAgnxaV5ORkrVu3Tm+88YbsdruOHj0qSQoLC1NQUJAvowH4fz99B56fn5+qqqqUkZGhRYsW6eGHH9bhw4fVoUMHXX311br99tslSfPmzdPBgwc1YsQIBQcHKykpSePGjXM/NN8Qx48f1+TJk1VcXKywsDDFxcUpLy9Pt9xyy0WdIwBz2SxPH/dvzIPbbHXOz8jI0NSpUy+4fWlpqcLCwlRSUiKHw9HgHBUVFe776uXl5dz6AQDAizx5/fb5rR8AAIBz4bt+ABjB6XQqNDT0nD9Op9PXEQH4gBHv+gGAyMjI83578pkPlgPQslBUABjB39/f/fkoAHAGt34AAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLonIB+fn5stlsOnnyZL23mTp1qsaNG+fRcb799ltNnDhRDodDbdu21X333afy8nLPwgIAcImhqBhi4sSJ+uyzz7R582b3N0knJSX5OhYAAD5FUfFQZmam2rZtq7y8PPXr10+hoaEaOXKkiouLJUmPPfaY1qxZozfeeEM2m002m035+fnn3eeePXuUm5url156SQkJCbr22mv17LPPKisrS0eOHGmCswIAwEwUlQb4/vvvtXTpUr3yyivaunWrnE6nUlJSJEkpKSkaP368u7wUFxdr6NCh593fBx98oLZt22rIkCHuecOHD1erVq304YcfevVcAAAwGZ9M2wCnT5/WCy+8oJiYGEnStGnT9Pjjj0uSQkNDFRQUJJfLpYiIiHrt7+jRo+rUqVOtef7+/goPD9fRo0cbNzwAAM0IV1QaIDg42F1SJKlLly46fvy4DxMBAHBpoqg0QOvWrWtN22w2WZbV4P1FREScVXSqqqr07bff1vuqDAAAlyKKihcEBASourq63usnJibq5MmTKigocM97++23VVNTo4SEBG9EBACgWaCoeEGPHj30j3/8Q3v37tU333yj06dPn3f9fv36aeTIkXrggQf00Ucfadu2bZo2bZp+8Ytf8NX2AIAWjaLiBQ888ID69OmjIUOGqGPHjtq2bdsFt1m7dq369u2rm2++WbfddpuuvfZarVq1qgnSAgBgLpt1MQ9X+FhpaanCwsJUUlIih8PR4P1UVFQoNDRUklReXq6QkJDGiggAAH7Ck9dvrqgAAABjUVSaQFpamkJDQ+v8GTVqlK/jAQBgrAZ94NvHH3+s7OxsOZ1OVVZW1lr22muvNUqwS8mDDz6o8ePH17ksKCioidMAANB8eFxUsrKyNHnyZI0YMUJ/+9vfdOutt2rfvn06duyY7rzzTm9kbPbCw8MVHh7u6xgAADQ7Ht/6SUtL0zPPPKP//d//VUBAgNLT0/XPf/5T48ePV7du3byREQAAtFAeF5UDBw5o9OjRkn74YLOKigrZbDbNmjWLt9MCAIBG5XFRadeuncrKyiRJUVFR+vTTTyVJJ0+e1Pfff9+46QAAQIvm8TMq1113nTZv3qzY2FjdddddmjFjht5++21t3rxZN998szcyAgCAFsrjovLcc8/p1KlTkqS5c+eqdevW2r59u/7jP/5D8+bNa/SAAACg5fK4qPz43SutWrXSo48+2qiBAAAAzvD4GZXMzMw651dVVWnOnDkXmwcAAMDN46Ly0EMP6a677tJ3333nnrd3714lJCRo/fr1jRoOAAC0bB4XlV27dumrr75SbGysNm/erJUrV2rQoEHq27ev/v73v3sjIwAAaKE8fkYlJiZG27Zt08yZMzVy5Ej5+flpzZo1mjBhgjfyAQCAFqxBX0q4adMmZWVlKTExUW3bttXLL7+sI0eONHY2AADQwnlcVH7961/rrrvuUmpqqt577z394x//UEBAgGJjY5Wdne2NjAAAoIXy+NbPtm3b9OGHH2rAgAGSpIiICL355ptauXKlfvWrX53zW4IBAAA85XFRKSgoUGBg4Fnzk5OTNXz48EYJBQAAIDXg1k9dJeWMPn36XFQYAACAH/P4iookbdiwQdnZ2XI6naqsrKy17JNPPmmUYAAAAB5fUfnDH/6ge++9V507d9auXbsUHx+v9u3b6+DBgxo1apQ3MgIAgBbK46Lyxz/+UatWrdKzzz6rgIAAPfLII9q8ebMeeughlZSUeCMjAABooTwuKk6nU0OHDpUkBQUFqaysTJI0adIkPkIfAAA0Ko+LSkREhL799ltJUrdu3bRjxw5J0qFDh2RZVuOmAwAALZrHReWmm27SX/7yF0nSvffeq1mzZumWW27R3XffrTvvvLPRAwIAgJbL43f9zJ07V1FRUZJ++OyU9u3ba/v27brjjjs0cuTIRg8IAABaLpvl4f0aPz8/FRcXq1OnTrXmnzhxQp06dVJ1dXWjBjyf0tJShYWFqaSkRA6Ho8H7qaioUGhoqCSpvLxcISEhjRURAAD8hCev3x7f+jlXrykvL1ebNm083R0AAMA51fvWz+zZsyVJNptNv//97xUcHOxeVl1drQ8//FBXXXVVowcEAAAtV72Lyq5duyT9cEWlqKhIAQEB7mUBAQEaMGCAUlJSGj8hAABosepdVN555x1JP7zTJz09/aKeCQEAAKgPj9/1k5GR4Y0cAAAAZ/H4YVoAAICmQlEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgBAE7MsS0lJSQoPD5fNZlNhYaGvIxmLogIAwI/k5+fLZrPp5MmTXjtGbm6uMjMztXHjRhUXF+vKK6+s97ZLliyRzWbTzJkzvZbPJD4tKs8//7zi4uLkcDjkcDiUmJiov/71r76MBACA1x04cEBdunTR0KFDFRERIX9//3ptt3PnTr344ouKi4vzckJz+LSodO3aVUuWLFFBQYE+/vhj3XTTTRo7dqw+++wzX8YCAOCcXC6XUlJSFBUVpZCQECUkJCg/P9+9/MSJE5owYYKioqIUHBys2NhYrV+/3r186tSpmj59upxOp2w2m3r06FGv45aXl2vixIlavXq12rVr18hnZa76VTgvGTNmTK3pJ598Us8//7x27Nih/v37+ygV0PJUVFT4OgLgEyEhIR5vM23aNO3evVtZWVmKjIxUTk6ORo4cqaKiIvXq1UunTp3S4MGDlZqaKofDoU2bNmnSpEmKiYlRfHy80tPTFRMTo1WrVmnnzp3y8/Or13GTk5M1evRoDR8+XIsWLfI4d3Pl06LyY9XV1Xr11VdVUVGhxMTEOtdxuVxyuVzu6dLS0qaKB1zSQkNDfR0B8AnLsjxa3+l0KiMjQ06nU5GRkZKklJQU5ebmKiMjQ2lpaYqKilJKSop7m+nTpysvL0/Z2dmKj49XWFiY7Ha7/Pz8FBERUa/jZmVl6ZNPPtHOnTs9ynsp8HlRKSoqUmJiok6dOqXQ0FDl5OToiiuuqHPdxYsXa+HChU2cELh0VVRUUFIADxQVFam6ulq9e/euNd/lcql9+/aSfviHd1pamrKzs3X48GFVVlbK5XIpODi4Qcf88ssvNWPGDG3evFlt2rS56HNobnxeVPr06aPCwkKVlJRow4YNmjJlit599906y8qcOXM0e/Zs93Rpaamio6ObMi5wyTp27FiDLoMDLUl5ebn8/PxUUFBw1i2bM6X/6aefVnp6ulasWKHY2FiFhIRo5syZqqysbNAxCwoKdPz4cQ0aNMg9r7q6Wlu3btVzzz0nl8tV79tHzZHPi0pAQIB69uwpSRo8eLB27typ9PR0vfjii2etGxgYqMDAwKaOCLQIISEhFBXgAgYOHKjq6modP35cw4YNq3Odbdu2aezYsbrnnnskSTU1Ndq3b9857xZcyM0336yioqJa8+6991717dtXqampl3RJkQwoKj9VU1NT6zkUAAB8oaioSHa73T1ts9k0YMAATZw4UZMnT9ayZcs0cOBAff3119qyZYvi4uI0evRo9erVSxs2bND27dvVrl07LV++XMeOHWtwUbHb7Wd9zkpISIjat2/v0eevNFc+LSpz5szRqFGj1K1bN5WVlWndunXKz89XXl6eL2MBAKDrrruu1rSfn5+qqqqUkZGhRYsW6eGHH9bhw4fVoUMHXX311br99tslSfPmzdPBgwc1YsQIBQcHKykpSePGjVNJSYkvTqPZs1mePvLciO677z5t2bJFxcXFCgsLU1xcnFJTU3XLLbfUa/vS0lKFhYWppKREDoejwTl+/EBheXk5l7/RYvC7D8AXPHn99ukVlZdfftmXhwcAAIbju34AAPAhp9Op0NDQc/44nU5fR/Qp4x6mBQCgJYmMjDzvtyef+WC5loqiAgCAD/n7+7s/pgNn49YPAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFQJ3y8/Nls9l08uTJem8zdepUjRs3zqPjPPnkkxo6dKiCg4PVtm1bj7YFcOmjqADwqcrKSt111136z//8T19HAWAgigqAesnMzFTbtm2Vl5enfv36KTQ0VCNHjlRxcbEk6bHHHtOaNWv0xhtvyGazyWazKT8//4L7XbhwoWbNmqXY2FgvnwGA5sjf1wFMU1FR4esIQJPx9Pf9+++/19KlS/XKK6+oVatWuueee5SSkqK1a9cqJSVFe/bsUWlpqTIyMiRJ4eHh3ogNoAWhqPxE586dfR0BMNbp06f1wgsvKCYmRpI0bdo0Pf7445Kk0NBQBQUFyeVyKSIiwpcxAVxCuPUjKSQkRJZl+ToGYLzg4GB3SZGkLl266Pjx4z5MBOBSxxWVHykvL/d1BKBJVVRUeHQVsXXr1rWmbTYbJR+AV1FUfiQkJMTXEYBmLSAgQNXV1b6OAeASQlEB0Gh69OihvLw87d27V+3bt1dYWNhZV2F+yul06ttvv5XT6VR1dbUKCwslST179lRoaGgTpAZgMp5RAdBoHnjgAfXp00dDhgxRx44dtW3btgtu8/vf/14DBw7UggULVF5eroEDB2rgwIH6+OOPmyAxANPZrGZ8g7m0tFRhYWEqKSmRw+HwdRyg2amoqHBftSgvL+f2J4Am4cnrN1dUAACAsSgqALwmLS1NoaGhdf6MGjXK1/EANAM8TAvAax588EGNHz++zmVBQUFNnAZAc0RRAeA14eHhfIw+gIvCrR8AAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAJqYZVlKSkpSeHi4bDabCgsLfR3JWBQVAAB+JD8/XzabTSdPnvTaMXJzc5WZmamNGzequLhYV1555XnXf+yxx2Sz2Wr99O3b12v5TOLv6wAAALQ0Bw4cUJcuXTR06NB6b9O/f3+99dZb7ml//5bxEm7MFZUlS5bIZrNp5syZvo4CAMA5uVwupaSkKCoqSiEhIUpISFB+fr57+YkTJzRhwgRFRUUpODhYsbGxWr9+vXv51KlTNX36dDmdTtlsNvXo0aNex/X391dERIT7p0OHDo18ZmYyoo7t3LlTL774ouLi4nwdBWixKioqfB0BaHIhISEebzNt2jTt3r1bWVlZioyMVE5OjkaOHKmioiL16tVLp06d0uDBg5WamiqHw6FNmzZp0qRJiomJUXx8vNLT0xUTE6NVq1Zp586d8vPzq9dxP//8c0VGRqpNmzZKTEzU4sWL1a1bN4/zNzc+Lyrl5eWaOHGiVq9erUWLFp13XZfLJZfL5Z4uLS31djygxejcubOvIwBNzrIsj9Z3Op3KyMiQ0+lUZGSkJCklJUW5ubnKyMhQWlqaoqKilJKS4t5m+vTpysvLU3Z2tuLj4xUWFia73S4/Pz9FRETU67gJCQnKzMxUnz59VFxcrIULF2rYsGH69NNPZbfbPTqH5sbnRSU5OVmjR4/W8OHDL1hUFi9erIULFzZRMuDSFxISIsuyZLPZfB0FaBaKiopUXV2t3r1715rvcrnUvn17SVJ1dbXS0tKUnZ2tw4cPq7KyUi6XS8HBwQ0+7qhRo9z/HRcXp4SEBHXv3l3Z2dm67777Grzf5sCnRSUrK0uffPKJdu7cWa/158yZo9mzZ7unS0tLFR0d7a14QItRXl7u6whAs1BeXi4/Pz8VFBScdcsmNDRUkvT0008rPT1dK1asUGxsrEJCQjRz5kxVVlY2Wo62bduqd+/e2r9/f6Pt01Q+KypffvmlZsyYoc2bN6tNmzb12iYwMFCBgYFeTga0PA25Tw+0RAMHDlR1dbWOHz+uYcOG1bnOtm3bNHbsWN1zzz2SpJqaGu3bt09XXHFFo+UoLy/XgQMHNGnSpEbbp6l8VlQKCgp0/PhxDRo0yD2vurpaW7du1XPPPSeXy1XvB4wAAGhsRUVFtZ7/sNlsGjBggCZOnKjJkydr2bJlGjhwoL7++mtt2bJFcXFxGj16tHr16qUNGzZo+/btateunZYvX65jx45dVFFJSUnRmDFj1L17dx05ckQLFiyQn5+fJkyY0BinajSfFZWbb75ZRUVFtebde++96tu3r1JTUykpAACfuu6662pN+/n5qaqqShkZGVq0aJEefvhhHT58WB06dNDVV1+t22+/XZI0b948HTx4UCNGjFBwcLCSkpI0btw4lZSUNDjLV199pQkTJujEiRPq2LGjrr32Wu3YsUMdO3a8qHNsDmyWp488e9ENN9ygq666SitWrKjX+qWlpQoLC1NJSYkcDod3wwEAgEbhyeu3MR/4BgAA8FM+f3vyj/34k/0AAGgJnE7neZ9f2b17d4v4YLdzMaqoAADQ0kRGRp7325PPfLBcS0VRAQDAh/z9/dWzZ09fxzAWz6gAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABjL39cBLoZlWZKk0tJSHycBAAD1deZ1+8zr+Pk066JSVlYmSYqOjvZxEgAA4KmysjKFhYWddx2bVZ86Y6iamhodOXJEdrtdNpvtovZVWlqq6Ohoffnll3I4HI2UsPljXM7GmJyNMakb43I2xuRsLXFMLMtSWVmZIiMj1arV+Z9CadZXVFq1aqWuXbs26j4dDkeL+UXxBONyNsbkbIxJ3RiXszEmZ2tpY3KhKyln8DAtAAAwFkUFAAAYi6Ly/wIDA7VgwQIFBgb6OopRGJezMSZnY0zqxricjTE5G2Nyfs36YVoAAHBp44oKAAAwFkUFAAAYi6ICAACMRVEBAADGalFFZeXKlerRo4fatGmjhIQEffTRR+dd/9VXX1Xfvn3Vpk0bxcbG6s0332yipE3Lk3FZvXq1hg0bpnbt2qldu3YaPnz4BcexOfL0d+WMrKws2Ww2jRs3zrsBfcDTMTl58qSSk5PVpUsXBQYGqnfv3pfk/0OejsuKFSvUp08fBQUFKTo6WrNmzdKpU6eaKK33bd26VWPGjFFkZKRsNptef/31C26Tn5+vQYMGKTAwUD179lRmZqbXczYlT8fktdde0y233KKOHTvK4XAoMTFReXl5TRPWRFYLkZWVZQUEBFj//d//bX322WfWAw88YLVt29Y6duxYnetv27bN8vPzs5566ilr9+7d1rx586zWrVtbRUVFTZzcuzwdl1/+8pfWypUrrV27dll79uyxpk6daoWFhVlfffVVEyf3Hk/H5IxDhw5ZUVFR1rBhw6yxY8c2Tdgm4umYuFwua8iQIdZtt91mvf/++9ahQ4es/Px8q7CwsImTe5en47J27VorMDDQWrt2rXXo0CErLy/P6tKlizVr1qwmTu49b775pjV37lzrtddesyRZOTk5513/4MGDVnBwsDV79mxr9+7d1rPPPmv5+flZubm5TRO4CXg6JjNmzLD+67/+y/roo4+sffv2WXPmzLFat25tffLJJ00T2DAtpqjEx8dbycnJ7unq6morMjLSWrx4cZ3rjx8/3ho9enSteQkJCdavf/1rr+Zsap6Oy09VVVVZdrvdWrNmjbciNrmGjElVVZU1dOhQ66WXXrKmTJlyyRUVT8fk+eefty6//HKrsrKyqSL6hKfjkpycbN1000215s2ePdu65pprvJrTV+rzovzII49Y/fv3rzXv7rvvtkaMGOHFZL5TnzGpyxVXXGEtXLiw8QM1Ay3i1k9lZaUKCgo0fPhw97xWrVpp+PDh+uCDD+rc5oMPPqi1viSNGDHinOs3Rw0Zl5/6/vvvdfr0aYWHh3srZpNq6Jg8/vjj6tSpk+67776miNmkGjImf/nLX5SYmKjk5GR17txZV155pdLS0lRdXd1Usb2uIeMydOhQFRQUuG8PHTx4UG+++aZuu+22Jslsopbwt/Zi1dTUqKys7JL5O+upZv2lhPX1zTffqLq6Wp07d641v3PnzvrnP/9Z5zZHjx6tc/2jR496LWdTa8i4/FRqaqoiIyPP+kPTXDVkTN5//329/PLLKiwsbIKETa8hY3Lw4EG9/fbbmjhxot58803t379fv/nNb3T69GktWLCgKWJ7XUPG5Ze//KW++eYbXXvttbIsS1VVVXrwwQf1u9/9rikiG+lcf2tLS0v173//W0FBQT5KZo6lS5eqvLxc48eP93UUn2gRV1TgHUuWLFFWVpZycnLUpk0bX8fxibKyMk2aNEmrV69Whw4dfB3HGDU1NerUqZNWrVqlwYMH6+6779bcuXP1wgsv+DqaT+Xn5ystLU1//OMf9cknn+i1117Tpk2b9MQTT/g6Ggy1bt06LVy4UNnZ2erUqZOv4/hEi7ii0qFDB/n5+enYsWO15h87dkwRERF1bhMREeHR+s1RQ8bljKVLl2rJkiV66623FBcX582YTcrTMTlw4IC++OILjRkzxj2vpqZGkuTv76+9e/cqJibGu6G9rCG/J126dFHr1q3l5+fnntevXz8dPXpUlZWVCggI8GrmptCQcZk/f74mTZqk+++/X5IUGxuriooKJSUlae7cuWrVquX92/Fcf2sdDkeLv5qSlZWl+++/X6+++uolc9W6IVrE/xUBAQEaPHiwtmzZ4p5XU1OjLVu2KDExsc5tEhMTa60vSZs3bz7n+s1RQ8ZFkp566ik98cQTys3N1ZAhQ5oiapPxdEz69u2roqIiFRYWun/uuOMO3XjjjSosLFR0dHRTxveKhvyeXHPNNdq/f7+7tEnSvn371KVLl0uipEgNG5fvv//+rDJypsxZLfRr11rC39qGWL9+ve69916tX79eo0eP9nUc3/L107xNJSsrywoMDLQyMzOt3bt3W0lJSVbbtm2to0ePWpZlWZMmTbIeffRR9/rbtm2z/P39raVLl1p79uyxFixYcMm+PdmTcVmyZIkVEBBgbdiwwSouLnb/lJWV+eoUGp2nY/JTl+K7fjwdE6fTadntdmvatGnW3r17rY0bN1qdOnWyFi1a5KtT8ApPx2XBggWW3W631q9fbx08eND629/+ZsXExFjjx4/31Sk0urKyMmvXrl3Wrl27LEnW8uXLrV27dln/+te/LMuyrEcffdSaNGmSe/0zb0/+7W9/a+3Zs8dauXLlJff2ZE/HZO3atZa/v7+1cuXKWn9nT5486atT8KkWU1Qsy7KeffZZq1u3blZAQIAVHx9v7dixw73s+uuvt6ZMmVJr/ezsbKt3795WQECA1b9/f2vTpk1NnLhpeDIu3bt3tySd9bNgwYKmD+5Fnv6u/NilWFQsy/Mx2b59u5WQkGAFBgZal19+ufXkk09aVVVVTZza+zwZl9OnT1uPPfaYFRMTY7Vp08aKjo62fvOb31jfffdd0wf3knfeeafOvxFnxmHKlCnW9ddff9Y2V111lRUQEGBdfvnlVkZGRpPn9iZPx+T6668/7/otjc2yWuj1RgAAYLwW8YwKAABonigqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVSAS9wNN9ygmTNn+jpGnb744gvZbDYVFhZ6tJ3NZtPrr7/ulUyeeuyxx3TVVVf5OgZwyaKoAEA9mVSQgJaCogLgLJWVlb6OAACSKCpAi1BVVaVp06YpLCxMHTp00Pz58/Xjr/nq0aOHnnjiCU2ePFkOh0NJSUmSpNTUVPXu3VvBwcG6/PLLNX/+fJ0+fdq93ZnbHq+88op69OihsLAw/eIXv1BZWZl7nZqaGj311FPq2bOnAgMD1a1bNz355JO18h08eFA33nijgoODNWDAAH3wwQcend+XX36p8ePHq23btgoPD9fYsWP1xRdfuJdPnTpV48aN09KlS9WlSxe1b99eycnJtc6luLhYo0ePVlBQkC677DKtW7dOPXr00IoVK9xjJEl33nmnbDabe/qM840BgIajqAAtwJo1a+Tv76+PPvpI6enpWr58uV566aVa6yxdulQDBgzQrl27NH/+fEmS3W5XZmamdu/erfT0dK1evVrPPPNMre0OHDig119/XRs3btTGjRv17rvvasmSJe7lc+bM0ZIlSzR//nzt3r1b69atU+fOnWvtY+7cuUpJSVFhYaF69+6tCRMmqKqqql7ndvr0aY0YMUJ2u13vvfeetm3bptDQUI0cObLWlaF33nlHBw4c0DvvvKM1a9YoMzNTmZmZ7uWTJ0/WkSNHlJ+fr//5n//RqlWrdPz4cffynTt3SpIyMjJUXFzsnq7PGAC4CD7+9mYAXnb99ddb/fr1s2pqatzzUlNTrX79+rmnu3fvbo0bN+6C+3r66aetwYMHu6cXLFhgBQcHW6Wlpe55v/3tb62EhATLsiyrtLTUCgwMtFavXl3n/g4dOmRJsl566SX3vM8++8ySZO3Zs+ecOSRZOTk5lmVZ1iuvvGL16dOn1vm5XC4rKCjIysvLsyzLsqZMmWJ1797dqqqqcq9z1113WXfffbdlWZa1Z88eS5K1c+dO9/LPP//ckmQ988wzdR63vmMA4OJwRQVoAa6++mrZbDb3dGJioj7//HNVV1e75w0ZMuSs7f785z/rmmuuUUREhEJDQzVv3jw5nc5a6/To0UN2u9093aVLF/eViD179sjlcunmm28+b764uLha20uqdTXjfP7+979r//79stvtCg0NVWhoqMLDw3Xq1CkdOHDAvV7//v3l5+dXZ869e/fK399fgwYNci/v2bOn2rVrV68M5xsDABfH39cBAJghJCSk1vQHH3ygiRMnauHChRoxYoTCwsKUlZWlZcuW1VqvdevWtaZtNptqamokSUFBQfU69o/3caZQndnHhZSXl2vw4MFau3btWcs6duxYr5wXy5v7Blo6igrQAnz44Ye1pnfs2KFevXrVusLwU9u3b1f37t01d+5c97x//etfHh23V69eCgoK0pYtW3T//fd7FrqeBg0apD//+c/q1KmTHA5Hg/bRp08fVVVVadeuXRo8eLAkaf/+/fruu+9qrde6detaV6EAeB+3foAWwOl0avbs2dq7d6/Wr1+vZ599VjNmzDjvNr169ZLT6VRWVpYOHDigP/zhD8rJyfHouG3atFFqaqoeeeQR/elPf9KBAwe0Y8cOvfzyyxdzOrVMnDhRHTp00NixY/Xee+/p0KFDys/P10MPPaSvvvqqXvvo27evhg8frqSkJH300UfatWuXkpKSFBQUVOuWWY8ePbRlyxYdPXr0rBIDwDsoKkALMHnyZP373/9WfHy8kpOTNWPGDPdbkM/ljjvu0KxZszRt2jRdddVV2r59u/vdQJ6YP3++Hn74Yf3+979Xv379dPfddzfq8xvBwcHaunWrunXrpp///Ofq16+f7rvvPp06dcqjKyx/+tOf1LlzZ1133XW688479cADD8hut6tNmzbudZYtW6bNmzcrOjpaAwcObLRzAHBuNsv60YcpAAAkSV999ZWio6P11ltvXfBhYADeQ1EBAElvv/22ysvLFRsbq+LiYj3yyCM6fPiw9u3bd9bDsgCaDg/TAoB++OC43/3udzp48KDsdruGDh2qtWvXUlIAH+OKCgAAMBYP0wIAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxvo/0ZCI4QyD6tMAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "newick_tree = open(\"./newick.txt\", \"r\")\n",
    "newick = newick_tree.read()\n",
    "newick_tree.close()\n",
    "tree = Phylo.read(StringIO(newick), \"newick\")\n",
    "\n",
    "# Change nodes names\n",
    "for idx, clade in enumerate(tree.get_nonterminals()):\n",
    "    clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "\n",
    "# Change leaves names\n",
    "for idx, clade in enumerate(tree.get_terminals()):\n",
    "    clade.name = \"Leaf_\" + clade.name \n",
    "\n",
    "Phylo.draw(tree)\n",
    "\n",
    "net = Phylo.to_networkx(tree) # to graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'directory' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[36], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m edges, node_distr, filename \u001b[39m=\u001b[39m simulate(net, \u001b[39m500\u001b[39m, directory) \u001b[39m# simulate data calling the module data_simulation.py \u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[39m# Preprocess leave alignments\u001b[39;00m\n\u001b[1;32m      4\u001b[0m leave_alignments \u001b[39m=\u001b[39m [i \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m SeqIO\u001b[39m.\u001b[39mparse(filename, \u001b[39m\"\u001b[39m\u001b[39mfasta\u001b[39m\u001b[39m\"\u001b[39m)]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'directory' is not defined"
     ]
    }
   ],
   "source": [
    "edges, node_distr, filename = simulate(net, 500) # simulate data calling the module data_simulation.py \n",
    "\n",
    "# Preprocess leave alignments\n",
    "leave_alignments = [i for i in SeqIO.parse(filename, \"fasta\")]\n",
    "for i in leave_alignments:\n",
    "    print(\"Seq>\", i.seq)\n",
    "# Codify the alphabet {A,G,C,T} -> {0,1,2,3}\n",
    "leave_sequences = []\n",
    "for i in leave_alignments:\n",
    "    seq = \"\"\n",
    "    for j in i.seq:\n",
    "        if j == \"A\":\n",
    "            seq += \"0\"\n",
    "        elif j == \"G\":\n",
    "            seq += \"1\"\n",
    "        elif j == \"C\":\n",
    "            seq += \"2\"\n",
    "        elif j == \"T\":\n",
    "            seq += \"3\"\n",
    "        else:\n",
    "            raise ValueError(\"Invalid nucleotide\")\n",
    "    leave_sequences.append(seq)\n",
    "#leave_sequences\n",
    "edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## **EM algorithm**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from evaluate import init_EM\n",
    "\n",
    "Results = init_EM(newick, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Source: Int_0\n",
      "Target: Leaf_2\n",
      "Matrix:\n",
      " [[0.80529814 0.06095358 0.09251042 0.04123786]\n",
      " [0.05611119 0.71483627 0.11102639 0.11802615]\n",
      " [0.0952762  0.07415636 0.731184   0.09938344]\n",
      " [0.02196288 0.0820518  0.08234946 0.81363586]]\n",
      "Source: Int_0\n",
      "Target: Leaf_3\n",
      "Matrix:\n",
      " [[0.63632807 0.06782976 0.1446222  0.15121997]\n",
      " [0.12810897 0.67208297 0.04706075 0.15274731]\n",
      " [0.10204507 0.09021686 0.74449869 0.06323937]\n",
      " [0.06570111 0.08178298 0.05155947 0.80095644]]\n",
      "Source: Int_0\n",
      "Target: Int_1\n",
      "Matrix:\n",
      " [[0.72577892 0.07421689 0.0831795  0.11682468]\n",
      " [0.08700214 0.7261134  0.06686419 0.12002027]\n",
      " [0.09519544 0.182137   0.68012648 0.04254107]\n",
      " [0.06761601 0.06509887 0.08961157 0.77767356]]\n",
      "Source: Int_1\n",
      "Target: Leaf_4\n",
      "Matrix:\n",
      " [[0.85639817 0.07234946 0.04036045 0.03089192]\n",
      " [0.0298384  0.73128255 0.0944944  0.14438465]\n",
      " [0.03533407 0.10420379 0.77836777 0.08209437]\n",
      " [0.06403097 0.03195277 0.06966545 0.8343508 ]]\n",
      "Source: Int_1\n",
      "Target: Leaf_5\n",
      "Matrix:\n",
      " [[0.80075352 0.06438893 0.11198018 0.02287738]\n",
      " [0.07400816 0.72459648 0.10712499 0.09427037]\n",
      " [0.14358855 0.08350079 0.74500083 0.02790983]\n",
      " [0.01210798 0.14166184 0.06442053 0.78180965]]\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(Results)):\n",
    "    print(\"Source:\", Results[i].source)\n",
    "    print(\"Target:\", Results[i].target)\n",
    "    print(\"Matrix:\\n\", Results[i].matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "l -73482.53888183684\n",
      "73482.53888183684\n",
      "--------\n",
      "2803.2235244028125\n",
      "--------\n",
      "403.42311980716477\n",
      "--------\n",
      "147.34679458629398\n",
      "--------\n",
      "65.81431201811938\n",
      "--------\n",
      "32.62458047835389\n",
      "--------\n",
      "17.4679848113592\n",
      "--------\n",
      "9.96598948755127\n",
      "--------\n",
      "5.979527266841615\n",
      "--------\n",
      "3.724567122422741\n",
      "--------\n",
      "2.382098244605004\n",
      "--------\n",
      "1.5508962869207608\n",
      "--------\n",
      "1.021128828782821\n",
      "--------\n",
      "0.6763189632765716\n",
      "--------\n",
      "0.44849936086393427\n",
      "--------\n",
      "0.29640033769828733\n",
      "--------\n",
      "0.19417478849936742\n",
      "--------\n",
      "0.12524008554464672\n",
      "--------\n",
      "0.07875408534891903\n",
      "--------\n",
      "0.047520216307020746\n",
      "--------\n",
      "0.02670328039675951\n",
      "--------\n",
      "0.01302253625181038\n",
      "--------\n",
      "0.004234299674862996\n",
      "--------\n",
      "0.0012054162216372788\n",
      "--------\n",
      "0.004364024440292269\n",
      "--------\n",
      "0.005981738388072699\n",
      "--------\n",
      "0.006572629208676517\n",
      "--------\n",
      "0.006493631357443519\n",
      "--------\n",
      "0.005992020829580724\n",
      "--------\n",
      "0.005238327459665015\n",
      "--------\n",
      "0.00434925795707386\n",
      "--------\n",
      "0.003403757931664586\n",
      "--------\n",
      "0.002454297398799099\n",
      "--------\n",
      "0.0015348267188528553\n",
      "--------\n",
      "Done!\n",
      "Number of iterations:  34\n"
     ]
    }
   ],
   "source": [
    "# Auxiliary functions\n",
    "\n",
    "def init_root_distribution():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the root distribution (all probabilities between 0.2 and 0.3)\n",
    "    \"\"\"\n",
    "    estimated_root_distribution = []\n",
    "    while True:\n",
    "        estimated_root_distribution = np.random.dirichlet([1,1,1,1])\n",
    "        i = 0\n",
    "        while i<4:\n",
    "            if estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3:\n",
    "                i += 1\n",
    "            else:\n",
    "                break\n",
    "        if i == 4:\n",
    "            assert(np.sum(estimated_root_distribution) < 1.0000001 and np.sum(estimated_root_distribution) > 0.9999999)\n",
    "            assert(estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3 for i in range(4))\n",
    "            break\n",
    "    return estimated_root_distribution\n",
    "\n",
    "def init_transition_matrix():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the transition matrices (which we estimate as LARGE-DIAGONAL-VALUE matrices)\n",
    "    \"\"\"\n",
    "    M = np.zeros((4,4))\n",
    "    i=0\n",
    "    while i<4:\n",
    "        R = np.random.dirichlet([1,1,1,1])\n",
    "        if R[i] > 0.6 and R[i] < 0.8:\n",
    "            M[i,:] = R \n",
    "            i = i + 1\n",
    "    assert (M[i,i] > 0.6 and M[i,i] < 0.8 for i in range(4))\n",
    "    assert (np.sum(M[i,:]) < 1.0000001 and np.sum(M[i,:]) > 0.9999999 for i in range(4))\n",
    "    return M\n",
    "\n",
    "def log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "    \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "    logL = 0\n",
    "    for obs in states:\n",
    "        p_i = 0 \n",
    "        observed_data = obs[n_int:]\n",
    "        if observed_data in u_i:\n",
    "            for state in states:\n",
    "                if state[n_int:] == observed_data:\n",
    "                    pi = root_distribution[int(state[0])]\n",
    "                    for p in params.items():\n",
    "                        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                        pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                    p_i += pi\n",
    "        logL += (u_i[observed_data] * math.log(p_i))\n",
    "    return logL\n",
    "\n",
    "############################################################################################################\n",
    "########################################## FELSENSTEIN ALGORITHM ###########################################\n",
    "############################################################################################################\n",
    "# # Felsentstein algorithm\n",
    "# # \"((4,5),(2,3))\"\n",
    "# newick_tree = \"(((4:1,5:1):1,(6:1,7:1):1):1,(8:1,9:1):1);\"\n",
    "# L_tree = Phylo.read(StringIO(newick_tree), \"newick\")\n",
    "# # Change nodes names\n",
    "# for idx, clade in enumerate(L_tree.get_nonterminals()):\n",
    "#     id = 0\n",
    "#     if idx == 2:\n",
    "#         clade.name = \"Int_Id_\" + str(id)\n",
    "#         id += 1\n",
    "#     else:\n",
    "#         clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "# # Change leaves names\n",
    "# for idx, clade in enumerate(L_tree.get_terminals()):\n",
    "#     clade.name = \"Leaf_\" + clade.name \n",
    "# L_net = Phylo.to_networkx(L_tree) # to graph\n",
    "# L_edges = []\n",
    "# for edge in L_net.edges():\n",
    "#     if edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     elif edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] != \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     elif edge[0].name.split(\"_\")[1] != \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     else:\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     L_edges.append((u,v))\n",
    "# # Create a dictionary to store the father and son information\n",
    "# family_tree = {}\n",
    "# # Loop through the edges and update the dictionary\n",
    "# for father, son in L_edges:\n",
    "#     if father not in family_tree:\n",
    "#         family_tree[father] = []\n",
    "#     family_tree[father].append(son)\n",
    "# # Create a DataFrame from the dictionary\n",
    "# tree_df = pd.DataFrame.from_dict(family_tree, orient='index')\n",
    "# # Fill NaN values with empty strings\n",
    "# tree_df = tree_df.fillna('')\n",
    "# # Change column names\n",
    "# tree_df.columns = ['child_1', 'child_2']\n",
    "\n",
    "# class TreeInference():\n",
    "#     def __init__(self, tree_data, leaf_values, params, root_distr):\n",
    "#         # Set parameters\n",
    "#         self.tree_data = tree_data.copy()\n",
    "#         self.leaf_values = leaf_values\n",
    "#         self.params = params\n",
    "#         self.root_distr = root_distr\n",
    "\n",
    "#         # Find IDs of internal nodes, all nodes and leaf nodes\n",
    "#         self.internal_node_ids = np.array(tree_data.index)\n",
    "#         self.all_node_ids = np.unique(np.concatenate((np.array(self.tree_data.index),\n",
    "#                                                       np.array(self.tree_data[\"child_1\"]),\n",
    "#                                                       np.array(self.tree_data[\"child_2\"]))))\n",
    "#         self.leaf_node_ids = list(set(self.all_node_ids) - set(self.internal_node_ids))\n",
    "        \n",
    "#         # Count internal nodes and leaves for each row in tree_data and set root node\n",
    "#         self.treeNodeCounts()\n",
    "#         self.root_node = 0\n",
    "        \n",
    "#         # Get indexes for matricies for states, e.g. A is row 0, G is row 2...\n",
    "#         states = [\"A\", \"G\", \"C\", \"T\"]\n",
    "#         self.matrix_idx = dict(zip(states, range(len(states))))\n",
    "        \n",
    "#         # Empty matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "#         # Log likelihood of whole tree\n",
    "#         self.L_root = None\n",
    "        \n",
    "\n",
    "#     \"\"\"\n",
    "#     Calculates the number of internal nodes and leaves below each internal node.\n",
    "#     This works by recursively iterating down the child nodes from the original internal node until a\n",
    "#     leaf node is found, and then returning upwards counting the leaves and nodes.\n",
    "#     The new derived information is then added as two new columns at the end of the data.\n",
    "#     \"\"\"\n",
    "#     def treeNodeCounts(self):\n",
    "#         n_nodes = np.array([])\n",
    "#         n_leaves = np.array([])\n",
    "\n",
    "#         # Set the number of internal nodes and leaves for each internal node row\n",
    "#         for n in self.internal_node_ids:\n",
    "#             num_nodes, num_leaves = self.checkNode(self.tree_data, n, self.internal_node_ids)\n",
    "#             n_nodes = np.append(n_nodes, num_nodes)\n",
    "#             n_leaves = np.append(n_leaves, num_leaves)\n",
    "\n",
    "#         # Add new information to dataframe\n",
    "#         self.tree_data[\"N_Internal\"] = n_nodes\n",
    "#         self.tree_data[\"N_Leaves\"] = n_leaves\n",
    "                \n",
    "#     # Recursively find the number of internal nodes and leaves given a node ID\n",
    "#     def checkNode(self, tree_data, node_id, node_array):\n",
    "#         # Record the number of internal nodes and leaves below\n",
    "#         num_nodes = 0\n",
    "#         num_leaves = 0\n",
    "\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[node_id][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[node_id][\"child_2\"]\n",
    "\n",
    "#         # Check if found an internal node to the left\n",
    "#         if child_1 in node_array:\n",
    "#             # Recursively check nodes below\n",
    "#             d1_nodes, d1_leaves = self.checkNode(self.tree_data, child_1, node_array)\n",
    "#             num_nodes += d1_nodes + 1\n",
    "#             num_leaves += d1_leaves\n",
    "\n",
    "#         # Check if found an internal node to the right\n",
    "#         if child_2 in node_array:\n",
    "#             d2_nodes, d2_leaves = self.checkNode(self.tree_data, child_2, node_array)\n",
    "#             num_nodes += d2_nodes + 1\n",
    "#             num_leaves += d2_leaves\n",
    "\n",
    "#         return num_nodes, num_leaves\n",
    "    \n",
    "    \n",
    "#     # Calculate likelihood for an internal node\n",
    "#     def nodeLikelihood(self, node_n, params):\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[str(node_n)][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[str(node_n)][\"child_2\"]\n",
    "        \n",
    "#         # Check if likelihood of left child node set\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(child_1, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_1), 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(int(child_1), params)\n",
    "#         # Check if likelihood of right child node set\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 self.nodeLikelihood(child_2, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_2), 0] is None):\n",
    "#                 self.nodeLikelihood(int(child_2), params)\n",
    "        \n",
    "#         M1, M2 = np.full((4,4),None), np.full((4,4),None)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             # Create transition matrix for left and right child nodes\n",
    "#             matrix_name_child1 = \"M_0_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_0_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_1 == \"Id_0\":\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = np.identity(4)\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_2 == \"Id_0\":\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = np.identity(4)\n",
    "#         else:\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "\n",
    "#         # Get conditional likelihood of child nodes\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             d1_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d1_likelihood = self.L[int(child_1)]\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             d2_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d2_likelihood = self.L[int(child_2)]\n",
    "        \n",
    "#         # Calculate the likelihood of each state for the node\n",
    "#         left_prob = np.sum(M1 * d1_likelihood, axis = 1)\n",
    "#         right_prob = np.sum(M2 * d2_likelihood, axis = 1)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             self.L[-1,:] = np.multiply(left_prob, right_prob)\n",
    "#         else:\n",
    "#             self.L[int(node_n),:] = np.multiply(left_prob, right_prob)\n",
    "    \n",
    "#     # Run Felsenstein’s algorithm to calculate the log likelihood of the tree\n",
    "#     def calculateLikelihood(self, params, Estep):\n",
    "#         # Clear matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "        \n",
    "#         # Set likelihood of leaf states\n",
    "#         for leaf_n in self.leaf_node_ids:\n",
    "#             if leaf_n == \"Id_0\":\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[-1, :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[-1, self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "#             else:\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[int(leaf_n), :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[int(leaf_n), self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "        \n",
    "#         for node_n in self.internal_node_ids:\n",
    "#             # Check likelihood has not been set\n",
    "#             if node_n == \"Id_0\":\n",
    "#                 if self.L[-1, 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "#             else:\n",
    "#                 if self.L[int(node_n), 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "        \n",
    "#         if Estep == True:\n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = np.sum(np.multiply(self.root_distr, self.L[self.root_node]))\n",
    "#         else: \n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = math.log(np.sum(np.multiply(self.root_distr, self.L[self.root_node])))\n",
    "        \n",
    "#         return self.L_root\n",
    "\n",
    "# def Felsenstein_log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "#     \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "#     logL = 0\n",
    "#     for obs in states:\n",
    "#         observed_data = obs[n_int:]\n",
    "#         tree_leaves = dict()\n",
    "#         for i in range(len(observed_data)):\n",
    "#             if observed_data[i] == str(0):\n",
    "#                 tree_leaves[n_int+i] = \"A\"\n",
    "#             elif observed_data[i] == str(1):\n",
    "#                 tree_leaves[n_int+i] = \"G\"\n",
    "#             elif observed_data[i] == str(2):\n",
    "#                 tree_leaves[n_int+i] = \"C\"\n",
    "#             elif observed_data[i] == str(3):\n",
    "#                 tree_leaves[n_int+i] = \"T\"\n",
    "#         tree_inference = TreeInference(tree_df, tree_leaves, params, root_distribution)\n",
    "#         tree_inference.calculateLikelihood(params, False)\n",
    "#         logL += (u_i[observed_data] * tree_inference.L_root)\n",
    "#         del tree_inference\n",
    "#     return logL\n",
    "\n",
    "\n",
    "############################################################################################################\n",
    "\n",
    "class Param:\n",
    "    \"\"\"\n",
    "    Class to store the parameters of the tree\n",
    "    \"\"\"\n",
    "    def __init__(self, edge, transition_matrix, alignment=None):\n",
    "        self.edge = edge\n",
    "        self.transition_matrix = transition_matrix\n",
    "        self.alignment = alignment # it will only be placed in the leaves\n",
    "\n",
    "# Obtain number of leaves and internal nodes\n",
    "n_leaves, n_int = 0, 0 \n",
    "for node in net.nodes():\n",
    "    if node.name.startswith(\"L\"):\n",
    "        n_leaves += 1\n",
    "    else:\n",
    "        n_int += 1\n",
    "\n",
    "# Count all the leaves ocurrences \n",
    "sequence_length = len(leave_sequences[0])\n",
    "number_of_sequences = len(leave_sequences)\n",
    "occurrences = []\n",
    "for i in range(sequence_length):\n",
    "    _ = ''\n",
    "    for j in range(number_of_sequences):\n",
    "        _ += leave_sequences[j][i]\n",
    "    occurrences.append(_)\n",
    "c = Counter(occurrences)\n",
    "u_i = OrderedDict(sorted(c.items()))\n",
    "\n",
    "# Define all possible (hidden) states of the internal nodes \n",
    "hidden_combinations = itertools.product([0,1,2,3], repeat=n_int)\n",
    "sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "for i in range(len(sorted_hidden_combinations)):\n",
    "    sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    " \n",
    "# Define all possible fully observed states (i.e. combining u_i + internal nodes)\n",
    "# '000000' would mean to have an 'A' in all nodes: Int_0, Int_1, Leaf_2, Leaf_3, Leaf_4, Leaf_5 \n",
    "states = list(itertools.product(list(sorted_hidden_combinations), list(u_i.keys())))\n",
    "states = [i[0]+i[1] for i in states]\n",
    "\n",
    "############################################################################################################\n",
    "###################################### STEP 0: Initialise parameters #######################################\n",
    "############################################################################################################\n",
    "estimated_root_distribution = init_root_distribution() # First estimation for the root distribution\n",
    "\n",
    "# PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "# for edge in net.edges():\n",
    "#     # If we are in a leaf, we need to specify the alignment\n",
    "#     iter_leaves = 0\n",
    "#     if edge[1].name.startswith(\"L\"):\n",
    "#         new_edge = Param(edge, init_transition_matrix(), leave_alignments[iter_leaves].seq)\n",
    "#         iter_leaves += 1\n",
    "#     # Otherwise, are in internal nodes\n",
    "#     else:\n",
    "#         new_edge = Param(edge, init_transition_matrix())\n",
    "#     u = str(edge[0].name.split(\"_\")[1])\n",
    "#     v = str(edge[1].name.split(\"_\")[1])\n",
    "#     name = \"M_\" + u + \"_to_\" + v\n",
    "#     PARAMS[name] = new_edge\n",
    "\n",
    "# ADRIA CODE\n",
    "i=0\n",
    "PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "for edge in net.edges():\n",
    "    # If we are in a leaf, we need to specify the alignment\n",
    "    iter_leaves = 0\n",
    "    if edge[1].name.startswith(\"L\"):\n",
    "        new_edge = Param(edge, Results[i].matrix, leave_alignments[iter_leaves].seq)\n",
    "        iter_leaves += 1\n",
    "    # Otherwise, are in internal nodes\n",
    "    else:\n",
    "        new_edge = Param(edge, Results[i].matrix)\n",
    "    u = str(edge[0].name.split(\"_\")[1])\n",
    "    v = str(edge[1].name.split(\"_\")[1])\n",
    "    name = \"M_\" + u + \"_to_\" + v\n",
    "    PARAMS[name] = new_edge\n",
    "    i += 1\n",
    "\n",
    "eps = 10**(-3)                  # epsilon for convergence\n",
    "iter = 0                        # iteration counter \n",
    "params = copy.copy(PARAMS)      # copy of the parameters (will be updated at each iteration)\n",
    "# logF = Felsenstein_log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "logL = log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "# print(logF)\n",
    "print(\"l\", logL)\n",
    "logL_ = 0\n",
    "\n",
    "while np.abs(logL_ - logL) > eps:\n",
    "    print(np.abs(logL_ - logL))\n",
    "    # print(np.abs(logF_ - logF))\n",
    "    print(\"--------\")\n",
    "    if iter > 0:\n",
    "        logL = logL_\n",
    "        # logF = logF_\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## E-step ##################################################\n",
    "    ############################################################################################################\n",
    "    # Define the expected hidden data matrix\n",
    "    U = dict()\n",
    "    root_distr = estimated_root_distribution ########\n",
    "\n",
    "    for obs in states:\n",
    "        U[obs] = u_i[obs[n_int:]] # initialise by computing u_i of the expected hidden data matrix formula\n",
    "        f_ij = estimated_root_distribution[int(obs[0])] \n",
    "        for p in params.items():\n",
    "            u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "            f_ij *= p[1].transition_matrix[int(obs[u]), int(obs[v])]\n",
    "        # Now f_ij := numerator of the expected hidden data matrix formula \n",
    "\n",
    "        # We need to marginalise to obtain f_i := denominator of the expected hidden data matrix formula\n",
    "        f_i = 0 # we initialise the partial data model\n",
    "        observed_data = obs[n_int:]\n",
    "        for state in states:\n",
    "            if state[n_int:] == observed_data:\n",
    "                pi = estimated_root_distribution[int(state[0])]\n",
    "                for p in params.items():\n",
    "                    u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                    pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                f_i += pi\n",
    "\n",
    "        # observed_data = obs[n_int:]\n",
    "        # tree_leaves = dict()\n",
    "        # for i in range(len(observed_data)):\n",
    "        #     if observed_data[i] == str(0):\n",
    "        #         tree_leaves[n_int+i] = \"A\"\n",
    "        #     elif observed_data[i] == str(1):\n",
    "        #         tree_leaves[n_int+i] = \"G\"\n",
    "        #     elif observed_data[i] == str(2):\n",
    "        #         tree_leaves[n_int+i] = \"C\"\n",
    "        #     elif observed_data[i] == str(3):\n",
    "        #         tree_leaves[n_int+i] = \"T\"\n",
    "        # tree_inference = TreeInference(tree_df, tree_leaves, params, root_distr)\n",
    "        # tree_inference.calculateLikelihood(params, True)\n",
    "\n",
    "        U[obs] *= f_ij/f_i\n",
    "        # U[obs] *= f_ij/tree_inference.L_root\n",
    "        # del tree_inference\n",
    "        # del tree_leaves\n",
    "\n",
    "\n",
    "    assert(len(U) <= 4**(n_int+n_leaves))\n",
    "    sum = 0\n",
    "    sum2 = 0\n",
    "    for i in U.items():\n",
    "        sum += i[1]\n",
    "    for i in u_i.items():\n",
    "        sum2 += i[1]\n",
    "    assert(sum < sequence_length+0.001 and sum > sequence_length-0.001)\n",
    "    assert(sum2 < sequence_length+0.001 and sum2 > sequence_length-0.001)\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## M-step ##################################################\n",
    "    ############################################################################################################\n",
    "    estimated_parameters = dict()\n",
    "    root_distr = np.zeros(4)\n",
    "    iter_root = 0 # counter for the root distribution, we only want to count once\n",
    "\n",
    "    for p in params.items():\n",
    "        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "        name = \"M_\" + str(u) + \"_to_\" + str(v)\n",
    "        estimated_parameters[name] = np.zeros((4,4))\n",
    "        for i in range(4):\n",
    "            for j in range(4):\n",
    "                ui = 0\n",
    "                for obs in U.items():\n",
    "                    # For the root distribution, we only need to consider the first internal node\n",
    "                    if iter_root == 0 and obs[0][0] == str(i):\n",
    "                        root_distr[i] += obs[1]\n",
    "                    # For the transition matrices:\n",
    "                    if obs[0][u] == str(i) and obs[0][v] == str(j):\n",
    "                        estimated_parameters[name][i][j] += obs[1]\n",
    "                    if obs[0][u] == str(i):\n",
    "                        ui += obs[1]\n",
    "                # Update root distribution\n",
    "                if iter_root == 0:\n",
    "                    root_distr[i] /= sequence_length\n",
    "                iter_root = 1\n",
    "                estimated_parameters[name][i][j] /= ui\n",
    "            iter_root = 0\n",
    "            \n",
    "        # Update parameters\n",
    "        params[name].transition_matrix = estimated_parameters[name]\n",
    "\n",
    "    ##########################################################################################################\n",
    "    ########################### Compute log-likelihood of the estimated parameters ###########################\n",
    "    ##########################################################################################################\n",
    "    # logF_ = Felsenstein_log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "    logL_ = log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "\n",
    "    iter += 1\n",
    "\n",
    "print(\"Done!\")\n",
    "print(\"Number of iterations: \", iter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################### DEBUGGING ############################\n",
    "\n",
    "# # All possible (theoretical) occurrences of nucleotides in the leaves\n",
    "# alphabet = [0, 1, 2, 3]\n",
    "# combinations = itertools.product(alphabet, repeat=n_leaves)\n",
    "# sorted_combinations = sorted(combinations)\n",
    "# for i in range(len(sorted_combinations)):\n",
    "#     sorted_combinations[i] = ''.join([str(s) for s in sorted_combinations[i]])\n",
    "\n",
    "# # All possible (hidden) states of the internal nodes (i.e. indexes of columns of U)\n",
    "# hidden_combinations = itertools.product(alphabet, repeat=n_int)\n",
    "# sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "# for i in range(len(sorted_hidden_combinations)):\n",
    "#     sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    "\n",
    "# # To check if an element is within a dictionary or not \n",
    "# if '0101' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')\n",
    "# if '0000' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix: [[0.64642183 0.02803076 0.27483013 0.05071728]\n",
      " [0.03848122 0.7757021  0.10189982 0.08391686]\n",
      " [0.03184807 0.07934462 0.82109495 0.06771236]\n",
      " [0.04702478 0.01879095 0.04076404 0.89342023]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix: [[0.55850753 0.13612552 0.2499751  0.05539185]\n",
      " [0.01791204 0.82605787 0.05907209 0.096958  ]\n",
      " [0.04534628 0.008074   0.85149918 0.09508054]\n",
      " [0.00646156 0.06355415 0.02348564 0.90649866]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix: [[0.75152183 0.09827146 0.07355421 0.0766525 ]\n",
      " [0.04888819 0.86735324 0.02989229 0.05386628]\n",
      " [0.05446854 0.02450495 0.75018255 0.17084396]\n",
      " [0.12039713 0.03979299 0.05715314 0.78265675]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix: [[0.82835309 0.03976936 0.05146899 0.08040856]\n",
      " [0.02327998 0.80873654 0.06043788 0.10754561]\n",
      " [0.11544022 0.13895831 0.71027209 0.03532938]\n",
      " [0.09988558 0.08065349 0.01737091 0.80209001]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix: [[0.78296278 0.06942125 0.08171685 0.06589912]\n",
      " [0.02166725 0.83743552 0.01932756 0.12156967]\n",
      " [0.11182788 0.03378373 0.78484422 0.06954417]\n",
      " [0.11063295 0.0881492  0.0495941  0.75162375]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23195748 0.24737645 0.23524372 0.28542235]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.6464218290229772"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the theoretical parameters:\n",
    "theoretical_matrices = dict()\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    u = edges[i].edge[0].name.split(\"_\")[1]\n",
    "    v = edges[i].edge[1].name.split(\"_\")[1]\n",
    "    matrix_name = \"M_\" + u + \"_to_\" + v\n",
    "    print(\"Transition matrix:\", edges[i].transition_matrix)\n",
    "    theoretical_matrices[matrix_name] = np.array(edges[i].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", node_distr['Int_0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'M_0_to_2': array([[0.64642183, 0.02803076, 0.27483013, 0.05071728],\n",
      "       [0.03848122, 0.7757021 , 0.10189982, 0.08391686],\n",
      "       [0.03184807, 0.07934462, 0.82109495, 0.06771236],\n",
      "       [0.04702478, 0.01879095, 0.04076404, 0.89342023]]), 'M_0_to_3': array([[0.55850753, 0.13612552, 0.2499751 , 0.05539185],\n",
      "       [0.01791204, 0.82605787, 0.05907209, 0.096958  ],\n",
      "       [0.04534628, 0.008074  , 0.85149918, 0.09508054],\n",
      "       [0.00646156, 0.06355415, 0.02348564, 0.90649866]]), 'M_0_to_1': array([[0.75152183, 0.09827146, 0.07355421, 0.0766525 ],\n",
      "       [0.04888819, 0.86735324, 0.02989229, 0.05386628],\n",
      "       [0.05446854, 0.02450495, 0.75018255, 0.17084396],\n",
      "       [0.12039713, 0.03979299, 0.05715314, 0.78265675]]), 'M_1_to_4': array([[0.82835309, 0.03976936, 0.05146899, 0.08040856],\n",
      "       [0.02327998, 0.80873654, 0.06043788, 0.10754561],\n",
      "       [0.11544022, 0.13895831, 0.71027209, 0.03532938],\n",
      "       [0.09988558, 0.08065349, 0.01737091, 0.80209001]]), 'M_1_to_5': array([[0.78296278, 0.06942125, 0.08171685, 0.06589912],\n",
      "       [0.02166725, 0.83743552, 0.01932756, 0.12156967],\n",
      "       [0.11182788, 0.03378373, 0.78484422, 0.06954417],\n",
      "       [0.11063295, 0.0881492 , 0.0495941 , 0.75162375]])}\n"
     ]
    }
   ],
   "source": [
    "# # save the dictionary to a file\n",
    "# np.save('data.npy', theoretical_matrices)\n",
    "# # load the dictionary from the file\n",
    "# loaded_dict = np.load('data.npy', allow_pickle=True).item()\n",
    "# # print the loaded dictionary\n",
    "# print(loaded_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M_0_to_2\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix:\n",
      " [[0.86753892 0.05213542 0.06946348 0.01086218]\n",
      " [0.08065343 0.85671968 0.04982893 0.01279796]\n",
      " [0.11219846 0.17464684 0.65461785 0.05853685]\n",
      " [0.02775028 0.08494961 0.04088151 0.8464186 ]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_3\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix:\n",
      " [[0.75527498 0.04249873 0.00019955 0.20202674]\n",
      " [0.10446698 0.73349645 0.10777451 0.05426206]\n",
      " [0.009178   0.09167603 0.88105832 0.01808765]\n",
      " [0.10672648 0.08298123 0.01949473 0.79079756]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_1\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix:\n",
      " [[0.69009493 0.15321495 0.12518281 0.03150731]\n",
      " [0.03390489 0.89612499 0.03748613 0.032484  ]\n",
      " [0.13612944 0.00355591 0.86031238 0.00000227]\n",
      " [0.13829849 0.07169377 0.04358193 0.74642581]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_4\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix:\n",
      " [[0.79787321 0.09123409 0.0748348  0.0360579 ]\n",
      " [0.1525167  0.62670853 0.12700414 0.09377063]\n",
      " [0.06182208 0.09423139 0.82703856 0.01690797]\n",
      " [0.09434814 0.05487633 0.01727208 0.83350345]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_5\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix:\n",
      " [[0.77670263 0.03884329 0.11981174 0.06464234]\n",
      " [0.07306013 0.6964252  0.14381485 0.08669983]\n",
      " [0.0001472  0.02174493 0.93060026 0.04750762]\n",
      " [0.01290462 0.10677068 0.06826479 0.81205991]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23508015 0.26506847 0.26101933 0.23983305]\n"
     ]
    }
   ],
   "source": [
    "# Check the estimated parameters:\n",
    "for i in params.items():\n",
    "    print(i[0])\n",
    "    print(\"Edge:\", i[1].edge)\n",
    "    print(\"Transition matrix:\\n\", i[1].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", root_distr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix difference:\n",
      " [[0.09960865 0.07093256 0.00498621 0.03366231]\n",
      " [0.01576357 0.02703813 0.00892483 0.03387688]\n",
      " [0.00533819 0.03058066 0.01301418 0.01222828]\n",
      " [0.00504287 0.04282721 0.00457214 0.05244223]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix difference:\n",
      " [[0.00386991 0.01431933 0.0324696  0.04291902]\n",
      " [0.03240956 0.00438465 0.00578344 0.03101077]\n",
      " [0.00129948 0.00595081 0.00836875 0.00111846]\n",
      " [0.00368219 0.0201782  0.01427531 0.03077132]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix difference:\n",
      " [[0.00043271 0.04586535 0.01971483 0.02571781]\n",
      " [0.04628465 0.03270525 0.01225697 0.00132243]\n",
      " [0.00405161 0.02596747 0.04979691 0.01977783]\n",
      " [0.05635943 0.0032355  0.00691523 0.06003916]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix difference:\n",
      " [[0.02206907 0.00438435 0.02045318 0.00276846]\n",
      " [0.04114356 0.07985678 0.0152662  0.02344702]\n",
      " [0.01766089 0.02629588 0.04017144 0.00378534]\n",
      " [0.00126844 0.00706299 0.01874386 0.02707528]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix difference:\n",
      " [[0.04539034 0.01368587 0.03621623 0.00451176]\n",
      " [0.00302602 0.00613968 0.01219578 0.00908212]\n",
      " [0.00442834 0.01538392 0.00158094 0.0213932 ]\n",
      " [0.02699704 0.00888034 0.00551462 0.03036276]]\n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Compute the matrix difference between the theoretical and estimated parameters\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    print(\"Transition matrix difference:\\n\", np.abs(edges[i].transition_matrix - params['M_' + str(edges[i].edge[0].name.split(\"_\")[1]) + '_to_' + str(edges[i].edge[1].name.split(\"_\")[1])].transition_matrix))\n",
    "    print(\"---\"*30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Real pi\n",
      "[0.22578033 0.29516582 0.23953496 0.2395189 ]\n",
      "Real M\n",
      "M_0_to_2 [[0.88695491 0.01480513 0.07887592 0.01936404]\n",
      " [0.05760886 0.89945784 0.02936219 0.01357111]\n",
      " [0.06832355 0.0188208  0.80682896 0.10602669]\n",
      " [0.04160246 0.03278386 0.0147921  0.91082158]]\n",
      "M_0_to_3 [[0.84386291 0.08878631 0.0350045  0.03234627]\n",
      " [0.05707543 0.89797564 0.01842245 0.02652647]\n",
      " [0.04032408 0.0117897  0.90369441 0.04419181]\n",
      " [0.0786903  0.03779026 0.0289028  0.85461664]]\n",
      "M_0_to_1 [[0.89742342 0.00859638 0.0223438  0.0716364 ]\n",
      " [0.04219306 0.87245251 0.0371326  0.04822183]\n",
      " [0.04946544 0.07224094 0.84511865 0.03317497]\n",
      " [0.06690734 0.04108518 0.00632532 0.88568216]]\n",
      "M_1_to_4 [[0.84795328 0.05666423 0.0586081  0.03677439]\n",
      " [0.0493495  0.88868651 0.02868821 0.03327577]\n",
      " [0.07970769 0.04121695 0.8507314  0.02834395]\n",
      " [0.04575331 0.01748524 0.02345293 0.91330852]]\n",
      "M_1_to_5 [[0.88348369 0.06689142 0.03093429 0.01869059]\n",
      " [0.05721516 0.85219077 0.0592512  0.03134287]\n",
      " [0.05243798 0.01737545 0.85906242 0.07112415]\n",
      " [0.02107062 0.04086701 0.03172546 0.90633691]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Estimated M\n",
      "M_0_to_2 [[0.89521249 0.01467245 0.05941563 0.03069943]\n",
      " [0.06135694 0.88388093 0.03315699 0.02160514]\n",
      " [0.05388241 0.02656353 0.81891552 0.10063854]\n",
      " [0.06536689 0.03290982 0.01775414 0.88396915]]\n",
      "M_0_to_3 [[0.87784947 0.07269322 0.01426851 0.03518881]\n",
      " [0.05173    0.91431141 0.01039767 0.02356092]\n",
      " [0.05786897 0.01230747 0.87128198 0.05854157]\n",
      " [0.08500699 0.05306368 0.02554498 0.83638436]]\n",
      "M_0_to_1 [[0.93587973 0.00509451 0.01550927 0.04351648]\n",
      " [0.03564753 0.86831877 0.02559711 0.07043659]\n",
      " [0.03108415 0.06221373 0.86263191 0.0440702 ]\n",
      " [0.0501674  0.04042687 0.00405169 0.90535404]]\n",
      "M_1_to_4 [[0.86262317 0.04398836 0.06270201 0.03068646]\n",
      " [0.08031754 0.86466146 0.01661779 0.03840321]\n",
      " [0.0406554  0.04770331 0.88564154 0.02599974]\n",
      " [0.04855123 0.02002234 0.00548088 0.92594556]]\n",
      "M_1_to_5 [[0.87693947 0.07748    0.03043578 0.01514475]\n",
      " [0.05978215 0.87823762 0.04640915 0.01557108]\n",
      " [0.07178473 0.01443748 0.87427928 0.03949852]\n",
      " [0.02053934 0.05007767 0.03364835 0.89573465]]\n",
      "Estimated l\n",
      "M_0_to_2 0.5784064448232823\n",
      "M_0_to_3 0.5488230670382537\n",
      "M_0_to_1 0.4700083541664077\n",
      "M_1_to_4 0.5062258252002148\n",
      "M_1_to_5 0.520054343377952\n",
      "Estimated pi\n",
      "[0.21308539 0.30644422 0.24321785 0.23825355]\n",
      "Niter\n",
      "35\n",
      "TExec\n",
      "27.80595374107361\n"
     ]
    }
   ],
   "source": [
    "print(\"Real pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a55_b55_1000_random/real_root_distr.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Real M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a55_b55_1000_random/real_matrices.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"---\"*30)\n",
    "print(\"Estimated M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a55_b55_1000_random/RESULTS/repetition_98/M_estimation.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated l\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a55_b55_1000_random/RESULTS/repetition_98/estimated_branch_lengths.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a55_b55_1000_random/RESULTS/repetition_98/root_estimation.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Niter\")\n",
    "file = open(\"./OUTPUTS/EM_a55_b55_1000_random/RESULTS/repetition_98/Niter.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)\n",
    "print(\"TExec\")\n",
    "file = open(\"./OUTPUTS/EM_a55_b55_1000_random/RESULTS/repetition_98/TExec.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File deleted successfully.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Specify the file path\n",
    "file_path = \"/Users/marticortadagarcia/Desktop/prova.fasta\"\n",
    "\n",
    "# Check if the file exists\n",
    "try:\n",
    "    if os.path.exists(file_path):\n",
    "        # Delete the file\n",
    "        os.remove(file_path)\n",
    "        print(\"File deleted successfully.\")\n",
    "except OSError as error:\n",
    "    print(\"File does not exist.\")\n",
    "    pass \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "debugger",
   "language": "python",
   "name": "debugger"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "03f790eb89590da6bbc9dab944ceea1669c4a69ff2bd32047ab9c0f81f755bd1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
