{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Expectation-Maximization algorithm** (hidden model)\n",
    "### Author: Mart√≠ Cortada Garcia\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import numpy as np\n",
    "from Bio import SeqIO, Phylo\n",
    "from io import StringIO\n",
    "from collections import Counter, OrderedDict\n",
    "import itertools\n",
    "import pandas as pd\n",
    "import math\n",
    "import copy\n",
    "\n",
    "# Import required functions from modules we have programmed\n",
    "from data_simulation import simulate"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Input:**\n",
    "We have a tree $T$ (where the root corresponds to the internal node codified as `Int_0`) and the alignments of the leaves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAGwCAYAAACHJU4LAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy88F64QAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAx/ElEQVR4nO3de1iUdeL//9cIgpwGxROCqIXHDPDQSljaQU2NTNrPla5rHlqL7btgmbKZq2aWqdemJltupXWB25US+Ynaj7aamWR5KCXZpXQ1tRpT1LLk1Aoy3L8//DkbeWKQYd7I83Fdc13d99yH1/2+aOblfd8zY7MsyxIAAICBmng7AAAAwMVQVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjOXr7QBXoqqqSkePHlVISIhsNpu34wAAgBqwLEslJSWKiIhQkyaXPmfSoIvK0aNHFRUV5e0YAACgFg4fPqz27dtfcpkGXVRCQkIknT1Qu93u5TQAAKAmiouLFRUV5Xofv5QGXVTOXe6x2+0UFQAAGpia3LbBzbQAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAADYZlWUpOTlZYWJhsNpvy8/O9HQkeRlEBANSJ3Nxc2Ww2nTp1ymP7WL9+vTIzM7V27VoVFhbq+uuvv+TyCxYs0K9+9SuFhISoTZs2SkpK0r59+zyWD3WPogIAaDAOHjyodu3aqX///goPD5evr+8ll//www+VkpKiHTt2aOPGjTpz5ozuuOMOlZWV1VNiXCmvFpUtW7ZoxIgRioiIkM1m09tvv+3NOAAADyovL1daWpoiIyMVFBSk+Ph45ebmup4/efKkxowZo8jISAUGBiomJkarV692PT9x4kRNnjxZDodDNptNnTp1uuw+169fr4kTJ6pnz56Ki4tTZmamHA6H8vLyPHCE8IRLV1EPKysrU1xcnH73u9/p17/+tTejuPIAAC4vKCjI7XVSU1O1Z88eZWVlKSIiQjk5ORo2bJgKCgrUpUsXnT59Wn379tX06dNlt9u1bt06jRs3TtHR0erXr5/S09MVHR2t5cuXa+fOnfLx8XE7Q1FRkSQpLCzM7XXhJZYhJFk5OTmXXOb06dNWUVGR63H48GFLklVUVFRnGXjw4MGDx+UfF7J582ZLkvXjjz+e99w333xj+fj4WEeOHKk2f9CgQdaMGTMu+rqcmJhoTZs2zTX93HPPWR07dqzVa7zT6bQSExOtm266qVbro+4UFRVZUs3ev716RsVdCxYs0Ny5c+t8u2VlZQoODq7z7QIAziooKJDT6VTXrl2rzS8vL1fLli0lSU6nU/Pnz1d2draOHDmiiooKlZeXKzAwsE4ypKSk6PPPP9fHH39cJ9tD/WhQRWXGjBmaOnWqa7q4uFhRUVF1uo/jx4/X6pQmAODiSktL5ePjo7y8vPMu2Zz7h+Kzzz6r9PR0LV26VDExMQoKCtKUKVNUUVFxxftPTU3V2rVrtWXLFrVv3/6Kt4f606CKir+/v/z9/T26j6CgIIoKANSx3r17y+l06sSJExowYMAFl9m6datGjhyp++67T5JUVVWl/fv367rrrqv1fi3L0uTJk5WTk6Pc3Fxdc801td4WvKNBFRUAgPkKCgoUEhLimrbZbIqLi9PYsWM1fvx4LV68WL1799Z3332nTZs2KTY2VomJierSpYvWrFmjbdu2qUWLFlqyZImOHz9+RUUlJSVFq1at0jvvvKOQkBAdO3ZMkhQaGqqAgIArPlZ4HkUFAFCnBg4cWG3ax8dHlZWVysjI0Lx58zRt2jQdOXJErVq10o033qi77rpLkjRr1iwdOnRIQ4cOVWBgoJKTk5WUlOT6pE5tvPjii5KkW2+9tdr8jIwMTZw4sdbbRf2xWZZleWvnpaWlOnDggKSzpwWXLFmi2267TWFhYerQocNl1y8uLlZoaKiKiopkt9trnePnN9OWlpZy6QcAAA9y5/3bq2dUdu3apdtuu801fe5G2QkTJigzM9NLqQAAgCm8WlRuvfVWefGEDgCgAXM4HJe8f2XPnj01OjsPs3GPCgCgQYqIiLjkrydHRETUXxh4DEUFANAg+fr6qnPnzt6OAQ/j15MBAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqqDOWZSk5OVlhYWGy2WzKz8/3diQAQANHUWkkcnNzZbPZdOrUKY/tY/369crMzNTatWtVWFio66+//pLLv/jii4qNjZXdbpfdbldCQoL+8Y9/eCwfAKDh8WpRWbBggX71q18pJCREbdq0UVJSkvbt2+fNSLgCBw8eVLt27dS/f3+Fh4fL19f3ksu3b99eCxcuVF5ennbt2qXbb79dI0eO1BdffFFPiQEApvNqUfnwww+VkpKiHTt2aOPGjTpz5ozuuOMOlZWVeTNWo1ReXq60tDRFRkYqKChI8fHxys3NdT1/8uRJjRkzRpGRkQoMDFRMTIxWr17ten7ixImaPHmyHA6HbDabOnXqdNl9jhgxQnfeeae6dOmirl276plnnlFwcLB27NjhgSMEADREl/4nr4etX7++2nRmZqbatGmjvLw8DRw40CuZroaSFBQU5PY6qamp2rNnj7KyshQREaGcnBwNGzZMBQUF6tKli06fPq2+fftq+vTpstvtWrduncaNG6fo6Gj169dP6enpio6O1vLly7Vz5075+Pi4tX+n06k333xTZWVlSkhIcDs/AODq5NWi8ktFRUWSpLCwsAs+X15ervLyctd0cXFxnWdo27ZtnW+zvlmW5dbyDodDGRkZcjgcioiIkCSlpaVp/fr1ysjI0Pz58xUZGam0tDTXOpMnT9aGDRuUnZ2tfv36KTQ0VCEhIfLx8VF4eHiN911QUKCEhASdPn1awcHBysnJ0XXXXedWfgDA1cuYolJVVaUpU6bopptuuuhNmAsWLNDcuXPrOdnVr6CgQE6nU127dq02v7y8XC1btpR09ozH/PnzlZ2drSNHjqiiokLl5eUKDAy8on1369ZN+fn5Kioq0po1azRhwgR9+OGHlBUAgCSDikpKSoo+//xzffzxxxddZsaMGZo6dapruri4WFFRUXWa4/jx47W6dNKQlZaWysfHR3l5eeddsgkODpYkPfvss0pPT9fSpUsVExOjoKAgTZkyRRUVFVe0bz8/P3Xu3FmS1LdvX+3cuVPp6el6+eWXr2i7AICrgxFFJTU1VWvXrtWWLVvUvn37iy7n7+8vf39/j2YJCgpqdEWld+/ecjqdOnHihAYMGHDBZbZu3aqRI0fqvvvuk3T2DNj+/fvr/MxHVVVVtct7AIDGzatFxbIsTZ48WTk5OcrNzdU111zjzTiNQkFBgUJCQlzTNptNcXFxGjt2rMaPH6/Fixerd+/e+u6777Rp0ybFxsYqMTFRXbp00Zo1a7Rt2za1aNFCS5Ys0fHjx6+oqMyYMUPDhw9Xhw4dVFJSolWrVik3N1cbNmyoi0MFAFwFvFpUUlJStGrVKr3zzjsKCQnRsWPHJEmhoaEKCAjwZrSr1i8/TeXj46PKykplZGRo3rx5mjZtmo4cOaJWrVrpxhtv1F133SVJmjVrlg4dOqShQ4cqMDBQycnJSkpKct0AXRsnTpzQ+PHjVVhYqNDQUMXGxmrDhg0aMmTIFR0jAODqYbPc/YhIXe7cZrvg/IyMDE2cOPGy6xcXFys0NFRFRUWy2+21zlFWVua6F6O0tLTRXfoBAKA+ufP+7fVLPwAAABfDb/3AIxwOh4KDgy/6cDgc3o4IAGgAjPjUD64+ERERl/z15HNfLAcAwKVQVOARvr6+ru9HAQCgtrj0AwAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBSVy8jNzZXNZtOpU6dqvM7EiROVlJTk1n5++OEHjR07Vna7Xc2bN9ekSZNUWlrqXlgAAK4yFBVDjB07Vl988YU2btzo+iXp5ORkb8cCAMCrKCpuyszMVPPmzbVhwwb16NFDwcHBGjZsmAoLCyVJTz75pFauXKl33nlHNptNNptNubm5l9zm3r17tX79er3yyiuKj4/XzTffrOeff15ZWVk6evRoPRwVAABmoqjUwk8//aRFixbptdde05YtW+RwOJSWliZJSktL06hRo1zlpbCwUP3797/k9rZv367mzZvrhhtucM0bPHiwmjRpok8++cSjxwIAgMn4ZtpaOHPmjF566SVFR0dLklJTU/XUU09JkoKDgxUQEKDy8nKFh4fXaHvHjh1TmzZtqs3z9fVVWFiYjh07VrfhAQBoQDijUguBgYGukiJJ7dq104kTJ7yYCACAqxNFpRaaNm1abdpms8myrFpvLzw8/LyiU1lZqR9++KHGZ2UAALgaUVQ8wM/PT06ns8bLJyQk6NSpU8rLy3PN++CDD1RVVaX4+HhPRAQAoEGgqHhAp06d9K9//Uv79u3T999/rzNnzlxy+R49emjYsGF68MEH9emnn2rr1q1KTU3Vb37zG0VERNRTagAAzENR8YAHH3xQ3bp10w033KDWrVtr69atl13n9ddfV/fu3TVo0CDdeeeduvnmm7V8+fJ6SAsAgLls1pXcXOFlxcXFCg0NVVFRkex2e623U1ZWpuDgYElSaWmpgoKC6ioiAAD4BXfevzmjAgAAjEVRqQfz589XcHDwBR/Dhw/3djwAAIxVqy9827Vrl7Kzs+VwOFRRUVHtubfeeqtOgl1NHnroIY0aNeqCzwUEBNRzGgAAGg63i0pWVpbGjx+voUOH6r333tMdd9yh/fv36/jx47rnnns8kbHBCwsLU1hYmLdjAADQ4Lh96Wf+/Pl67rnn9H//93/y8/NTenq6/v3vf2vUqFHq0KGDJzICAIBGyu2icvDgQSUmJko6+8VmZWVlstlsevTRR/k4LQAAqFNuF5UWLVqopKREkhQZGanPP/9cknTq1Cn99NNPdZsOAAA0am7fozJw4EBt3LhRMTExuvfee/XII4/ogw8+0MaNGzVo0CBPZAQAAI2U20XlhRde0OnTpyVJM2fOVNOmTbVt2zb9z//8j2bNmlXnAQEAQOPldlH5+adXmjRposcff7xOAwEAAJzj9j0qmZmZF5xfWVmpGTNmXGkeAAAAF7eLysMPP6x7771XP/74o2vevn37FB8fr9WrV9dpOAAA0Li5XVR2796tb7/9VjExMdq4caOWLVumPn36qHv37vrnP//piYwAAKCRcvselejoaG3dulVTpkzRsGHD5OPjo5UrV2rMmDGeyAcAABqxWv0o4bp165SVlaWEhAQ1b95cr776qo4ePVrX2QAAQCPndlH5/e9/r3vvvVfTp0/XRx99pH/961/y8/NTTEyMsrOzPZERAAA0Um5f+tm6das++eQTxcXFSZLCw8P17rvvatmyZfrd73530V8JBgAAcJfbRSUvL0/+/v7nzU9JSdHgwYPrJBQAAIBUi0s/Fyop53Tr1u2KwgAAAPyc22dUJGnNmjXKzs6Ww+FQRUVFtec+++yzOgkGAADg9hmVv/zlL7r//vvVtm1b7d69W/369VPLli116NAhDR8+3BMZAQBAI+V2UfnrX/+q5cuX6/nnn5efn58ee+wxbdy4UQ8//LCKioo8kREAADRSbhcVh8Oh/v37S5ICAgJUUlIiSRo3bhxfoQ8AAOqU20UlPDxcP/zwgySpQ4cO2rFjhyTpq6++kmVZdZsOAAA0am4Xldtvv11///vfJUn333+/Hn30UQ0ZMkSjR4/WPffcU+cBAQBA4+X2p35mzpypyMhISWe/O6Vly5batm2b7r77bg0bNqzOAwIAgMbLZrl5vcbHx0eFhYVq06ZNtfknT55UmzZt5HQ66zTgpRQXFys0NFRFRUWy2+213k5ZWZmCg4MlSaWlpQoKCqqriAAA4Bfcef92+9LPxXpNaWmpmjVr5u7mAAAALqrGl36mTp0qSbLZbHriiScUGBjoes7pdOqTTz5Rr1696jwgAABovGpcVHbv3i3p7BmVgoIC+fn5uZ7z8/NTXFyc0tLS6j4hAABotGpcVDZv3izp7Cd90tPTr+ieEAAAgJpw+1M/GRkZnsgBAABwHrdvpgUAAKgvFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADUM8uylJycrLCwMNlsNuXn53s7krEoKgAA/Exubq5sNptOnTrlsX2sX79emZmZWrt2rQoLC3X99dfXeN2FCxfKZrNpypQpHstnEq8WlRdffFGxsbGy2+2y2+1KSEjQP/7xD29GAgDA4w4ePKh27dqpf//+Cg8Pl6+vb43W27lzp15++WXFxsZ6OKE5vFpU2rdvr4ULFyovL0+7du3S7bffrpEjR+qLL77wZiwAAC6qvLxcaWlpioyMVFBQkOLj45Wbm+t6/uTJkxozZowiIyMVGBiomJgYrV692vX8xIkTNXnyZDkcDtlsNnXq1KlG+y0tLdXYsWO1YsUKtWjRoo6Pylw1q3AeMmLEiGrTzzzzjF588UXt2LFDPXv29FKqhq+srMzbEQCgQQgKCnJ7ndTUVO3Zs0dZWVmKiIhQTk6Ohg0bpoKCAnXp0kWnT59W3759NX36dNntdq1bt07jxo1TdHS0+vXrp/T0dEVHR2v58uXauXOnfHx8arTflJQUJSYmavDgwZo3b57buRsqrxaVn3M6nXrzzTdVVlamhISECy5TXl6u8vJy13RxcXF9xWtQgoODvR0BABoEy7LcWt7hcCgjI0MOh0MRERGSpLS0NK1fv14ZGRmaP3++IiMjlZaW5lpn8uTJ2rBhg7Kzs9WvXz+FhoYqJCREPj4+Cg8Pr9F+s7Ky9Nlnn2nnzp1u5b0aeL2oFBQUKCEhQadPn1ZwcLBycnJ03XXXXXDZBQsWaO7cufWcsOEoKyujpACABxUUFMjpdKpr167V5peXl6tly5aSzv7De/78+crOztaRI0dUUVGh8vJyBQYG1mqfhw8f1iOPPKKNGzeqWbNmV3wMDY3Xi0q3bt2Un5+voqIirVmzRhMmTNCHH354wbIyY8YMTZ061TVdXFysqKio+ozbYBw/frxWpzQBABdXWloqHx8f5eXlnXfJ5tw/FJ999lmlp6dr6dKliomJUVBQkKZMmaKKiopa7TMvL08nTpxQnz59XPOcTqe2bNmiF154QeXl5TW+fNQQeb2o+Pn5qXPnzpKkvn37aufOnUpPT9fLL7983rL+/v7y9/ev74gNUlBQEEUFAOpY79695XQ6deLECQ0YMOCCy2zdulUjR47UfffdJ0mqqqrS/v37L3q14HIGDRqkgoKCavPuv/9+de/eXdOnT7+qS4pkQFH5paqqqmr3oQAA4A0FBQUKCQlxTdtsNsXFxWns2LEaP368Fi9erN69e+u7777Tpk2bFBsbq8TERHXp0kVr1qzRtm3b1KJFCy1ZskTHjx+vdVEJCQk573tWgoKC1LJlS7e+f6Wh8mpRmTFjhoYPH64OHTqopKREq1atUm5urjZs2ODNWAAAaODAgdWmfXx8VFlZqYyMDM2bN0/Tpk3TkSNH1KpVK91444266667JEmzZs3SoUOHNHToUAUGBio5OVlJSUkqKiryxmE0eDbL3Vue69CkSZO0adMmFRYWKjQ0VLGxsZo+fbqGDBlSo/WLi4sVGhqqoqIi2e32Wuf4+U2opaWlDfaSydVyHACAq5s7799ePaPy6quvenP3AADAcPzWDwAAXuRwOBQcHHzRh8Ph8HZErzLuZloAABqTiIiIS/568rkvlmusKCoAAHiRr6+v62s6cD4u/QAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVBqJ3Nxc2Ww2nTp1qsbrTJw4UUlJSW7t55lnnlH//v0VGBio5s2bu7UuAAC/RFFBnaqoqNC9996r//f//p+3owAArgIUlUYqMzNTzZs314YNG9SjRw8FBwdr2LBhKiwslCQ9+eSTWrlypd555x3ZbDbZbDbl5uZedrtz587Vo48+qpiYGA8fAQCgMfD1dgDTlJWVeTtCrbmb/aefftKiRYv02muvqUmTJrrvvvuUlpam119/XWlpadq7d6+Ki4uVkZEhSQoLC/NEbAAALoqi8gtt27b1doR6c+bMGb300kuKjo6WJKWmpuqpp56SJAUHBysgIEDl5eUKDw/3ZkwAQCPGpR9JQUFBsizL2zHqXWBgoKukSFK7du104sQJLyYCAKA6zqj8TGlpqbcjXJGysjK3zgg1bdq02rTNZmuUhQ0AYC6Kys8EBQV5O4JR/Pz85HQ6vR0DANCIUVRwUZ06ddKGDRu0b98+tWzZUqGhoeedhfklh8OhH374QQ6HQ06nU/n5+ZKkzp07Kzg4uB5SAwCuJtyjgot68MEH1a1bN91www1q3bq1tm7detl1nnjiCfXu3Vtz5sxRaWmpevfurd69e2vXrl31kBgAcLWxWQ34poTi4mKFhoaqqKhIdrvd23G8rqyszHXWorS0lEtZAAAjufP+zRkVAABgLIoKamz+/PkKDg6+4GP48OHejgcAuApxMy1q7KGHHtKoUaMu+FxAQEA9pwEANAYUFdRYWFgYX6MPAKhXXPoBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBfsGyLCUnJyssLEw2m035+fnejgQAjRZFBQ1Kbm6ubDabTp065bF9rF+/XpmZmVq7dq0KCwt1/fXXX3L5J598Ujabrdqje/fuHssHAI2Jr7cDAKY5ePCg2rVrp/79+9d4nZ49e+r99993Tfv68r8WANQFY86oLFy4UDabTVOmTPF2FDRg5eXlSktLU2RkpIKCghQfH6/c3FzX8ydPntSYMWMUGRmpwMBAxcTEaPXq1a7nJ06cqMmTJ8vhcMhms6lTp0412q+vr6/Cw8Ndj1atWtXxkQFA42TEP/t27typl19+WbGxsd6OctUoKyvzdoQrFhQU5PY6qamp2rNnj7KyshQREaGcnBwNGzZMBQUF6tKli06fPq2+fftq+vTpstvtWrduncaNG6fo6Gj169dP6enpio6O1vLly7Vz5075+PjUaL9ffvmlIiIi1KxZMyUkJGjBggXq0KGD2/kBANV5vaiUlpZq7NixWrFihebNm3fJZcvLy1VeXu6aLi4u9nS8Bqtt27bejnDFLMtya3mHw6GMjAw5HA5FRERIktLS0rR+/XplZGRo/vz5ioyMVFpammudyZMna8OGDcrOzla/fv0UGhqqkJAQ+fj4KDw8vEb7jY+PV2Zmprp166bCwkLNnTtXAwYM0Oeff66QkBC3jgEAUJ3Xi0pKSooSExM1ePDgyxaVBQsWaO7cufWUrOEJCgqSZVmy2WzejuIVBQUFcjqd6tq1a7X55eXlatmypSTJ6XRq/vz5ys7O1pEjR1RRUaHy8nIFBgbWer/Dhw93/XdsbKzi4+PVsWNHZWdna9KkSbXeLgDAy0UlKytLn332mXbu3Fmj5WfMmKGpU6e6pouLixUVFeWpeA1WaWmptyN4RWlpqXx8fJSXl3feJZvg4GBJ0rPPPqv09HQtXbpUMTExCgoK0pQpU1RRUVFnOZo3b66uXbvqwIEDdbZNAGisvFZUDh8+rEceeUQbN25Us2bNarSOv7+//P39PZys4avNvR1Xg969e8vpdOrEiRMaMGDABZfZunWrRo4cqfvuu0+SVFVVpf379+u6666rsxylpaU6ePCgxo0bV2fbBIDGymtFJS8vTydOnFCfPn1c85xOp7Zs2aIXXnhB5eXlNb6REY1PQUFBtfs/bDab4uLiNHbsWI0fP16LFy9W79699d1332nTpk2KjY1VYmKiunTpojVr1mjbtm1q0aKFlixZouPHj19RUUlLS9OIESPUsWNHHT16VHPmzJGPj4/GjBlTF4cKAI2a14rKoEGDVFBQUG3e/fffr+7du2v69OmUFFzSwIEDq037+PiosrJSGRkZmjdvnqZNm6YjR46oVatWuvHGG3XXXXdJkmbNmqVDhw5p6NChCgwMVHJyspKSklRUVFTrLN9++63GjBmjkydPqnXr1rr55pu1Y8cOtW7d+oqOEQAg2Sx3P1rhQbfeeqt69eqlpUuX1mj54uJihYaGqqioSHa73bPhAABAnXDn/duYL3wDAAD4Ja9/PPnnfv4NooAJHA7HJe9f2bNnD1/sBgAeZFRRAUwTERFxyV9PPvfFcgAAz6CoAJfg6+urzp07ezsGADRa3KMCAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsX28HuBKWZUmSiouLvZwEAADU1Ln37XPv45fSoItKSUmJJCkqKsrLSQAAgLtKSkoUGhp6yWVsVk3qjKGqqqp09OhRhYSEyGazXdG2iouLFRUVpcOHD8tut9dRwoaJsTiLcTiLcfgvxuIsxuEsxuG/3B0Ly7JUUlKiiIgINWly6btQGvQZlSZNmqh9+/Z1uk273d7o/+DOYSzOYhzOYhz+i7E4i3E4i3H4L3fG4nJnUs7hZloAAGAsigoAADAWReX/5+/vrzlz5sjf39/bUbyOsTiLcTiLcfgvxuIsxuEsxuG/PDkWDfpmWgAAcHXjjAoAADAWRQUAABiLogIAAIxFUQEAAMZqVEVl2bJl6tSpk5o1a6b4+Hh9+umnl1z+zTffVPfu3dWsWTPFxMTo3XffraeknufOWKxYsUIDBgxQixYt1KJFCw0ePPiyY9dQuPs3cU5WVpZsNpuSkpI8G7CeuDsOp06dUkpKitq1ayd/f3917dr1qvn/w92xWLp0qbp166aAgABFRUXp0Ucf1enTp+sprWds2bJFI0aMUEREhGw2m95+++3LrpObm6s+ffrI399fnTt3VmZmpsdzepq74/DWW29pyJAhat26tex2uxISErRhw4b6CetBtfl7OGfr1q3y9fVVr169ar3/RlNU3njjDU2dOlVz5szRZ599pri4OA0dOlQnTpy44PLbtm3TmDFjNGnSJO3evVtJSUlKSkrS559/Xs/J6567Y5Gbm6sxY8Zo8+bN2r59u6KionTHHXfoyJEj9Zy8brk7Dud8/fXXSktL04ABA+opqWe5Ow4VFRUaMmSIvv76a61Zs0b79u3TihUrFBkZWc/J6567Y7Fq1So9/vjjmjNnjvbu3atXX31Vb7zxhv70pz/Vc/K6VVZWpri4OC1btqxGy3/11VdKTEzUbbfdpvz8fE2ZMkUPPPBAg3+TdncctmzZoiFDhujdd99VXl6ebrvtNo0YMUK7d+/2cFLPcncczjl16pTGjx+vQYMGXVkAq5Ho16+flZKS4pp2Op1WRESEtWDBggsuP2rUKCsxMbHavPj4eOv3v/+9R3PWB3fH4pcqKyutkJAQa+XKlZ6KWC9qMw6VlZVW//79rVdeecWaMGGCNXLkyHpI6lnujsOLL75oXXvttVZFRUV9Raw37o5FSkqKdfvtt1ebN3XqVOumm27yaM76JMnKycm55DKPPfaY1bNnz2rzRo8ebQ0dOtSDyepXTcbhQq677jpr7ty5dR/IS9wZh9GjR1uzZs2y5syZY8XFxdV6n43ijEpFRYXy8vI0ePBg17wmTZpo8ODB2r59+wXX2b59e7XlJWno0KEXXb6hqM1Y/NJPP/2kM2fOKCwszFMxPa624/DUU0+pTZs2mjRpUn3E9LjajMPf//53JSQkKCUlRW3bttX111+v+fPny+l01ldsj6jNWPTv3195eXmuy0OHDh3Su+++qzvvvLNeMpvian29vFJVVVUqKSlp0K+VtZWRkaFDhw5pzpw5V7ytBv2jhDX1/fffy+l0qm3bttXmt23bVv/+978vuM6xY8cuuPyxY8c8lrM+1GYsfmn69OmKiIg474WpIanNOHz88cd69dVXlZ+fXw8J60dtxuHQoUP64IMPNHbsWL377rs6cOCA/vCHP+jMmTN18qLkLbUZi9/+9rf6/vvvdfPNN8uyLFVWVuqhhx5q8Jd+3HWx18vi4mL95z//UUBAgJeSedeiRYtUWlqqUaNGeTtKvfryyy/1+OOP66OPPpKv75XXjEZxRgV1Z+HChcrKylJOTo6aNWvm7Tj1pqSkROPGjdOKFSvUqlUrb8fxqqqqKrVp00bLly9X3759NXr0aM2cOVMvvfSSt6PVu9zcXM2fP19//etf9dlnn+mtt97SunXr9PTTT3s7Grxs1apVmjt3rrKzs9WmTRtvx6k3TqdTv/3tbzV37lx17dq1TrbZKM6otGrVSj4+Pjp+/Hi1+cePH1d4ePgF1wkPD3dr+YaiNmNxzqJFi7Rw4UK9//77io2N9WRMj3N3HA4ePKivv/5aI0aMcM2rqqqSJPn6+mrfvn2Kjo72bGgPqM3fQ7t27dS0aVP5+Pi45vXo0UPHjh1TRUWF/Pz8PJrZU2ozFrNnz9a4ceP0wAMPSJJiYmJUVlam5ORkzZw5U02aNI5/C17s9dJutzfKsylZWVl64IEH9OabbzboM8+1UVJSol27dmn37t1KTU2VdPa10rIs+fr66r333tPtt9/u1jYbxf9Ffn5+6tu3rzZt2uSaV1VVpU2bNikhIeGC6yQkJFRbXpI2btx40eUbitqMhST9+c9/1tNPP63169frhhtuqI+oHuXuOHTv3l0FBQXKz893Pe6++27XpxyioqLqM36dqc3fw0033aQDBw64ipok7d+/X+3atWuwJUWq3Vj89NNP55WRcwXOakQ/o3a1vl7WxurVq3X//fdr9erVSkxM9Haceme32897rXzooYfUrVs35efnKz4+3v2N1vo23AYmKyvL8vf3tzIzM609e/ZYycnJVvPmza1jx45ZlmVZ48aNsx5//HHX8lu3brV8fX2tRYsWWXv37rXmzJljNW3a1CooKPDWIdQZd8di4cKFlp+fn7VmzRqrsLDQ9SgpKfHWIdQJd8fhl66WT/24Ow4Oh8MKCQmxUlNTrX379llr16612rRpY82bN89bh1Bn3B2LOXPmWCEhIdbq1autQ4cOWe+9954VHR1tjRo1yluHUCdKSkqs3bt3W7t377YkWUuWLLF2795tffPNN5ZlWdbjjz9ujRs3zrX8oUOHrMDAQOuPf/yjtXfvXmvZsmWWj4+PtX79em8dQp1wdxxef/11y9fX11q2bFm118pTp0556xDqhLvj8EtX+qmfRlNULMuynn/+eatDhw6Wn5+f1a9fP2vHjh2u52655RZrwoQJ1ZbPzs62unbtavn5+Vk9e/a01q1bV8+JPcedsejYsaMl6bzHnDlz6j94HXP3b+LnrpaiYlnuj8O2bdus+Ph4y9/f37r22mutZ555xqqsrKzn1J7hzlicOXPGevLJJ63o6GirWbNmVlRUlPWHP/zB+vHHH+s/eB3avHnzBf+fP3fsEyZMsG655Zbz1unVq5fl5+dnXXvttVZGRka9565r7o7DLbfccsnlG6ra/D383JUWFZtlNaLzkwAAoEFpFPeoAACAhomiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFuMrdeuutmjJlirdjXNDXX38tm82m/Px8t9az2Wx6++23PZLJXU8++aR69erl7RjAVYuiAgA1ZFJBAhoLigqA81RUVHg7AgBIoqgAjUJlZaVSU1MVGhqqVq1aafbs2fr5z3x16tRJTz/9tMaPHy+73a7k5GRJ0vTp09W1a1cFBgbq2muv1ezZs3XmzBnXeucue7z22mvq1KmTQkND9Zvf/EYlJSWuZaqqqvTnP/9ZnTt3lr+/vzp06KBnnnmmWr5Dhw7ptttuU2BgoOLi4rR9+3a3ju/w4cMaNWqUmjdvrrCwMI0cOVJff/216/mJEycqKSlJixYtUrt27dSyZUulpKRUO5bCwkIlJiYqICBA11xzjVatWqVOnTpp6dKlrjGSpHvuuUc2m801fc6lxgBA7VFUgEZg5cqV8vX11aeffqr09HQtWbJEr7zySrVlFi1apLi4OO3evVuzZ8+WJIWEhCgzM1N79uxRenq6VqxYoeeee67aegcPHtTbb7+ttWvXau3atfrwww+1cOFC1/MzZszQwoULNXv2bO3Zs0erVq1S27Ztq21j5syZSktLU35+vrp27aoxY8aosrKyRsd25swZDR06VCEhIfroo4+0detWBQcHa9iwYdXODG3evFkHDx7U5s2btXLlSmVmZiozM9P1/Pjx43X06FHl5ubqf//3f7V8+XKdOHHC9fzOnTslSRkZGSosLHRN12QMAFyBWv/uMoAG4ZZbbrF69OhhVVVVueZNnz7d6tGjh2u6Y8eOVlJS0mW39eyzz1p9+/Z1Tc+ZM8cKDAy0iouLXfP++Mc/WvHx8ZZlWVZxcbHl7+9vrVix4oLb++qrryxJ1iuvvOKa98UXX1iSrL179140hyQrJyfHsizLeu2116xu3bpVO77y8nIrICDA2rBhg2VZZ3+GvmPHjlZlZaVrmXvvvdcaPXq0ZVmWtXfvXkuStXPnTtfzX375pSXJeu655y6435qOAYArwxkVoBG48cYbZbPZXNMJCQn68ssv5XQ6XfNuuOGG89Z74403dNNNNyk8PFzBwcGaNWuWHA5HtWU6deqkkJAQ13S7du1cZyL27t2r8vJyDRo06JL5YmNjq60vqdrZjEv55z//qQMHDigkJETBwcEKDg5WWFiYTp8+rYMHD7qW69mzp3x8fC6Yc9++ffL19VWfPn1cz3fu3FktWrSoUYZLjQGAK+Pr7QAAzBAUFFRtevv27Ro7dqzmzp2roUOHKjQ0VFlZWVq8eHG15Zo2bVpt2mazqaqqSpIUEBBQo33/fBvnCtW5bVxOaWmp+vbtq9dff/2851q3bl2jnFfKk9sGGjuKCtAIfPLJJ9Wmd+zYoS5dulQ7w/BL27ZtU8eOHTVz5kzXvG+++cat/Xbp0kUBAQHatGmTHnjgAfdC11CfPn30xhtvqE2bNrLb7bXaRrdu3VRZWandu3erb9++kqQDBw7oxx9/rLZc06ZNq52FAuB5XPoBGgGHw6GpU6dq3759Wr16tZ5//nk98sgjl1ynS5cucjgcysrK0sGDB/WXv/xFOTk5bu23WbNmmj59uh577DH97W9/08GDB7Vjxw69+uqrV3I41YwdO1atWrXSyJEj9dFHH+mrr75Sbm6uHn74YX377bc12kb37t01ePBgJScn69NPP9Xu3buVnJysgICAapfMOnXqpE2bNunYsWPnlRgAnkFRARqB8ePH6z//+Y/69eunlJQUPfLII66PIF/M3XffrUcffVSpqanq1auXtm3b5vo0kDtmz56tadOm6YknnlCPHj00evToOr1/IzAwUFu2bFGHDh3061//Wj169NCkSZN0+vRpt86w/O1vf1Pbtm01cOBA3XPPPXrwwQcVEhKiZs2auZZZvHixNm7cqKioKPXu3bvOjgHAxdks62dfpgAAkCR9++23ioqK0vvvv3/Zm4EBeA5FBQAkffDBByotLVVMTIwKCwv12GOP6ciRI9q/f/95N8sCqD/cTAsAOvvFcX/605906NAhhYSEqH///nr99dcpKYCXcUYFAAAYi5tpAQCAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABj/X9lUHWDme1/lgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "newick_tree = open(\"./newick.txt\", \"r\")\n",
    "newick = newick_tree.read()\n",
    "newick_tree.close()\n",
    "tree = Phylo.read(StringIO(newick), \"newick\")\n",
    "\n",
    "# Change nodes names\n",
    "for idx, clade in enumerate(tree.get_nonterminals()):\n",
    "    clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "\n",
    "# Change leaves names\n",
    "for idx, clade in enumerate(tree.get_terminals()):\n",
    "    clade.name = \"Leaf_\" + clade.name \n",
    "\n",
    "Phylo.draw(tree)\n",
    "\n",
    "net = Phylo.to_networkx(tree) # to graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "simulate() missing 1 required positional argument: 'directory'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m edges, node_distr, filename \u001b[39m=\u001b[39m simulate(net, \u001b[39m500\u001b[39;49m) \u001b[39m# simulate data calling the module data_simulation.py \u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[39m# Preprocess leave alignments\u001b[39;00m\n\u001b[1;32m      4\u001b[0m leave_alignments \u001b[39m=\u001b[39m [i \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m SeqIO\u001b[39m.\u001b[39mparse(filename, \u001b[39m\"\u001b[39m\u001b[39mfasta\u001b[39m\u001b[39m\"\u001b[39m)]\n",
      "\u001b[0;31mTypeError\u001b[0m: simulate() missing 1 required positional argument: 'directory'"
     ]
    }
   ],
   "source": [
    "edges, node_distr, filename = simulate(net, 500) # simulate data calling the module data_simulation.py \n",
    "\n",
    "# Preprocess leave alignments\n",
    "leave_alignments = [i for i in SeqIO.parse(filename, \"fasta\")]\n",
    "for i in leave_alignments:\n",
    "    print(\"Seq>\", i.seq)\n",
    "# Codify the alphabet {A,G,C,T} -> {0,1,2,3}\n",
    "leave_sequences = []\n",
    "for i in leave_alignments:\n",
    "    seq = \"\"\n",
    "    for j in i.seq:\n",
    "        if j == \"A\":\n",
    "            seq += \"0\"\n",
    "        elif j == \"G\":\n",
    "            seq += \"1\"\n",
    "        elif j == \"C\":\n",
    "            seq += \"2\"\n",
    "        elif j == \"T\":\n",
    "            seq += \"3\"\n",
    "        else:\n",
    "            raise ValueError(\"Invalid nucleotide\")\n",
    "    leave_sequences.append(seq)\n",
    "#leave_sequences\n",
    "edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## **EM algorithm**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from evaluate import init_EM\n",
    "\n",
    "Results = init_EM(newick, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Source: Int_0\n",
      "Target: Leaf_2\n",
      "Matrix:\n",
      " [[0.80529814 0.06095358 0.09251042 0.04123786]\n",
      " [0.05611119 0.71483627 0.11102639 0.11802615]\n",
      " [0.0952762  0.07415636 0.731184   0.09938344]\n",
      " [0.02196288 0.0820518  0.08234946 0.81363586]]\n",
      "Source: Int_0\n",
      "Target: Leaf_3\n",
      "Matrix:\n",
      " [[0.63632807 0.06782976 0.1446222  0.15121997]\n",
      " [0.12810897 0.67208297 0.04706075 0.15274731]\n",
      " [0.10204507 0.09021686 0.74449869 0.06323937]\n",
      " [0.06570111 0.08178298 0.05155947 0.80095644]]\n",
      "Source: Int_0\n",
      "Target: Int_1\n",
      "Matrix:\n",
      " [[0.72577892 0.07421689 0.0831795  0.11682468]\n",
      " [0.08700214 0.7261134  0.06686419 0.12002027]\n",
      " [0.09519544 0.182137   0.68012648 0.04254107]\n",
      " [0.06761601 0.06509887 0.08961157 0.77767356]]\n",
      "Source: Int_1\n",
      "Target: Leaf_4\n",
      "Matrix:\n",
      " [[0.85639817 0.07234946 0.04036045 0.03089192]\n",
      " [0.0298384  0.73128255 0.0944944  0.14438465]\n",
      " [0.03533407 0.10420379 0.77836777 0.08209437]\n",
      " [0.06403097 0.03195277 0.06966545 0.8343508 ]]\n",
      "Source: Int_1\n",
      "Target: Leaf_5\n",
      "Matrix:\n",
      " [[0.80075352 0.06438893 0.11198018 0.02287738]\n",
      " [0.07400816 0.72459648 0.10712499 0.09427037]\n",
      " [0.14358855 0.08350079 0.74500083 0.02790983]\n",
      " [0.01210798 0.14166184 0.06442053 0.78180965]]\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(Results)):\n",
    "    print(\"Source:\", Results[i].source)\n",
    "    print(\"Target:\", Results[i].target)\n",
    "    print(\"Matrix:\\n\", Results[i].matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "l -73482.53888183684\n",
      "73482.53888183684\n",
      "--------\n",
      "2803.2235244028125\n",
      "--------\n",
      "403.42311980716477\n",
      "--------\n",
      "147.34679458629398\n",
      "--------\n",
      "65.81431201811938\n",
      "--------\n",
      "32.62458047835389\n",
      "--------\n",
      "17.4679848113592\n",
      "--------\n",
      "9.96598948755127\n",
      "--------\n",
      "5.979527266841615\n",
      "--------\n",
      "3.724567122422741\n",
      "--------\n",
      "2.382098244605004\n",
      "--------\n",
      "1.5508962869207608\n",
      "--------\n",
      "1.021128828782821\n",
      "--------\n",
      "0.6763189632765716\n",
      "--------\n",
      "0.44849936086393427\n",
      "--------\n",
      "0.29640033769828733\n",
      "--------\n",
      "0.19417478849936742\n",
      "--------\n",
      "0.12524008554464672\n",
      "--------\n",
      "0.07875408534891903\n",
      "--------\n",
      "0.047520216307020746\n",
      "--------\n",
      "0.02670328039675951\n",
      "--------\n",
      "0.01302253625181038\n",
      "--------\n",
      "0.004234299674862996\n",
      "--------\n",
      "0.0012054162216372788\n",
      "--------\n",
      "0.004364024440292269\n",
      "--------\n",
      "0.005981738388072699\n",
      "--------\n",
      "0.006572629208676517\n",
      "--------\n",
      "0.006493631357443519\n",
      "--------\n",
      "0.005992020829580724\n",
      "--------\n",
      "0.005238327459665015\n",
      "--------\n",
      "0.00434925795707386\n",
      "--------\n",
      "0.003403757931664586\n",
      "--------\n",
      "0.002454297398799099\n",
      "--------\n",
      "0.0015348267188528553\n",
      "--------\n",
      "Done!\n",
      "Number of iterations:  34\n"
     ]
    }
   ],
   "source": [
    "# Auxiliary functions\n",
    "\n",
    "def init_root_distribution():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the root distribution (all probabilities between 0.2 and 0.3)\n",
    "    \"\"\"\n",
    "    estimated_root_distribution = []\n",
    "    while True:\n",
    "        estimated_root_distribution = np.random.dirichlet([1,1,1,1])\n",
    "        i = 0\n",
    "        while i<4:\n",
    "            if estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3:\n",
    "                i += 1\n",
    "            else:\n",
    "                break\n",
    "        if i == 4:\n",
    "            assert(np.sum(estimated_root_distribution) < 1.0000001 and np.sum(estimated_root_distribution) > 0.9999999)\n",
    "            assert(estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3 for i in range(4))\n",
    "            break\n",
    "    return estimated_root_distribution\n",
    "\n",
    "def init_transition_matrix():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the transition matrices (which we estimate as LARGE-DIAGONAL-VALUE matrices)\n",
    "    \"\"\"\n",
    "    M = np.zeros((4,4))\n",
    "    i=0\n",
    "    while i<4:\n",
    "        R = np.random.dirichlet([1,1,1,1])\n",
    "        if R[i] > 0.6 and R[i] < 0.8:\n",
    "            M[i,:] = R \n",
    "            i = i + 1\n",
    "    assert (M[i,i] > 0.6 and M[i,i] < 0.8 for i in range(4))\n",
    "    assert (np.sum(M[i,:]) < 1.0000001 and np.sum(M[i,:]) > 0.9999999 for i in range(4))\n",
    "    return M\n",
    "\n",
    "def log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "    \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "    logL = 0\n",
    "    for obs in states:\n",
    "        p_i = 0 \n",
    "        observed_data = obs[n_int:]\n",
    "        if observed_data in u_i:\n",
    "            for state in states:\n",
    "                if state[n_int:] == observed_data:\n",
    "                    pi = root_distribution[int(state[0])]\n",
    "                    for p in params.items():\n",
    "                        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                        pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                    p_i += pi\n",
    "        logL += (u_i[observed_data] * math.log(p_i))\n",
    "    return logL\n",
    "\n",
    "############################################################################################################\n",
    "########################################## FELSENSTEIN ALGORITHM ###########################################\n",
    "############################################################################################################\n",
    "# # Felsentstein algorithm\n",
    "# # \"((4,5),(2,3))\"\n",
    "# newick_tree = \"(((4:1,5:1):1,(6:1,7:1):1):1,(8:1,9:1):1);\"\n",
    "# L_tree = Phylo.read(StringIO(newick_tree), \"newick\")\n",
    "# # Change nodes names\n",
    "# for idx, clade in enumerate(L_tree.get_nonterminals()):\n",
    "#     id = 0\n",
    "#     if idx == 2:\n",
    "#         clade.name = \"Int_Id_\" + str(id)\n",
    "#         id += 1\n",
    "#     else:\n",
    "#         clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "# # Change leaves names\n",
    "# for idx, clade in enumerate(L_tree.get_terminals()):\n",
    "#     clade.name = \"Leaf_\" + clade.name \n",
    "# L_net = Phylo.to_networkx(L_tree) # to graph\n",
    "# L_edges = []\n",
    "# for edge in L_net.edges():\n",
    "#     if edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     elif edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] != \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     elif edge[0].name.split(\"_\")[1] != \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     else:\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     L_edges.append((u,v))\n",
    "# # Create a dictionary to store the father and son information\n",
    "# family_tree = {}\n",
    "# # Loop through the edges and update the dictionary\n",
    "# for father, son in L_edges:\n",
    "#     if father not in family_tree:\n",
    "#         family_tree[father] = []\n",
    "#     family_tree[father].append(son)\n",
    "# # Create a DataFrame from the dictionary\n",
    "# tree_df = pd.DataFrame.from_dict(family_tree, orient='index')\n",
    "# # Fill NaN values with empty strings\n",
    "# tree_df = tree_df.fillna('')\n",
    "# # Change column names\n",
    "# tree_df.columns = ['child_1', 'child_2']\n",
    "\n",
    "# class TreeInference():\n",
    "#     def __init__(self, tree_data, leaf_values, params, root_distr):\n",
    "#         # Set parameters\n",
    "#         self.tree_data = tree_data.copy()\n",
    "#         self.leaf_values = leaf_values\n",
    "#         self.params = params\n",
    "#         self.root_distr = root_distr\n",
    "\n",
    "#         # Find IDs of internal nodes, all nodes and leaf nodes\n",
    "#         self.internal_node_ids = np.array(tree_data.index)\n",
    "#         self.all_node_ids = np.unique(np.concatenate((np.array(self.tree_data.index),\n",
    "#                                                       np.array(self.tree_data[\"child_1\"]),\n",
    "#                                                       np.array(self.tree_data[\"child_2\"]))))\n",
    "#         self.leaf_node_ids = list(set(self.all_node_ids) - set(self.internal_node_ids))\n",
    "        \n",
    "#         # Count internal nodes and leaves for each row in tree_data and set root node\n",
    "#         self.treeNodeCounts()\n",
    "#         self.root_node = 0\n",
    "        \n",
    "#         # Get indexes for matricies for states, e.g. A is row 0, G is row 2...\n",
    "#         states = [\"A\", \"G\", \"C\", \"T\"]\n",
    "#         self.matrix_idx = dict(zip(states, range(len(states))))\n",
    "        \n",
    "#         # Empty matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "#         # Log likelihood of whole tree\n",
    "#         self.L_root = None\n",
    "        \n",
    "\n",
    "#     \"\"\"\n",
    "#     Calculates the number of internal nodes and leaves below each internal node.\n",
    "#     This works by recursively iterating down the child nodes from the original internal node until a\n",
    "#     leaf node is found, and then returning upwards counting the leaves and nodes.\n",
    "#     The new derived information is then added as two new columns at the end of the data.\n",
    "#     \"\"\"\n",
    "#     def treeNodeCounts(self):\n",
    "#         n_nodes = np.array([])\n",
    "#         n_leaves = np.array([])\n",
    "\n",
    "#         # Set the number of internal nodes and leaves for each internal node row\n",
    "#         for n in self.internal_node_ids:\n",
    "#             num_nodes, num_leaves = self.checkNode(self.tree_data, n, self.internal_node_ids)\n",
    "#             n_nodes = np.append(n_nodes, num_nodes)\n",
    "#             n_leaves = np.append(n_leaves, num_leaves)\n",
    "\n",
    "#         # Add new information to dataframe\n",
    "#         self.tree_data[\"N_Internal\"] = n_nodes\n",
    "#         self.tree_data[\"N_Leaves\"] = n_leaves\n",
    "                \n",
    "#     # Recursively find the number of internal nodes and leaves given a node ID\n",
    "#     def checkNode(self, tree_data, node_id, node_array):\n",
    "#         # Record the number of internal nodes and leaves below\n",
    "#         num_nodes = 0\n",
    "#         num_leaves = 0\n",
    "\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[node_id][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[node_id][\"child_2\"]\n",
    "\n",
    "#         # Check if found an internal node to the left\n",
    "#         if child_1 in node_array:\n",
    "#             # Recursively check nodes below\n",
    "#             d1_nodes, d1_leaves = self.checkNode(self.tree_data, child_1, node_array)\n",
    "#             num_nodes += d1_nodes + 1\n",
    "#             num_leaves += d1_leaves\n",
    "\n",
    "#         # Check if found an internal node to the right\n",
    "#         if child_2 in node_array:\n",
    "#             d2_nodes, d2_leaves = self.checkNode(self.tree_data, child_2, node_array)\n",
    "#             num_nodes += d2_nodes + 1\n",
    "#             num_leaves += d2_leaves\n",
    "\n",
    "#         return num_nodes, num_leaves\n",
    "    \n",
    "    \n",
    "#     # Calculate likelihood for an internal node\n",
    "#     def nodeLikelihood(self, node_n, params):\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[str(node_n)][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[str(node_n)][\"child_2\"]\n",
    "        \n",
    "#         # Check if likelihood of left child node set\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(child_1, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_1), 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(int(child_1), params)\n",
    "#         # Check if likelihood of right child node set\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 self.nodeLikelihood(child_2, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_2), 0] is None):\n",
    "#                 self.nodeLikelihood(int(child_2), params)\n",
    "        \n",
    "#         M1, M2 = np.full((4,4),None), np.full((4,4),None)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             # Create transition matrix for left and right child nodes\n",
    "#             matrix_name_child1 = \"M_0_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_0_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_1 == \"Id_0\":\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = np.identity(4)\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_2 == \"Id_0\":\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = np.identity(4)\n",
    "#         else:\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "\n",
    "#         # Get conditional likelihood of child nodes\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             d1_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d1_likelihood = self.L[int(child_1)]\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             d2_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d2_likelihood = self.L[int(child_2)]\n",
    "        \n",
    "#         # Calculate the likelihood of each state for the node\n",
    "#         left_prob = np.sum(M1 * d1_likelihood, axis = 1)\n",
    "#         right_prob = np.sum(M2 * d2_likelihood, axis = 1)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             self.L[-1,:] = np.multiply(left_prob, right_prob)\n",
    "#         else:\n",
    "#             self.L[int(node_n),:] = np.multiply(left_prob, right_prob)\n",
    "    \n",
    "#     # Run Felsenstein‚Äôs algorithm to calculate the log likelihood of the tree\n",
    "#     def calculateLikelihood(self, params, Estep):\n",
    "#         # Clear matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "        \n",
    "#         # Set likelihood of leaf states\n",
    "#         for leaf_n in self.leaf_node_ids:\n",
    "#             if leaf_n == \"Id_0\":\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[-1, :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[-1, self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "#             else:\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[int(leaf_n), :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[int(leaf_n), self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "        \n",
    "#         for node_n in self.internal_node_ids:\n",
    "#             # Check likelihood has not been set\n",
    "#             if node_n == \"Id_0\":\n",
    "#                 if self.L[-1, 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "#             else:\n",
    "#                 if self.L[int(node_n), 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "        \n",
    "#         if Estep == True:\n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = np.sum(np.multiply(self.root_distr, self.L[self.root_node]))\n",
    "#         else: \n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = math.log(np.sum(np.multiply(self.root_distr, self.L[self.root_node])))\n",
    "        \n",
    "#         return self.L_root\n",
    "\n",
    "# def Felsenstein_log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "#     \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "#     logL = 0\n",
    "#     for obs in states:\n",
    "#         observed_data = obs[n_int:]\n",
    "#         tree_leaves = dict()\n",
    "#         for i in range(len(observed_data)):\n",
    "#             if observed_data[i] == str(0):\n",
    "#                 tree_leaves[n_int+i] = \"A\"\n",
    "#             elif observed_data[i] == str(1):\n",
    "#                 tree_leaves[n_int+i] = \"G\"\n",
    "#             elif observed_data[i] == str(2):\n",
    "#                 tree_leaves[n_int+i] = \"C\"\n",
    "#             elif observed_data[i] == str(3):\n",
    "#                 tree_leaves[n_int+i] = \"T\"\n",
    "#         tree_inference = TreeInference(tree_df, tree_leaves, params, root_distribution)\n",
    "#         tree_inference.calculateLikelihood(params, False)\n",
    "#         logL += (u_i[observed_data] * tree_inference.L_root)\n",
    "#         del tree_inference\n",
    "#     return logL\n",
    "\n",
    "\n",
    "############################################################################################################\n",
    "\n",
    "class Param:\n",
    "    \"\"\"\n",
    "    Class to store the parameters of the tree\n",
    "    \"\"\"\n",
    "    def __init__(self, edge, transition_matrix, alignment=None):\n",
    "        self.edge = edge\n",
    "        self.transition_matrix = transition_matrix\n",
    "        self.alignment = alignment # it will only be placed in the leaves\n",
    "\n",
    "# Obtain number of leaves and internal nodes\n",
    "n_leaves, n_int = 0, 0 \n",
    "for node in net.nodes():\n",
    "    if node.name.startswith(\"L\"):\n",
    "        n_leaves += 1\n",
    "    else:\n",
    "        n_int += 1\n",
    "\n",
    "# Count all the leaves ocurrences \n",
    "sequence_length = len(leave_sequences[0])\n",
    "number_of_sequences = len(leave_sequences)\n",
    "occurrences = []\n",
    "for i in range(sequence_length):\n",
    "    _ = ''\n",
    "    for j in range(number_of_sequences):\n",
    "        _ += leave_sequences[j][i]\n",
    "    occurrences.append(_)\n",
    "c = Counter(occurrences)\n",
    "u_i = OrderedDict(sorted(c.items()))\n",
    "\n",
    "# Define all possible (hidden) states of the internal nodes \n",
    "hidden_combinations = itertools.product([0,1,2,3], repeat=n_int)\n",
    "sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "for i in range(len(sorted_hidden_combinations)):\n",
    "    sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    " \n",
    "# Define all possible fully observed states (i.e. combining u_i + internal nodes)\n",
    "# '000000' would mean to have an 'A' in all nodes: Int_0, Int_1, Leaf_2, Leaf_3, Leaf_4, Leaf_5 \n",
    "states = list(itertools.product(list(sorted_hidden_combinations), list(u_i.keys())))\n",
    "states = [i[0]+i[1] for i in states]\n",
    "\n",
    "############################################################################################################\n",
    "###################################### STEP 0: Initialise parameters #######################################\n",
    "############################################################################################################\n",
    "estimated_root_distribution = init_root_distribution() # First estimation for the root distribution\n",
    "\n",
    "# PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "# for edge in net.edges():\n",
    "#     # If we are in a leaf, we need to specify the alignment\n",
    "#     iter_leaves = 0\n",
    "#     if edge[1].name.startswith(\"L\"):\n",
    "#         new_edge = Param(edge, init_transition_matrix(), leave_alignments[iter_leaves].seq)\n",
    "#         iter_leaves += 1\n",
    "#     # Otherwise, are in internal nodes\n",
    "#     else:\n",
    "#         new_edge = Param(edge, init_transition_matrix())\n",
    "#     u = str(edge[0].name.split(\"_\")[1])\n",
    "#     v = str(edge[1].name.split(\"_\")[1])\n",
    "#     name = \"M_\" + u + \"_to_\" + v\n",
    "#     PARAMS[name] = new_edge\n",
    "\n",
    "# ADRIA CODE\n",
    "i=0\n",
    "PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "for edge in net.edges():\n",
    "    # If we are in a leaf, we need to specify the alignment\n",
    "    iter_leaves = 0\n",
    "    if edge[1].name.startswith(\"L\"):\n",
    "        new_edge = Param(edge, Results[i].matrix, leave_alignments[iter_leaves].seq)\n",
    "        iter_leaves += 1\n",
    "    # Otherwise, are in internal nodes\n",
    "    else:\n",
    "        new_edge = Param(edge, Results[i].matrix)\n",
    "    u = str(edge[0].name.split(\"_\")[1])\n",
    "    v = str(edge[1].name.split(\"_\")[1])\n",
    "    name = \"M_\" + u + \"_to_\" + v\n",
    "    PARAMS[name] = new_edge\n",
    "    i += 1\n",
    "\n",
    "eps = 10**(-3)                  # epsilon for convergence\n",
    "iter = 0                        # iteration counter \n",
    "params = copy.copy(PARAMS)      # copy of the parameters (will be updated at each iteration)\n",
    "# logF = Felsenstein_log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "logL = log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "# print(logF)\n",
    "print(\"l\", logL)\n",
    "logL_ = 0\n",
    "\n",
    "while np.abs(logL_ - logL) > eps:\n",
    "    print(np.abs(logL_ - logL))\n",
    "    # print(np.abs(logF_ - logF))\n",
    "    print(\"--------\")\n",
    "    if iter > 0:\n",
    "        logL = logL_\n",
    "        # logF = logF_\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## E-step ##################################################\n",
    "    ############################################################################################################\n",
    "    # Define the expected hidden data matrix\n",
    "    U = dict()\n",
    "    root_distr = estimated_root_distribution ########\n",
    "\n",
    "    for obs in states:\n",
    "        U[obs] = u_i[obs[n_int:]] # initialise by computing u_i of the expected hidden data matrix formula\n",
    "        f_ij = estimated_root_distribution[int(obs[0])] \n",
    "        for p in params.items():\n",
    "            u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "            f_ij *= p[1].transition_matrix[int(obs[u]), int(obs[v])]\n",
    "        # Now f_ij := numerator of the expected hidden data matrix formula \n",
    "\n",
    "        # We need to marginalise to obtain f_i := denominator of the expected hidden data matrix formula\n",
    "        f_i = 0 # we initialise the partial data model\n",
    "        observed_data = obs[n_int:]\n",
    "        for state in states:\n",
    "            if state[n_int:] == observed_data:\n",
    "                pi = estimated_root_distribution[int(state[0])]\n",
    "                for p in params.items():\n",
    "                    u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                    pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                f_i += pi\n",
    "\n",
    "        # observed_data = obs[n_int:]\n",
    "        # tree_leaves = dict()\n",
    "        # for i in range(len(observed_data)):\n",
    "        #     if observed_data[i] == str(0):\n",
    "        #         tree_leaves[n_int+i] = \"A\"\n",
    "        #     elif observed_data[i] == str(1):\n",
    "        #         tree_leaves[n_int+i] = \"G\"\n",
    "        #     elif observed_data[i] == str(2):\n",
    "        #         tree_leaves[n_int+i] = \"C\"\n",
    "        #     elif observed_data[i] == str(3):\n",
    "        #         tree_leaves[n_int+i] = \"T\"\n",
    "        # tree_inference = TreeInference(tree_df, tree_leaves, params, root_distr)\n",
    "        # tree_inference.calculateLikelihood(params, True)\n",
    "\n",
    "        U[obs] *= f_ij/f_i\n",
    "        # U[obs] *= f_ij/tree_inference.L_root\n",
    "        # del tree_inference\n",
    "        # del tree_leaves\n",
    "\n",
    "\n",
    "    assert(len(U) <= 4**(n_int+n_leaves))\n",
    "    sum = 0\n",
    "    sum2 = 0\n",
    "    for i in U.items():\n",
    "        sum += i[1]\n",
    "    for i in u_i.items():\n",
    "        sum2 += i[1]\n",
    "    assert(sum < sequence_length+0.001 and sum > sequence_length-0.001)\n",
    "    assert(sum2 < sequence_length+0.001 and sum2 > sequence_length-0.001)\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## M-step ##################################################\n",
    "    ############################################################################################################\n",
    "    estimated_parameters = dict()\n",
    "    root_distr = np.zeros(4)\n",
    "    iter_root = 0 # counter for the root distribution, we only want to count once\n",
    "\n",
    "    for p in params.items():\n",
    "        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "        name = \"M_\" + str(u) + \"_to_\" + str(v)\n",
    "        estimated_parameters[name] = np.zeros((4,4))\n",
    "        for i in range(4):\n",
    "            for j in range(4):\n",
    "                ui = 0\n",
    "                for obs in U.items():\n",
    "                    # For the root distribution, we only need to consider the first internal node\n",
    "                    if iter_root == 0 and obs[0][0] == str(i):\n",
    "                        root_distr[i] += obs[1]\n",
    "                    # For the transition matrices:\n",
    "                    if obs[0][u] == str(i) and obs[0][v] == str(j):\n",
    "                        estimated_parameters[name][i][j] += obs[1]\n",
    "                    if obs[0][u] == str(i):\n",
    "                        ui += obs[1]\n",
    "                # Update root distribution\n",
    "                if iter_root == 0:\n",
    "                    root_distr[i] /= sequence_length\n",
    "                iter_root = 1\n",
    "                estimated_parameters[name][i][j] /= ui\n",
    "            iter_root = 0\n",
    "            \n",
    "        # Update parameters\n",
    "        params[name].transition_matrix = estimated_parameters[name]\n",
    "\n",
    "    ##########################################################################################################\n",
    "    ########################### Compute log-likelihood of the estimated parameters ###########################\n",
    "    ##########################################################################################################\n",
    "    # logF_ = Felsenstein_log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "    logL_ = log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "\n",
    "    iter += 1\n",
    "\n",
    "print(\"Done!\")\n",
    "print(\"Number of iterations: \", iter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################### DEBUGGING ############################\n",
    "\n",
    "# # All possible (theoretical) occurrences of nucleotides in the leaves\n",
    "# alphabet = [0, 1, 2, 3]\n",
    "# combinations = itertools.product(alphabet, repeat=n_leaves)\n",
    "# sorted_combinations = sorted(combinations)\n",
    "# for i in range(len(sorted_combinations)):\n",
    "#     sorted_combinations[i] = ''.join([str(s) for s in sorted_combinations[i]])\n",
    "\n",
    "# # All possible (hidden) states of the internal nodes (i.e. indexes of columns of U)\n",
    "# hidden_combinations = itertools.product(alphabet, repeat=n_int)\n",
    "# sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "# for i in range(len(sorted_hidden_combinations)):\n",
    "#     sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    "\n",
    "# # To check if an element is within a dictionary or not \n",
    "# if '0101' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')\n",
    "# if '0000' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix: [[0.64642183 0.02803076 0.27483013 0.05071728]\n",
      " [0.03848122 0.7757021  0.10189982 0.08391686]\n",
      " [0.03184807 0.07934462 0.82109495 0.06771236]\n",
      " [0.04702478 0.01879095 0.04076404 0.89342023]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix: [[0.55850753 0.13612552 0.2499751  0.05539185]\n",
      " [0.01791204 0.82605787 0.05907209 0.096958  ]\n",
      " [0.04534628 0.008074   0.85149918 0.09508054]\n",
      " [0.00646156 0.06355415 0.02348564 0.90649866]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix: [[0.75152183 0.09827146 0.07355421 0.0766525 ]\n",
      " [0.04888819 0.86735324 0.02989229 0.05386628]\n",
      " [0.05446854 0.02450495 0.75018255 0.17084396]\n",
      " [0.12039713 0.03979299 0.05715314 0.78265675]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix: [[0.82835309 0.03976936 0.05146899 0.08040856]\n",
      " [0.02327998 0.80873654 0.06043788 0.10754561]\n",
      " [0.11544022 0.13895831 0.71027209 0.03532938]\n",
      " [0.09988558 0.08065349 0.01737091 0.80209001]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix: [[0.78296278 0.06942125 0.08171685 0.06589912]\n",
      " [0.02166725 0.83743552 0.01932756 0.12156967]\n",
      " [0.11182788 0.03378373 0.78484422 0.06954417]\n",
      " [0.11063295 0.0881492  0.0495941  0.75162375]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23195748 0.24737645 0.23524372 0.28542235]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.6464218290229772"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the theoretical parameters:\n",
    "theoretical_matrices = dict()\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    u = edges[i].edge[0].name.split(\"_\")[1]\n",
    "    v = edges[i].edge[1].name.split(\"_\")[1]\n",
    "    matrix_name = \"M_\" + u + \"_to_\" + v\n",
    "    print(\"Transition matrix:\", edges[i].transition_matrix)\n",
    "    theoretical_matrices[matrix_name] = np.array(edges[i].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", node_distr['Int_0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'M_0_to_2': array([[0.64642183, 0.02803076, 0.27483013, 0.05071728],\n",
      "       [0.03848122, 0.7757021 , 0.10189982, 0.08391686],\n",
      "       [0.03184807, 0.07934462, 0.82109495, 0.06771236],\n",
      "       [0.04702478, 0.01879095, 0.04076404, 0.89342023]]), 'M_0_to_3': array([[0.55850753, 0.13612552, 0.2499751 , 0.05539185],\n",
      "       [0.01791204, 0.82605787, 0.05907209, 0.096958  ],\n",
      "       [0.04534628, 0.008074  , 0.85149918, 0.09508054],\n",
      "       [0.00646156, 0.06355415, 0.02348564, 0.90649866]]), 'M_0_to_1': array([[0.75152183, 0.09827146, 0.07355421, 0.0766525 ],\n",
      "       [0.04888819, 0.86735324, 0.02989229, 0.05386628],\n",
      "       [0.05446854, 0.02450495, 0.75018255, 0.17084396],\n",
      "       [0.12039713, 0.03979299, 0.05715314, 0.78265675]]), 'M_1_to_4': array([[0.82835309, 0.03976936, 0.05146899, 0.08040856],\n",
      "       [0.02327998, 0.80873654, 0.06043788, 0.10754561],\n",
      "       [0.11544022, 0.13895831, 0.71027209, 0.03532938],\n",
      "       [0.09988558, 0.08065349, 0.01737091, 0.80209001]]), 'M_1_to_5': array([[0.78296278, 0.06942125, 0.08171685, 0.06589912],\n",
      "       [0.02166725, 0.83743552, 0.01932756, 0.12156967],\n",
      "       [0.11182788, 0.03378373, 0.78484422, 0.06954417],\n",
      "       [0.11063295, 0.0881492 , 0.0495941 , 0.75162375]])}\n"
     ]
    }
   ],
   "source": [
    "# # save the dictionary to a file\n",
    "# np.save('data.npy', theoretical_matrices)\n",
    "# # load the dictionary from the file\n",
    "# loaded_dict = np.load('data.npy', allow_pickle=True).item()\n",
    "# # print the loaded dictionary\n",
    "# print(loaded_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M_0_to_2\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix:\n",
      " [[0.86753892 0.05213542 0.06946348 0.01086218]\n",
      " [0.08065343 0.85671968 0.04982893 0.01279796]\n",
      " [0.11219846 0.17464684 0.65461785 0.05853685]\n",
      " [0.02775028 0.08494961 0.04088151 0.8464186 ]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_3\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix:\n",
      " [[0.75527498 0.04249873 0.00019955 0.20202674]\n",
      " [0.10446698 0.73349645 0.10777451 0.05426206]\n",
      " [0.009178   0.09167603 0.88105832 0.01808765]\n",
      " [0.10672648 0.08298123 0.01949473 0.79079756]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_1\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix:\n",
      " [[0.69009493 0.15321495 0.12518281 0.03150731]\n",
      " [0.03390489 0.89612499 0.03748613 0.032484  ]\n",
      " [0.13612944 0.00355591 0.86031238 0.00000227]\n",
      " [0.13829849 0.07169377 0.04358193 0.74642581]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_4\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix:\n",
      " [[0.79787321 0.09123409 0.0748348  0.0360579 ]\n",
      " [0.1525167  0.62670853 0.12700414 0.09377063]\n",
      " [0.06182208 0.09423139 0.82703856 0.01690797]\n",
      " [0.09434814 0.05487633 0.01727208 0.83350345]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_5\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix:\n",
      " [[0.77670263 0.03884329 0.11981174 0.06464234]\n",
      " [0.07306013 0.6964252  0.14381485 0.08669983]\n",
      " [0.0001472  0.02174493 0.93060026 0.04750762]\n",
      " [0.01290462 0.10677068 0.06826479 0.81205991]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23508015 0.26506847 0.26101933 0.23983305]\n"
     ]
    }
   ],
   "source": [
    "# Check the estimated parameters:\n",
    "for i in params.items():\n",
    "    print(i[0])\n",
    "    print(\"Edge:\", i[1].edge)\n",
    "    print(\"Transition matrix:\\n\", i[1].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", root_distr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix difference:\n",
      " [[0.09960865 0.07093256 0.00498621 0.03366231]\n",
      " [0.01576357 0.02703813 0.00892483 0.03387688]\n",
      " [0.00533819 0.03058066 0.01301418 0.01222828]\n",
      " [0.00504287 0.04282721 0.00457214 0.05244223]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix difference:\n",
      " [[0.00386991 0.01431933 0.0324696  0.04291902]\n",
      " [0.03240956 0.00438465 0.00578344 0.03101077]\n",
      " [0.00129948 0.00595081 0.00836875 0.00111846]\n",
      " [0.00368219 0.0201782  0.01427531 0.03077132]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix difference:\n",
      " [[0.00043271 0.04586535 0.01971483 0.02571781]\n",
      " [0.04628465 0.03270525 0.01225697 0.00132243]\n",
      " [0.00405161 0.02596747 0.04979691 0.01977783]\n",
      " [0.05635943 0.0032355  0.00691523 0.06003916]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix difference:\n",
      " [[0.02206907 0.00438435 0.02045318 0.00276846]\n",
      " [0.04114356 0.07985678 0.0152662  0.02344702]\n",
      " [0.01766089 0.02629588 0.04017144 0.00378534]\n",
      " [0.00126844 0.00706299 0.01874386 0.02707528]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix difference:\n",
      " [[0.04539034 0.01368587 0.03621623 0.00451176]\n",
      " [0.00302602 0.00613968 0.01219578 0.00908212]\n",
      " [0.00442834 0.01538392 0.00158094 0.0213932 ]\n",
      " [0.02699704 0.00888034 0.00551462 0.03036276]]\n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Compute the matrix difference between the theoretical and estimated parameters\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    print(\"Transition matrix difference:\\n\", np.abs(edges[i].transition_matrix - params['M_' + str(edges[i].edge[0].name.split(\"_\")[1]) + '_to_' + str(edges[i].edge[1].name.split(\"_\")[1])].transition_matrix))\n",
    "    print(\"---\"*30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Real pi\n",
      "[0.29440662 0.23200733 0.22880334 0.24478271]\n",
      "Real M\n",
      "M_0_to_2 [[0.75398651 0.06603219 0.14941675 0.03056456]\n",
      " [0.02974708 0.79252121 0.13562902 0.04210269]\n",
      " [0.17559149 0.04519857 0.71486412 0.06434582]\n",
      " [0.05052381 0.01284161 0.01744751 0.91918708]]\n",
      "M_0_to_3 [[9.52927492e-01 5.66850386e-03 1.82199061e-02 2.31840982e-02]\n",
      " [7.40205978e-03 9.83902535e-01 4.52640925e-04 8.24276385e-03]\n",
      " [2.08108476e-02 1.61602785e-02 9.46762453e-01 1.62664208e-02]\n",
      " [1.66087970e-02 1.06333037e-02 1.86824161e-03 9.70889658e-01]]\n",
      "M_0_to_1 [[0.97407016 0.00174372 0.00644578 0.01774034]\n",
      " [0.01050996 0.95666087 0.0230865  0.00974267]\n",
      " [0.01650874 0.02044945 0.9477316  0.01531021]\n",
      " [0.00563024 0.01018917 0.00933656 0.97484402]]\n",
      "M_1_to_4 [[0.83443031 0.03428078 0.02692816 0.10436075]\n",
      " [0.00937298 0.8344091  0.09065986 0.06555807]\n",
      " [0.09218889 0.11819048 0.71194926 0.07767136]\n",
      " [0.11701922 0.04809655 0.05908477 0.77579946]]\n",
      "M_1_to_5 [[0.96473858 0.01308674 0.0065822  0.01559248]\n",
      " [0.03343483 0.95352297 0.00621211 0.0068301 ]\n",
      " [0.0259327  0.01771879 0.95451401 0.0018345 ]\n",
      " [0.01380366 0.00409431 0.00459092 0.9775111 ]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Estimated M\n",
      "M_0_to_2 [[7.91587641e-01 4.43457502e-02 1.31369341e-01 3.26972673e-02]\n",
      " [4.33305732e-02 7.44446816e-01 1.72673510e-01 3.95491007e-02]\n",
      " [1.80631430e-01 4.09232249e-02 7.51000419e-01 2.74449264e-02]\n",
      " [9.24744467e-02 2.07520063e-41 1.60212342e-02 8.91504319e-01]]\n",
      "M_0_to_3 [[9.29840100e-01 7.79635745e-03 2.16251527e-02 4.07383899e-02]\n",
      " [1.33025376e-02 9.77946234e-01 1.55524126e-28 8.75122872e-03]\n",
      " [2.10214001e-02 3.31910232e-02 9.31691466e-01 1.40961110e-02]\n",
      " [1.71210998e-02 6.07759788e-51 7.32007652e-03 9.75558824e-01]]\n",
      "M_0_to_1 [[9.85347471e-01 2.95535820e-37 4.18269828e-30 1.46525290e-02]\n",
      " [2.57657902e-02 9.45379300e-01 2.85566714e-02 2.98238657e-04]\n",
      " [3.70513824e-14 1.58197391e-02 9.84180261e-01 3.02982334e-44]\n",
      " [6.90078286e-03 8.00286087e-03 9.82580378e-69 9.85096356e-01]]\n",
      "M_1_to_4 [[0.82626673 0.01515987 0.03356376 0.12500964]\n",
      " [0.02643604 0.82242768 0.09573112 0.05540517]\n",
      " [0.10559098 0.13293681 0.70272167 0.05875054]\n",
      " [0.09187696 0.06041985 0.03021267 0.81749053]]\n",
      "M_1_to_5 [[9.77304723e-001 2.26952774e-002 4.90546828e-049 8.59338875e-017]\n",
      " [9.30360295e-003 9.26110848e-001 3.15912451e-002 3.29943037e-002]\n",
      " [1.42363220e-002 1.45071710e-002 9.71256507e-001 8.23795404e-035]\n",
      " [1.59747482e-002 7.02719993e-047 5.50044495e-107 9.84025252e-001]]\n",
      "Estimated l\n",
      "M_0_to_2 0.978307511635952\n",
      "M_0_to_3 0.18901354998030023\n",
      "M_0_to_1 0.10223286266941137\n",
      "M_1_to_4 0.9795586289570805\n",
      "M_1_to_5 0.14375315377301118\n",
      "Estimated pi\n",
      "[0.26672125 0.24649383 0.22186212 0.2669268 ]\n",
      "Niter\n",
      "42\n",
      "TExec\n",
      "14.575841188430786\n"
     ]
    }
   ],
   "source": [
    "print(\"Real pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_random/real_root_distr.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Real M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_random/real_matrices.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"---\"*30)\n",
    "print(\"Estimated M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_random/RESULTS/repetition_98/M_estimation.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated l\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_random/RESULTS/repetition_98/estimated_branch_lengths.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_random/RESULTS/repetition_98/root_estimation.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Niter\")\n",
    "file = open(\"./OUTPUTS/EM_a15_b100_500_random/RESULTS/repetition_98/Niter.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)\n",
    "print(\"TExec\")\n",
    "file = open(\"./OUTPUTS/EM_a15_b100_500_random/RESULTS/repetition_98/TExec.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "debugger",
   "language": "python",
   "name": "debugger"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "03f790eb89590da6bbc9dab944ceea1669c4a69ff2bd32047ab9c0f81f755bd1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
