{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Expectation-Maximization algorithm** (hidden model)\n",
    "### Author: Martí Cortada Garcia\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import numpy as np\n",
    "from Bio import SeqIO, Phylo\n",
    "from io import StringIO\n",
    "from collections import Counter, OrderedDict\n",
    "import itertools\n",
    "import pandas as pd\n",
    "import math\n",
    "import copy\n",
    "\n",
    "# Import required functions from modules we have programmed\n",
    "from data_simulation import simulate"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Input:**\n",
    "We have a tree $T$ (where the root corresponds to the internal node codified as `Int_0`) and the alignments of the leaves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAGwCAYAAACHJU4LAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy88F64QAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAzqUlEQVR4nO3dfXQU9b3H8c+QkJDnEJ7yYAAN4UFMYsAmhopiQQlFJfYeUQ4PYtW0PUBFTKVcH5BWgVZE4sNVsTZBr5BG2si92IYiutAGUIzQBqEoAbsKASxKko1lA8ncP7zsYQVCNiTZX8j7dc4emJnfzHy/mSz7YWZ217Jt2xYAAICBuvi7AAAAgHMhqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGCvQ3wVciMbGRh08eFARERGyLMvf5QAAgGawbVu1tbWKj49Xly5NnzPp0EHl4MGDSkxM9HcZAACgBT777DNdcsklTY7p0EElIiJC0jeNRkZG+rkaAADQHDU1NUpMTPS8jjelQweVU5d7IiMjCSoAAHQwzbltg5tpAQCAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgA8BvbtpWbm6uYmBhZlqUdO3b4uyQAhiGoADgrh8Mhy7J07NixNttHaWmpCgsLtXbtWlVVVemKK65ocvyiRYv0ne98RxEREerdu7dycnK0Z8+eNqsPgP8RVAD4TWVlpeLi4jRixAjFxsYqMDCwyfEbN27UjBkztHXrVq1fv14nTpzQjTfeqLq6unaqGEB782tQ2bRpk26++WbFx8fLsiy9+eab/iwHgA/cbrfy8vKUkJCgsLAwZWZmyuFweJYfPXpUkyZNUkJCgkJDQ5WSkqJVq1Z5lk+fPl2zZs2S0+mUZVnq37//efdZWlqq6dOna+jQoUpLS1NhYaGcTqfKy8vboEMAJmj6vy9trK6uTmlpafrhD3+oH/zgB/4sxVMP0BmFhYX5vM7MmTO1a9cuFRUVKT4+XiUlJcrOzlZFRYWSk5N1/PhxDR8+XHPnzlVkZKTeeustTZ06VUlJScrIyFB+fr6SkpK0fPlybdu2TQEBAT7XUF1dLUmKiYnxeV0AHYNl27bt7yIkybIslZSUKCcn55xj3G633G63Z7qmpkaJiYmqrq5WZGRkq9QAdEZn+2fA4XDo+uuv11dffaXo6GivZU6nU5dddpmcTqfi4+M988eMGaOMjAwtXLjwrPu56aabNHjwYC1ZskSStGzZMi1btkyffvqpzzU3Njbqlltu0bFjx/TXv/7V5/UB+E9NTY2ioqKa9frt1zMqvlq0aJEWLFjQ6tutq6tTeHh4q28XuFhVVFSooaFBAwcO9JrvdrvVo0cPSVJDQ4MWLlyo4uJiHThwQPX19XK73QoNDW2VGmbMmKGdO3cSUoCLXIcKKvPmzdOcOXM806fOqLSmw4cPt+g0ONCZuFwuBQQEqLy8/IxLNqdC/5NPPqn8/HwtW7ZMKSkpCgsL0+zZs1VfX3/B+585c6bWrl2rTZs26ZJLLrng7QEwV4cKKsHBwQoODm7TfYSFhRFUgPNIT09XQ0ODjhw5opEjR551TFlZmSZMmKApU6ZI+uZSzccff6zLL7+8xfu1bVuzZs1SSUmJHA6HLr300hZvC0DH0KGCCoD2V1FRoYiICM+0ZVlKS0vT5MmTNW3aND311FNKT0/XF198oQ0bNig1NVXjx49XcnKyVq9erc2bN6t79+5aunSpDh8+fEFBZcaMGVq5cqXWrFmjiIgIHTp0SJIUFRWlkJCQC+4VgHkIKgCadO2113pNBwQE6OTJkyooKNDjjz+uBx54QAcOHFDPnj119dVX66abbpIkPfzww9q3b5/Gjh2r0NBQ5ebmKicnx/NOnZZ44YUXJEmjRo3yml9QUKDp06e3eLsAzOXXd/24XC7t3btX0jenkpcuXarrr79eMTEx6tu373nX9+Wu4aacfjOty+Xi0g8AAG2ow7zr54MPPtD111/vmT51o+ydd96pwsJCP1UFAABM4degMmrUqLN+fgOAzsfpdDZ5/8quXbuadaYVwMWFe1QAGCE+Pr7Jb08+/YPlAHQeBBUARggMDNSAAQP8XQYAw/DtyQAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAfmPbtnJzcxUTEyPLsrRjxw5/lwTAMAQVAGflcDhkWZaOHTvWZvsoLS1VYWGh1q5dq6qqKl1xxRVNjn/hhReUmpqqyMhIRUZGKisrS3/605/arD4A/ufXoLJo0SJ95zvfUUREhHr37q2cnBzt2bPHnyUBaEeVlZWKi4vTiBEjFBsbq8DAwCbHX3LJJVq8eLHKy8v1wQcf6Hvf+54mTJigjz76qJ0qBtDe/BpUNm7cqBkzZmjr1q1av369Tpw4oRtvvFF1dXX+LAtAM7jdbuXl5SkhIUFhYWHKzMyUw+HwLD969KgmTZqkhIQEhYaGKiUlRatWrfIsnz59umbNmiWn0ynLstS/f//z7vPmm2/W97//fSUnJ2vgwIF64oknFB4erq1bt7ZBhwBM0PR/X9pYaWmp13RhYaF69+6t8vJyXXvttX6piZCEzigsLMzndWbOnKldu3apqKhI8fHxKikpUXZ2tioqKpScnKzjx49r+PDhmjt3riIjI/XWW29p6tSpSkpKUkZGhvLz85WUlKTly5dr27ZtCggI8Gn/DQ0NeuONN1RXV6esrCyf6wfQMfg1qHxbdXW1JCkmJuasy91ut9xut2e6pqam1Wvo06dPq28TMJ1t2z6NdzqdKigokNPpVHx8vCQpLy9PpaWlKigo0MKFC5WQkKC8vDzPOrNmzdK6detUXFysjIwMRUVFKSIiQgEBAYqNjW32visqKpSVlaXjx48rPDxcJSUluvzyy32qH0DHYUxQaWxs1OzZs/Xd7373nDfULVq0SAsWLGjnygB8W0VFhRoaGjRw4ECv+W63Wz169JD0zRmPhQsXqri4WAcOHFB9fb3cbrdCQ0MvaN+DBg3Sjh07VF1drdWrV+vOO+/Uxo0bCSvARcqYoDJjxgzt3LlTf/3rX885Zt68eZozZ45nuqamRomJia1ax+HDh1t0GhzoTFwulwICAlReXn7GJZvw8HBJ0pNPPqn8/HwtW7ZMKSkpCgsL0+zZs1VfX39B+w4KCtKAAQMkScOHD9e2bduUn5+vl1566YK2C8BMRgSVmTNnau3atdq0aZMuueSSc44LDg5WcHBwm9YSFhZGUAHOIz09XQ0NDTpy5IhGjhx51jFlZWWaMGGCpkyZIumbs6Yff/xxq5/5aGxs9LokDODi4tegYtu2Zs2apZKSEjkcDl166aX+LAfAWVRUVCgiIsIzbVmW0tLSNHnyZE2bNk1PPfWU0tPT9cUXX2jDhg1KTU3V+PHjlZycrNWrV2vz5s3q3r27li5dqsOHD19QUJk3b57GjRunvn37qra2VitXrpTD4dC6detao1UABvJrUJkxY4ZWrlypNWvWKCIiQocOHZIkRUVFKSQkxJ+lAfh/334HXkBAgE6ePKmCggI9/vjjeuCBB3TgwAH17NlTV199tW666SZJ0sMPP6x9+/Zp7NixCg0NVW5urnJycjw3zbfEkSNHNG3aNFVVVSkqKkqpqalat26dbrjhhgvqEYC5LNvX2/1bc+eWddb5BQUFmj59+nnXr6mpUVRUlKqrqxUZGdniOurq6jzX1V0uF5d+AABoQ768fvv90g8AAMC58F0/AIzgdDoVHh5+zofT6fR3iQD8wIh3/QBAfHx8k9+efOqD5QB0LgQVAEYIDAz0fD4KAJzCpR8AAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqJyHw+GQZVk6duxYs9eZPn26cnJyfNrPl19+qcmTJysyMlLR0dG6++675XK5fCsWAICLDEHFEJMnT9ZHH32k9evXe75JOjc3199lAQDgVwQVHxUWFio6Olrr1q3TkCFDFB4eruzsbFVVVUmSHnvsMa1YsUJr1qyRZVmyLEsOh6PJbe7evVulpaX6zW9+o8zMTF1zzTV69tlnVVRUpIMHD7ZDVwAAmImg0gJff/21lixZotdee02bNm2S0+lUXl6eJCkvL08TJ070hJeqqiqNGDGiye1t2bJF0dHRuuqqqzzzxowZoy5duui9995r014AADAZn0zbAidOnNCLL76opKQkSdLMmTP1i1/8QpIUHh6ukJAQud1uxcbGNmt7hw4dUu/evb3mBQYGKiYmRocOHWrd4gEA6EA4o9ICoaGhnpAiSXFxcTpy5IgfKwIA4OJEUGmBrl27ek1bliXbtlu8vdjY2DOCzsmTJ/Xll182+6wMAAAXI4JKGwgKClJDQ0Ozx2dlZenYsWMqLy/3zHvnnXfU2NiozMzMtigRAIAOgaDSBvr376+///3v2rNnj/71r3/pxIkTTY4fMmSIsrOzde+99+r9999XWVmZZs6cqTvuuIOvtgcAdGoElTZw7733atCgQbrqqqvUq1cvlZWVnXed119/XYMHD9bo0aP1/e9/X9dcc42WL1/eDtUCAGAuy76Qmyv8rKamRlFRUaqurlZkZGSLt1NXV6fw8HBJksvlUlhYWGuVCAAAvsWX12/OqAAAAGMRVNrBwoULFR4eftbHuHHj/F0eAADGatEHvn3wwQcqLi6W0+lUfX2917I//OEPrVLYxeTHP/6xJk6ceNZlISEh7VwNAAAdh89BpaioSNOmTdPYsWP15z//WTfeeKM+/vhjHT58WLfeemtb1NjhxcTEKCYmxt9lAADQ4fh86WfhwoV6+umn9b//+78KCgpSfn6+/vGPf2jixInq27dvW9QIAAA6KZ+DSmVlpcaPHy/pmw82q6urk2VZuv/++3k7LQAAaFU+B5Xu3burtrZWkpSQkKCdO3dKko4dO6avv/66dasDAACdms/3qFx77bVav369UlJSdNttt+m+++7TO++8o/Xr12v06NFtUSMAAOikfA4qzz33nI4fPy5Jeuihh9S1a1dt3rxZ//Ef/6GHH3641QsEAACdl89B5fR3r3Tp0kU///nPW7UgAACAU3y+R6WwsPCs80+ePKl58+ZdaD0AAAAePgeVn/70p7rtttv01Vdfeebt2bNHmZmZWrVqVasWBwAAOjefg8r27dv1+eefKyUlRevXr9fzzz+vYcOGafDgwfrb3/7WFjUCAIBOyud7VJKSklRWVqbZs2crOztbAQEBWrFihSZNmtQW9QEAgE6sRV9K+NZbb6moqEhZWVmKjo7WK6+8ooMHD7Z2bQAAoJPzOaj86Ec/0m233aa5c+fqL3/5i/7+978rKChIKSkpKi4ubosaAQBAJ+XzpZ+ysjK99957SktLkyTFxsbqj3/8o55//nn98Ic/POe3BAMAAPjK56BSXl6u4ODgM+bPmDFDY8aMaZWiAAAApBZc+jlbSDll0KBBF1QMAADA6Xw+oyJJq1evVnFxsZxOp+rr672Wffjhh61SGAAAgM9nVJ555hnddddd6tOnj7Zv366MjAz16NFD+/bt07hx49qiRgAA0En5HFT+67/+S8uXL9ezzz6roKAgPfjgg1q/fr1++tOfqrq6ui1qBAAAnZTPQcXpdGrEiBGSpJCQENXW1kqSpk6dykfoAwCAVuVzUImNjdWXX34pSerbt6+2bt0qSdq/f79s227d6gAAQKfmc1D53ve+p//5n/+RJN111126//77dcMNN+j222/Xrbfe2uoFAgCAzsvnd/089NBDSkhIkPTNZ6f06NFDmzdv1i233KLs7OxWLxAAAHRelu3j9ZqAgABVVVWpd+/eXvOPHj2q3r17q6GhoVULbEpNTY2ioqJUXV2tyMjIFm+nrq5O4eHhkiSXy6WwsLDWKhEAAHyLL6/fPl/6OVeucblc6tatm6+bAwAAOKdmX/qZM2eOJMmyLD366KMKDQ31LGtoaNB7772nK6+8stULBAAAnVezg8r27dslfXNGpaKiQkFBQZ5lQUFBSktLU15eXutXCAAAOq1mB5V3331X0jfv9MnPz7+ge0IAAACaw+d3/RQUFLRFHQAAAGfw+WZaAACA9kJQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAQDuzbVu5ubmKiYmRZVnasWOHv0syFkEFAIDTOBwOWZalY8eOtdk+SktLVVhYqLVr16qqqkpXXHFFs9ddvHixLMvS7Nmz26w+k/g1qLzwwgtKTU1VZGSkIiMjlZWVpT/96U/+LAkAgDZXWVmpuLg4jRgxQrGxsQoMDGzWetu2bdNLL72k1NTUNq7QHH4NKpdccokWL16s8vJyffDBB/re976nCRMm6KOPPvJnWQAAnJPb7VZeXp4SEhIUFhamzMxMORwOz/KjR49q0qRJSkhIUGhoqFJSUrRq1SrP8unTp2vWrFlyOp2yLEv9+/dv1n5dLpcmT56sl19+Wd27d2/lrszVvAjXRm6++Wav6SeeeEIvvPCCtm7dqqFDh/qpKqDzqaur83cJgF+EhYX5vM7MmTO1a9cuFRUVKT4+XiUlJcrOzlZFRYWSk5N1/PhxDR8+XHPnzlVkZKTeeustTZ06VUlJScrIyFB+fr6SkpK0fPlybdu2TQEBAc3a74wZMzR+/HiNGTNGjz/+uM91d1R+DSqna2ho0BtvvKG6ujplZWWddYzb7Zbb7fZM19TUtFd5wEUtPDzc3yUAfmHbtk/jnU6nCgoK5HQ6FR8fL0nKy8tTaWmpCgoKtHDhQiUkJCgvL8+zzqxZs7Ru3ToVFxcrIyNDUVFRioiIUEBAgGJjY5u136KiIn344Yfatm2bT/VeDPweVCoqKpSVlaXjx48rPDxcJSUluvzyy886dtGiRVqwYEE7VwhcvOrq6ggpgA8qKirU0NCggQMHes13u93q0aOHpG/+471w4UIVFxfrwIEDqq+vl9vtVmhoaIv2+dlnn+m+++7T+vXr1a1btwvuoaPxe1AZNGiQduzYoerqaq1evVp33nmnNm7ceNawMm/ePM2ZM8czXVNTo8TExPYsF7hoHT58uEWnwYHOxOVyKSAgQOXl5WdcsjkV+p988knl5+dr2bJlSklJUVhYmGbPnq36+voW7bO8vFxHjhzRsGHDPPMaGhq0adMmPffcc3K73c2+fNQR+T2oBAUFacCAAZKk4cOHa9u2bcrPz9dLL710xtjg4GAFBwe3d4lApxAWFkZQAc4jPT1dDQ0NOnLkiEaOHHnWMWVlZZowYYKmTJkiSWpsbNTHH398zqsF5zN69GhVVFR4zbvrrrs0ePBgzZ0796IOKZIBQeXbGhsbve5DAQDAHyoqKhQREeGZtixLaWlpmjx5sqZNm6annnpK6enp+uKLL7RhwwalpqZq/PjxSk5O1urVq7V582Z1795dS5cu1eHDh1scVCIiIs74nJWwsDD16NHDp89f6aj8GlTmzZuncePGqW/fvqqtrdXKlSvlcDi0bt06f5YFAICuvfZar+mAgACdPHlSBQUFevzxx/XAAw/owIED6tmzp66++mrddNNNkqSHH35Y+/bt09ixYxUaGqrc3Fzl5OSourraH210eJbt6y3Prejuu+/Whg0bVFVVpaioKKWmpmru3Lm64YYbmrV+TU2NoqKiVF1drcjIyBbXcfoNhS6Xi9Pf6DT43QfgD768fvv1jMorr7ziz90DAADD8V0/AAD4kdPpVHh4+DkfTqfT3yX6lXE30wIA0JnEx8c3+e3Jpz5YrrMiqAAA4EeBgYGej+nAmbj0AwAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAXBWDodDlmXp2LFjzV5n+vTpysnJ8Wk/TzzxhEaMGKHQ0FBFR0f7tC6Aix9BBYBf1dfX67bbbtNPfvITf5cCwEAEFQDNUlhYqOjoaK1bt05DhgxReHi4srOzVVVVJUl67LHHtGLFCq1Zs0aWZcmyLDkcjvNud8GCBbr//vuVkpLSxh0A6IgC/V2Aaerq6vxdAtBufP19//rrr7VkyRK99tpr6tKli6ZMmaK8vDy9/vrrysvL0+7du1VTU6OCggJJUkxMTFuUDaATIah8S58+ffxdAmCsEydO6MUXX1RSUpIkaebMmfrFL34hSQoPD1dISIjcbrdiY2P9WSaAiwiXfiSFhYXJtm1/lwEYLzQ01BNSJCkuLk5HjhzxY0UALnacUTmNy+XydwlAu6qrq/PpLGLXrl29pi3LIuQDaFMEldOEhYX5uwSgQwsKClJDQ4O/ywBwESGoAGg1/fv317p167Rnzx716NFDUVFRZ5yF+Tan06kvv/xSTqdTDQ0N2rFjhyRpwIABCg8Pb4eqAZiMe1QAtJp7771XgwYN0lVXXaVevXqprKzsvOs8+uijSk9P1/z58+VyuZSenq709HR98MEH7VAxANNZdge+wFxTU6OoqChVV1crMjLS3+UAHU5dXZ3nrIXL5eLyJ4B24cvrN2dUAACAsQgqANrMwoULFR4eftbHuHHj/F0egA6Am2kBtJkf//jHmjhx4lmXhYSEtHM1ADoiggqANhMTE8PH6AO4IFz6AQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQCgndm2rdzcXMXExMiyLO3YscPfJRmLoAIAwGkcDocsy9KxY8fabB+lpaUqLCzU2rVrVVVVpSuuuKLJ8Y899pgsy/J6DB48uM3qM0mgvwsAAKCzqaysVFxcnEaMGNHsdYYOHaq3337bMx0Y2Dlewo05o7J48WJZlqXZs2f7uxQAAM7J7XYrLy9PCQkJCgsLU2ZmphwOh2f50aNHNWnSJCUkJCg0NFQpKSlatWqVZ/n06dM1a9YsOZ1OWZal/v37N2u/gYGBio2N9Tx69uzZyp2ZyYg4tm3bNr300ktKTU31dylAp1VXV+fvEoB2FxYW5vM6M2fO1K5du1RUVKT4+HiVlJQoOztbFRUVSk5O1vHjxzV8+HDNnTtXkZGReuuttzR16lQlJSUpIyND+fn5SkpK0vLly7Vt2zYFBAQ0a7+ffPKJ4uPj1a1bN2VlZWnRokXq27evz/V3NH4PKi6XS5MnT9bLL7+sxx9/vMmxbrdbbrfbM11TU9PW5QGdRp8+ffxdAtDubNv2abzT6VRBQYGcTqfi4+MlSXl5eSotLVVBQYEWLlyohIQE5eXledaZNWuW1q1bp+LiYmVkZCgqKkoREREKCAhQbGxss/abmZmpwsJCDRo0SFVVVVqwYIFGjhypnTt3KiIiwqceOhq/B5UZM2Zo/PjxGjNmzHmDyqJFi7RgwYJ2qgy4+IWFhcm2bVmW5e9SgA6hoqJCDQ0NGjhwoNd8t9utHj16SJIaGhq0cOFCFRcX68CBA6qvr5fb7VZoaGiL9ztu3DjP31NTU5WZmal+/fqpuLhYd999d4u32xH4NagUFRXpww8/1LZt25o1ft68eZozZ45nuqamRomJiW1VHtBpuFwuf5cAdAgul0sBAQEqLy8/45JNeHi4JOnJJ59Ufn6+li1bppSUFIWFhWn27Nmqr69vtTqio6M1cOBA7d27t9W2aSq/BZXPPvtM9913n9avX69u3bo1a53g4GAFBwe3cWVA59OS6/RAZ5Senq6GhgYdOXJEI0eOPOuYsrIyTZgwQVOmTJEkNTY26uOPP9bll1/eanW4XC5VVlZq6tSprbZNU/ktqJSXl+vIkSMaNmyYZ15DQ4M2bdqk5557Tm63u9k3GAEA0NoqKiq87v+wLEtpaWmaPHmypk2bpqeeekrp6en64osvtGHDBqWmpmr8+PFKTk7W6tWrtXnzZnXv3l1Lly7V4cOHLyio5OXl6eabb1a/fv108OBBzZ8/XwEBAZo0aVJrtGo0vwWV0aNHq6KiwmveXXfdpcGDB2vu3LmEFACAX1177bVe0wEBATp58qQKCgr0+OOP64EHHtCBAwfUs2dPXX311brpppskSQ8//LD27dunsWPHKjQ0VLm5ucrJyVF1dXWLa/n88881adIkHT16VL169dI111yjrVu3qlevXhfUY0dg2b7e8tyGRo0apSuvvFLLli1r1viamhpFRUWpurpakZGRbVscAABoFb68fhvzgW8AAADf5ve3J5/u9E/2AwCgM3A6nU3ev7Jr165O8cFu52JUUAEAoLOJj49v8tuTT32wXGdFUAEAwI8CAwM1YMAAf5dhLO5RAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwFkEFAAAYi6ACAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABiLoAIAAIxFUAEAAMYiqAAAAGMRVAAAgLEIKgAAwFgEFQAAYCyCCgAAMBZBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKAAAwVqC/C7gQtm1LkmpqavxcCQAAaK5Tr9unXseb0qGDSm1trSQpMTHRz5UAAABf1dbWKioqqskxlt2cOGOoxsZGHTx4UBEREbIs64K2VVNTo8TERH322WeKjIxspQo7js7cf2fuXaL/ztx/Z+5d6tz9+7t327ZVW1ur+Ph4denS9F0oHfqMSpcuXXTJJZe06jYjIyM73S/s6Tpz/525d4n+O3P/nbl3qXP378/ez3cm5RRupgUAAMYiqAAAAGMRVP5fcHCw5s+fr+DgYH+X4heduf/O3LtE/525/87cu9S5++9IvXfom2kBAMDFjTMqAADAWAQVAABgLIIKAAAwFkEFAAAY66IOKs8//7z69++vbt26KTMzU++//36T49944w0NHjxY3bp1U0pKiv74xz96LbdtW48++qji4uIUEhKiMWPG6JNPPmnLFlrMl95ffvlljRw5Ut27d1f37t01ZsyYM8ZPnz5dlmV5PbKzs9u6jRbzpf/CwsIzeuvWrZvXmI507CXf+h81atQZ/VuWpfHjx3vGdJTjv2nTJt18882Kj4+XZVl68803z7uOw+HQsGHDFBwcrAEDBqiwsPCMMb7+W+IPvvb+hz/8QTfccIN69eqlyMhIZWVlad26dV5jHnvssTOO++DBg9uwi5bztX+Hw3HW3/tDhw55jesIx17yvf+zPacty9LQoUM9Y0w5/hdtUPnd736nOXPmaP78+frwww+VlpamsWPH6siRI2cdv3nzZk2aNEl33323tm/frpycHOXk5Gjnzp2eMb/+9a/1zDPP6MUXX9R7772nsLAwjR07VsePH2+vtprF194dDocmTZqkd999V1u2bFFiYqJuvPFGHThwwGtcdna2qqqqPI9Vq1a1Rzs+87V/6ZtPZzy9t3/+859eyzvKsZd87/8Pf/iDV+87d+5UQECAbrvtNq9xHeH419XVKS0tTc8//3yzxu/fv1/jx4/X9ddfrx07dmj27Nm65557vF6wW/L75A++9r5p0ybdcMMN+uMf/6jy8nJdf/31uvnmm7V9+3avcUOHDvU67n/961/bovwL5mv/p+zZs8erv969e3uWdZRjL/nef35+vlffn332mWJiYs543htx/O2LVEZGhj1jxgzPdENDgx0fH28vWrTorOMnTpxojx8/3mteZmam/aMf/ci2bdtubGy0Y2Nj7SeffNKz/NixY3ZwcLC9atWqNuig5Xzt/dtOnjxpR0RE2CtWrPDMu/POO+0JEya0dqltwtf+CwoK7KioqHNuryMde9u+8OP/9NNP2xEREbbL5fLM60jH/xRJdklJSZNjHnzwQXvo0KFe826//XZ77NixnukL/Xn6Q3N6P5vLL7/cXrBggWd6/vz5dlpaWusV1k6a0/+7775rS7K/+uqrc47piMfetlt2/EtKSmzLsuxPP/3UM8+U439RnlGpr69XeXm5xowZ45nXpUsXjRkzRlu2bDnrOlu2bPEaL0ljx471jN+/f78OHTrkNSYqKkqZmZnn3KY/tKT3b/v666914sQJxcTEeM13OBzq3bu3Bg0apJ/85Cc6evRoq9beGlrav8vlUr9+/ZSYmKgJEyboo48+8izrKMdeap3j/8orr+iOO+5QWFiY1/yOcPx9db7nfWv8PDuKxsZG1dbWnvG8/+STTxQfH6/LLrtMkydPltPp9FOFbePKK69UXFycbrjhBpWVlXnmd6ZjL33zvB8zZoz69evnNd+E439RBpV//etfamhoUJ8+fbzm9+nT54zrj6ccOnSoyfGn/vRlm/7Qkt6/be7cuYqPj/d6gmZnZ+vVV1/Vhg0b9Ktf/UobN27UuHHj1NDQ0Kr1X6iW9D9o0CD99re/1Zo1a/Tf//3famxs1IgRI/T5559L6jjHXrrw4//+++9r586duueee7zmd5Tj76tzPe9ramr073//u1WeTx3FkiVL5HK5NHHiRM+8zMxMFRYWqrS0VC+88IL279+vkSNHqra21o+Vto64uDi9+OKL+v3vf6/f//73SkxM1KhRo/Thhx9Kap1/SzuKgwcP6k9/+tMZz3tTjn+H/vZktL7FixerqKhIDofD64bSO+64w/P3lJQUpaamKikpSQ6HQ6NHj/ZHqa0mKytLWVlZnukRI0ZoyJAheumll/TLX/7Sj5W1v1deeUUpKSnKyMjwmn8xH39IK1eu1IIFC7RmzRqvezTGjRvn+XtqaqoyMzPVr18/FRcX6+677/ZHqa1m0KBBGjRokGd6xIgRqqys1NNPP63XXnvNj5W1vxUrVig6Olo5OTle8005/hflGZWePXsqICBAhw8f9pp/+PBhxcbGnnWd2NjYJsef+tOXbfpDS3o/ZcmSJVq8eLH+/Oc/KzU1tcmxl112mXr27Km9e/decM2t6UL6P6Vr165KT0/39NZRjr10Yf3X1dWpqKioWf8AmXr8fXWu531kZKRCQkJa5ffJdEVFRbrnnntUXFx8xmWwb4uOjtbAgQM7/HE/l4yMDE9vneHYS9+8o/G3v/2tpk6dqqCgoCbH+uv4X5RBJSgoSMOHD9eGDRs88xobG7Vhwwav/zmfLisry2u8JK1fv94z/tJLL1VsbKzXmJqaGr333nvn3KY/tKR36Zt3tfzyl79UaWmprrrqqvPu5/PPP9fRo0cVFxfXKnW3lpb2f7qGhgZVVFR4eusox166sP7feOMNud1uTZky5bz7MfX4++p8z/vW+H0y2apVq3TXXXdp1apVXm9HPxeXy6XKysoOf9zPZceOHZ7eLvZjf8rGjRu1d+/eZv0HxW/H399387aVoqIiOzg42C4sLLR37dpl5+bm2tHR0fahQ4ds27btqVOn2j//+c8948vKyuzAwEB7yZIl9u7du+358+fbXbt2tSsqKjxjFi9ebEdHR9tr1qyx//73v9sTJkywL730Uvvf//53u/fXFF97X7x4sR0UFGSvXr3arqqq8jxqa2tt27bt2tpaOy8vz96yZYu9f/9+++2337aHDRtmJycn28ePH/dLj03xtf8FCxbY69atsysrK+3y8nL7jjvusLt162Z/9NFHnjEd5djbtu/9n3LNNdfYt99++xnzO9Lxr62ttbdv325v377dlmQvXbrU3r59u/3Pf/7Ttm3b/vnPf25PnTrVM37fvn12aGio/bOf/czevXu3/fzzz9sBAQF2aWmpZ8z5fp6m8LX3119/3Q4MDLSff/55r+f9sWPHPGMeeOAB2+Fw2Pv377fLysrsMWPG2D179rSPHDnS7v2dj6/9P/300/abb75pf/LJJ3ZFRYV933332V26dLHffvttz5iOcuxt2/f+T5kyZYqdmZl51m2acvwv2qBi27b97LPP2n379rWDgoLsjIwMe+vWrZ5l1113nX3nnXd6jS8uLrYHDhxoBwUF2UOHDrXfeustr+WNjY32I488Yvfp08cODg62R48ebe/Zs6c9WvGZL73369fPlnTGY/78+bZt2/bXX39t33jjjXavXr3srl272v369bPvvfdeI5+sp/jS/+zZsz1j+/TpY3//+9+3P/zwQ6/tdaRjb9u+/+7/4x//sCXZf/7zn8/YVkc6/qfecvrtx6l+77zzTvu66647Y50rr7zSDgoKsi+77DK7oKDgjO029fM0ha+9X3fddU2Ot+1v3qodFxdnBwUF2QkJCfbtt99u7927t30bayZf+//Vr35lJyUl2d26dbNjYmLsUaNG2e+8884Z2+0Ix962W/a7f+zYMTskJMRevnz5WbdpyvG3bNu22/ikDQAAQItclPeoAACAiwNBBQAAGIugAgAAjEVQAQAAxiKoAAAAYxFUAACAsQgqAADAWAQVAABgLIIKcJEbNWqUZs+e7e8yzurTTz+VZVnasWOHT+tZlqU333yzTWry1WOPPaYrr7zS32UAFy2CCgA0k0kBCegsCCoAzlBfX+/vEgBAEkEF6BROnjypmTNnKioqSj179tQjjzyi07/mq3///vrlL3+padOmKTIyUrm5uZKkuXPnauDAgQoNDdVll12mRx55RCdOnPCsd+qyx2uvvab+/fsrKipKd9xxh2praz1jGhsb9etf/1oDBgxQcHCw+vbtqyeeeMKrvn379un6669XaGio0tLStGXLFp/6++yzzzRx4kRFR0crJiZGEyZM0KeffupZPn36dOXk5GjJkiWKi4tTjx49NGPGDK9eqqqqNH78eIWEhOjSSy/VypUr1b9/fy1btszzM5KkW2+9VZZleaZPaepnAKDlCCpAJ7BixQoFBgbq/fffV35+vpYuXarf/OY3XmOWLFmitLQ0bd++XY888ogkKSIiQoWFhdq1a5fy8/P18ssv6+mnn/Zar7KyUm+++abWrl2rtWvXauPGjVq8eLFn+bx587R48WI98sgj2rVrl1auXKk+ffp4beOhhx5SXl6eduzYoYEDB2rSpEk6efJks3o7ceKExo4dq4iICP3lL39RWVmZwsPDlZ2d7XVm6N1331VlZaXeffddrVixQoWFhSosLPQsnzZtmg4ePCiHw6Hf//73Wr58uY4cOeJZvm3bNklSQUGBqqqqPNPN+RkAuADt/n3NANrVddddZw8ZMsRubGz0zJs7d649ZMgQz3S/fv3snJyc827rySeftIcPH+6Znj9/vh0aGmrX1NR45v3sZz+zMzMzbdu27ZqaGjs4ONh++eWXz7q9/fv325Ls3/zmN555H330kS3J3r179znrkGSXlJTYtm3br732mj1o0CCv/txutx0SEmKvW7fOtu1vvuK+X79+9smTJz1jbrvtNvv222+3bdu2d+/ebUuyt23b5ln+ySef2JLsp59++qz7be7PAMCF4YwK0AlcffXVsizLM52VlaVPPvlEDQ0NnnlXXXXVGev97ne/03e/+13FxsYqPDxcDz/8sJxOp9eY/v37KyIiwjMdFxfnOROxe/duud1ujR49usn6UlNTvdaX5HU2oyl/+9vftHfvXkVERCg8PFzh4eGKiYnR8ePHVVlZ6Rk3dOhQBQQEnLXOPXv2KDAwUMOGDfMsHzBggLp3796sGpr6GQC4MIH+LgCAGcLCwrymt2zZosmTJ2vBggUaO3asoqKiVFRUpKeeesprXNeuXb2mLctSY2OjJCkkJKRZ+z59G6cC1altnI/L5dLw4cP1+uuvn7GsV69ezarzQrXltoHOjqACdALvvfee1/TWrVuVnJzsdYbh2zZv3qx+/frpoYce8sz75z//6dN+k5OTFRISog0bNuiee+7xrehmGjZsmH73u9+pd+/eioyMbNE2Bg0apJMnT2r79u0aPny4JGnv3r366quvvMZ17drV6ywUgLbHpR+gE3A6nZozZ4727NmjVatW6dlnn9V9993X5DrJyclyOp0qKipSZWWlnnnmGZWUlPi0327dumnu3Ll68MEH9eqrr6qyslJbt27VK6+8ciHteJk8ebJ69uypCRMm6C9/+Yv2798vh8Ohn/70p/r888+btY3BgwdrzJgxys3N1fvvv6/t27crNzdXISEhXpfM+vfvrw0bNujQoUNnhBgAbYOgAnQC06ZN07///W9lZGRoxowZuu+++zxvQT6XW265Rffff79mzpypK6+8Ups3b/a8G8gXjzzyiB544AE9+uijGjJkiG6//fZWvX8jNDRUmzZtUt++ffWDH/xAQ4YM0d13363jx4/7dIbl1VdfVZ8+fXTttdfq1ltv1b333quIiAh169bNM+app57S+vXrlZiYqPT09FbrAcC5WbZ92ocpAAAkSZ9//rkSExP19ttvn/dmYABth6ACAJLeeecduVwupaSkqKqqSg8++KAOHDigjz/++IybZQG0H26mBQB988Fx//mf/6l9+/YpIiJCI0aM0Ouvv05IAfyMMyoAAMBY3EwLAACMRVABAADGIqgAAABjEVQAAICxCCoAAMBYBBUAAGAsggoAADAWQQUAABjr/wDdrlLBA7YYDwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "newick_tree = open(\"./newick.txt\", \"r\")\n",
    "newick = newick_tree.read()\n",
    "newick_tree.close()\n",
    "tree = Phylo.read(StringIO(newick), \"newick\")\n",
    "\n",
    "# Change nodes names\n",
    "for idx, clade in enumerate(tree.get_nonterminals()):\n",
    "    clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "\n",
    "# Change leaves names\n",
    "for idx, clade in enumerate(tree.get_terminals()):\n",
    "    clade.name = \"Leaf_\" + clade.name \n",
    "\n",
    "Phylo.draw(tree)\n",
    "\n",
    "net = Phylo.to_networkx(tree) # to graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "simulate() missing 1 required positional argument: 'directory'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m edges, node_distr, filename \u001b[39m=\u001b[39m simulate(net, \u001b[39m500\u001b[39;49m) \u001b[39m# simulate data calling the module data_simulation.py \u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[39m# Preprocess leave alignments\u001b[39;00m\n\u001b[1;32m      4\u001b[0m leave_alignments \u001b[39m=\u001b[39m [i \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m SeqIO\u001b[39m.\u001b[39mparse(filename, \u001b[39m\"\u001b[39m\u001b[39mfasta\u001b[39m\u001b[39m\"\u001b[39m)]\n",
      "\u001b[0;31mTypeError\u001b[0m: simulate() missing 1 required positional argument: 'directory'"
     ]
    }
   ],
   "source": [
    "edges, node_distr, filename = simulate(net, 500) # simulate data calling the module data_simulation.py \n",
    "\n",
    "# Preprocess leave alignments\n",
    "leave_alignments = [i for i in SeqIO.parse(filename, \"fasta\")]\n",
    "for i in leave_alignments:\n",
    "    print(\"Seq>\", i.seq)\n",
    "# Codify the alphabet {A,G,C,T} -> {0,1,2,3}\n",
    "leave_sequences = []\n",
    "for i in leave_alignments:\n",
    "    seq = \"\"\n",
    "    for j in i.seq:\n",
    "        if j == \"A\":\n",
    "            seq += \"0\"\n",
    "        elif j == \"G\":\n",
    "            seq += \"1\"\n",
    "        elif j == \"C\":\n",
    "            seq += \"2\"\n",
    "        elif j == \"T\":\n",
    "            seq += \"3\"\n",
    "        else:\n",
    "            raise ValueError(\"Invalid nucleotide\")\n",
    "    leave_sequences.append(seq)\n",
    "#leave_sequences\n",
    "edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## **EM algorithm**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from evaluate import init_EM\n",
    "\n",
    "Results = init_EM(newick, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Source: Int_0\n",
      "Target: Leaf_2\n",
      "Matrix:\n",
      " [[0.80529814 0.06095358 0.09251042 0.04123786]\n",
      " [0.05611119 0.71483627 0.11102639 0.11802615]\n",
      " [0.0952762  0.07415636 0.731184   0.09938344]\n",
      " [0.02196288 0.0820518  0.08234946 0.81363586]]\n",
      "Source: Int_0\n",
      "Target: Leaf_3\n",
      "Matrix:\n",
      " [[0.63632807 0.06782976 0.1446222  0.15121997]\n",
      " [0.12810897 0.67208297 0.04706075 0.15274731]\n",
      " [0.10204507 0.09021686 0.74449869 0.06323937]\n",
      " [0.06570111 0.08178298 0.05155947 0.80095644]]\n",
      "Source: Int_0\n",
      "Target: Int_1\n",
      "Matrix:\n",
      " [[0.72577892 0.07421689 0.0831795  0.11682468]\n",
      " [0.08700214 0.7261134  0.06686419 0.12002027]\n",
      " [0.09519544 0.182137   0.68012648 0.04254107]\n",
      " [0.06761601 0.06509887 0.08961157 0.77767356]]\n",
      "Source: Int_1\n",
      "Target: Leaf_4\n",
      "Matrix:\n",
      " [[0.85639817 0.07234946 0.04036045 0.03089192]\n",
      " [0.0298384  0.73128255 0.0944944  0.14438465]\n",
      " [0.03533407 0.10420379 0.77836777 0.08209437]\n",
      " [0.06403097 0.03195277 0.06966545 0.8343508 ]]\n",
      "Source: Int_1\n",
      "Target: Leaf_5\n",
      "Matrix:\n",
      " [[0.80075352 0.06438893 0.11198018 0.02287738]\n",
      " [0.07400816 0.72459648 0.10712499 0.09427037]\n",
      " [0.14358855 0.08350079 0.74500083 0.02790983]\n",
      " [0.01210798 0.14166184 0.06442053 0.78180965]]\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(Results)):\n",
    "    print(\"Source:\", Results[i].source)\n",
    "    print(\"Target:\", Results[i].target)\n",
    "    print(\"Matrix:\\n\", Results[i].matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "l -73482.53888183684\n",
      "73482.53888183684\n",
      "--------\n",
      "2803.2235244028125\n",
      "--------\n",
      "403.42311980716477\n",
      "--------\n",
      "147.34679458629398\n",
      "--------\n",
      "65.81431201811938\n",
      "--------\n",
      "32.62458047835389\n",
      "--------\n",
      "17.4679848113592\n",
      "--------\n",
      "9.96598948755127\n",
      "--------\n",
      "5.979527266841615\n",
      "--------\n",
      "3.724567122422741\n",
      "--------\n",
      "2.382098244605004\n",
      "--------\n",
      "1.5508962869207608\n",
      "--------\n",
      "1.021128828782821\n",
      "--------\n",
      "0.6763189632765716\n",
      "--------\n",
      "0.44849936086393427\n",
      "--------\n",
      "0.29640033769828733\n",
      "--------\n",
      "0.19417478849936742\n",
      "--------\n",
      "0.12524008554464672\n",
      "--------\n",
      "0.07875408534891903\n",
      "--------\n",
      "0.047520216307020746\n",
      "--------\n",
      "0.02670328039675951\n",
      "--------\n",
      "0.01302253625181038\n",
      "--------\n",
      "0.004234299674862996\n",
      "--------\n",
      "0.0012054162216372788\n",
      "--------\n",
      "0.004364024440292269\n",
      "--------\n",
      "0.005981738388072699\n",
      "--------\n",
      "0.006572629208676517\n",
      "--------\n",
      "0.006493631357443519\n",
      "--------\n",
      "0.005992020829580724\n",
      "--------\n",
      "0.005238327459665015\n",
      "--------\n",
      "0.00434925795707386\n",
      "--------\n",
      "0.003403757931664586\n",
      "--------\n",
      "0.002454297398799099\n",
      "--------\n",
      "0.0015348267188528553\n",
      "--------\n",
      "Done!\n",
      "Number of iterations:  34\n"
     ]
    }
   ],
   "source": [
    "# Auxiliary functions\n",
    "\n",
    "def init_root_distribution():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the root distribution (all probabilities between 0.2 and 0.3)\n",
    "    \"\"\"\n",
    "    estimated_root_distribution = []\n",
    "    while True:\n",
    "        estimated_root_distribution = np.random.dirichlet([1,1,1,1])\n",
    "        i = 0\n",
    "        while i<4:\n",
    "            if estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3:\n",
    "                i += 1\n",
    "            else:\n",
    "                break\n",
    "        if i == 4:\n",
    "            assert(np.sum(estimated_root_distribution) < 1.0000001 and np.sum(estimated_root_distribution) > 0.9999999)\n",
    "            assert(estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3 for i in range(4))\n",
    "            break\n",
    "    return estimated_root_distribution\n",
    "\n",
    "def init_transition_matrix():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the transition matrices (which we estimate as LARGE-DIAGONAL-VALUE matrices)\n",
    "    \"\"\"\n",
    "    M = np.zeros((4,4))\n",
    "    i=0\n",
    "    while i<4:\n",
    "        R = np.random.dirichlet([1,1,1,1])\n",
    "        if R[i] > 0.6 and R[i] < 0.8:\n",
    "            M[i,:] = R \n",
    "            i = i + 1\n",
    "    assert (M[i,i] > 0.6 and M[i,i] < 0.8 for i in range(4))\n",
    "    assert (np.sum(M[i,:]) < 1.0000001 and np.sum(M[i,:]) > 0.9999999 for i in range(4))\n",
    "    return M\n",
    "\n",
    "def log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "    \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "    logL = 0\n",
    "    for obs in states:\n",
    "        p_i = 0 \n",
    "        observed_data = obs[n_int:]\n",
    "        if observed_data in u_i:\n",
    "            for state in states:\n",
    "                if state[n_int:] == observed_data:\n",
    "                    pi = root_distribution[int(state[0])]\n",
    "                    for p in params.items():\n",
    "                        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                        pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                    p_i += pi\n",
    "        logL += (u_i[observed_data] * math.log(p_i))\n",
    "    return logL\n",
    "\n",
    "############################################################################################################\n",
    "########################################## FELSENSTEIN ALGORITHM ###########################################\n",
    "############################################################################################################\n",
    "# # Felsentstein algorithm\n",
    "# # \"((4,5),(2,3))\"\n",
    "# newick_tree = \"(((4:1,5:1):1,(6:1,7:1):1):1,(8:1,9:1):1);\"\n",
    "# L_tree = Phylo.read(StringIO(newick_tree), \"newick\")\n",
    "# # Change nodes names\n",
    "# for idx, clade in enumerate(L_tree.get_nonterminals()):\n",
    "#     id = 0\n",
    "#     if idx == 2:\n",
    "#         clade.name = \"Int_Id_\" + str(id)\n",
    "#         id += 1\n",
    "#     else:\n",
    "#         clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "# # Change leaves names\n",
    "# for idx, clade in enumerate(L_tree.get_terminals()):\n",
    "#     clade.name = \"Leaf_\" + clade.name \n",
    "# L_net = Phylo.to_networkx(L_tree) # to graph\n",
    "# L_edges = []\n",
    "# for edge in L_net.edges():\n",
    "#     if edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     elif edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] != \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     elif edge[0].name.split(\"_\")[1] != \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     else:\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     L_edges.append((u,v))\n",
    "# # Create a dictionary to store the father and son information\n",
    "# family_tree = {}\n",
    "# # Loop through the edges and update the dictionary\n",
    "# for father, son in L_edges:\n",
    "#     if father not in family_tree:\n",
    "#         family_tree[father] = []\n",
    "#     family_tree[father].append(son)\n",
    "# # Create a DataFrame from the dictionary\n",
    "# tree_df = pd.DataFrame.from_dict(family_tree, orient='index')\n",
    "# # Fill NaN values with empty strings\n",
    "# tree_df = tree_df.fillna('')\n",
    "# # Change column names\n",
    "# tree_df.columns = ['child_1', 'child_2']\n",
    "\n",
    "# class TreeInference():\n",
    "#     def __init__(self, tree_data, leaf_values, params, root_distr):\n",
    "#         # Set parameters\n",
    "#         self.tree_data = tree_data.copy()\n",
    "#         self.leaf_values = leaf_values\n",
    "#         self.params = params\n",
    "#         self.root_distr = root_distr\n",
    "\n",
    "#         # Find IDs of internal nodes, all nodes and leaf nodes\n",
    "#         self.internal_node_ids = np.array(tree_data.index)\n",
    "#         self.all_node_ids = np.unique(np.concatenate((np.array(self.tree_data.index),\n",
    "#                                                       np.array(self.tree_data[\"child_1\"]),\n",
    "#                                                       np.array(self.tree_data[\"child_2\"]))))\n",
    "#         self.leaf_node_ids = list(set(self.all_node_ids) - set(self.internal_node_ids))\n",
    "        \n",
    "#         # Count internal nodes and leaves for each row in tree_data and set root node\n",
    "#         self.treeNodeCounts()\n",
    "#         self.root_node = 0\n",
    "        \n",
    "#         # Get indexes for matricies for states, e.g. A is row 0, G is row 2...\n",
    "#         states = [\"A\", \"G\", \"C\", \"T\"]\n",
    "#         self.matrix_idx = dict(zip(states, range(len(states))))\n",
    "        \n",
    "#         # Empty matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "#         # Log likelihood of whole tree\n",
    "#         self.L_root = None\n",
    "        \n",
    "\n",
    "#     \"\"\"\n",
    "#     Calculates the number of internal nodes and leaves below each internal node.\n",
    "#     This works by recursively iterating down the child nodes from the original internal node until a\n",
    "#     leaf node is found, and then returning upwards counting the leaves and nodes.\n",
    "#     The new derived information is then added as two new columns at the end of the data.\n",
    "#     \"\"\"\n",
    "#     def treeNodeCounts(self):\n",
    "#         n_nodes = np.array([])\n",
    "#         n_leaves = np.array([])\n",
    "\n",
    "#         # Set the number of internal nodes and leaves for each internal node row\n",
    "#         for n in self.internal_node_ids:\n",
    "#             num_nodes, num_leaves = self.checkNode(self.tree_data, n, self.internal_node_ids)\n",
    "#             n_nodes = np.append(n_nodes, num_nodes)\n",
    "#             n_leaves = np.append(n_leaves, num_leaves)\n",
    "\n",
    "#         # Add new information to dataframe\n",
    "#         self.tree_data[\"N_Internal\"] = n_nodes\n",
    "#         self.tree_data[\"N_Leaves\"] = n_leaves\n",
    "                \n",
    "#     # Recursively find the number of internal nodes and leaves given a node ID\n",
    "#     def checkNode(self, tree_data, node_id, node_array):\n",
    "#         # Record the number of internal nodes and leaves below\n",
    "#         num_nodes = 0\n",
    "#         num_leaves = 0\n",
    "\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[node_id][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[node_id][\"child_2\"]\n",
    "\n",
    "#         # Check if found an internal node to the left\n",
    "#         if child_1 in node_array:\n",
    "#             # Recursively check nodes below\n",
    "#             d1_nodes, d1_leaves = self.checkNode(self.tree_data, child_1, node_array)\n",
    "#             num_nodes += d1_nodes + 1\n",
    "#             num_leaves += d1_leaves\n",
    "\n",
    "#         # Check if found an internal node to the right\n",
    "#         if child_2 in node_array:\n",
    "#             d2_nodes, d2_leaves = self.checkNode(self.tree_data, child_2, node_array)\n",
    "#             num_nodes += d2_nodes + 1\n",
    "#             num_leaves += d2_leaves\n",
    "\n",
    "#         return num_nodes, num_leaves\n",
    "    \n",
    "    \n",
    "#     # Calculate likelihood for an internal node\n",
    "#     def nodeLikelihood(self, node_n, params):\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[str(node_n)][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[str(node_n)][\"child_2\"]\n",
    "        \n",
    "#         # Check if likelihood of left child node set\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(child_1, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_1), 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(int(child_1), params)\n",
    "#         # Check if likelihood of right child node set\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 self.nodeLikelihood(child_2, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_2), 0] is None):\n",
    "#                 self.nodeLikelihood(int(child_2), params)\n",
    "        \n",
    "#         M1, M2 = np.full((4,4),None), np.full((4,4),None)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             # Create transition matrix for left and right child nodes\n",
    "#             matrix_name_child1 = \"M_0_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_0_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_1 == \"Id_0\":\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = np.identity(4)\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_2 == \"Id_0\":\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = np.identity(4)\n",
    "#         else:\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "\n",
    "#         # Get conditional likelihood of child nodes\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             d1_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d1_likelihood = self.L[int(child_1)]\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             d2_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d2_likelihood = self.L[int(child_2)]\n",
    "        \n",
    "#         # Calculate the likelihood of each state for the node\n",
    "#         left_prob = np.sum(M1 * d1_likelihood, axis = 1)\n",
    "#         right_prob = np.sum(M2 * d2_likelihood, axis = 1)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             self.L[-1,:] = np.multiply(left_prob, right_prob)\n",
    "#         else:\n",
    "#             self.L[int(node_n),:] = np.multiply(left_prob, right_prob)\n",
    "    \n",
    "#     # Run Felsenstein’s algorithm to calculate the log likelihood of the tree\n",
    "#     def calculateLikelihood(self, params, Estep):\n",
    "#         # Clear matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "        \n",
    "#         # Set likelihood of leaf states\n",
    "#         for leaf_n in self.leaf_node_ids:\n",
    "#             if leaf_n == \"Id_0\":\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[-1, :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[-1, self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "#             else:\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[int(leaf_n), :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[int(leaf_n), self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "        \n",
    "#         for node_n in self.internal_node_ids:\n",
    "#             # Check likelihood has not been set\n",
    "#             if node_n == \"Id_0\":\n",
    "#                 if self.L[-1, 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "#             else:\n",
    "#                 if self.L[int(node_n), 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "        \n",
    "#         if Estep == True:\n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = np.sum(np.multiply(self.root_distr, self.L[self.root_node]))\n",
    "#         else: \n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = math.log(np.sum(np.multiply(self.root_distr, self.L[self.root_node])))\n",
    "        \n",
    "#         return self.L_root\n",
    "\n",
    "# def Felsenstein_log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "#     \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "#     logL = 0\n",
    "#     for obs in states:\n",
    "#         observed_data = obs[n_int:]\n",
    "#         tree_leaves = dict()\n",
    "#         for i in range(len(observed_data)):\n",
    "#             if observed_data[i] == str(0):\n",
    "#                 tree_leaves[n_int+i] = \"A\"\n",
    "#             elif observed_data[i] == str(1):\n",
    "#                 tree_leaves[n_int+i] = \"G\"\n",
    "#             elif observed_data[i] == str(2):\n",
    "#                 tree_leaves[n_int+i] = \"C\"\n",
    "#             elif observed_data[i] == str(3):\n",
    "#                 tree_leaves[n_int+i] = \"T\"\n",
    "#         tree_inference = TreeInference(tree_df, tree_leaves, params, root_distribution)\n",
    "#         tree_inference.calculateLikelihood(params, False)\n",
    "#         logL += (u_i[observed_data] * tree_inference.L_root)\n",
    "#         del tree_inference\n",
    "#     return logL\n",
    "\n",
    "\n",
    "############################################################################################################\n",
    "\n",
    "class Param:\n",
    "    \"\"\"\n",
    "    Class to store the parameters of the tree\n",
    "    \"\"\"\n",
    "    def __init__(self, edge, transition_matrix, alignment=None):\n",
    "        self.edge = edge\n",
    "        self.transition_matrix = transition_matrix\n",
    "        self.alignment = alignment # it will only be placed in the leaves\n",
    "\n",
    "# Obtain number of leaves and internal nodes\n",
    "n_leaves, n_int = 0, 0 \n",
    "for node in net.nodes():\n",
    "    if node.name.startswith(\"L\"):\n",
    "        n_leaves += 1\n",
    "    else:\n",
    "        n_int += 1\n",
    "\n",
    "# Count all the leaves ocurrences \n",
    "sequence_length = len(leave_sequences[0])\n",
    "number_of_sequences = len(leave_sequences)\n",
    "occurrences = []\n",
    "for i in range(sequence_length):\n",
    "    _ = ''\n",
    "    for j in range(number_of_sequences):\n",
    "        _ += leave_sequences[j][i]\n",
    "    occurrences.append(_)\n",
    "c = Counter(occurrences)\n",
    "u_i = OrderedDict(sorted(c.items()))\n",
    "\n",
    "# Define all possible (hidden) states of the internal nodes \n",
    "hidden_combinations = itertools.product([0,1,2,3], repeat=n_int)\n",
    "sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "for i in range(len(sorted_hidden_combinations)):\n",
    "    sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    " \n",
    "# Define all possible fully observed states (i.e. combining u_i + internal nodes)\n",
    "# '000000' would mean to have an 'A' in all nodes: Int_0, Int_1, Leaf_2, Leaf_3, Leaf_4, Leaf_5 \n",
    "states = list(itertools.product(list(sorted_hidden_combinations), list(u_i.keys())))\n",
    "states = [i[0]+i[1] for i in states]\n",
    "\n",
    "############################################################################################################\n",
    "###################################### STEP 0: Initialise parameters #######################################\n",
    "############################################################################################################\n",
    "estimated_root_distribution = init_root_distribution() # First estimation for the root distribution\n",
    "\n",
    "# PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "# for edge in net.edges():\n",
    "#     # If we are in a leaf, we need to specify the alignment\n",
    "#     iter_leaves = 0\n",
    "#     if edge[1].name.startswith(\"L\"):\n",
    "#         new_edge = Param(edge, init_transition_matrix(), leave_alignments[iter_leaves].seq)\n",
    "#         iter_leaves += 1\n",
    "#     # Otherwise, are in internal nodes\n",
    "#     else:\n",
    "#         new_edge = Param(edge, init_transition_matrix())\n",
    "#     u = str(edge[0].name.split(\"_\")[1])\n",
    "#     v = str(edge[1].name.split(\"_\")[1])\n",
    "#     name = \"M_\" + u + \"_to_\" + v\n",
    "#     PARAMS[name] = new_edge\n",
    "\n",
    "# ADRIA CODE\n",
    "i=0\n",
    "PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "for edge in net.edges():\n",
    "    # If we are in a leaf, we need to specify the alignment\n",
    "    iter_leaves = 0\n",
    "    if edge[1].name.startswith(\"L\"):\n",
    "        new_edge = Param(edge, Results[i].matrix, leave_alignments[iter_leaves].seq)\n",
    "        iter_leaves += 1\n",
    "    # Otherwise, are in internal nodes\n",
    "    else:\n",
    "        new_edge = Param(edge, Results[i].matrix)\n",
    "    u = str(edge[0].name.split(\"_\")[1])\n",
    "    v = str(edge[1].name.split(\"_\")[1])\n",
    "    name = \"M_\" + u + \"_to_\" + v\n",
    "    PARAMS[name] = new_edge\n",
    "    i += 1\n",
    "\n",
    "eps = 10**(-3)                  # epsilon for convergence\n",
    "iter = 0                        # iteration counter \n",
    "params = copy.copy(PARAMS)      # copy of the parameters (will be updated at each iteration)\n",
    "# logF = Felsenstein_log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "logL = log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "# print(logF)\n",
    "print(\"l\", logL)\n",
    "logL_ = 0\n",
    "\n",
    "while np.abs(logL_ - logL) > eps:\n",
    "    print(np.abs(logL_ - logL))\n",
    "    # print(np.abs(logF_ - logF))\n",
    "    print(\"--------\")\n",
    "    if iter > 0:\n",
    "        logL = logL_\n",
    "        # logF = logF_\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## E-step ##################################################\n",
    "    ############################################################################################################\n",
    "    # Define the expected hidden data matrix\n",
    "    U = dict()\n",
    "    root_distr = estimated_root_distribution ########\n",
    "\n",
    "    for obs in states:\n",
    "        U[obs] = u_i[obs[n_int:]] # initialise by computing u_i of the expected hidden data matrix formula\n",
    "        f_ij = estimated_root_distribution[int(obs[0])] \n",
    "        for p in params.items():\n",
    "            u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "            f_ij *= p[1].transition_matrix[int(obs[u]), int(obs[v])]\n",
    "        # Now f_ij := numerator of the expected hidden data matrix formula \n",
    "\n",
    "        # We need to marginalise to obtain f_i := denominator of the expected hidden data matrix formula\n",
    "        f_i = 0 # we initialise the partial data model\n",
    "        observed_data = obs[n_int:]\n",
    "        for state in states:\n",
    "            if state[n_int:] == observed_data:\n",
    "                pi = estimated_root_distribution[int(state[0])]\n",
    "                for p in params.items():\n",
    "                    u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                    pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                f_i += pi\n",
    "\n",
    "        # observed_data = obs[n_int:]\n",
    "        # tree_leaves = dict()\n",
    "        # for i in range(len(observed_data)):\n",
    "        #     if observed_data[i] == str(0):\n",
    "        #         tree_leaves[n_int+i] = \"A\"\n",
    "        #     elif observed_data[i] == str(1):\n",
    "        #         tree_leaves[n_int+i] = \"G\"\n",
    "        #     elif observed_data[i] == str(2):\n",
    "        #         tree_leaves[n_int+i] = \"C\"\n",
    "        #     elif observed_data[i] == str(3):\n",
    "        #         tree_leaves[n_int+i] = \"T\"\n",
    "        # tree_inference = TreeInference(tree_df, tree_leaves, params, root_distr)\n",
    "        # tree_inference.calculateLikelihood(params, True)\n",
    "\n",
    "        U[obs] *= f_ij/f_i\n",
    "        # U[obs] *= f_ij/tree_inference.L_root\n",
    "        # del tree_inference\n",
    "        # del tree_leaves\n",
    "\n",
    "\n",
    "    assert(len(U) <= 4**(n_int+n_leaves))\n",
    "    sum = 0\n",
    "    sum2 = 0\n",
    "    for i in U.items():\n",
    "        sum += i[1]\n",
    "    for i in u_i.items():\n",
    "        sum2 += i[1]\n",
    "    assert(sum < sequence_length+0.001 and sum > sequence_length-0.001)\n",
    "    assert(sum2 < sequence_length+0.001 and sum2 > sequence_length-0.001)\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## M-step ##################################################\n",
    "    ############################################################################################################\n",
    "    estimated_parameters = dict()\n",
    "    root_distr = np.zeros(4)\n",
    "    iter_root = 0 # counter for the root distribution, we only want to count once\n",
    "\n",
    "    for p in params.items():\n",
    "        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "        name = \"M_\" + str(u) + \"_to_\" + str(v)\n",
    "        estimated_parameters[name] = np.zeros((4,4))\n",
    "        for i in range(4):\n",
    "            for j in range(4):\n",
    "                ui = 0\n",
    "                for obs in U.items():\n",
    "                    # For the root distribution, we only need to consider the first internal node\n",
    "                    if iter_root == 0 and obs[0][0] == str(i):\n",
    "                        root_distr[i] += obs[1]\n",
    "                    # For the transition matrices:\n",
    "                    if obs[0][u] == str(i) and obs[0][v] == str(j):\n",
    "                        estimated_parameters[name][i][j] += obs[1]\n",
    "                    if obs[0][u] == str(i):\n",
    "                        ui += obs[1]\n",
    "                # Update root distribution\n",
    "                if iter_root == 0:\n",
    "                    root_distr[i] /= sequence_length\n",
    "                iter_root = 1\n",
    "                estimated_parameters[name][i][j] /= ui\n",
    "            iter_root = 0\n",
    "            \n",
    "        # Update parameters\n",
    "        params[name].transition_matrix = estimated_parameters[name]\n",
    "\n",
    "    ##########################################################################################################\n",
    "    ########################### Compute log-likelihood of the estimated parameters ###########################\n",
    "    ##########################################################################################################\n",
    "    # logF_ = Felsenstein_log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "    logL_ = log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "\n",
    "    iter += 1\n",
    "\n",
    "print(\"Done!\")\n",
    "print(\"Number of iterations: \", iter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################### DEBUGGING ############################\n",
    "\n",
    "# # All possible (theoretical) occurrences of nucleotides in the leaves\n",
    "# alphabet = [0, 1, 2, 3]\n",
    "# combinations = itertools.product(alphabet, repeat=n_leaves)\n",
    "# sorted_combinations = sorted(combinations)\n",
    "# for i in range(len(sorted_combinations)):\n",
    "#     sorted_combinations[i] = ''.join([str(s) for s in sorted_combinations[i]])\n",
    "\n",
    "# # All possible (hidden) states of the internal nodes (i.e. indexes of columns of U)\n",
    "# hidden_combinations = itertools.product(alphabet, repeat=n_int)\n",
    "# sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "# for i in range(len(sorted_hidden_combinations)):\n",
    "#     sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    "\n",
    "# # To check if an element is within a dictionary or not \n",
    "# if '0101' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')\n",
    "# if '0000' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix: [[0.64642183 0.02803076 0.27483013 0.05071728]\n",
      " [0.03848122 0.7757021  0.10189982 0.08391686]\n",
      " [0.03184807 0.07934462 0.82109495 0.06771236]\n",
      " [0.04702478 0.01879095 0.04076404 0.89342023]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix: [[0.55850753 0.13612552 0.2499751  0.05539185]\n",
      " [0.01791204 0.82605787 0.05907209 0.096958  ]\n",
      " [0.04534628 0.008074   0.85149918 0.09508054]\n",
      " [0.00646156 0.06355415 0.02348564 0.90649866]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix: [[0.75152183 0.09827146 0.07355421 0.0766525 ]\n",
      " [0.04888819 0.86735324 0.02989229 0.05386628]\n",
      " [0.05446854 0.02450495 0.75018255 0.17084396]\n",
      " [0.12039713 0.03979299 0.05715314 0.78265675]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix: [[0.82835309 0.03976936 0.05146899 0.08040856]\n",
      " [0.02327998 0.80873654 0.06043788 0.10754561]\n",
      " [0.11544022 0.13895831 0.71027209 0.03532938]\n",
      " [0.09988558 0.08065349 0.01737091 0.80209001]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix: [[0.78296278 0.06942125 0.08171685 0.06589912]\n",
      " [0.02166725 0.83743552 0.01932756 0.12156967]\n",
      " [0.11182788 0.03378373 0.78484422 0.06954417]\n",
      " [0.11063295 0.0881492  0.0495941  0.75162375]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23195748 0.24737645 0.23524372 0.28542235]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.6464218290229772"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the theoretical parameters:\n",
    "theoretical_matrices = dict()\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    u = edges[i].edge[0].name.split(\"_\")[1]\n",
    "    v = edges[i].edge[1].name.split(\"_\")[1]\n",
    "    matrix_name = \"M_\" + u + \"_to_\" + v\n",
    "    print(\"Transition matrix:\", edges[i].transition_matrix)\n",
    "    theoretical_matrices[matrix_name] = np.array(edges[i].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", node_distr['Int_0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'M_0_to_2': array([[0.64642183, 0.02803076, 0.27483013, 0.05071728],\n",
      "       [0.03848122, 0.7757021 , 0.10189982, 0.08391686],\n",
      "       [0.03184807, 0.07934462, 0.82109495, 0.06771236],\n",
      "       [0.04702478, 0.01879095, 0.04076404, 0.89342023]]), 'M_0_to_3': array([[0.55850753, 0.13612552, 0.2499751 , 0.05539185],\n",
      "       [0.01791204, 0.82605787, 0.05907209, 0.096958  ],\n",
      "       [0.04534628, 0.008074  , 0.85149918, 0.09508054],\n",
      "       [0.00646156, 0.06355415, 0.02348564, 0.90649866]]), 'M_0_to_1': array([[0.75152183, 0.09827146, 0.07355421, 0.0766525 ],\n",
      "       [0.04888819, 0.86735324, 0.02989229, 0.05386628],\n",
      "       [0.05446854, 0.02450495, 0.75018255, 0.17084396],\n",
      "       [0.12039713, 0.03979299, 0.05715314, 0.78265675]]), 'M_1_to_4': array([[0.82835309, 0.03976936, 0.05146899, 0.08040856],\n",
      "       [0.02327998, 0.80873654, 0.06043788, 0.10754561],\n",
      "       [0.11544022, 0.13895831, 0.71027209, 0.03532938],\n",
      "       [0.09988558, 0.08065349, 0.01737091, 0.80209001]]), 'M_1_to_5': array([[0.78296278, 0.06942125, 0.08171685, 0.06589912],\n",
      "       [0.02166725, 0.83743552, 0.01932756, 0.12156967],\n",
      "       [0.11182788, 0.03378373, 0.78484422, 0.06954417],\n",
      "       [0.11063295, 0.0881492 , 0.0495941 , 0.75162375]])}\n"
     ]
    }
   ],
   "source": [
    "# # save the dictionary to a file\n",
    "# np.save('data.npy', theoretical_matrices)\n",
    "# # load the dictionary from the file\n",
    "# loaded_dict = np.load('data.npy', allow_pickle=True).item()\n",
    "# # print the loaded dictionary\n",
    "# print(loaded_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M_0_to_2\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix:\n",
      " [[0.86753892 0.05213542 0.06946348 0.01086218]\n",
      " [0.08065343 0.85671968 0.04982893 0.01279796]\n",
      " [0.11219846 0.17464684 0.65461785 0.05853685]\n",
      " [0.02775028 0.08494961 0.04088151 0.8464186 ]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_3\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix:\n",
      " [[0.75527498 0.04249873 0.00019955 0.20202674]\n",
      " [0.10446698 0.73349645 0.10777451 0.05426206]\n",
      " [0.009178   0.09167603 0.88105832 0.01808765]\n",
      " [0.10672648 0.08298123 0.01949473 0.79079756]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_1\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix:\n",
      " [[0.69009493 0.15321495 0.12518281 0.03150731]\n",
      " [0.03390489 0.89612499 0.03748613 0.032484  ]\n",
      " [0.13612944 0.00355591 0.86031238 0.00000227]\n",
      " [0.13829849 0.07169377 0.04358193 0.74642581]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_4\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix:\n",
      " [[0.79787321 0.09123409 0.0748348  0.0360579 ]\n",
      " [0.1525167  0.62670853 0.12700414 0.09377063]\n",
      " [0.06182208 0.09423139 0.82703856 0.01690797]\n",
      " [0.09434814 0.05487633 0.01727208 0.83350345]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_5\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix:\n",
      " [[0.77670263 0.03884329 0.11981174 0.06464234]\n",
      " [0.07306013 0.6964252  0.14381485 0.08669983]\n",
      " [0.0001472  0.02174493 0.93060026 0.04750762]\n",
      " [0.01290462 0.10677068 0.06826479 0.81205991]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23508015 0.26506847 0.26101933 0.23983305]\n"
     ]
    }
   ],
   "source": [
    "# Check the estimated parameters:\n",
    "for i in params.items():\n",
    "    print(i[0])\n",
    "    print(\"Edge:\", i[1].edge)\n",
    "    print(\"Transition matrix:\\n\", i[1].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", root_distr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix difference:\n",
      " [[0.09960865 0.07093256 0.00498621 0.03366231]\n",
      " [0.01576357 0.02703813 0.00892483 0.03387688]\n",
      " [0.00533819 0.03058066 0.01301418 0.01222828]\n",
      " [0.00504287 0.04282721 0.00457214 0.05244223]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix difference:\n",
      " [[0.00386991 0.01431933 0.0324696  0.04291902]\n",
      " [0.03240956 0.00438465 0.00578344 0.03101077]\n",
      " [0.00129948 0.00595081 0.00836875 0.00111846]\n",
      " [0.00368219 0.0201782  0.01427531 0.03077132]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix difference:\n",
      " [[0.00043271 0.04586535 0.01971483 0.02571781]\n",
      " [0.04628465 0.03270525 0.01225697 0.00132243]\n",
      " [0.00405161 0.02596747 0.04979691 0.01977783]\n",
      " [0.05635943 0.0032355  0.00691523 0.06003916]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix difference:\n",
      " [[0.02206907 0.00438435 0.02045318 0.00276846]\n",
      " [0.04114356 0.07985678 0.0152662  0.02344702]\n",
      " [0.01766089 0.02629588 0.04017144 0.00378534]\n",
      " [0.00126844 0.00706299 0.01874386 0.02707528]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix difference:\n",
      " [[0.04539034 0.01368587 0.03621623 0.00451176]\n",
      " [0.00302602 0.00613968 0.01219578 0.00908212]\n",
      " [0.00442834 0.01538392 0.00158094 0.0213932 ]\n",
      " [0.02699704 0.00888034 0.00551462 0.03036276]]\n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Compute the matrix difference between the theoretical and estimated parameters\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    print(\"Transition matrix difference:\\n\", np.abs(edges[i].transition_matrix - params['M_' + str(edges[i].edge[0].name.split(\"_\")[1]) + '_to_' + str(edges[i].edge[1].name.split(\"_\")[1])].transition_matrix))\n",
    "    print(\"---\"*30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Real pi\n",
      "[0.20402651 0.28098141 0.2943494  0.22064268]\n",
      "Real M\n",
      "M_0_to_2 [[0.62313662 0.04703023 0.05612722 0.27370592]\n",
      " [0.02652672 0.87016895 0.03646515 0.06683919]\n",
      " [0.02116387 0.0350012  0.86900428 0.07483066]\n",
      " [0.10259202 0.03959218 0.04234033 0.81547547]]\n",
      "M_0_to_3 [[0.97759178 0.0094502  0.00494207 0.00801595]\n",
      " [0.00511038 0.96404682 0.02481176 0.00603104]\n",
      " [0.01346532 0.01706666 0.95259448 0.01687354]\n",
      " [0.01580069 0.00971656 0.01189318 0.96258957]]\n",
      "M_0_to_1 [[0.94297467 0.04160616 0.01275057 0.00266859]\n",
      " [0.00206066 0.96572139 0.00720044 0.02501751]\n",
      " [0.00730919 0.01122769 0.9696116  0.01185151]\n",
      " [0.00480198 0.00154014 0.02165493 0.97200295]]\n",
      "M_1_to_4 [[0.74524268 0.18944609 0.02843996 0.03687127]\n",
      " [0.12749976 0.82099596 0.04440439 0.00709988]\n",
      " [0.10194701 0.04680566 0.7369428  0.11430453]\n",
      " [0.01023244 0.09172772 0.02934055 0.8686993 ]]\n",
      "M_1_to_5 [[0.963656   0.01486956 0.00794318 0.01353126]\n",
      " [0.02468435 0.95450804 0.00844202 0.01236559]\n",
      " [0.00573546 0.00531638 0.9772312  0.01171696]\n",
      " [0.01814195 0.00350424 0.01549612 0.96285769]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Estimated M\n",
      "M_0_to_2 [[0.626842   0.0718278  0.02845728 0.27287293]\n",
      " [0.02247744 0.89009638 0.00798747 0.07943871]\n",
      " [0.01286524 0.02290314 0.89469118 0.06954044]\n",
      " [0.11821687 0.02688332 0.03730715 0.81759266]]\n",
      "M_0_to_3 [[9.62448697e-001 2.14918626e-002 1.44574001e-029 1.60594404e-002]\n",
      " [4.18648274e-003 9.52149152e-001 4.36643653e-002 3.56263895e-056]\n",
      " [9.90843526e-100 2.45750159e-002 9.61273533e-001 1.41514510e-002]\n",
      " [9.34885163e-003 4.17489859e-028 2.87836912e-002 9.61867457e-001]]\n",
      "M_0_to_1 [[9.42483585e-01 4.71597637e-02 1.03566515e-02 1.68353088e-16]\n",
      " [3.35462227e-81 9.83217585e-01 1.57954588e-10 1.67824151e-02]\n",
      " [1.32177769e-02 8.70193589e-03 9.65148462e-01 1.29318250e-02]\n",
      " [9.95506243e-04 8.29529373e-03 9.43833391e-03 9.81270866e-01]]\n",
      "M_1_to_4 [[0.73385467 0.19595607 0.04033229 0.02985697]\n",
      " [0.10324607 0.85752341 0.03120931 0.00802121]\n",
      " [0.10342276 0.03680517 0.74758907 0.112183  ]\n",
      " [0.00853965 0.03086488 0.04576015 0.91483532]]\n",
      "M_1_to_5 [[9.72165273e-01 1.94052180e-02 5.97367099e-70 8.42950857e-03]\n",
      " [9.73694315e-60 9.72270581e-01 1.93252285e-02 8.40419023e-03]\n",
      " [6.48129073e-03 1.52469903e-02 9.63435878e-01 1.48358410e-02]\n",
      " [2.09456602e-02 2.01656191e-02 9.25403657e-03 9.49634684e-01]]\n",
      "Estimated l\n",
      "M_0_to_2 0.9473903384139062\n",
      "M_0_to_3 0.1632315143144635\n",
      "M_0_to_1 0.13015767013667762\n",
      "M_1_to_4 0.9031646413494592\n",
      "M_1_to_5 0.14633294967882585\n",
      "Estimated pi\n",
      "[0.21119349 0.25871453 0.31101784 0.22107815]\n",
      "Niter\n",
      "63\n",
      "TExec\n",
      "25.41742992401123\n"
     ]
    }
   ],
   "source": [
    "print(\"Real pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_spectral/real_root_distr.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Real M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_spectral/real_matrices.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"---\"*30)\n",
    "print(\"Estimated M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_spectral/RESULTS/repetition_98/M_estimation.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated l\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_spectral/RESULTS/repetition_98/estimated_branch_lengths.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a15_b100_500_spectral/RESULTS/repetition_98/root_estimation.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Niter\")\n",
    "file = open(\"./OUTPUTS/EM_a15_b100_500_spectral/RESULTS/repetition_98/Niter.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)\n",
    "print(\"TExec\")\n",
    "file = open(\"./OUTPUTS/EM_a15_b100_500_spectral/RESULTS/repetition_98/TExec.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "debugger",
   "language": "python",
   "name": "debugger"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "03f790eb89590da6bbc9dab944ceea1669c4a69ff2bd32047ab9c0f81f755bd1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
