{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Expectation-Maximization algorithm** (hidden model)\n",
    "### Author: Mart√≠ Cortada Garcia\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import numpy as np\n",
    "from Bio import SeqIO, Phylo\n",
    "from io import StringIO\n",
    "from collections import Counter, OrderedDict\n",
    "import itertools\n",
    "import pandas as pd\n",
    "import math\n",
    "import copy\n",
    "\n",
    "# Import required functions from modules we have programmed\n",
    "from data_simulation import simulate"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Input:**\n",
    "We have a tree $T$ (where the root corresponds to the internal node codified as `Int_0`) and the alignments of the leaves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjUAAAGwCAYAAABRgJRuAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy88F64QAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAvQ0lEQVR4nO3dfVRVdaL/8c8WBHlWNEXwgUJ8yADRgrAyK0vNSpq7slzmQ9PEzCwwyZjMm2XOmHpvajLVrawuNN2UyC51r87IJQstzCLUGXy4WlidUpTGEoTGg8L+/dH1/MR84OCBI1/er7VYy73Pd+/92Xuddc7Hvfc5x7Jt2xYAAEA718nbAQAAADyBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgAAYARfbwe4EI2NjTpw4IBCQkJkWZa34wAAgGawbVtHjx5VZGSkOnXy3PmVdl1qDhw4oL59+3o7BgAAaIFvvvlGffr08dj62nWpCQkJkfTTQQkNDfVyGgAA0Bw1NTXq27ev633cU9p1qTl5ySk0NJRSAwBAO+PpW0e4URgAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgAAYARKDQAAMAKlBgAAGIFSAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqUGAAAYgVIDAACMQKkBAABGoNQAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgAAYARKDQAAMAKlBgAAGIFSAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqUGAAAYgVIDAACMQKkBAABGoNQAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgAAYARKDQCvsW1baWlpCg8Pl2VZ2r59u7cjAWjHKDUAzqi4uFiWZenIkSOtto3169crNzdXa9euVWVlpa644opzjl+8eLGuuuoqhYSEqGfPnkpNTdWePXtaLR+A9oVSA8BrKioq1Lt3b40cOVIRERHy9fU95/iNGzcqPT1dW7ZsUVFRkY4fP65bbrlFdXV1bZQYwMXMq6Vm06ZNuv322xUZGSnLsvTOO+94Mw4ANzidTmVlZSkqKkpBQUFKTk5WcXGx6/HDhw9r8uTJioqKUmBgoOLi4rR69WrX4zNmzNDMmTPlcDhkWZaio6PPu83169drxowZGjp0qBISEpSbmyuHw6GysrJW2EMA7c25/1vUyurq6pSQkKBf/vKX+sUvfuHNKK48QEcUFBTk9jIZGRnatWuX8vLyFBkZqYKCAo0bN07l5eWKjY3VsWPHNGLECM2ZM0ehoaFat26dpk6dqpiYGCUlJSk7O1sxMTFauXKlSktL5ePj43aG6upqSVJ4eLjbywIwj2Xbtu3tEJJkWZYKCgqUmpp61jFOp1NOp9M1XVNTo759+6q6ulqhoaEeyQB0RGd6GSguLtYNN9ygH374QV27dm3ymMPh0GWXXSaHw6HIyEjX/DFjxigpKUmLFi0643Zuu+02DR48WEuXLpUkrVixQitWrNBXX33ldubGxkbdcccdOnLkiD766CO3lwfgPTU1NQoLC/PY+/dJXj1T467FixdrwYIFHl9vXV2dgoODPb5ewFTl5eVqaGjQwIEDm8x3Op3q3r27JKmhoUGLFi1Sfn6+9u/fr/r6ejmdTgUGBnokQ3p6unbs2EGhAeDSrkrN3LlzNXv2bNf0yTM1nnTo0KEWnYoHOpLa2lr5+PiorKzsZ5eNTv4H4emnn1Z2drZWrFihuLg4BQUFKTMzU/X19Re8/YyMDK1du1abNm1Snz59Lnh9AMzQrkqNv7+//P39W3UbQUFBlBrgPBITE9XQ0KCqqipdd911ZxxTUlKiiRMn6t5775X00+WivXv36vLLL2/xdm3b1syZM1VQUKDi4mJdeumlLV4XAPO0q1IDoO2Vl5crJCTENW1ZlhISEjRlyhRNmzZNy5YtU2Jior777jtt2LBB8fHxmjBhgmJjY7VmzRpt3rxZ3bp10/Lly3Xo0KELKjXp6elatWqV3n33XYWEhOjgwYOSpLCwMAUEBFzwvgJo3yg1AM5p1KhRTaZ9fHx04sQJ5eTkaOHChXr44Ye1f/9+9ejRQ1dffbVuu+02SdK8efO0b98+jR07VoGBgUpLS1NqaqrrE0st8cILL0iSRo8e3WR+Tk6OZsyY0eL1AjCDVz/9VFtbqy+++ELST6ezly9frhtuuEHh4eHq16/feZf31N3Tp94oXFtby+UnAABakZGffvrss890ww03uKZP3gQ8ffp05ebmeikVAABoj7xaakaPHn3G78cA0PE4HI5z3m+za9euZp3BBdBxcU8NgItCZGTkOX+l+9Qv+QOAM6HUALgo+Pr6asCAAd6OAaAd41e6AQCAESg1AADACJQaAABgBEoNAAAwAqUGAAAYgVIDAACMQKkBAABGoNQAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgAAYARKDQAAMAKlBgAAGIFSAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqUGAAAYgVIDAACMQKkBAABGoNQAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgAAYARKDQAAMAKlBgAAGIFSAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqUGAAAYgVIDAACMQKkBAABGoNQAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQC8xrZtpaWlKTw8XJZlafv27d6OBKAdo9QAOKPi4mJZlqUjR4602jbWr1+v3NxcrV27VpWVlbriiivOOf6FF15QfHy8QkNDFRoaqpSUFP3lL39ptXwA2hevlprFixfrqquuUkhIiHr27KnU1FTt2bPHm5EAtKGKigr17t1bI0eOVEREhHx9fc85vk+fPlqyZInKysr02Wef6cYbb9TEiRO1c+fONkoM4GLm1VKzceNGpaena8uWLSoqKtLx48d1yy23qK6uzpuxADSD0+lUVlaWoqKiFBQUpOTkZBUXF7seP3z4sCZPnqyoqCgFBgYqLi5Oq1evdj0+Y8YMzZw5Uw6HQ5ZlKTo6+rzbvP3223XrrbcqNjZWAwcO1FNPPaXg4GBt2bKlFfYQQHtz7v8WtbL169c3mc7NzVXPnj1VVlamUaNGeSUThQodUVBQkNvLZGRkaNeuXcrLy1NkZKQKCgo0btw4lZeXKzY2VseOHdOIESM0Z84chYaGat26dZo6dapiYmKUlJSk7OxsxcTEaOXKlSotLZWPj49b229oaNBbb72luro6paSkuJ0fgHm8WmpOV11dLUkKDw8/4+NOp1NOp9M1XVNT4/EMvXr18vg6gYudbdtujXc4HMrJyZHD4VBkZKQkKSsrS+vXr1dOTo4WLVqkqKgoZWVluZaZOXOmCgsLlZ+fr6SkJIWFhSkkJEQ+Pj6KiIho9rbLy8uVkpKiY8eOKTg4WAUFBbr88svdyg/ATBdNqWlsbFRmZqauueaas94suHjxYi1YsKCNkwE4XXl5uRoaGjRw4MAm851Op7p37y7ppzMpixYtUn5+vvbv36/6+no5nU4FBgZe0LYHDRqk7du3q7q6WmvWrNH06dO1ceNGig2Ai6fUpKena8eOHfroo4/OOmbu3LmaPXu2a7qmpkZ9+/b1aI5Dhw616FQ80JHU1tbKx8dHZWVlP7tsFBwcLEl6+umnlZ2drRUrViguLk5BQUHKzMxUfX39BW3bz89PAwYMkCSNGDFCpaWlys7O1ksvvXRB6wXQ/l0UpSYjI0Nr167Vpk2b1KdPn7OO8/f3l7+/f6tmCQoKotQA55GYmKiGhgZVVVXpuuuuO+OYkpISTZw4Uffee6+kn87G7t271+NnVBobG5tclgbQcXm11Ni2rZkzZ6qgoEDFxcW69NJLvRkHwBmUl5crJCTENW1ZlhISEjRlyhRNmzZNy5YtU2Jior777jtt2LBB8fHxmjBhgmJjY7VmzRpt3rxZ3bp10/Lly3Xo0KELKjVz587V+PHj1a9fPx09elSrVq1ScXGxCgsLPbGrANo5r5aa9PR0rVq1Su+++65CQkJ08OBBSVJYWJgCAgK8GQ3A/zn9k4g+Pj46ceKEcnJytHDhQj388MPav3+/evTooauvvlq33XabJGnevHnat2+fxo4dq8DAQKWlpSk1NdX1gYCWqKqq0rRp01RZWamwsDDFx8ersLBQN9988wXtIwAzWLa7H3vw5MYt64zzc3JyNGPGjPMuX1NTo7CwMFVXVys0NLTFOerq6lz3AdTW1nL5CQCAVuSp9+/Tef3yEwAAgCfw208ALgoOh0PBwcFn/XM4HN6OCOAid1F8+gkAIiMjz/kr3Se/5A8AzoZSA+Ci4Ovr6/r+GQBoCS4/AQAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqXmPIqLi2VZlo4cOdLsZWbMmKHU1FS3tvP9999rypQpCg0NVdeuXXX//fertrbWvbAAAHRglJqLxJQpU7Rz504VFRW5frE8LS3N27EAAGg3KDVuys3NVdeuXVVYWKghQ4YoODhY48aNU2VlpSTpySef1GuvvaZ3331XlmXJsiwVFxefc527d+/W+vXr9corryg5OVnXXnutnn32WeXl5enAgQNtsFcAALR/lJoW+PHHH7V06VK9/vrr2rRpkxwOh7KysiRJWVlZmjRpkqvoVFZWauTIkedc38cff6yuXbvqyiuvdM0bM2aMOnXqpE8++aRV9wUAAFPwjcItcPz4cb344ouKiYmRJGVkZOj3v/+9JCk4OFgBAQFyOp2KiIho1voOHjyonj17Npnn6+ur8PBwHTx40LPhAQAwFGdqWiAwMNBVaCSpd+/eqqqq8mIiAABAqWmBzp07N5m2LEu2bbd4fRERET8rRSdOnND333/f7LM9AAB0dJSaVuDn56eGhoZmj09JSdGRI0dUVlbmmvf++++rsbFRycnJrRERAADjUGpaQXR0tP72t79pz549+vvf/67jx4+fc/yQIUM0btw4PfDAA/r0009VUlKijIwM3XPPPYqMjGyj1AAAtG+UmlbwwAMPaNCgQbryyit1ySWXqKSk5LzLvPHGGxo8eLBuuukm3Xrrrbr22mu1cuXKNkgLAIAZLPtCbgbxspqaGoWFham6ulqhoaEtXk9dXZ2Cg4MlSbW1tQoKCvJURAAAcBpPvX+fjjM1AADACJSaNrBo0SIFBwef8W/8+PHejgcAgBFa9OV7n332mfLz8+VwOFRfX9/ksf/8z//0SDCT/OY3v9GkSZPO+FhAQEAbpwEAwExul5q8vDxNmzZNY8eO1f/8z//olltu0d69e3Xo0CHdeeedrZGx3QsPD1d4eLi3YwAAYDS3Lz8tWrRIzzzzjP77v/9bfn5+ys7O1v/+7/9q0qRJ6tevX2tkBAAAOC+3S01FRYUmTJgg6acvmaurq5NlWXrooYf4CDIAAPAat0tNt27ddPToUUlSVFSUduzYIUk6cuSIfvzxR8+mAwAAaCa376kZNWqUioqKFBcXp7vuukuzZs3S+++/r6KiIt10002tkREAAOC83C41zz33nI4dOyZJeuyxx9S5c2dt3rxZ//RP/6R58+Z5PCAAAEBzuF1qTv0UT6dOnfToo496NBAAAEBLuH1PTW5u7hnnnzhxQnPnzr3QPAAAAC3idql58MEHddddd+mHH35wzduzZ4+Sk5O1evVqj4YDAABoLrdLzbZt2/Ttt98qLi5ORUVFev755zV8+HANHjxYf/3rX1sjIwAAwHm5fU9NTEyMSkpKlJmZqXHjxsnHx0evvfaaJk+e3Br5AAAAmqVFP2i5bt065eXlKSUlRV27dtWrr76qAwcOeDobAABAs7ldan7961/rrrvu0pw5c/Thhx/qb3/7m/z8/BQXF6f8/PzWyAgAAHBebl9+Kikp0SeffKKEhARJUkREhP785z/r+eef1y9/+cuz/ho1AABAa3K71JSVlcnf3/9n89PT0zVmzBiPhAIAAHCX25efzlRoTho0aNAFhQEAAGgpt8/USNKaNWuUn58vh8Oh+vr6Jo9t3brVI8EAAADc4faZmj/+8Y+677771KtXL23btk1JSUnq3r279u3bp/Hjx7dGRgAAgPNyu9T827/9m1auXKlnn31Wfn5+euSRR1RUVKQHH3xQ1dXVrZERAADgvNwuNQ6HQyNHjpQkBQQE6OjRo5KkqVOn8jMJAADAa9wuNREREfr+++8lSf369dOWLVskSV9++aVs2/ZsOgAAgGZyu9TceOON+q//+i9J0n333aeHHnpIN998s+6++27deeedHg8IAADQHG5/+umxxx5TVFSUpJ++m6Z79+7avHmz7rjjDo0bN87jAQEAAJrDst28ZuTj46PKykr17NmzyfzDhw+rZ8+eamho8GjAc6mpqVFYWJiqq6sVGhra4vXU1dUpODhYklRbW6ugoCBPRQQAAKfx1Pv36dy+/HS2DlRbW6suXbpccCAAAICWaPblp9mzZ0uSLMvSE088ocDAQNdjDQ0N+uSTTzRs2DCPBwQAAGiOZpeabdu2SfrpTE15ebn8/Pxcj/n5+SkhIUFZWVmeTwgAANAMzS41H3zwgaSfPvGUnZ3t0WtgAAAAF8rtTz/l5OS0Rg4AAIAL4vaNwgAAABcjSg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgAAYARKDQAAMAKlBgAAGIFSAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAADamG3bSktLU3h4uCzL0vbt270dyQiUGgAATlFcXCzLsnTkyJFW28b69euVm5urtWvXqrKyUldccUWzl12yZIksy1JmZmar5WuvvFpqXnjhBcXHxys0NFShoaFKSUnRX/7yF29GAgCg1VVUVKh3794aOXKkIiIi5Ovr26zlSktL9dJLLyk+Pr6VE7ZPXi01ffr00ZIlS1RWVqbPPvtMN954oyZOnKidO3d6MxYAAGfldDqVlZWlqKgoBQUFKTk5WcXFxa7HDx8+rMmTJysqKkqBgYGKi4vT6tWrXY/PmDFDM2fOlMPhkGVZio6ObtZ2a2trNWXKFL388svq1q2bh/fKDM2rhq3k9ttvbzL91FNP6YUXXtCWLVs0dOhQL6UCOp66ujpvRwC8IigoyO1lMjIytGvXLuXl5SkyMlIFBQUaN26cysvLFRsbq2PHjmnEiBGaM2eOQkNDtW7dOk2dOlUxMTFKSkpSdna2YmJitHLlSpWWlsrHx6dZ201PT9eECRM0ZswYLVy40O3cHYFXS82pGhoa9NZbb6murk4pKSlnHON0OuV0Ol3TNTU1bRUPMFpwcLC3IwBeYdu2W+MdDodycnLkcDgUGRkpScrKytL69euVk5OjRYsWKSoqSllZWa5lZs6cqcLCQuXn5yspKUlhYWEKCQmRj4+PIiIimrXdvLw8bd26VaWlpW7l7Wi8XmrKy8uVkpKiY8eOKTg4WAUFBbr88svPOHbx4sVasGBBGycEzFVXV0ehAdxQXl6uhoYGDRw4sMl8p9Op7t27S/rpP+mLFi1Sfn6+9u/fr/r6ejmdTgUGBrZom998841mzZqloqIidenS5YL3wWReLzWDBg3S9u3bVV1drTVr1mj69OnauHHjGYvN3LlzNXv2bNd0TU2N+vbt25ZxAWMdOnSoRafigY6ktrZWPj4+Kisr+9llo5P/QXj66aeVnZ2tFStWKC4uTkFBQcrMzFR9fX2LtllWVqaqqioNHz7cNa+hoUGbNm3Sc889J6fT2exLWKbzeqnx8/PTgAEDJEkjRoxQaWmpsrOz9dJLL/1srL+/v/z9/ds6ItAhBAUFUWqA80hMTFRDQ4Oqqqp03XXXnXFMSUmJJk6cqHvvvVeS1NjYqL179571KsT53HTTTSovL28y77777tPgwYM1Z84cCs0pvF5qTtfY2NjkvhkAALyhvLxcISEhrmnLspSQkKApU6Zo2rRpWrZsmRITE/Xdd99pw4YNio+P14QJExQbG6s1a9Zo8+bN6tatm5YvX65Dhw61uNSEhIT87HtsgoKC1L17d7e+36Yj8GqpmTt3rsaPH69+/frp6NGjWrVqlYqLi1VYWOjNWAAAaNSoUU2mfXx8dOLECeXk5GjhwoV6+OGHtX//fvXo0UNXX321brvtNknSvHnztG/fPo0dO1aBgYFKS0tTamqqqqurvbEbHYplu3vrtwfdf//92rBhgyorKxUWFqb4+HjNmTNHN998c7OWr6mpUVhYmKqrqxUaGtriHKfeLFlbW8speHQYPPcBeIOn3r9P59UzNa+++qo3Nw8AAAzCbz8BAOBFDodDwcHBZ/1zOBzejthuXHQ3CgMA0JFERkae81e6T37JH86PUgMAgBf5+vq6vtoEF4bLTwAAwAiUGgAAYARKDQAAMAKlBgAAGIFSAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqUGAAAYgVIDAACMQKkBAABGoNQAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgBnVFxcLMuydOTIkWYvM2PGDKWmprq1naeeekojR45UYGCgunbt6tayAHAqSg0Ar6qvr9ddd92l3/72t96OAqCdo9QAaJbc3Fx17dpVhYWFGjJkiIKDgzVu3DhVVlZKkp588km99tprevfdd2VZlizLUnFx8XnXu2DBAj300EOKi4tr5T0AYDpfbwe42NTV1Xk7AtBm3H2+//jjj1q6dKlef/11derUSffee6+ysrL0xhtvKCsrS7t371ZNTY1ycnIkSeHh4a0RGwDOiFJzml69enk7AnDROn78uF588UXFxMRIkjIyMvT73/9ekhQcHKyAgAA5nU5FRER4MyaADorLT5KCgoJk27a3YwAXvcDAQFehkaTevXurqqrKi4kA4P/jTM0pamtrvR0BaFN1dXVunZ3s3Llzk2nLsvgPAYCLBqXmFEFBQd6OALRrfn5+amho8HYMAB0UpQaAx0RHR6uwsFB79uxR9+7dFRYW9rOzO6dzOBz6/vvv5XA41NDQoO3bt0uSBgwYoODg4DZIDcAU3FMDwGMeeOABDRo0SFdeeaUuueQSlZSUnHeZJ554QomJiZo/f75qa2uVmJioxMREffbZZ22QGIBJLLsdXxCvqalRWFiYqqurFRoa6u04QLtTV1fnOhtSW1vLJVgAbaK13r85UwMAAIxAqQHQahYtWqTg4OAz/o0fP97b8QAYhhuFAbSa3/zmN5o0adIZHwsICGjjNABMR6kB0GrCw8P5qQQAbYbLTwAAwAiUGgAAYARKDQAAMAKlBgAAGIFSAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqUGAAAYgVIDAACMQKkBAABGoNQAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAtDHbtpWWlqbw8HBZlqXt27d7O5IRKDUAAJyiuLhYlmXpyJEjrbaN9evXKzc3V2vXrlVlZaWuuOKKc45/8sknZVlWk7/Bgwe3Wr72ytfbAQAA6GgqKirUu3dvjRw5stnLDB06VO+9955r2teXt/DTXTRnapYsWSLLspSZmentKAAAnJXT6VRWVpaioqIUFBSk5ORkFRcXux4/fPiwJk+erKioKAUGBiouLk6rV692PT5jxgzNnDlTDodDlmUpOjq6Wdv19fVVRESE669Hjx4e3rP276KoeaWlpXrppZcUHx/v7ShAh1VXV+ftCECbCwoKcnuZjIwM7dq1S3l5eYqMjFRBQYHGjRun8vJyxcbG6tixYxoxYoTmzJmj0NBQrVu3TlOnTlVMTIySkpKUnZ2tmJgYrVy5UqWlpfLx8WnWdj///HNFRkaqS5cuSklJ0eLFi9WvXz+385vM66WmtrZWU6ZM0csvv6yFCxeec6zT6ZTT6XRN19TUtHY8oMPo1auXtyMAbc62bbfGOxwO5eTkyOFwKDIyUpKUlZWl9evXKycnR4sWLVJUVJSysrJcy8ycOVOFhYXKz89XUlKSwsLCFBISIh8fH0VERDRru8nJycrNzdWgQYNUWVmpBQsW6LrrrtOOHTsUEhLi1j6YzOulJj09XRMmTNCYMWPOW2oWL16sBQsWtFEywHxBQUGybVuWZXk7CtAulJeXq6GhQQMHDmwy3+l0qnv37pKkhoYGLVq0SPn5+dq/f7/q6+vldDoVGBjY4u2OHz/e9e/4+HglJyerf//+ys/P1/3339/i9ZrGq6UmLy9PW7duVWlpabPGz507V7Nnz3ZN19TUqG/fvq0VD+gwamtrvR0BaBdqa2vl4+OjsrKyn102Cg4OliQ9/fTTys7O1ooVKxQXF6egoCBlZmaqvr7eYzm6du2qgQMH6osvvvDYOk3gtVLzzTffaNasWSoqKlKXLl2atYy/v7/8/f1bORnQ8bTkvgKgI0pMTFRDQ4Oqqqp03XXXnXFMSUmJJk6cqHvvvVeS1NjYqL179+ryyy/3WI7a2lpVVFRo6tSpHlunCbxWasrKylRVVaXhw4e75jU0NGjTpk167rnn5HQ6m33zFAAAnlZeXt7kfhXLspSQkKApU6Zo2rRpWrZsmRITE/Xdd99pw4YNio+P14QJExQbG6s1a9Zo8+bN6tatm5YvX65Dhw5dUKnJysrS7bffrv79++vAgQOaP3++fHx8NHnyZE/sqjG8VmpuuukmlZeXN5l33333afDgwZozZw6FBgDgVaNGjWoy7ePjoxMnTignJ0cLFy7Uww8/rP3796tHjx66+uqrddttt0mS5s2bp3379mns2LEKDAxUWlqaUlNTVV1d3eIs3377rSZPnqzDhw/rkksu0bXXXqstW7bokksuuaB9NI1lu3vrdysaPXq0hg0bphUrVjRrfE1NjcLCwlRdXa3Q0NDWDQcAADyitd6/L5ov3wMAALgQXv9I96lO/UZGAAA6AofDcc77bXbt2sWX7DXTRVVqAADoaCIjI8/5K90nv+QP50epAQDAi3x9fTVgwABvxzAC99QAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgAAYARKDQAAMAKlBgAAGIFSAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqUGAAAYgVIDAACMQKkBAABGoNQAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAiUGgAAYARKDQAAMAKlBgAAGIFSAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqUGAAAYgVIDAACMQKkBAABGoNQAAAAjUGoAAIARKDUAAMAIlBoAAGAESg0AADACpQYAABjB19sBLoRt25KkmpoaLycBAADNdfJ9++T7uKe061Jz9OhRSVLfvn29nAQAALjr8OHDCgsL89j6LNvTNakNNTY26sCBAwoJCZFlWRe0rpqaGvXt21fffPONQkNDPZQQzcXx9x6Ovfdw7L2L4+891dXV6tevn3744Qd17drVY+tt12dqOnXqpD59+nh0naGhoTy5vYjj7z0ce+/h2HsXx997OnXy7K293CgMAACMQKkBAABGoNT8H39/f82fP1/+/v7ejtIhcfy9h2PvPRx77+L4e09rHft2faMwAADASZypAQAARqDUAAAAI1BqAACAESg1AADACB2q1Dz//POKjo5Wly5dlJycrE8//fSc49966y0NHjxYXbp0UVxcnP785z+3UVIzuXP8c3NzZVlWk78uXbq0YVpzbNq0SbfffrsiIyNlWZbeeeed8y5TXFys4cOHy9/fXwMGDFBubm6r5zSRu8e+uLj4Z897y7J08ODBtglskMWLF+uqq65SSEiIevbsqdTUVO3Zs+e8y/G6f+Facuw99ZrfYUrNm2++qdmzZ2v+/PnaunWrEhISNHbsWFVVVZ1x/ObNmzV58mTdf//92rZtm1JTU5WamqodO3a0cXIzuHv8pZ++5bOystL19/XXX7dhYnPU1dUpISFBzz//fLPGf/nll5owYYJuuOEGbd++XZmZmfrVr36lwsLCVk5qHneP/Ul79uxp8tzv2bNnKyU018aNG5Wenq4tW7aoqKhIx48f1y233KK6urqzLsPrvme05NhLHnrNtzuIpKQkOz093TXd0NBgR0ZG2osXLz7j+EmTJtkTJkxoMi85Odn+9a9/3ao5TeXu8c/JybHDwsLaKF3HIckuKCg455hHHnnEHjp0aJN5d999tz127NhWTGa+5hz7Dz74wJZk//DDD22SqSOpqqqyJdkbN2486xhe91tHc469p17zO8SZmvr6epWVlWnMmDGueZ06ddKYMWP08ccfn3GZjz/+uMl4SRo7duxZx+PsWnL8Jam2tlb9+/dX3759NXHiRO3cubMt4nZ4PPe9b9iwYerdu7duvvlmlZSUeDuOEaqrqyVJ4eHhZx3Dc791NOfYS555ze8Qpebvf/+7Ghoa1KtXrybze/XqddZr1QcPHnRrPM6uJcd/0KBB+vd//3e9++67+o//+A81NjZq5MiR+vbbb9sicod2tud+TU2N/vGPf3gpVcfQu3dvvfjii3r77bf19ttvq2/fvho9erS2bt3q7WjtWmNjozIzM3XNNdfoiiuuOOs4Xvc9r7nH3lOv+e36V7phrpSUFKWkpLimR44cqSFDhuill17SH/7wBy8mA1rPoEGDNGjQINf0yJEjVVFRoWeeeUavv/66F5O1b+np6dqxY4c++ugjb0fpcJp77D31mt8hztT06NFDPj4+OnToUJP5hw4dUkRExBmXiYiIcGs8zq4lx/90nTt3VmJior744ovWiIhTnO25HxoaqoCAAC+l6riSkpJ43l+AjIwMrV27Vh988IH69OlzzrG87nuWO8f+dC19ze8QpcbPz08jRozQhg0bXPMaGxu1YcOGJs3wVCkpKU3GS1JRUdFZx+PsWnL8T9fQ0KDy8nL17t27tWLi//Dcv7hs376d530L2LatjIwMFRQU6P3339ell1563mV47ntGS4796Vr8mn/Btxq3E3l5eba/v7+dm5tr79q1y05LS7O7du1qHzx40LZt2546dar96KOPusaXlJTYvr6+9tKlS+3du3fb8+fPtzt37myXl5d7axfaNXeP/4IFC+zCwkK7oqLCLisrs++55x67S5cu9s6dO721C+3W0aNH7W3bttnbtm2zJdnLly+3t23bZn/99de2bdv2o48+ak+dOtU1ft++fXZgYKD9u9/9zt69e7f9/PPP2z4+Pvb69eu9tQvtlrvH/plnnrHfeecd+/PPP7fLy8vtWbNm2Z06dbLfe+89b+1Cu/Xb3/7WDgsLs4uLi+3KykrX348//ugaw+t+62jJsffUa36HKTW2bdvPPvus3a9fP9vPz89OSkqyt2zZ4nrs+uuvt6dPn95kfH5+vj1w4EDbz8/PHjp0qL1u3bo2TmwWd45/Zmama2yvXr3sW2+91d66dasXUrd/Jz8mfPrfyeM9ffp0+/rrr//ZMsOGDbP9/Pzsyy67zM7JyWnz3CZw99j/y7/8ix0TE2N36dLFDg8Pt0ePHm2///773gnfzp3puEtq8lzmdb91tOTYe+o13/q/AAAAAO1ah7inBgAAmI9SAwAAjECpAQAARqDUAAAAI1BqAACAESg1AADACJQaAABgBEoNAAAwAqUGMNzo0aOVmZnp7Rhn9NVXX8myLG3fvt2t5SzL0jvvvNMqmdz15JNPatiwYd6OAUCUGgBotoupTAH4OUoNgJ+pr6/3dgQAcBulBugATpw4oYyMDIWFhalHjx56/PHHderPvkVHR+sPf/iDpk2bptDQUKWlpUmS5syZo4EDByowMFCXXXaZHn/8cR0/fty13MlLL6+//rqio6MVFhame+65R0ePHnWNaWxs1L/+679qwIAB8vf3V79+/fTUU081ybdv3z7dcMMNCgwMVEJCgj7++GO39u+bb77RpEmT1LVrV4WHh2vixIn66quvXI/PmDFDqampWrp0qXr37q3u3bsrPT29yb5UVlZqwoQJCggI0KWXXqpVq1YpOjpaK1ascB0jSbrzzjtlWZZr+qRzHQMAbYNSA3QAr732mnx9ffXpp58qOztby5cv1yuvvNJkzNKlS5WQkKBt27bp8ccflySFhIQoNzdXu3btUnZ2tl5++WU988wzTZarqKjQO++8o7Vr12rt2rXauHGjlixZ4np87ty5WrJkiR5//HHt2rVLq1atUq9evZqs47HHHlNWVpa2b9+ugQMHavLkyTpx4kSz9u348eMaO3asQkJC9OGHH6qkpETBwcEaN25ckzNOH3zwgSoqKvTBBx/otddeU25urnJzc12PT5s2TQcOHFBxcbHefvttrVy5UlVVVa7HS0tLJUk5OTmqrKx0TTfnGABoIxf6E+MALm7XX3+9PWTIELuxsdE1b86cOfaQIUNc0/3797dTU1PPu66nn37aHjFihGt6/vz5dmBgoF1TU+Oa97vf/c5OTk62bdu2a2pqbH9/f/vll18+4/q+/PJLW5L9yiuvuObt3LnTlmTv3r37rDkk2QUFBbZt2/brr79uDxo0qMn+OZ1OOyAgwC4sLLRt27anT59u9+/f3z5x4oRrzF133WXffffdtm3b9u7du21Jdmlpqevxzz//3JZkP/PMM2fcbnOPAYC2w5kaoAO4+uqrZVmWazolJUWff/65GhoaXPOuvPLKny335ptv6pprrlFERISCg4M1b948ORyOJmOio6MVEhLimu7du7frDMfu3bvldDp10003nTNffHx8k+UlNTlLci5//etf9cUXXygkJETBwcEKDg5WeHi4jh07poqKCte4oUOHysfH54w59+zZI19fXw0fPtz1+IABA9StW7dmZTjXMQDQdny9HQDAxSEoKKjJ9Mcff6wpU6ZowYIFGjt2rMLCwpSXl6dly5Y1Gde5c+cm05ZlqbGxUZIUEBDQrG2fuo6T5evkOs6ntrZWI0aM0BtvvPGzxy655JJm5bxQrbluAM1HqQE6gE8++aTJ9JYtWxQbG9vkzMXpNm/erP79++uxxx5zzfv666/d2m5sbKwCAgK0YcMG/epXv3IvdDMNHz5cb775pnr27KnQ0NAWrWPQoEE6ceKEtm3bphEjRkiSvvjiC/3www9NxnXu3LnJ2S0AFxcuPwEdgMPh0OzZs7Vnzx6tXr1azz77rGbNmnXOZWJjY+VwOJSXl6eKigr98Y9/VEFBgVvb7dKli+bMmaNHHnlEf/rTn1RRUaEtW7bo1VdfvZDdaWLKlCnq0aOHJk6cqA8//FBffvmliouL9eCDD+rbb79t1joGDx6sMWPGKC0tTZ9++qm2bdumtLQ0BQQENLlsFx0drQ0bNujgwYM/KzwAvI9SA3QA06ZN0z/+8Q8lJSUpPT1ds2bNcn1s+2zuuOMOPfTQQ8rIyNCwYcO0efNm16ei3PH444/r4Ycf1hNPPKEhQ4bo7rvv9uj9JoGBgdq0aZP69eunX/ziFxoyZIjuv/9+HTt2zK0zN3/605/Uq1cvjRo1SnfeeaceeOABhYSEqEuXLq4xy5YtU1FRkfr27avExESP7QMAz7Bs+5QvqwAASJK+/fZb9e3bV++99955b3QGcHGg1ACApPfff1+1tbWKi4tTZWWlHnnkEe3fv1979+792Y3AAC5O3CgMAPrpS/z++Z//Wfv27VNISIhGjhypN954g0IDtCOcqQEAAEbgRmEAAGAESg0AADACpQYAABiBUgMAAIxAqQEAAEag1AAAACNQagAAgBEoNQAAwAj/DzmlPS6XxppnAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "newick_tree = open(\"./newick.txt\", \"r\")\n",
    "newick = newick_tree.read()\n",
    "newick_tree.close()\n",
    "tree = Phylo.read(StringIO(newick), \"newick\")\n",
    "\n",
    "# Change nodes names\n",
    "for idx, clade in enumerate(tree.get_nonterminals()):\n",
    "    clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "\n",
    "# Change leaves names\n",
    "for idx, clade in enumerate(tree.get_terminals()):\n",
    "    clade.name = \"Leaf_\" + clade.name \n",
    "\n",
    "Phylo.draw(tree)\n",
    "\n",
    "net = Phylo.to_networkx(tree) # to graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "simulate() missing 1 required positional argument: 'directory'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m edges, node_distr, filename \u001b[39m=\u001b[39m simulate(net, \u001b[39m500\u001b[39;49m) \u001b[39m# simulate data calling the module data_simulation.py \u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[39m# Preprocess leave alignments\u001b[39;00m\n\u001b[1;32m      4\u001b[0m leave_alignments \u001b[39m=\u001b[39m [i \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m SeqIO\u001b[39m.\u001b[39mparse(filename, \u001b[39m\"\u001b[39m\u001b[39mfasta\u001b[39m\u001b[39m\"\u001b[39m)]\n",
      "\u001b[0;31mTypeError\u001b[0m: simulate() missing 1 required positional argument: 'directory'"
     ]
    }
   ],
   "source": [
    "edges, node_distr, filename = simulate(net, 500) # simulate data calling the module data_simulation.py \n",
    "\n",
    "# Preprocess leave alignments\n",
    "leave_alignments = [i for i in SeqIO.parse(filename, \"fasta\")]\n",
    "for i in leave_alignments:\n",
    "    print(\"Seq>\", i.seq)\n",
    "# Codify the alphabet {A,G,C,T} -> {0,1,2,3}\n",
    "leave_sequences = []\n",
    "for i in leave_alignments:\n",
    "    seq = \"\"\n",
    "    for j in i.seq:\n",
    "        if j == \"A\":\n",
    "            seq += \"0\"\n",
    "        elif j == \"G\":\n",
    "            seq += \"1\"\n",
    "        elif j == \"C\":\n",
    "            seq += \"2\"\n",
    "        elif j == \"T\":\n",
    "            seq += \"3\"\n",
    "        else:\n",
    "            raise ValueError(\"Invalid nucleotide\")\n",
    "    leave_sequences.append(seq)\n",
    "#leave_sequences\n",
    "edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## **EM algorithm**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from evaluate import init_EM\n",
    "\n",
    "Results = init_EM(newick, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Source: Int_0\n",
      "Target: Leaf_2\n",
      "Matrix:\n",
      " [[0.80529814 0.06095358 0.09251042 0.04123786]\n",
      " [0.05611119 0.71483627 0.11102639 0.11802615]\n",
      " [0.0952762  0.07415636 0.731184   0.09938344]\n",
      " [0.02196288 0.0820518  0.08234946 0.81363586]]\n",
      "Source: Int_0\n",
      "Target: Leaf_3\n",
      "Matrix:\n",
      " [[0.63632807 0.06782976 0.1446222  0.15121997]\n",
      " [0.12810897 0.67208297 0.04706075 0.15274731]\n",
      " [0.10204507 0.09021686 0.74449869 0.06323937]\n",
      " [0.06570111 0.08178298 0.05155947 0.80095644]]\n",
      "Source: Int_0\n",
      "Target: Int_1\n",
      "Matrix:\n",
      " [[0.72577892 0.07421689 0.0831795  0.11682468]\n",
      " [0.08700214 0.7261134  0.06686419 0.12002027]\n",
      " [0.09519544 0.182137   0.68012648 0.04254107]\n",
      " [0.06761601 0.06509887 0.08961157 0.77767356]]\n",
      "Source: Int_1\n",
      "Target: Leaf_4\n",
      "Matrix:\n",
      " [[0.85639817 0.07234946 0.04036045 0.03089192]\n",
      " [0.0298384  0.73128255 0.0944944  0.14438465]\n",
      " [0.03533407 0.10420379 0.77836777 0.08209437]\n",
      " [0.06403097 0.03195277 0.06966545 0.8343508 ]]\n",
      "Source: Int_1\n",
      "Target: Leaf_5\n",
      "Matrix:\n",
      " [[0.80075352 0.06438893 0.11198018 0.02287738]\n",
      " [0.07400816 0.72459648 0.10712499 0.09427037]\n",
      " [0.14358855 0.08350079 0.74500083 0.02790983]\n",
      " [0.01210798 0.14166184 0.06442053 0.78180965]]\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(Results)):\n",
    "    print(\"Source:\", Results[i].source)\n",
    "    print(\"Target:\", Results[i].target)\n",
    "    print(\"Matrix:\\n\", Results[i].matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "l -73482.53888183684\n",
      "73482.53888183684\n",
      "--------\n",
      "2803.2235244028125\n",
      "--------\n",
      "403.42311980716477\n",
      "--------\n",
      "147.34679458629398\n",
      "--------\n",
      "65.81431201811938\n",
      "--------\n",
      "32.62458047835389\n",
      "--------\n",
      "17.4679848113592\n",
      "--------\n",
      "9.96598948755127\n",
      "--------\n",
      "5.979527266841615\n",
      "--------\n",
      "3.724567122422741\n",
      "--------\n",
      "2.382098244605004\n",
      "--------\n",
      "1.5508962869207608\n",
      "--------\n",
      "1.021128828782821\n",
      "--------\n",
      "0.6763189632765716\n",
      "--------\n",
      "0.44849936086393427\n",
      "--------\n",
      "0.29640033769828733\n",
      "--------\n",
      "0.19417478849936742\n",
      "--------\n",
      "0.12524008554464672\n",
      "--------\n",
      "0.07875408534891903\n",
      "--------\n",
      "0.047520216307020746\n",
      "--------\n",
      "0.02670328039675951\n",
      "--------\n",
      "0.01302253625181038\n",
      "--------\n",
      "0.004234299674862996\n",
      "--------\n",
      "0.0012054162216372788\n",
      "--------\n",
      "0.004364024440292269\n",
      "--------\n",
      "0.005981738388072699\n",
      "--------\n",
      "0.006572629208676517\n",
      "--------\n",
      "0.006493631357443519\n",
      "--------\n",
      "0.005992020829580724\n",
      "--------\n",
      "0.005238327459665015\n",
      "--------\n",
      "0.00434925795707386\n",
      "--------\n",
      "0.003403757931664586\n",
      "--------\n",
      "0.002454297398799099\n",
      "--------\n",
      "0.0015348267188528553\n",
      "--------\n",
      "Done!\n",
      "Number of iterations:  34\n"
     ]
    }
   ],
   "source": [
    "# Auxiliary functions\n",
    "\n",
    "def init_root_distribution():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the root distribution (all probabilities between 0.2 and 0.3)\n",
    "    \"\"\"\n",
    "    estimated_root_distribution = []\n",
    "    while True:\n",
    "        estimated_root_distribution = np.random.dirichlet([1,1,1,1])\n",
    "        i = 0\n",
    "        while i<4:\n",
    "            if estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3:\n",
    "                i += 1\n",
    "            else:\n",
    "                break\n",
    "        if i == 4:\n",
    "            assert(np.sum(estimated_root_distribution) < 1.0000001 and np.sum(estimated_root_distribution) > 0.9999999)\n",
    "            assert(estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3 for i in range(4))\n",
    "            break\n",
    "    return estimated_root_distribution\n",
    "\n",
    "def init_transition_matrix():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the transition matrices (which we estimate as LARGE-DIAGONAL-VALUE matrices)\n",
    "    \"\"\"\n",
    "    M = np.zeros((4,4))\n",
    "    i=0\n",
    "    while i<4:\n",
    "        R = np.random.dirichlet([1,1,1,1])\n",
    "        if R[i] > 0.6 and R[i] < 0.8:\n",
    "            M[i,:] = R \n",
    "            i = i + 1\n",
    "    assert (M[i,i] > 0.6 and M[i,i] < 0.8 for i in range(4))\n",
    "    assert (np.sum(M[i,:]) < 1.0000001 and np.sum(M[i,:]) > 0.9999999 for i in range(4))\n",
    "    return M\n",
    "\n",
    "def log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "    \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "    logL = 0\n",
    "    for obs in states:\n",
    "        p_i = 0 \n",
    "        observed_data = obs[n_int:]\n",
    "        if observed_data in u_i:\n",
    "            for state in states:\n",
    "                if state[n_int:] == observed_data:\n",
    "                    pi = root_distribution[int(state[0])]\n",
    "                    for p in params.items():\n",
    "                        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                        pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                    p_i += pi\n",
    "        logL += (u_i[observed_data] * math.log(p_i))\n",
    "    return logL\n",
    "\n",
    "############################################################################################################\n",
    "########################################## FELSENSTEIN ALGORITHM ###########################################\n",
    "############################################################################################################\n",
    "# # Felsentstein algorithm\n",
    "# # \"((4,5),(2,3))\"\n",
    "# newick_tree = \"(((4:1,5:1):1,(6:1,7:1):1):1,(8:1,9:1):1);\"\n",
    "# L_tree = Phylo.read(StringIO(newick_tree), \"newick\")\n",
    "# # Change nodes names\n",
    "# for idx, clade in enumerate(L_tree.get_nonterminals()):\n",
    "#     id = 0\n",
    "#     if idx == 2:\n",
    "#         clade.name = \"Int_Id_\" + str(id)\n",
    "#         id += 1\n",
    "#     else:\n",
    "#         clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "# # Change leaves names\n",
    "# for idx, clade in enumerate(L_tree.get_terminals()):\n",
    "#     clade.name = \"Leaf_\" + clade.name \n",
    "# L_net = Phylo.to_networkx(L_tree) # to graph\n",
    "# L_edges = []\n",
    "# for edge in L_net.edges():\n",
    "#     if edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     elif edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] != \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     elif edge[0].name.split(\"_\")[1] != \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     else:\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     L_edges.append((u,v))\n",
    "# # Create a dictionary to store the father and son information\n",
    "# family_tree = {}\n",
    "# # Loop through the edges and update the dictionary\n",
    "# for father, son in L_edges:\n",
    "#     if father not in family_tree:\n",
    "#         family_tree[father] = []\n",
    "#     family_tree[father].append(son)\n",
    "# # Create a DataFrame from the dictionary\n",
    "# tree_df = pd.DataFrame.from_dict(family_tree, orient='index')\n",
    "# # Fill NaN values with empty strings\n",
    "# tree_df = tree_df.fillna('')\n",
    "# # Change column names\n",
    "# tree_df.columns = ['child_1', 'child_2']\n",
    "\n",
    "# class TreeInference():\n",
    "#     def __init__(self, tree_data, leaf_values, params, root_distr):\n",
    "#         # Set parameters\n",
    "#         self.tree_data = tree_data.copy()\n",
    "#         self.leaf_values = leaf_values\n",
    "#         self.params = params\n",
    "#         self.root_distr = root_distr\n",
    "\n",
    "#         # Find IDs of internal nodes, all nodes and leaf nodes\n",
    "#         self.internal_node_ids = np.array(tree_data.index)\n",
    "#         self.all_node_ids = np.unique(np.concatenate((np.array(self.tree_data.index),\n",
    "#                                                       np.array(self.tree_data[\"child_1\"]),\n",
    "#                                                       np.array(self.tree_data[\"child_2\"]))))\n",
    "#         self.leaf_node_ids = list(set(self.all_node_ids) - set(self.internal_node_ids))\n",
    "        \n",
    "#         # Count internal nodes and leaves for each row in tree_data and set root node\n",
    "#         self.treeNodeCounts()\n",
    "#         self.root_node = 0\n",
    "        \n",
    "#         # Get indexes for matricies for states, e.g. A is row 0, G is row 2...\n",
    "#         states = [\"A\", \"G\", \"C\", \"T\"]\n",
    "#         self.matrix_idx = dict(zip(states, range(len(states))))\n",
    "        \n",
    "#         # Empty matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "#         # Log likelihood of whole tree\n",
    "#         self.L_root = None\n",
    "        \n",
    "\n",
    "#     \"\"\"\n",
    "#     Calculates the number of internal nodes and leaves below each internal node.\n",
    "#     This works by recursively iterating down the child nodes from the original internal node until a\n",
    "#     leaf node is found, and then returning upwards counting the leaves and nodes.\n",
    "#     The new derived information is then added as two new columns at the end of the data.\n",
    "#     \"\"\"\n",
    "#     def treeNodeCounts(self):\n",
    "#         n_nodes = np.array([])\n",
    "#         n_leaves = np.array([])\n",
    "\n",
    "#         # Set the number of internal nodes and leaves for each internal node row\n",
    "#         for n in self.internal_node_ids:\n",
    "#             num_nodes, num_leaves = self.checkNode(self.tree_data, n, self.internal_node_ids)\n",
    "#             n_nodes = np.append(n_nodes, num_nodes)\n",
    "#             n_leaves = np.append(n_leaves, num_leaves)\n",
    "\n",
    "#         # Add new information to dataframe\n",
    "#         self.tree_data[\"N_Internal\"] = n_nodes\n",
    "#         self.tree_data[\"N_Leaves\"] = n_leaves\n",
    "                \n",
    "#     # Recursively find the number of internal nodes and leaves given a node ID\n",
    "#     def checkNode(self, tree_data, node_id, node_array):\n",
    "#         # Record the number of internal nodes and leaves below\n",
    "#         num_nodes = 0\n",
    "#         num_leaves = 0\n",
    "\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[node_id][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[node_id][\"child_2\"]\n",
    "\n",
    "#         # Check if found an internal node to the left\n",
    "#         if child_1 in node_array:\n",
    "#             # Recursively check nodes below\n",
    "#             d1_nodes, d1_leaves = self.checkNode(self.tree_data, child_1, node_array)\n",
    "#             num_nodes += d1_nodes + 1\n",
    "#             num_leaves += d1_leaves\n",
    "\n",
    "#         # Check if found an internal node to the right\n",
    "#         if child_2 in node_array:\n",
    "#             d2_nodes, d2_leaves = self.checkNode(self.tree_data, child_2, node_array)\n",
    "#             num_nodes += d2_nodes + 1\n",
    "#             num_leaves += d2_leaves\n",
    "\n",
    "#         return num_nodes, num_leaves\n",
    "    \n",
    "    \n",
    "#     # Calculate likelihood for an internal node\n",
    "#     def nodeLikelihood(self, node_n, params):\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[str(node_n)][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[str(node_n)][\"child_2\"]\n",
    "        \n",
    "#         # Check if likelihood of left child node set\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(child_1, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_1), 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(int(child_1), params)\n",
    "#         # Check if likelihood of right child node set\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 self.nodeLikelihood(child_2, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_2), 0] is None):\n",
    "#                 self.nodeLikelihood(int(child_2), params)\n",
    "        \n",
    "#         M1, M2 = np.full((4,4),None), np.full((4,4),None)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             # Create transition matrix for left and right child nodes\n",
    "#             matrix_name_child1 = \"M_0_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_0_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_1 == \"Id_0\":\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = np.identity(4)\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_2 == \"Id_0\":\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = np.identity(4)\n",
    "#         else:\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "\n",
    "#         # Get conditional likelihood of child nodes\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             d1_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d1_likelihood = self.L[int(child_1)]\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             d2_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d2_likelihood = self.L[int(child_2)]\n",
    "        \n",
    "#         # Calculate the likelihood of each state for the node\n",
    "#         left_prob = np.sum(M1 * d1_likelihood, axis = 1)\n",
    "#         right_prob = np.sum(M2 * d2_likelihood, axis = 1)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             self.L[-1,:] = np.multiply(left_prob, right_prob)\n",
    "#         else:\n",
    "#             self.L[int(node_n),:] = np.multiply(left_prob, right_prob)\n",
    "    \n",
    "#     # Run Felsenstein‚Äôs algorithm to calculate the log likelihood of the tree\n",
    "#     def calculateLikelihood(self, params, Estep):\n",
    "#         # Clear matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "        \n",
    "#         # Set likelihood of leaf states\n",
    "#         for leaf_n in self.leaf_node_ids:\n",
    "#             if leaf_n == \"Id_0\":\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[-1, :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[-1, self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "#             else:\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[int(leaf_n), :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[int(leaf_n), self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "        \n",
    "#         for node_n in self.internal_node_ids:\n",
    "#             # Check likelihood has not been set\n",
    "#             if node_n == \"Id_0\":\n",
    "#                 if self.L[-1, 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "#             else:\n",
    "#                 if self.L[int(node_n), 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "        \n",
    "#         if Estep == True:\n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = np.sum(np.multiply(self.root_distr, self.L[self.root_node]))\n",
    "#         else: \n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = math.log(np.sum(np.multiply(self.root_distr, self.L[self.root_node])))\n",
    "        \n",
    "#         return self.L_root\n",
    "\n",
    "# def Felsenstein_log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "#     \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "#     logL = 0\n",
    "#     for obs in states:\n",
    "#         observed_data = obs[n_int:]\n",
    "#         tree_leaves = dict()\n",
    "#         for i in range(len(observed_data)):\n",
    "#             if observed_data[i] == str(0):\n",
    "#                 tree_leaves[n_int+i] = \"A\"\n",
    "#             elif observed_data[i] == str(1):\n",
    "#                 tree_leaves[n_int+i] = \"G\"\n",
    "#             elif observed_data[i] == str(2):\n",
    "#                 tree_leaves[n_int+i] = \"C\"\n",
    "#             elif observed_data[i] == str(3):\n",
    "#                 tree_leaves[n_int+i] = \"T\"\n",
    "#         tree_inference = TreeInference(tree_df, tree_leaves, params, root_distribution)\n",
    "#         tree_inference.calculateLikelihood(params, False)\n",
    "#         logL += (u_i[observed_data] * tree_inference.L_root)\n",
    "#         del tree_inference\n",
    "#     return logL\n",
    "\n",
    "\n",
    "############################################################################################################\n",
    "\n",
    "class Param:\n",
    "    \"\"\"\n",
    "    Class to store the parameters of the tree\n",
    "    \"\"\"\n",
    "    def __init__(self, edge, transition_matrix, alignment=None):\n",
    "        self.edge = edge\n",
    "        self.transition_matrix = transition_matrix\n",
    "        self.alignment = alignment # it will only be placed in the leaves\n",
    "\n",
    "# Obtain number of leaves and internal nodes\n",
    "n_leaves, n_int = 0, 0 \n",
    "for node in net.nodes():\n",
    "    if node.name.startswith(\"L\"):\n",
    "        n_leaves += 1\n",
    "    else:\n",
    "        n_int += 1\n",
    "\n",
    "# Count all the leaves ocurrences \n",
    "sequence_length = len(leave_sequences[0])\n",
    "number_of_sequences = len(leave_sequences)\n",
    "occurrences = []\n",
    "for i in range(sequence_length):\n",
    "    _ = ''\n",
    "    for j in range(number_of_sequences):\n",
    "        _ += leave_sequences[j][i]\n",
    "    occurrences.append(_)\n",
    "c = Counter(occurrences)\n",
    "u_i = OrderedDict(sorted(c.items()))\n",
    "\n",
    "# Define all possible (hidden) states of the internal nodes \n",
    "hidden_combinations = itertools.product([0,1,2,3], repeat=n_int)\n",
    "sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "for i in range(len(sorted_hidden_combinations)):\n",
    "    sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    " \n",
    "# Define all possible fully observed states (i.e. combining u_i + internal nodes)\n",
    "# '000000' would mean to have an 'A' in all nodes: Int_0, Int_1, Leaf_2, Leaf_3, Leaf_4, Leaf_5 \n",
    "states = list(itertools.product(list(sorted_hidden_combinations), list(u_i.keys())))\n",
    "states = [i[0]+i[1] for i in states]\n",
    "\n",
    "############################################################################################################\n",
    "###################################### STEP 0: Initialise parameters #######################################\n",
    "############################################################################################################\n",
    "estimated_root_distribution = init_root_distribution() # First estimation for the root distribution\n",
    "\n",
    "# PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "# for edge in net.edges():\n",
    "#     # If we are in a leaf, we need to specify the alignment\n",
    "#     iter_leaves = 0\n",
    "#     if edge[1].name.startswith(\"L\"):\n",
    "#         new_edge = Param(edge, init_transition_matrix(), leave_alignments[iter_leaves].seq)\n",
    "#         iter_leaves += 1\n",
    "#     # Otherwise, are in internal nodes\n",
    "#     else:\n",
    "#         new_edge = Param(edge, init_transition_matrix())\n",
    "#     u = str(edge[0].name.split(\"_\")[1])\n",
    "#     v = str(edge[1].name.split(\"_\")[1])\n",
    "#     name = \"M_\" + u + \"_to_\" + v\n",
    "#     PARAMS[name] = new_edge\n",
    "\n",
    "# ADRIA CODE\n",
    "i=0\n",
    "PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "for edge in net.edges():\n",
    "    # If we are in a leaf, we need to specify the alignment\n",
    "    iter_leaves = 0\n",
    "    if edge[1].name.startswith(\"L\"):\n",
    "        new_edge = Param(edge, Results[i].matrix, leave_alignments[iter_leaves].seq)\n",
    "        iter_leaves += 1\n",
    "    # Otherwise, are in internal nodes\n",
    "    else:\n",
    "        new_edge = Param(edge, Results[i].matrix)\n",
    "    u = str(edge[0].name.split(\"_\")[1])\n",
    "    v = str(edge[1].name.split(\"_\")[1])\n",
    "    name = \"M_\" + u + \"_to_\" + v\n",
    "    PARAMS[name] = new_edge\n",
    "    i += 1\n",
    "\n",
    "eps = 10**(-3)                  # epsilon for convergence\n",
    "iter = 0                        # iteration counter \n",
    "params = copy.copy(PARAMS)      # copy of the parameters (will be updated at each iteration)\n",
    "# logF = Felsenstein_log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "logL = log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "# print(logF)\n",
    "print(\"l\", logL)\n",
    "logL_ = 0\n",
    "\n",
    "while np.abs(logL_ - logL) > eps:\n",
    "    print(np.abs(logL_ - logL))\n",
    "    # print(np.abs(logF_ - logF))\n",
    "    print(\"--------\")\n",
    "    if iter > 0:\n",
    "        logL = logL_\n",
    "        # logF = logF_\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## E-step ##################################################\n",
    "    ############################################################################################################\n",
    "    # Define the expected hidden data matrix\n",
    "    U = dict()\n",
    "    root_distr = estimated_root_distribution ########\n",
    "\n",
    "    for obs in states:\n",
    "        U[obs] = u_i[obs[n_int:]] # initialise by computing u_i of the expected hidden data matrix formula\n",
    "        f_ij = estimated_root_distribution[int(obs[0])] \n",
    "        for p in params.items():\n",
    "            u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "            f_ij *= p[1].transition_matrix[int(obs[u]), int(obs[v])]\n",
    "        # Now f_ij := numerator of the expected hidden data matrix formula \n",
    "\n",
    "        # We need to marginalise to obtain f_i := denominator of the expected hidden data matrix formula\n",
    "        f_i = 0 # we initialise the partial data model\n",
    "        observed_data = obs[n_int:]\n",
    "        for state in states:\n",
    "            if state[n_int:] == observed_data:\n",
    "                pi = estimated_root_distribution[int(state[0])]\n",
    "                for p in params.items():\n",
    "                    u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                    pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                f_i += pi\n",
    "\n",
    "        # observed_data = obs[n_int:]\n",
    "        # tree_leaves = dict()\n",
    "        # for i in range(len(observed_data)):\n",
    "        #     if observed_data[i] == str(0):\n",
    "        #         tree_leaves[n_int+i] = \"A\"\n",
    "        #     elif observed_data[i] == str(1):\n",
    "        #         tree_leaves[n_int+i] = \"G\"\n",
    "        #     elif observed_data[i] == str(2):\n",
    "        #         tree_leaves[n_int+i] = \"C\"\n",
    "        #     elif observed_data[i] == str(3):\n",
    "        #         tree_leaves[n_int+i] = \"T\"\n",
    "        # tree_inference = TreeInference(tree_df, tree_leaves, params, root_distr)\n",
    "        # tree_inference.calculateLikelihood(params, True)\n",
    "\n",
    "        U[obs] *= f_ij/f_i\n",
    "        # U[obs] *= f_ij/tree_inference.L_root\n",
    "        # del tree_inference\n",
    "        # del tree_leaves\n",
    "\n",
    "\n",
    "    assert(len(U) <= 4**(n_int+n_leaves))\n",
    "    sum = 0\n",
    "    sum2 = 0\n",
    "    for i in U.items():\n",
    "        sum += i[1]\n",
    "    for i in u_i.items():\n",
    "        sum2 += i[1]\n",
    "    assert(sum < sequence_length+0.001 and sum > sequence_length-0.001)\n",
    "    assert(sum2 < sequence_length+0.001 and sum2 > sequence_length-0.001)\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## M-step ##################################################\n",
    "    ############################################################################################################\n",
    "    estimated_parameters = dict()\n",
    "    root_distr = np.zeros(4)\n",
    "    iter_root = 0 # counter for the root distribution, we only want to count once\n",
    "\n",
    "    for p in params.items():\n",
    "        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "        name = \"M_\" + str(u) + \"_to_\" + str(v)\n",
    "        estimated_parameters[name] = np.zeros((4,4))\n",
    "        for i in range(4):\n",
    "            for j in range(4):\n",
    "                ui = 0\n",
    "                for obs in U.items():\n",
    "                    # For the root distribution, we only need to consider the first internal node\n",
    "                    if iter_root == 0 and obs[0][0] == str(i):\n",
    "                        root_distr[i] += obs[1]\n",
    "                    # For the transition matrices:\n",
    "                    if obs[0][u] == str(i) and obs[0][v] == str(j):\n",
    "                        estimated_parameters[name][i][j] += obs[1]\n",
    "                    if obs[0][u] == str(i):\n",
    "                        ui += obs[1]\n",
    "                # Update root distribution\n",
    "                if iter_root == 0:\n",
    "                    root_distr[i] /= sequence_length\n",
    "                iter_root = 1\n",
    "                estimated_parameters[name][i][j] /= ui\n",
    "            iter_root = 0\n",
    "            \n",
    "        # Update parameters\n",
    "        params[name].transition_matrix = estimated_parameters[name]\n",
    "\n",
    "    ##########################################################################################################\n",
    "    ########################### Compute log-likelihood of the estimated parameters ###########################\n",
    "    ##########################################################################################################\n",
    "    # logF_ = Felsenstein_log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "    logL_ = log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "\n",
    "    iter += 1\n",
    "\n",
    "print(\"Done!\")\n",
    "print(\"Number of iterations: \", iter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################### DEBUGGING ############################\n",
    "\n",
    "# # All possible (theoretical) occurrences of nucleotides in the leaves\n",
    "# alphabet = [0, 1, 2, 3]\n",
    "# combinations = itertools.product(alphabet, repeat=n_leaves)\n",
    "# sorted_combinations = sorted(combinations)\n",
    "# for i in range(len(sorted_combinations)):\n",
    "#     sorted_combinations[i] = ''.join([str(s) for s in sorted_combinations[i]])\n",
    "\n",
    "# # All possible (hidden) states of the internal nodes (i.e. indexes of columns of U)\n",
    "# hidden_combinations = itertools.product(alphabet, repeat=n_int)\n",
    "# sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "# for i in range(len(sorted_hidden_combinations)):\n",
    "#     sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    "\n",
    "# # To check if an element is within a dictionary or not \n",
    "# if '0101' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')\n",
    "# if '0000' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix: [[0.64642183 0.02803076 0.27483013 0.05071728]\n",
      " [0.03848122 0.7757021  0.10189982 0.08391686]\n",
      " [0.03184807 0.07934462 0.82109495 0.06771236]\n",
      " [0.04702478 0.01879095 0.04076404 0.89342023]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix: [[0.55850753 0.13612552 0.2499751  0.05539185]\n",
      " [0.01791204 0.82605787 0.05907209 0.096958  ]\n",
      " [0.04534628 0.008074   0.85149918 0.09508054]\n",
      " [0.00646156 0.06355415 0.02348564 0.90649866]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix: [[0.75152183 0.09827146 0.07355421 0.0766525 ]\n",
      " [0.04888819 0.86735324 0.02989229 0.05386628]\n",
      " [0.05446854 0.02450495 0.75018255 0.17084396]\n",
      " [0.12039713 0.03979299 0.05715314 0.78265675]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix: [[0.82835309 0.03976936 0.05146899 0.08040856]\n",
      " [0.02327998 0.80873654 0.06043788 0.10754561]\n",
      " [0.11544022 0.13895831 0.71027209 0.03532938]\n",
      " [0.09988558 0.08065349 0.01737091 0.80209001]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix: [[0.78296278 0.06942125 0.08171685 0.06589912]\n",
      " [0.02166725 0.83743552 0.01932756 0.12156967]\n",
      " [0.11182788 0.03378373 0.78484422 0.06954417]\n",
      " [0.11063295 0.0881492  0.0495941  0.75162375]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23195748 0.24737645 0.23524372 0.28542235]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.6464218290229772"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the theoretical parameters:\n",
    "theoretical_matrices = dict()\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    u = edges[i].edge[0].name.split(\"_\")[1]\n",
    "    v = edges[i].edge[1].name.split(\"_\")[1]\n",
    "    matrix_name = \"M_\" + u + \"_to_\" + v\n",
    "    print(\"Transition matrix:\", edges[i].transition_matrix)\n",
    "    theoretical_matrices[matrix_name] = np.array(edges[i].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", node_distr['Int_0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'M_0_to_2': array([[0.64642183, 0.02803076, 0.27483013, 0.05071728],\n",
      "       [0.03848122, 0.7757021 , 0.10189982, 0.08391686],\n",
      "       [0.03184807, 0.07934462, 0.82109495, 0.06771236],\n",
      "       [0.04702478, 0.01879095, 0.04076404, 0.89342023]]), 'M_0_to_3': array([[0.55850753, 0.13612552, 0.2499751 , 0.05539185],\n",
      "       [0.01791204, 0.82605787, 0.05907209, 0.096958  ],\n",
      "       [0.04534628, 0.008074  , 0.85149918, 0.09508054],\n",
      "       [0.00646156, 0.06355415, 0.02348564, 0.90649866]]), 'M_0_to_1': array([[0.75152183, 0.09827146, 0.07355421, 0.0766525 ],\n",
      "       [0.04888819, 0.86735324, 0.02989229, 0.05386628],\n",
      "       [0.05446854, 0.02450495, 0.75018255, 0.17084396],\n",
      "       [0.12039713, 0.03979299, 0.05715314, 0.78265675]]), 'M_1_to_4': array([[0.82835309, 0.03976936, 0.05146899, 0.08040856],\n",
      "       [0.02327998, 0.80873654, 0.06043788, 0.10754561],\n",
      "       [0.11544022, 0.13895831, 0.71027209, 0.03532938],\n",
      "       [0.09988558, 0.08065349, 0.01737091, 0.80209001]]), 'M_1_to_5': array([[0.78296278, 0.06942125, 0.08171685, 0.06589912],\n",
      "       [0.02166725, 0.83743552, 0.01932756, 0.12156967],\n",
      "       [0.11182788, 0.03378373, 0.78484422, 0.06954417],\n",
      "       [0.11063295, 0.0881492 , 0.0495941 , 0.75162375]])}\n"
     ]
    }
   ],
   "source": [
    "# # save the dictionary to a file\n",
    "# np.save('data.npy', theoretical_matrices)\n",
    "# # load the dictionary from the file\n",
    "# loaded_dict = np.load('data.npy', allow_pickle=True).item()\n",
    "# # print the loaded dictionary\n",
    "# print(loaded_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M_0_to_2\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix:\n",
      " [[0.86753892 0.05213542 0.06946348 0.01086218]\n",
      " [0.08065343 0.85671968 0.04982893 0.01279796]\n",
      " [0.11219846 0.17464684 0.65461785 0.05853685]\n",
      " [0.02775028 0.08494961 0.04088151 0.8464186 ]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_3\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix:\n",
      " [[0.75527498 0.04249873 0.00019955 0.20202674]\n",
      " [0.10446698 0.73349645 0.10777451 0.05426206]\n",
      " [0.009178   0.09167603 0.88105832 0.01808765]\n",
      " [0.10672648 0.08298123 0.01949473 0.79079756]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_1\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix:\n",
      " [[0.69009493 0.15321495 0.12518281 0.03150731]\n",
      " [0.03390489 0.89612499 0.03748613 0.032484  ]\n",
      " [0.13612944 0.00355591 0.86031238 0.00000227]\n",
      " [0.13829849 0.07169377 0.04358193 0.74642581]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_4\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix:\n",
      " [[0.79787321 0.09123409 0.0748348  0.0360579 ]\n",
      " [0.1525167  0.62670853 0.12700414 0.09377063]\n",
      " [0.06182208 0.09423139 0.82703856 0.01690797]\n",
      " [0.09434814 0.05487633 0.01727208 0.83350345]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_5\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix:\n",
      " [[0.77670263 0.03884329 0.11981174 0.06464234]\n",
      " [0.07306013 0.6964252  0.14381485 0.08669983]\n",
      " [0.0001472  0.02174493 0.93060026 0.04750762]\n",
      " [0.01290462 0.10677068 0.06826479 0.81205991]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23508015 0.26506847 0.26101933 0.23983305]\n"
     ]
    }
   ],
   "source": [
    "# Check the estimated parameters:\n",
    "for i in params.items():\n",
    "    print(i[0])\n",
    "    print(\"Edge:\", i[1].edge)\n",
    "    print(\"Transition matrix:\\n\", i[1].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", root_distr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix difference:\n",
      " [[0.09960865 0.07093256 0.00498621 0.03366231]\n",
      " [0.01576357 0.02703813 0.00892483 0.03387688]\n",
      " [0.00533819 0.03058066 0.01301418 0.01222828]\n",
      " [0.00504287 0.04282721 0.00457214 0.05244223]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix difference:\n",
      " [[0.00386991 0.01431933 0.0324696  0.04291902]\n",
      " [0.03240956 0.00438465 0.00578344 0.03101077]\n",
      " [0.00129948 0.00595081 0.00836875 0.00111846]\n",
      " [0.00368219 0.0201782  0.01427531 0.03077132]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix difference:\n",
      " [[0.00043271 0.04586535 0.01971483 0.02571781]\n",
      " [0.04628465 0.03270525 0.01225697 0.00132243]\n",
      " [0.00405161 0.02596747 0.04979691 0.01977783]\n",
      " [0.05635943 0.0032355  0.00691523 0.06003916]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix difference:\n",
      " [[0.02206907 0.00438435 0.02045318 0.00276846]\n",
      " [0.04114356 0.07985678 0.0152662  0.02344702]\n",
      " [0.01766089 0.02629588 0.04017144 0.00378534]\n",
      " [0.00126844 0.00706299 0.01874386 0.02707528]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix difference:\n",
      " [[0.04539034 0.01368587 0.03621623 0.00451176]\n",
      " [0.00302602 0.00613968 0.01219578 0.00908212]\n",
      " [0.00442834 0.01538392 0.00158094 0.0213932 ]\n",
      " [0.02699704 0.00888034 0.00551462 0.03036276]]\n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Compute the matrix difference between the theoretical and estimated parameters\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    print(\"Transition matrix difference:\\n\", np.abs(edges[i].transition_matrix - params['M_' + str(edges[i].edge[0].name.split(\"_\")[1]) + '_to_' + str(edges[i].edge[1].name.split(\"_\")[1])].transition_matrix))\n",
    "    print(\"---\"*30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Real pi\n",
      "[0.23497177 0.27362806 0.2594391  0.23196106]\n",
      "Real M\n",
      "M_0_to_2 [[0.87343151 0.0578093  0.01269275 0.05606644]\n",
      " [0.02401471 0.78877947 0.12724148 0.05996434]\n",
      " [0.04061902 0.16591748 0.76563328 0.02783022]\n",
      " [0.0252505  0.12880773 0.12099257 0.72494921]]\n",
      "M_0_to_3 [[0.80443485 0.00766047 0.09144638 0.0964583 ]\n",
      " [0.02477192 0.86693172 0.05904228 0.04925408]\n",
      " [0.10642242 0.12640001 0.70419191 0.06298566]\n",
      " [0.12416814 0.01237958 0.07212055 0.79133173]]\n",
      "M_0_to_1 [[0.89984046 0.02385079 0.04364508 0.03266367]\n",
      " [0.01761317 0.81819334 0.06176861 0.10242488]\n",
      " [0.05079037 0.15904707 0.72963898 0.06052358]\n",
      " [0.01608905 0.16483136 0.10365304 0.71542655]]\n",
      "M_1_to_4 [[0.83783739 0.0992672  0.04316275 0.01973266]\n",
      " [0.02001914 0.81657682 0.13038749 0.03301655]\n",
      " [0.04255222 0.10268669 0.80360584 0.05115526]\n",
      " [0.08617845 0.15064439 0.09009475 0.67308241]]\n",
      "M_1_to_5 [[0.81592343 0.16015358 0.01925916 0.00466383]\n",
      " [0.11334925 0.77935332 0.0753095  0.03198792]\n",
      " [0.08043757 0.06355713 0.76146017 0.09454513]\n",
      " [0.16077809 0.01586654 0.03524409 0.78811127]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Estimated M\n",
      "M_0_to_2 [[8.23843964e-01 6.75938183e-02 2.66439543e-14 1.08562218e-01]\n",
      " [4.08407783e-02 7.86454831e-01 1.18936269e-01 5.37681224e-02]\n",
      " [2.28686098e-02 1.83727733e-01 7.38900067e-01 5.45035898e-02]\n",
      " [5.38864906e-02 1.82404730e-01 1.55612440e-01 6.08096339e-01]]\n",
      "M_0_to_3 [[7.64140455e-01 1.12028805e-17 1.22248195e-01 1.13611350e-01]\n",
      " [1.31416432e-02 9.15046846e-01 6.98450410e-02 1.96646985e-03]\n",
      " [1.10685038e-01 1.82598805e-01 6.36140201e-01 7.05759558e-02]\n",
      " [7.97748575e-02 4.12826508e-02 1.06079988e-01 7.72862504e-01]]\n",
      "M_0_to_1 [[8.67386863e-01 5.24000068e-05 9.53342826e-02 3.72264541e-02]\n",
      " [3.62516684e-02 7.92650265e-01 7.09949180e-03 1.63998574e-01]\n",
      " [5.97831945e-02 2.57382409e-01 6.80377557e-01 2.45683955e-03]\n",
      " [5.39480057e-03 1.01914606e-01 3.82457399e-08 8.92690555e-01]]\n",
      "M_1_to_4 [[8.46566384e-01 1.02217163e-01 2.77833250e-02 2.34331279e-02]\n",
      " [3.02510003e-12 7.92307355e-01 1.16762441e-01 9.09302044e-02]\n",
      " [5.44664768e-02 1.09640542e-01 8.27350184e-01 8.54279666e-03]\n",
      " [1.00757534e-01 3.04344040e-01 7.31669987e-02 5.21731427e-01]]\n",
      "M_1_to_5 [[7.83706985e-01 1.93855430e-01 2.24375850e-02 1.93280032e-17]\n",
      " [8.22460187e-02 8.04055553e-01 1.13698425e-01 3.92860067e-09]\n",
      " [8.67890925e-02 1.53853660e-01 7.19945256e-01 3.94119919e-02]\n",
      " [1.61577964e-01 1.16081043e-03 8.61193868e-02 7.51141839e-01]]\n",
      "Estimated l\n",
      "M_0_to_2 1.2985518260767415\n",
      "M_0_to_3 1.1430628483558387\n",
      "M_0_to_1 0.9040381680544624\n",
      "M_1_to_4 1.2700145756005907\n",
      "M_1_to_5 1.1136800570824406\n",
      "Estimated pi\n",
      "[0.25748927 0.24340192 0.27706463 0.22404819]\n",
      "Niter\n",
      "36\n",
      "TExec\n",
      "37.1959388256073\n"
     ]
    }
   ],
   "source": [
    "print(\"Real pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a100_b100_500_spectral/real_root_distr.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Real M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a100_b100_500_spectral/real_matrices.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"---\"*30)\n",
    "print(\"Estimated M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a100_b100_500_spectral/RESULTS/repetition_99/M_estimation.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated l\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a100_b100_500_spectral/RESULTS/repetition_99/estimated_branch_lengths.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a100_b100_500_spectral/RESULTS/repetition_99/root_estimation.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Niter\")\n",
    "file = open(\"./OUTPUTS/EM_a100_b100_500_spectral/RESULTS/repetition_99/Niter.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)\n",
    "print(\"TExec\")\n",
    "file = open(\"./OUTPUTS/EM_a100_b100_500_spectral/RESULTS/repetition_99/TExec.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "debugger",
   "language": "python",
   "name": "debugger"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "03f790eb89590da6bbc9dab944ceea1669c4a69ff2bd32047ab9c0f81f755bd1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
