{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Expectation-Maximization algorithm** (hidden model)\n",
    "### Author: Martí Cortada Garcia\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "import numpy as np\n",
    "from Bio import SeqIO, Phylo\n",
    "from io import StringIO\n",
    "from collections import Counter, OrderedDict\n",
    "import itertools\n",
    "import pandas as pd\n",
    "import math\n",
    "import copy\n",
    "\n",
    "# Import required functions from modules we have programmed\n",
    "from data_simulation import simulate"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Input:**\n",
    "We have a tree $T$ (where the root corresponds to the internal node codified as `Int_0`) and the alignments of the leaves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAGwCAYAAACHJU4LAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy88F64QAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAwh0lEQVR4nO3de1SVdaLG8We7EeSOaMpFlEK8VKCkA2FpN0vNSppZWR7z0nRi5iywzDgxnmzMMnTlJZnqNGkNOJ2UIU/WHG1wzCILs4xkhsSjiXZ2KmpjCULHjWze80fHvSRvbAT2D/l+1tpr9d6f97dY7qf3fffeNsuyLAEAABioi7cDAAAAnAtFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWD7eDnAxGhsbdfDgQQUHB8tms3k7DgAAaAbLsnT8+HFFRUWpS5fzXzPp0EXl4MGDiomJ8XYMAADQAt9884369Olz3nU6dFEJDg6W9OOJhoSEeDkNAABojpqaGsXExLjfx8+nQxeVU7d7QkJCKCoAAHQwzXlsg4dpAQCAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgA8BrLspSenq7w8HDZbDaVlZV5OxIAw1BUAJxVcXGxbDabjh071mbHKCoqUn5+vtatW6eqqipdffXV511/wYIF+tnPfqbg4GD16tVLaWlp2rVrV5vlA+B9FBUAXlNZWanIyEiNGDFCERER8vHxOe/6H374oTIyMrR161Zt3LhRJ0+e1G233aa6urp2SgygvXm1qGzevFl33nmnoqKiZLPZ9Pbbb3szDgAPOJ1OZWVlKTo6WoGBgUpJSVFxcbF7+dGjRzVp0iRFR0crICBACQkJWr16tXv59OnTNWPGDDkcDtlsNsXGxl7wmEVFRZo+fbquuuoqDRkyRPn5+XI4HCotLW2DMwRggvP/70sbq6ur05AhQ/TLX/5SP//5z70ZxZ0H6IwCAwM93iYzM1MVFRUqKChQVFSU1q5dq7Fjx6q8vFzx8fE6ceKEhg0bpuzsbIWEhGj9+vWaMmWK4uLilJycrNzcXMXFxWn58uXatm2b7Ha7xxmqq6slSeHh4R5vC6BjsFmWZXk7hCTZbDatXbtWaWlp51zH6XTK6XS6p2tqahQTE6Pq6mqFhIS0SgagMzrbPwPFxcW66aab9P333yssLKzJMofDoSuuuEIOh0NRUVHu+aNHj1ZycrJycnLOepw77rhDgwYN0uLFiyVJy5Yt07Jly/T11197nLmxsVF33XWXjh07po8//tjj7QF4T01NjUJDQ5v1/u3VKyqeWrBggebNm9fq+62rq1NQUFCr7xe4VJWXl8vlcmnAgAFN5judTvXo0UOS5HK5lJOTo8LCQh04cED19fVyOp0KCAholQwZGRn68ssvKSnAJa5DFZXZs2dr1qxZ7ulTV1Ra0+HDh1t0GRzoTGpra2W321VaWnrGLZtTpX/RokXKzc3VsmXLlJCQoMDAQM2cOVP19fUXffzMzEytW7dOmzdvVp8+fS56fwDM1aGKip+fn/z8/Nr0GIGBgRQV4AKSkpLkcrl05MgRjRw58qzrlJSUaMKECbr//vsl/XirZvfu3bryyitbfFzLsjRjxgytXbtWxcXFuvzyy1u8LwAdQ4cqKgDaX3l5uYKDg93TNptNQ4YM0eTJkzV16lQtWbJESUlJ+vbbb7Vp0yYlJiZq/Pjxio+P15o1a7RlyxZ1795dS5cu1eHDhy+qqGRkZGjVqlV65513FBwcrEOHDkmSQkND5e/vf9HnCsA8FBUA5zVq1Kgm03a7XQ0NDcrLy9P8+fP12GOP6cCBA+rZs6euvfZa3XHHHZKkOXPmaO/evRozZowCAgKUnp6utLQ09yd1WuLll1+WJN14441N5ufl5Wn69Okt3i8Ac3n1Uz+1tbXas2ePpB8vJS9dulQ33XSTwsPD1bdv3wtu78lTw+dz+sO0tbW13PoBAKANdZhP/Xz++ee66aab3NOnHpSdNm2a8vPzvZQKAACYwqtF5cYbbzzr9zcA6HwcDsd5n1+pqKho1pVWAJcWnlEBYISoqKjz/nry6V8sB6DzoKgAMIKPj4/69+/v7RgADMOvJwMAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAeI1lWUpPT1d4eLhsNpvKysq8HQmAYSgqAM6quLhYNptNx44da7NjFBUVKT8/X+vWrVNVVZWuvvrq867/8ssvKzExUSEhIQoJCVFqaqr+8pe/tFk+AN7n1aKyYMEC/exnP1NwcLB69eqltLQ07dq1y5uRALSjyspKRUZGasSIEYqIiJCPj8951+/Tp48WLlyo0tJSff7557r55ps1YcIE7dixo50SA2hvXi0qH374oTIyMrR161Zt3LhRJ0+e1G233aa6ujpvxgLQDE6nU1lZWYqOjlZgYKBSUlJUXFzsXn706FFNmjRJ0dHRCggIUEJCglavXu1ePn36dM2YMUMOh0M2m02xsbEXPOadd96p22+/XfHx8RowYICeffZZBQUFaevWrW1whgBMcP7/fWljRUVFTabz8/PVq1cvlZaWatSoUV7JRElCZxQYGOjxNpmZmaqoqFBBQYGioqK0du1ajR07VuXl5YqPj9eJEyc0bNgwZWdnKyQkROvXr9eUKVMUFxen5ORk5ebmKi4uTsuXL9e2bdtkt9s9Or7L5dKbb76puro6paamepwfQMfg1aLyU9XV1ZKk8PDwsy53Op1yOp3u6ZqamlbP0Lt371bfJ2A6y7I8Wt/hcCgvL08Oh0NRUVGSpKysLBUVFSkvL085OTmKjo5WVlaWe5sZM2Zow4YNKiwsVHJyskJDQxUcHCy73a6IiIhmH7u8vFypqak6ceKEgoKCtHbtWl155ZUe5QfQcRhTVBobGzVz5kxdd91153ygbsGCBZo3b147JwPwU+Xl5XK5XBowYECT+U6nUz169JD04xWPnJwcFRYW6sCBA6qvr5fT6VRAQMBFHXvgwIEqKytTdXW11qxZo2nTpunDDz+krACXKGOKSkZGhr788kt9/PHH51xn9uzZmjVrlnu6pqZGMTExrZrj8OHDLboMDnQmtbW1stvtKi0tPeOWTVBQkCRp0aJFys3N1bJly5SQkKDAwEDNnDlT9fX1F3VsX19f9e/fX5I0bNgwbdu2Tbm5uXrllVcuar8AzGREUcnMzNS6deu0efNm9enT55zr+fn5yc/Pr02zBAYGUlSAC0hKSpLL5dKRI0c0cuTIs65TUlKiCRMm6P7775f041XT3bt3t/qVj8bGxia3hAFcWrxaVCzL0owZM7R27VoVFxfr8ssv92YcAGdRXl6u4OBg97TNZtOQIUM0efJkTZ06VUuWLFFSUpK+/fZbbdq0SYmJiRo/frzi4+O1Zs0abdmyRd27d9fSpUt1+PDhiyoqs2fP1rhx49S3b18dP35cq1atUnFxsTZs2NAapwrAQF4tKhkZGVq1apXeeecdBQcH69ChQ5Kk0NBQ+fv7ezMagP/300/g2e12NTQ0KC8vT/Pnz9djjz2mAwcOqGfPnrr22mt1xx13SJLmzJmjvXv3asyYMQoICFB6errS0tLcD823xJEjRzR16lRVVVUpNDRUiYmJ2rBhg2699daLOkcA5rJZnj7u35oHt9nOOj8vL0/Tp0+/4PY1NTUKDQ1VdXW1QkJCWpyjrq7OfV+9traWWz8AALQhT96/vX7rBwAA4Fz4rR8ARnA4HAoKCjrny+FweDsiAC8w4lM/ABAVFXXeX08+9cVyADoXigoAI/j4+Li/HwUATuHWDwAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FULqC4uFg2m03Hjh1r9jbTp09XWlqaR8f57rvvNHnyZIWEhCgsLEwPPvigamtrPQsLAMAlhqJiiMmTJ2vHjh3auHGj+5ek09PTvR0LAACvoqh4KD8/X2FhYdqwYYMGDx6soKAgjR07VlVVVZKkp556SitXrtQ777wjm80mm82m4uLi8+5z586dKioq0quvvqqUlBRdf/31euGFF1RQUKCDBw+2w1kBAGAmikoL/PDDD1q8eLFef/11bd68WQ6HQ1lZWZKkrKwsTZw40V1eqqqqNGLEiPPu75NPPlFYWJiGDx/unjd69Gh16dJFn376aZueCwAAJuObaVvg5MmT+v3vf6+4uDhJUmZmpp5++mlJUlBQkPz9/eV0OhUREdGs/R06dEi9evVqMs/Hx0fh4eE6dOhQ64YHAKAD4YpKCwQEBLhLiiRFRkbqyJEjXkwEAMCliaLSAl27dm0ybbPZZFlWi/cXERFxRtFpaGjQd9991+yrMgAAXIooKm3A19dXLper2eunpqbq2LFjKi0tdc97//331djYqJSUlLaICABAh0BRaQOxsbH6+9//rl27dukf//iHTp48ed71Bw8erLFjx+qhhx7SZ599ppKSEmVmZuq+++7jp+0BAJ0aRaUNPPTQQxo4cKCGDx+uyy67TCUlJRfc5o033tCgQYN0yy236Pbbb9f111+v5cuXt0NaAADMZbMu5uEKL6upqVFoaKiqq6sVEhLS4v3U1dUpKChIklRbW6vAwMDWiggAAH7Ck/dvrqgAAABjUVTaQU5OjoKCgs76GjdunLfjAQBgrBZ94dvnn3+uwsJCORwO1dfXN1n21ltvtUqwS8mvf/1rTZw48azL/P392zkNAAAdh8dFpaCgQFOnTtWYMWP017/+Vbfddpt2796tw4cP6+67726LjB1eeHi4wsPDvR0DAIAOx+NbPzk5OXr++ef1X//1X/L19VVubq7++7//WxMnTlTfvn3bIiMAAOikPC4qlZWVGj9+vKQfv9isrq5ONptNjz76KB+nBQAArcrjotK9e3cdP35ckhQdHa0vv/xSknTs2DH98MMPrZsOAAB0ah4/ozJq1Cht3LhRCQkJuueee/TII4/o/fff18aNG3XLLbe0RUYAANBJeVxUXnzxRZ04cUKS9MQTT6hr167asmWLfvGLX2jOnDmtHhAAAHReHheV0z+90qVLF/3mN79p1UAAAACnePyMSn5+/lnnNzQ0aPbs2RebBwAAwM3jovLwww/rnnvu0ffff++et2vXLqWkpGj16tWtGg4AAHRuHheV7du3a//+/UpISNDGjRv10ksv6ZprrtGgQYP0t7/9rS0yAgCATsrjZ1Ti4uJUUlKimTNnauzYsbLb7Vq5cqUmTZrUFvkAAEAn1qIfJVy/fr0KCgqUmpqqsLAwvfbaazp48GBrZwMAAJ2cx0XlV7/6le655x5lZ2fro48+0t///nf5+voqISFBhYWFbZERAAB0Uh7f+ikpKdGnn36qIUOGSJIiIiL07rvv6qWXXtIvf/nLc/5KMAAAgKc8LiqlpaXy8/M7Y35GRoZGjx7dKqEAAACkFtz6OVtJOWXgwIEXFQYAAOB0Hl9RkaQ1a9aosLBQDodD9fX1TZZ98cUXrRIMAADA4ysqv/vd7/TAAw+od+/e2r59u5KTk9WjRw/t3btX48aNa4uMAACgk/K4qPz7v/+7li9frhdeeEG+vr56/PHHtXHjRj388MOqrq5ui4wAAKCT8rioOBwOjRgxQpLk7++v48ePS5KmTJnCV+gDAIBW5XFRiYiI0HfffSdJ6tu3r7Zu3SpJ2rdvnyzLat10AACgU/O4qNx8883685//LEl64IEH9Oijj+rWW2/Vvffeq7vvvrvVAwIAgM7L40/9PPHEE4qOjpb043en9OjRQ1u2bNFdd92lsWPHtnpAAADQedksD+/X2O12VVVVqVevXk3mHz16VL169ZLL5WrVgOdTU1Oj0NBQVVdXKyQkpMX7qaurU1BQkCSptrZWgYGBrRURAAD8hCfv3x7f+jlXr6mtrVW3bt083R0AAMA5NfvWz6xZsyRJNptNv/3tbxUQEOBe5nK59Omnn2ro0KGtHhAAAHRezS4q27dvl/TjFZXy8nL5+vq6l/n6+mrIkCHKyspq/YQAAKDTanZR+eCDDyT9+Emf3Nzci3omBAAAoDk8/tRPXl5eW+QAAAA4g8cP0wIAALQXigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAADamWVZSk9PV3h4uGw2m8rKyrwdyVgUFQAATlNcXCybzaZjx4612TGKioqUn5+vdevWqaqqSldffXWzt124cKFsNptmzpzZZvlM4tWi8vLLLysxMVEhISEKCQlRamqq/vKXv3gzEgAAba6yslKRkZEaMWKEIiIi5OPj06zttm3bpldeeUWJiYltnNAcXi0qffr00cKFC1VaWqrPP/9cN998syZMmKAdO3Z4MxYAAOfkdDqVlZWl6OhoBQYGKiUlRcXFxe7lR48e1aRJkxQdHa2AgAAlJCRo9erV7uXTp0/XjBkz5HA4ZLPZFBsb26zj1tbWavLkyVqxYoW6d+/eymdlruZVuDZy5513Npl+9tln9fLLL2vr1q266qqrvJQK6Hzq6uq8HQHwisDAQI+3yczMVEVFhQoKChQVFaW1a9dq7NixKi8vV3x8vE6cOKFhw4YpOztbISEhWr9+vaZMmaK4uDglJycrNzdXcXFxWr58ubZt2ya73d6s42ZkZGj8+PEaPXq05s+f73HujsqrReV0LpdLb775purq6pSamnrWdZxOp5xOp3u6pqamveIBl7SgoCBvRwC8wrIsj9Z3OBzKy8uTw+FQVFSUJCkrK0tFRUXKy8tTTk6OoqOjlZWV5d5mxowZ2rBhgwoLC5WcnKzQ0FAFBwfLbrcrIiKiWcctKCjQF198oW3btnmU91Lg9aJSXl6u1NRUnThxQkFBQVq7dq2uvPLKs667YMECzZs3r50TApeuuro6SgrggfLycrlcLg0YMKDJfKfTqR49ekj68X+8c3JyVFhYqAMHDqi+vl5Op1MBAQEtOuY333yjRx55RBs3blS3bt0u+hw6Gq8XlYEDB6qsrEzV1dVas2aNpk2bpg8//PCsZWX27NmaNWuWe7qmpkYxMTHtGRe4ZB0+fLhFl8GBzqS2tlZ2u12lpaVn3LI5VfoXLVqk3NxcLVu2TAkJCQoMDNTMmTNVX1/fomOWlpbqyJEjuuaaa9zzXC6XNm/erBdffFFOp7PZt486Iq8XFV9fX/Xv31+SNGzYMG3btk25ubl65ZVXzljXz89Pfn5+7R0R6BQCAwMpKsAFJCUlyeVy6ciRIxo5cuRZ1ykpKdGECRN0//33S5IaGxu1e/fuc94tuJBbbrlF5eXlTeY98MADGjRokLKzsy/pkiIZUFR+qrGxsclzKAAAeEN5ebmCg4Pd0zabTUOGDNHkyZM1depULVmyRElJSfr222+1adMmJSYmavz48YqPj9eaNWu0ZcsWde/eXUuXLtXhw4dbXFSCg4PP+J6VwMBA9ejRw6PvX+movFpUZs+erXHjxqlv3746fvy4Vq1apeLiYm3YsMGbsQAA0KhRo5pM2+12NTQ0KC8vT/Pnz9djjz2mAwcOqGfPnrr22mt1xx13SJLmzJmjvXv3asyYMQoICFB6errS0tJUXV3tjdPo8GyWp488t6IHH3xQmzZtUlVVlUJDQ5WYmKjs7Gzdeuutzdq+pqZGoaGhqq6uVkhISItznP5AYW1tLZe/0Wnwtw/AGzx5//bqFZXXXnvNm4cHAACG47d+AADwIofDoaCgoHO+HA6HtyN6lXEP0wIA0JlERUWd99eTT32xXGdFUQEAwIt8fHzcX9OBM3HrBwAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAuCsiouLZbPZdOzYsWZvM336dKWlpXl0nGeffVYjRoxQQECAwsLCPNoWwKWPogLAq+rr63XPPffoX/7lX7wdBYCBKCoAmiU/P19hYWHasGGDBg8erKCgII0dO1ZVVVWSpKeeekorV67UO++8I5vNJpvNpuLi4gvud968eXr00UeVkJDQxmcAoCPy8XYA09TV1Xk7AtBuPP17/+GHH7R48WK9/vrr6tKli+6//35lZWXpjTfeUFZWlnbu3Kmamhrl5eVJksLDw9siNoBOhKLyE7179/Z2BMBYJ0+e1O9//3vFxcVJkjIzM/X0009LkoKCguTv7y+n06mIiAhvxgRwCeHWj6TAwEBZluXtGIDxAgIC3CVFkiIjI3XkyBEvJgJwqeOKymlqa2u9HQFoV3V1dR5dRezatWuTaZvNRskH0KYoKqcJDAz0dgSgQ/P19ZXL5fJ2DACXEIoKgFYTGxurDRs2aNeuXerRo4dCQ0PPuArzUw6HQ999950cDodcLpfKysokSf3791dQUFA7pAZgMp5RAdBqHnroIQ0cOFDDhw/XZZddppKSkgtu89vf/lZJSUmaO3euamtrlZSUpKSkJH3++eftkBiA6WxWB77BXFNTo9DQUFVXVyskJMTbcYAOp66uzn3Vora2ltufANqFJ+/fXFEBAADGoqgAaDM5OTkKCgo662vcuHHejgegA+BhWgBt5te//rUmTpx41mX+/v7tnAZAR0RRAdBmwsPD+Rp9ABeFWz8AAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAALQzy7KUnp6u8PBw2Ww2lZWVeTuSsSgqAACcpri4WDabTceOHWuzYxQVFSk/P1/r1q1TVVWVrr766vOu/9RTT8lmszV5DRo0qM3ymcTH2wEAAOhsKisrFRkZqREjRjR7m6uuukrvvfeee9rHp3O8hRtzRWXhwoWy2WyaOXOmt6MAAHBOTqdTWVlZio6OVmBgoFJSUlRcXOxefvToUU2aNEnR0dEKCAhQQkKCVq9e7V4+ffp0zZgxQw6HQzabTbGxsc06ro+PjyIiItyvnj17tvKZmcmIOrZt2za98sorSkxM9HYUoNOqq6vzdgSg3QUGBnq8TWZmpioqKlRQUKCoqCitXbtWY8eOVXl5ueLj43XixAkNGzZM2dnZCgkJ0fr16zVlyhTFxcUpOTlZubm5iouL0/Lly7Vt2zbZ7fZmHferr75SVFSUunXrptTUVC1YsEB9+/b1OH9H4/WiUltbq8mTJ2vFihWaP3/+edd1Op1yOp3u6ZqamraOB3QavXv39nYEoN1ZluXR+g6HQ3l5eXI4HIqKipIkZWVlqaioSHl5ecrJyVF0dLSysrLc28yYMUMbNmxQYWGhkpOTFRoaquDgYNntdkVERDTruCkpKcrPz9fAgQNVVVWlefPmaeTIkfryyy8VHBzs0Tl0NF4vKhkZGRo/frxGjx59waKyYMECzZs3r52SAZe+wMBAWZYlm83m7ShAh1BeXi6Xy6UBAwY0me90OtWjRw9JksvlUk5OjgoLC3XgwAHV19fL6XQqICCgxccdN26c+78TExOVkpKifv36qbCwUA8++GCL99sReLWoFBQU6IsvvtC2bduatf7s2bM1a9Ys93RNTY1iYmLaKh7QadTW1no7AtAh1NbWym63q7S09IxbNkFBQZKkRYsWKTc3V8uWLVNCQoICAwM1c+ZM1dfXt1qOsLAwDRgwQHv27Gm1fZrKa0Xlm2++0SOPPKKNGzeqW7duzdrGz89Pfn5+bZwM6Hxacp8e6IySkpLkcrl05MgRjRw58qzrlJSUaMKECbr//vslSY2Njdq9e7euvPLKVstRW1uryspKTZkypdX2aSqvFZXS0lIdOXJE11xzjXuey+XS5s2b9eKLL8rpdDb7ASMAAFpbeXl5k+c/bDabhgwZosmTJ2vq1KlasmSJkpKS9O2332rTpk1KTEzU+PHjFR8frzVr1mjLli3q3r27li5dqsOHD19UUcnKytKdd96pfv366eDBg5o7d67sdrsmTZrUGqdqNK8VlVtuuUXl5eVN5j3wwAMaNGiQsrOzKSkAAK8aNWpUk2m73a6Ghgbl5eVp/vz5euyxx3TgwAH17NlT1157re644w5J0pw5c7R3716NGTNGAQEBSk9PV1pamqqrq1ucZf/+/Zo0aZKOHj2qyy67TNdff722bt2qyy677KLOsSOwWZ4+8tyGbrzxRg0dOlTLli1r1vo1NTUKDQ1VdXW1QkJC2jYcAABoFZ68fxvzhW8AAAA/5fWPJ5/u9G/2AwCgM3A4HOd9fqWioqJTfLHbuRhVVAAA6GyioqLO++vJp75YrrOiqAAA4EU+Pj7q37+/t2MYi2dUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACMRVEBAADGoqgAAABjUVQAAICxKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYCyKCgAAMBZFBQAAGIuiAgAAjEVRAQAAxqKoAAAAY1FUAACAsSgqAADAWBQVAABgLIoKAAAwFkUFAAAYi6ICAACM5ePtABfDsixJUk1NjZeTAACA5jr1vn3qffx8OnRROX78uCQpJibGy0kAAICnjh8/rtDQ0POuY7OaU2cM1djYqIMHDyo4OFg2m+2i9lVTU6OYmBh98803CgkJaaWElzbGzDOMl2cYL88wXp5jzDzTmuNlWZaOHz+uqKgodely/qdQOvQVlS5duqhPnz6tus+QkBD+YD3EmHmG8fIM4+UZxstzjJlnWmu8LnQl5RQepgUAAMaiqAAAAGNRVP6fn5+f5s6dKz8/P29H6TAYM88wXp5hvDzDeHmOMfOMt8arQz9MCwAALm1cUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFidqqi89NJLio2NVbdu3ZSSkqLPPvvsvOu/+eabGjRokLp166aEhAS9++677ZTUDJ6M144dO/SLX/xCsbGxstlsWrZsWfsFNYgnY7ZixQqNHDlS3bt3V/fu3TV69OgL/k1eajwZr7feekvDhw9XWFiYAgMDNXToUL3++uvtmNb7PP037JSCggLZbDalpaW1bUADeTJm+fn5stlsTV7dunVrx7Te5+nf2LFjx5SRkaHIyEj5+flpwIABrf9eaXUSBQUFlq+vr/WHP/zB2rFjh/XQQw9ZYWFh1uHDh8+6fklJiWW3263nnnvOqqiosObMmWN17drVKi8vb+fk3uHpeH322WdWVlaWtXr1aisiIsJ6/vnn2zewATwds3/6p3+yXnrpJWv79u3Wzp07renTp1uhoaHW/v372zm5d3g6Xh988IH11ltvWRUVFdaePXusZcuWWXa73SoqKmrn5N7h6Xidsm/fPis6OtoaOXKkNWHChPYJawhPxywvL88KCQmxqqqq3K9Dhw61c2rv8XS8nE6nNXz4cOv222+3Pv74Y2vfvn1WcXGxVVZW1qq5Ok1RSU5OtjIyMtzTLpfLioqKshYsWHDW9SdOnGiNHz++ybyUlBTrV7/6VZvmNIWn43W6fv36dcqicjFjZlmW1dDQYAUHB1srV65sq4hGudjxsizLSkpKsubMmdMW8YzTkvFqaGiwRowYYb366qvWtGnTOl1R8XTM8vLyrNDQ0HZKZx5Px+vll1+2rrjiCqu+vr5Nc3WKWz/19fUqLS3V6NGj3fO6dOmi0aNH65NPPjnrNp988kmT9SVpzJgx51z/UtKS8ersWmPMfvjhB508eVLh4eFtFdMYFztelmVp06ZN2rVrl0aNGtWWUY3Q0vF6+umn1atXLz344IPtEdMoLR2z2tpa9evXTzExMZowYYJ27NjRHnG9riXj9ec//1mpqanKyMhQ7969dfXVVysnJ0cul6tVs3WKovKPf/xDLpdLvXv3bjK/d+/eOnTo0Fm3OXTokEfrX0paMl6dXWuMWXZ2tqKios4oyJeilo5XdXW1goKC5Ovrq/Hjx+uFF17Qrbfe2tZxva4l4/Xxxx/rtdde04oVK9ojonFaMmYDBw7UH/7wB73zzjv6j//4DzU2NmrEiBHav39/e0T2qpaM1969e7VmzRq5XC69++67evLJJ7VkyRLNnz+/VbN16F9PBi4VCxcuVEFBgYqLizvdw3ueCA4OVllZmWpra7Vp0ybNmjVLV1xxhW688UZvRzPK8ePHNWXKFK1YsUI9e/b0dpwOIzU1Vampqe7pESNGaPDgwXrllVf0zDPPeDGZmRobG9WrVy8tX75cdrtdw4YN04EDB7Ro0SLNnTu31Y7TKYpKz549Zbfbdfjw4SbzDx8+rIiIiLNuExER4dH6l5KWjFdndzFjtnjxYi1cuFDvvfeeEhMT2zKmMVo6Xl26dFH//v0lSUOHDtXOnTu1YMGCS76oeDpelZWV+vrrr3XnnXe65zU2NkqSfHx8tGvXLsXFxbVtaC9rjX/HunbtqqSkJO3Zs6ctIhqlJeMVGRmprl27ym63u+cNHjxYhw4dUn19vXx9fVslW6e49ePr66thw4Zp06ZN7nmNjY3atGlTk/Z8utTU1CbrS9LGjRvPuf6lpCXj1dm1dMyee+45PfPMMyoqKtLw4cPbI6oRWutvrLGxUU6nsy0iGsXT8Ro0aJDKy8tVVlbmft1111266aabVFZWppiYmPaM7xWt8TfmcrlUXl6uyMjItoppjJaM13XXXac9e/a4S7Ak7d69W5GRka1WUiR1ro8n+/n5Wfn5+VZFRYWVnp5uhYWFuT96NmXKFOs3v/mNe/2SkhLLx8fHWrx4sbVz505r7ty5ne7jyZ6Ml9PptLZv325t377dioyMtLKysqzt27dbX331lbdOod15OmYLFy60fH19rTVr1jT5OOTx48e9dQrtytPxysnJsf76179alZWVVkVFhbV48WLLx8fHWrFihbdOoV15Ol4/1Rk/9ePpmM2bN8/asGGDVVlZaZWWllr33Xef1a1bN2vHjh3eOoV25el4ORwOKzg42MrMzLR27dplrVu3zurVq5c1f/78Vs3VaYqKZVnWCy+8YPXt29fy9fW1kpOTra1bt7qX3XDDDda0adOarF9YWGgNGDDA8vX1ta666ipr/fr17ZzYuzwZr3379lmSznjdcMMN7R/cizwZs379+p11zObOndv+wb3Ek/F64oknrP79+1vdunWzunfvbqWmploFBQVeSO09nv4bdrrOWFQsy7Mxmzlzpnvd3r17W7fffrv1xRdfeCG193j6N7ZlyxYrJSXF8vPzs6644grr2WeftRoaGlo1k82yLKv1rs8AAAC0nk7xjAoAAOiYKCoAAMBYFBUAAGAsigoAADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVIBL3I033qiZM2d6O8ZZff3117LZbCorK/NoO5vNprfffrtNMnnqqaee0tChQ70dA7hkUVQAoJlMKkhAZ0FRAXCG+vp6b0cAAEkUFaBTaGhoUGZmpkJDQ9WzZ089+eSTOv1nvmJjY/XMM89o6tSpCgkJUXp6uiQpOztbAwYMUEBAgK644go9+eSTOnnypHu7U7c9Xn/9dcXGxio0NFT33Xefjh8/7l6nsbFRzz33nPr37y8/Pz/17dtXzz77bJN8e/fu1U033aSAgAANGTJEn3zyiUfn980332jixIkKCwtTeHi4JkyYoK+//tq9fPr06UpLS9PixYsVGRmpHj16KCMjo8m5VFVVafz48fL399fll1+uVatWKTY2VsuWLXOPkSTdfffdstls7ulTzjcGAFqOogJ0AitXrpSPj48+++wz5ebmaunSpXr11VebrLN48WINGTJE27dv15NPPilJCg4OVn5+vioqKpSbm6sVK1bo+eefb7JdZWWl3n77ba1bt07r1q3Thx9+qIULF7qXz549WwsXLtSTTz6piooKrVq1Sr17926yjyeeeEJZWVkqKyvTgAEDNGnSJDU0NDTr3E6ePKkxY8YoODhYH330kUpKShQUFKSxY8c2uTL0wQcfqLKyUh988IFWrlyp/Px85efnu5dPnTpVBw8eVHFxsf7zP/9Ty5cv15EjR9zLt23bJknKy8tTVVWVe7o5YwDgIrTqbzEDMM4NN9xgDR482GpsbHTPy87OtgYPHuye7tevn5WWlnbBfS1atMgaNmyYe3ru3LlWQECAVVNT4573r//6r1ZKSoplWZZVU1Nj+fn5WStWrDjr/vbt22dJsl599VX3vB07dliSrJ07d54zhyRr7dq1lmVZ1uuvv24NHDiwyfk5nU7L39/f2rBhg2VZljVt2jSrX79+TX5+/p577rHuvfdey7Isa+fOnZYka9u2be7lX331lSXJev7558963OaOAYCLwxUVoBO49tprZbPZ3NOpqan66quv5HK53POGDx9+xnZ/+tOfdN111ykiIkJBQUGaM2eOHA5Hk3ViY2MVHBzsno6MjHRfidi5c6ecTqduueWW8+ZLTExssr2kJlczzudvf/ub9uzZo+DgYAUFBSkoKEjh4eE6ceKEKisr3etdddVVstvtZ825a9cu+fj46JprrnEv79+/v7p3796sDOcbAwAXx8fbAQCYITAwsMn0J598osmTJ2vevHkaM2aMQkNDVVBQoCVLljRZr2vXrk2mbTabGhsbJUn+/v7NOvbp+zhVqE7t40Jqa2s1bNgwvfHGG2csu+yyy5qV82K15b6Bzo6iAnQCn376aZPprVu3Kj4+vskVhp/asmWL+vXrpyeeeMI973/+5388Om58fLz8/f21adMm/fM//7NnoZvpmmuu0Z/+9Cf16tVLISEhLdrHwIED1dDQoO3bt2vYsGGSpD179uj7779vsl7Xrl2bXIUC0Pa49QN0Ag6HQ7NmzdKuXbu0evVqvfDCC3rkkUfOu018fLwcDocKCgpUWVmp3/3ud1q7dq1Hx+3WrZuys7P1+OOP649//KMqKyu1detWvfbaaxdzOk1MnjxZPXv21IQJE/TRRx9p3759Ki4u1sMPP6z9+/c3ax+DBg3S6NGjlZ6ers8++0zbt29Xenq6/P39m9wyi42N1aZNm3To0KEzSgyAtkFRATqBqVOn6n//93+VnJysjIwMPfLII+6PIJ/LXXfdpUcffVSZmZkaOnSotmzZ4v40kCeefPJJPfbYY/rtb3+rwYMH6957723V5zcCAgK0efNm9e3bVz//+c81ePBgPfjggzpx4oRHV1j++Mc/qnfv3ho1apTuvvtuPfTQQwoODla3bt3c6yxZskQbN25UTEyMkpKSWu0cAJybzbJO+zIFAIAkaf/+/YqJidF77713wYeBAbQdigoASHr//fdVW1urhIQEVVVV6fHHH9eBAwe0e/fuMx6WBdB+eJgWAPTjF8f927/9m/bu3avg4GCNGDFCb7zxBiUF8DKuqAAAAGPxMC0AADAWRQUAABiLogIAAIxFUQEAAMaiqAAAAGNRVAAAgLEoKgAAwFgUFQAAYKz/A08Xqy5zVZCkAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "newick_tree = open(\"./newick.txt\", \"r\")\n",
    "newick = newick_tree.read()\n",
    "newick_tree.close()\n",
    "tree = Phylo.read(StringIO(newick), \"newick\")\n",
    "\n",
    "# Change nodes names\n",
    "for idx, clade in enumerate(tree.get_nonterminals()):\n",
    "    clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "\n",
    "# Change leaves names\n",
    "for idx, clade in enumerate(tree.get_terminals()):\n",
    "    clade.name = \"Leaf_\" + clade.name \n",
    "\n",
    "Phylo.draw(tree)\n",
    "\n",
    "net = Phylo.to_networkx(tree) # to graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq> GACATTCAGCCGATCGAGGTAACGTGAACGTGCGAGGCGTGCCCCCCTTTCTATGTGATATTCTTCCGCGGTACGCATGTATCCAGGTGTTTTAATTAGAGCTGGCCTTGTTCCGTATGTCGTACTAAGCTGGCTTCCGCCGTGATTTCTGTGTAACCGAGTTTTTGTGTTAACGATCCATACCATTTATCCGATCGACGTTTTCTCGGTAACCAGCCGGCTTACTTGACGCACACCATTTTTCCGCGTTCTAGTTCCTTGACGGGAGTTGCACCACCCGACCTTTTCTGGCATTCCTGCCGCACCGCGCGCGGCGTTCCCTGGATCCGTGGGCACGTTAAAAGCATGGGTCGGCACGTCTTGTGGCCCAAGAGACTGTTTTTGGCATTTTTTTCGCCGTTATGTGTTATTCCACCGATCATGCTGCTCCTCACCCCATACAGCGCCTACAGGCTTTCGGCTCGTCTATGGTCCGGCCTCTACGATTCTGTTCCTCCCTC\n",
      "Seq> GGCCTGCAGAGGTTTTAGGGGACGTAGACCTGCGAGCCGTGAGCCCATTTCTTTAGGCTATTCTTCCGTAGTTCCCCTTTTTCGAGGTTTTTTCACCTCATCTAGACCTCTTCCGTCACTCCTCCTGCGATGGCTTCCGTCGTGAACTCTGTGTTGCCGAGTCTTTCCGTCAGCGACCTTTCCAATTTATTAGATTGCGCTTTTGTTGGTCCTTTGCCGGCTTGCTTGACGAACCGCATGTTTCCGCGTGGGTATTCCTTGAAGGCAGTTCTCCCACCCGACCTTTTATCGCATCAGTGACGCCCCGCGCGAGGAGTTTCATGGACCAGTGGGCGCCCTAAGGTTATGGGGCGGCCTGTCTCGTGGCCCGCGTGAAAGGTTTTTGTATTGTTTTCACTCTTGTGTGATCTGCGAAGGATTTGGGTGGTAGTCCCCCCCTACATCGCCTACAGCCGGTCGGATTGTCTCTCCTCTGCCCTATAAAGTTCCGTTCCTGGCCC\n",
      "Seq> GAAATTCAGCGGCTCCAATTCAAGTTACCATGCGAGGGCTGGGTGGAATCTTGACGGATAATCATGCCTGGTCCGGAGGTAACCAGGTTTATACACAGTCAACACAATTGCTCAGAAAATACTCCAGACATGACAACGGCCGTTAACTCTTTGTGAATGGCGGTATGCCTCAGGGAACTATCCTAATTCTTCGATTGATTTATTTTTCGTAGCGGGACGGCTTCCTCGACGCAGAATCTGTCTTCGCGTGGTTACTGCTTCATGCCAGTACATAAGCGGTACCATTGATGGCATGAGTGAAGCCCGTACCAAGGGAATGCAATGATACGCGGGGGCCTTATGGTTATGGGGAGGAAAGTCTACCGCCCTAAGCGGTAATTATGGTCGGATTTTTCCGTGGTTTGTGATATACGAGAACGCATCGAGGTAGGCAGATCATAAATCGCTGAAAAGCGTTCCTCTCGACAAAGGTTTGACATATAGAATTATGGTCGTACCCC\n",
      "Seq> AACCATCATAGGGTCCACGAAAAGTGACATTGAGATACCTGAGTGCAATTTTTTAGGATAACGTTCACTGGTGAGCAGGTAACCTTGCTTTATGACTTGAACCCGACTTGATCAAGAATGCATCTTGACTTGGCGTACGCCTAGATCTCAGTGTGTATGGTCGTTAGCGTTAGAGCACTATCCTAATTTCCTTCTCGAGTTATTGTTGGTAAACGGCCTGCTTGCTCGAAGCTGAATATGTATTCGCGTGGCACCTTCTGGGTGCGATATCATCAGCTGGACCTTTTATGGAATCTGTTACGCCCCGCGCAATGGAGCGCATGGATACGCGGGCGCCTTATAGTTATGGGGCGGCAAGTCTTGCGGCCTGAGCTGAACTTTTTGTCGCATTATTCCCTGTTTTGCGACTTTCGCGCACGCATGGCGGAAAGAAAATTGTAAATCGATTAAATGCGTTCGGCTCGTCGAGCGTCTGCCCAAGAGAATTCTGGTTCAACGCT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['10203320122103210113002131002131210112131222222333230313103033233221211302120313032201131333300330101231122331332213031321302300123112332212213103332313130022101333331313300210322030220333032210321021333323211300220122112330233102120202203333322121332301332233102111013312022022210223333231120332231221202212121211213322231103221311120213300001203111321120213233131122200101023133333112033333332122133031313303322022103203123123223202222030201212230201123332112321323031132211223230210332313322322232',\n",
       " '11223120101133330111102130102231210122131012220333233301123033233221301332222333332101133333320223203230102232332213202322322312103112332213213100232313133122101323332213201210223332200333033010331212333313311322333122112331233102100221203133322121311130332233100112013323222022210223333032120320131021222212121011013332031102201311121222300113303111121122313232131122212131000113333313033133332023233131310323121001103331113113013222222230203212230201221132110331323232232312223030001332213322311222',\n",
       " '10003320121123220033200133022031210111231113110032331021103003203122311322110113002201133303020201320020200331232010000302322010203102002112213300232333131003112113031223201110023032230033233210331033303333321301211102112332232102120100323132332121311330231233203122013020300121130220331031120310131001222130220011100312003103021211111223303113303111101100013230221222300121130033031132110333332213113331310303021010021203210113011201032030003212310000121332232321020001133310203030100330311321302222',\n",
       " '00220320301113220210000131020331010302231013120033333301103002133202311310120113002233123330310233100222102331032001003120323310233112130212230103232013131303113213301213301012023032230033322332321013303313311300021122312331232100123100303130332121311202233231113121030320320123110223333031100323133021222212120031101212031103021211121223303013303111121120013233121122310123100233333132120330332223133331210233321212021203112110001000033130003210330003121332112321321012132312220010100332311332002123']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edges, node_distr, filename = simulate(net, 500) # simulate data calling the module data_simulation.py \n",
    "\n",
    "# Preprocess leave alignments\n",
    "leave_alignments = [i for i in SeqIO.parse(filename, \"fasta\")]\n",
    "for i in leave_alignments:\n",
    "    print(\"Seq>\", i.seq)\n",
    "# Codify the alphabet {A,G,C,T} -> {0,1,2,3}\n",
    "leave_sequences = []\n",
    "for i in leave_alignments:\n",
    "    seq = \"\"\n",
    "    for j in i.seq:\n",
    "        if j == \"A\":\n",
    "            seq += \"0\"\n",
    "        elif j == \"G\":\n",
    "            seq += \"1\"\n",
    "        elif j == \"C\":\n",
    "            seq += \"2\"\n",
    "        elif j == \"T\":\n",
    "            seq += \"3\"\n",
    "        else:\n",
    "            raise ValueError(\"Invalid nucleotide\")\n",
    "    leave_sequences.append(seq)\n",
    "leave_sequences"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## **EM algorithm**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from evaluate import init_EM\n",
    "\n",
    "Results = init_EM(newick, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Source: Int_0\n",
      "Target: Leaf_2\n",
      "Matrix:\n",
      " [[0.80529814 0.06095358 0.09251042 0.04123786]\n",
      " [0.05611119 0.71483627 0.11102639 0.11802615]\n",
      " [0.0952762  0.07415636 0.731184   0.09938344]\n",
      " [0.02196288 0.0820518  0.08234946 0.81363586]]\n",
      "Source: Int_0\n",
      "Target: Leaf_3\n",
      "Matrix:\n",
      " [[0.63632807 0.06782976 0.1446222  0.15121997]\n",
      " [0.12810897 0.67208297 0.04706075 0.15274731]\n",
      " [0.10204507 0.09021686 0.74449869 0.06323937]\n",
      " [0.06570111 0.08178298 0.05155947 0.80095644]]\n",
      "Source: Int_0\n",
      "Target: Int_1\n",
      "Matrix:\n",
      " [[0.72577892 0.07421689 0.0831795  0.11682468]\n",
      " [0.08700214 0.7261134  0.06686419 0.12002027]\n",
      " [0.09519544 0.182137   0.68012648 0.04254107]\n",
      " [0.06761601 0.06509887 0.08961157 0.77767356]]\n",
      "Source: Int_1\n",
      "Target: Leaf_4\n",
      "Matrix:\n",
      " [[0.85639817 0.07234946 0.04036045 0.03089192]\n",
      " [0.0298384  0.73128255 0.0944944  0.14438465]\n",
      " [0.03533407 0.10420379 0.77836777 0.08209437]\n",
      " [0.06403097 0.03195277 0.06966545 0.8343508 ]]\n",
      "Source: Int_1\n",
      "Target: Leaf_5\n",
      "Matrix:\n",
      " [[0.80075352 0.06438893 0.11198018 0.02287738]\n",
      " [0.07400816 0.72459648 0.10712499 0.09427037]\n",
      " [0.14358855 0.08350079 0.74500083 0.02790983]\n",
      " [0.01210798 0.14166184 0.06442053 0.78180965]]\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(Results)):\n",
    "    print(\"Source:\", Results[i].source)\n",
    "    print(\"Target:\", Results[i].target)\n",
    "    print(\"Matrix:\\n\", Results[i].matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "l -73482.53888183684\n",
      "73482.53888183684\n",
      "--------\n",
      "2803.2235244028125\n",
      "--------\n",
      "403.42311980716477\n",
      "--------\n",
      "147.34679458629398\n",
      "--------\n",
      "65.81431201811938\n",
      "--------\n",
      "32.62458047835389\n",
      "--------\n",
      "17.4679848113592\n",
      "--------\n",
      "9.96598948755127\n",
      "--------\n",
      "5.979527266841615\n",
      "--------\n",
      "3.724567122422741\n",
      "--------\n",
      "2.382098244605004\n",
      "--------\n",
      "1.5508962869207608\n",
      "--------\n",
      "1.021128828782821\n",
      "--------\n",
      "0.6763189632765716\n",
      "--------\n",
      "0.44849936086393427\n",
      "--------\n",
      "0.29640033769828733\n",
      "--------\n",
      "0.19417478849936742\n",
      "--------\n",
      "0.12524008554464672\n",
      "--------\n",
      "0.07875408534891903\n",
      "--------\n",
      "0.047520216307020746\n",
      "--------\n",
      "0.02670328039675951\n",
      "--------\n",
      "0.01302253625181038\n",
      "--------\n",
      "0.004234299674862996\n",
      "--------\n",
      "0.0012054162216372788\n",
      "--------\n",
      "0.004364024440292269\n",
      "--------\n",
      "0.005981738388072699\n",
      "--------\n",
      "0.006572629208676517\n",
      "--------\n",
      "0.006493631357443519\n",
      "--------\n",
      "0.005992020829580724\n",
      "--------\n",
      "0.005238327459665015\n",
      "--------\n",
      "0.00434925795707386\n",
      "--------\n",
      "0.003403757931664586\n",
      "--------\n",
      "0.002454297398799099\n",
      "--------\n",
      "0.0015348267188528553\n",
      "--------\n",
      "Done!\n",
      "Number of iterations:  34\n"
     ]
    }
   ],
   "source": [
    "# Auxiliary functions\n",
    "\n",
    "def init_root_distribution():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the root distribution (all probabilities between 0.2 and 0.3)\n",
    "    \"\"\"\n",
    "    estimated_root_distribution = []\n",
    "    while True:\n",
    "        estimated_root_distribution = np.random.dirichlet([1,1,1,1])\n",
    "        i = 0\n",
    "        while i<4:\n",
    "            if estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3:\n",
    "                i += 1\n",
    "            else:\n",
    "                break\n",
    "        if i == 4:\n",
    "            assert(np.sum(estimated_root_distribution) < 1.0000001 and np.sum(estimated_root_distribution) > 0.9999999)\n",
    "            assert(estimated_root_distribution[i] > 0.2 and estimated_root_distribution[i] < 0.3 for i in range(4))\n",
    "            break\n",
    "    return estimated_root_distribution\n",
    "\n",
    "def init_transition_matrix():\n",
    "    \"\"\"\n",
    "    Initialise estimation for the transition matrices (which we estimate as LARGE-DIAGONAL-VALUE matrices)\n",
    "    \"\"\"\n",
    "    M = np.zeros((4,4))\n",
    "    i=0\n",
    "    while i<4:\n",
    "        R = np.random.dirichlet([1,1,1,1])\n",
    "        if R[i] > 0.6 and R[i] < 0.8:\n",
    "            M[i,:] = R \n",
    "            i = i + 1\n",
    "    assert (M[i,i] > 0.6 and M[i,i] < 0.8 for i in range(4))\n",
    "    assert (np.sum(M[i,:]) < 1.0000001 and np.sum(M[i,:]) > 0.9999999 for i in range(4))\n",
    "    return M\n",
    "\n",
    "def log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "    \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "    logL = 0\n",
    "    for obs in states:\n",
    "        p_i = 0 \n",
    "        observed_data = obs[n_int:]\n",
    "        if observed_data in u_i:\n",
    "            for state in states:\n",
    "                if state[n_int:] == observed_data:\n",
    "                    pi = root_distribution[int(state[0])]\n",
    "                    for p in params.items():\n",
    "                        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                        pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                    p_i += pi\n",
    "        logL += (u_i[observed_data] * math.log(p_i))\n",
    "    return logL\n",
    "\n",
    "############################################################################################################\n",
    "########################################## FELSENSTEIN ALGORITHM ###########################################\n",
    "############################################################################################################\n",
    "# # Felsentstein algorithm\n",
    "# # \"((4,5),(2,3))\"\n",
    "# newick_tree = \"(((4:1,5:1):1,(6:1,7:1):1):1,(8:1,9:1):1);\"\n",
    "# L_tree = Phylo.read(StringIO(newick_tree), \"newick\")\n",
    "# # Change nodes names\n",
    "# for idx, clade in enumerate(L_tree.get_nonterminals()):\n",
    "#     id = 0\n",
    "#     if idx == 2:\n",
    "#         clade.name = \"Int_Id_\" + str(id)\n",
    "#         id += 1\n",
    "#     else:\n",
    "#         clade.name = \"Int_\" + str(idx) if idx > 0 else \"Int_0\" \n",
    "# # Change leaves names\n",
    "# for idx, clade in enumerate(L_tree.get_terminals()):\n",
    "#     clade.name = \"Leaf_\" + clade.name \n",
    "# L_net = Phylo.to_networkx(L_tree) # to graph\n",
    "# L_edges = []\n",
    "# for edge in L_net.edges():\n",
    "#     if edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     elif edge[0].name.split(\"_\")[1] == \"Id\" and edge[1].name.split(\"_\")[1] != \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1]) + \"_\" + str(edge[0].name.split(\"_\")[2])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     elif edge[0].name.split(\"_\")[1] != \"Id\" and edge[1].name.split(\"_\")[1] == \"Id\":\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1]) + \"_\" + str(edge[1].name.split(\"_\")[2])\n",
    "#     else:\n",
    "#         u = str(edge[0].name.split(\"_\")[1])\n",
    "#         v = str(edge[1].name.split(\"_\")[1])\n",
    "#     L_edges.append((u,v))\n",
    "# # Create a dictionary to store the father and son information\n",
    "# family_tree = {}\n",
    "# # Loop through the edges and update the dictionary\n",
    "# for father, son in L_edges:\n",
    "#     if father not in family_tree:\n",
    "#         family_tree[father] = []\n",
    "#     family_tree[father].append(son)\n",
    "# # Create a DataFrame from the dictionary\n",
    "# tree_df = pd.DataFrame.from_dict(family_tree, orient='index')\n",
    "# # Fill NaN values with empty strings\n",
    "# tree_df = tree_df.fillna('')\n",
    "# # Change column names\n",
    "# tree_df.columns = ['child_1', 'child_2']\n",
    "\n",
    "# class TreeInference():\n",
    "#     def __init__(self, tree_data, leaf_values, params, root_distr):\n",
    "#         # Set parameters\n",
    "#         self.tree_data = tree_data.copy()\n",
    "#         self.leaf_values = leaf_values\n",
    "#         self.params = params\n",
    "#         self.root_distr = root_distr\n",
    "\n",
    "#         # Find IDs of internal nodes, all nodes and leaf nodes\n",
    "#         self.internal_node_ids = np.array(tree_data.index)\n",
    "#         self.all_node_ids = np.unique(np.concatenate((np.array(self.tree_data.index),\n",
    "#                                                       np.array(self.tree_data[\"child_1\"]),\n",
    "#                                                       np.array(self.tree_data[\"child_2\"]))))\n",
    "#         self.leaf_node_ids = list(set(self.all_node_ids) - set(self.internal_node_ids))\n",
    "        \n",
    "#         # Count internal nodes and leaves for each row in tree_data and set root node\n",
    "#         self.treeNodeCounts()\n",
    "#         self.root_node = 0\n",
    "        \n",
    "#         # Get indexes for matricies for states, e.g. A is row 0, G is row 2...\n",
    "#         states = [\"A\", \"G\", \"C\", \"T\"]\n",
    "#         self.matrix_idx = dict(zip(states, range(len(states))))\n",
    "        \n",
    "#         # Empty matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "#         # Log likelihood of whole tree\n",
    "#         self.L_root = None\n",
    "        \n",
    "\n",
    "#     \"\"\"\n",
    "#     Calculates the number of internal nodes and leaves below each internal node.\n",
    "#     This works by recursively iterating down the child nodes from the original internal node until a\n",
    "#     leaf node is found, and then returning upwards counting the leaves and nodes.\n",
    "#     The new derived information is then added as two new columns at the end of the data.\n",
    "#     \"\"\"\n",
    "#     def treeNodeCounts(self):\n",
    "#         n_nodes = np.array([])\n",
    "#         n_leaves = np.array([])\n",
    "\n",
    "#         # Set the number of internal nodes and leaves for each internal node row\n",
    "#         for n in self.internal_node_ids:\n",
    "#             num_nodes, num_leaves = self.checkNode(self.tree_data, n, self.internal_node_ids)\n",
    "#             n_nodes = np.append(n_nodes, num_nodes)\n",
    "#             n_leaves = np.append(n_leaves, num_leaves)\n",
    "\n",
    "#         # Add new information to dataframe\n",
    "#         self.tree_data[\"N_Internal\"] = n_nodes\n",
    "#         self.tree_data[\"N_Leaves\"] = n_leaves\n",
    "                \n",
    "#     # Recursively find the number of internal nodes and leaves given a node ID\n",
    "#     def checkNode(self, tree_data, node_id, node_array):\n",
    "#         # Record the number of internal nodes and leaves below\n",
    "#         num_nodes = 0\n",
    "#         num_leaves = 0\n",
    "\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[node_id][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[node_id][\"child_2\"]\n",
    "\n",
    "#         # Check if found an internal node to the left\n",
    "#         if child_1 in node_array:\n",
    "#             # Recursively check nodes below\n",
    "#             d1_nodes, d1_leaves = self.checkNode(self.tree_data, child_1, node_array)\n",
    "#             num_nodes += d1_nodes + 1\n",
    "#             num_leaves += d1_leaves\n",
    "\n",
    "#         # Check if found an internal node to the right\n",
    "#         if child_2 in node_array:\n",
    "#             d2_nodes, d2_leaves = self.checkNode(self.tree_data, child_2, node_array)\n",
    "#             num_nodes += d2_nodes + 1\n",
    "#             num_leaves += d2_leaves\n",
    "\n",
    "#         return num_nodes, num_leaves\n",
    "    \n",
    "    \n",
    "#     # Calculate likelihood for an internal node\n",
    "#     def nodeLikelihood(self, node_n, params):\n",
    "#         # Find the left and right daughter nodes\n",
    "#         child_1 = self.tree_data.loc[str(node_n)][\"child_1\"]\n",
    "#         child_2 = self.tree_data.loc[str(node_n)][\"child_2\"]\n",
    "        \n",
    "#         # Check if likelihood of left child node set\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(child_1, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_1), 0] is None):\n",
    "#                 # If not set then calculate child likelihood first\n",
    "#                 self.nodeLikelihood(int(child_1), params)\n",
    "#         # Check if likelihood of right child node set\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             if (self.L[-1, 0] is None):\n",
    "#                 self.nodeLikelihood(child_2, params)\n",
    "#         else:\n",
    "#             if (self.L[int(child_2), 0] is None):\n",
    "#                 self.nodeLikelihood(int(child_2), params)\n",
    "        \n",
    "#         M1, M2 = np.full((4,4),None), np.full((4,4),None)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             # Create transition matrix for left and right child nodes\n",
    "#             matrix_name_child1 = \"M_0_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_0_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_1 == \"Id_0\":\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = np.identity(4)\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "#         elif child_2 == \"Id_0\":\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = np.identity(4)\n",
    "#         else:\n",
    "#             matrix_name_child1 = \"M_\" + str(node_n) + \"_to_\" + str(child_1)\n",
    "#             matrix_name_child2 = \"M_\" + str(node_n) + \"_to_\" + str(child_2)\n",
    "#             M1 = self.params[matrix_name_child1].transition_matrix\n",
    "#             M2 = self.params[matrix_name_child2].transition_matrix\n",
    "\n",
    "#         # Get conditional likelihood of child nodes\n",
    "#         if child_1 == \"Id_0\":\n",
    "#             d1_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d1_likelihood = self.L[int(child_1)]\n",
    "#         if child_2 == \"Id_0\":\n",
    "#             d2_likelihood = self.L[-1]\n",
    "#         else:\n",
    "#             d2_likelihood = self.L[int(child_2)]\n",
    "        \n",
    "#         # Calculate the likelihood of each state for the node\n",
    "#         left_prob = np.sum(M1 * d1_likelihood, axis = 1)\n",
    "#         right_prob = np.sum(M2 * d2_likelihood, axis = 1)\n",
    "#         if node_n == \"Id_0\":\n",
    "#             self.L[-1,:] = np.multiply(left_prob, right_prob)\n",
    "#         else:\n",
    "#             self.L[int(node_n),:] = np.multiply(left_prob, right_prob)\n",
    "    \n",
    "#     # Run Felsenstein’s algorithm to calculate the log likelihood of the tree\n",
    "#     def calculateLikelihood(self, params, Estep):\n",
    "#         # Clear matrix of likelihood for each node\n",
    "#         self.L = np.full((len(self.all_node_ids), len(self.matrix_idx)), None)\n",
    "        \n",
    "#         # Set likelihood of leaf states\n",
    "#         for leaf_n in self.leaf_node_ids:\n",
    "#             if leaf_n == \"Id_0\":\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[-1, :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[-1, self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "#             else:\n",
    "#                 # Zero for non-matching base\n",
    "#                 self.L[int(leaf_n), :] = 0\n",
    "#                 # One for matching base\n",
    "#                 self.L[int(leaf_n), self.matrix_idx[self.leaf_values[int(leaf_n)]]] = 1\n",
    "        \n",
    "#         for node_n in self.internal_node_ids:\n",
    "#             # Check likelihood has not been set\n",
    "#             if node_n == \"Id_0\":\n",
    "#                 if self.L[-1, 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "#             else:\n",
    "#                 if self.L[int(node_n), 0] is None:\n",
    "#                     # Calculate conditional likelihood of each state for the node\n",
    "#                     self.nodeLikelihood(node_n, params)\n",
    "        \n",
    "#         if Estep == True:\n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = np.sum(np.multiply(self.root_distr, self.L[self.root_node]))\n",
    "#         else: \n",
    "#             # Calculate likelihood of the whole tree\n",
    "#             self.L_root = math.log(np.sum(np.multiply(self.root_distr, self.L[self.root_node])))\n",
    "        \n",
    "#         return self.L_root\n",
    "\n",
    "# def Felsenstein_log_likelihood(states, u_i, params, root_distribution, n_int):\n",
    "#     \"\"\"Compute the log-likehood of the tree\"\"\"\n",
    "#     logL = 0\n",
    "#     for obs in states:\n",
    "#         observed_data = obs[n_int:]\n",
    "#         tree_leaves = dict()\n",
    "#         for i in range(len(observed_data)):\n",
    "#             if observed_data[i] == str(0):\n",
    "#                 tree_leaves[n_int+i] = \"A\"\n",
    "#             elif observed_data[i] == str(1):\n",
    "#                 tree_leaves[n_int+i] = \"G\"\n",
    "#             elif observed_data[i] == str(2):\n",
    "#                 tree_leaves[n_int+i] = \"C\"\n",
    "#             elif observed_data[i] == str(3):\n",
    "#                 tree_leaves[n_int+i] = \"T\"\n",
    "#         tree_inference = TreeInference(tree_df, tree_leaves, params, root_distribution)\n",
    "#         tree_inference.calculateLikelihood(params, False)\n",
    "#         logL += (u_i[observed_data] * tree_inference.L_root)\n",
    "#         del tree_inference\n",
    "#     return logL\n",
    "\n",
    "\n",
    "############################################################################################################\n",
    "\n",
    "class Param:\n",
    "    \"\"\"\n",
    "    Class to store the parameters of the tree\n",
    "    \"\"\"\n",
    "    def __init__(self, edge, transition_matrix, alignment=None):\n",
    "        self.edge = edge\n",
    "        self.transition_matrix = transition_matrix\n",
    "        self.alignment = alignment # it will only be placed in the leaves\n",
    "\n",
    "# Obtain number of leaves and internal nodes\n",
    "n_leaves, n_int = 0, 0 \n",
    "for node in net.nodes():\n",
    "    if node.name.startswith(\"L\"):\n",
    "        n_leaves += 1\n",
    "    else:\n",
    "        n_int += 1\n",
    "\n",
    "# Count all the leaves ocurrences \n",
    "sequence_length = len(leave_sequences[0])\n",
    "number_of_sequences = len(leave_sequences)\n",
    "occurrences = []\n",
    "for i in range(sequence_length):\n",
    "    _ = ''\n",
    "    for j in range(number_of_sequences):\n",
    "        _ += leave_sequences[j][i]\n",
    "    occurrences.append(_)\n",
    "c = Counter(occurrences)\n",
    "u_i = OrderedDict(sorted(c.items()))\n",
    "\n",
    "# Define all possible (hidden) states of the internal nodes \n",
    "hidden_combinations = itertools.product([0,1,2,3], repeat=n_int)\n",
    "sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "for i in range(len(sorted_hidden_combinations)):\n",
    "    sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    " \n",
    "# Define all possible fully observed states (i.e. combining u_i + internal nodes)\n",
    "# '000000' would mean to have an 'A' in all nodes: Int_0, Int_1, Leaf_2, Leaf_3, Leaf_4, Leaf_5 \n",
    "states = list(itertools.product(list(sorted_hidden_combinations), list(u_i.keys())))\n",
    "states = [i[0]+i[1] for i in states]\n",
    "\n",
    "############################################################################################################\n",
    "###################################### STEP 0: Initialise parameters #######################################\n",
    "############################################################################################################\n",
    "estimated_root_distribution = init_root_distribution() # First estimation for the root distribution\n",
    "\n",
    "# PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "# for edge in net.edges():\n",
    "#     # If we are in a leaf, we need to specify the alignment\n",
    "#     iter_leaves = 0\n",
    "#     if edge[1].name.startswith(\"L\"):\n",
    "#         new_edge = Param(edge, init_transition_matrix(), leave_alignments[iter_leaves].seq)\n",
    "#         iter_leaves += 1\n",
    "#     # Otherwise, are in internal nodes\n",
    "#     else:\n",
    "#         new_edge = Param(edge, init_transition_matrix())\n",
    "#     u = str(edge[0].name.split(\"_\")[1])\n",
    "#     v = str(edge[1].name.split(\"_\")[1])\n",
    "#     name = \"M_\" + u + \"_to_\" + v\n",
    "#     PARAMS[name] = new_edge\n",
    "\n",
    "# ADRIA CODE\n",
    "i=0\n",
    "PARAMS = dict() # Dictionary to store the estimation matrices for each edge\n",
    "for edge in net.edges():\n",
    "    # If we are in a leaf, we need to specify the alignment\n",
    "    iter_leaves = 0\n",
    "    if edge[1].name.startswith(\"L\"):\n",
    "        new_edge = Param(edge, Results[i].matrix, leave_alignments[iter_leaves].seq)\n",
    "        iter_leaves += 1\n",
    "    # Otherwise, are in internal nodes\n",
    "    else:\n",
    "        new_edge = Param(edge, Results[i].matrix)\n",
    "    u = str(edge[0].name.split(\"_\")[1])\n",
    "    v = str(edge[1].name.split(\"_\")[1])\n",
    "    name = \"M_\" + u + \"_to_\" + v\n",
    "    PARAMS[name] = new_edge\n",
    "    i += 1\n",
    "\n",
    "eps = 10**(-3)                  # epsilon for convergence\n",
    "iter = 0                        # iteration counter \n",
    "params = copy.copy(PARAMS)      # copy of the parameters (will be updated at each iteration)\n",
    "# logF = Felsenstein_log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "logL = log_likelihood(states, u_i, params, estimated_root_distribution, n_int)\n",
    "# print(logF)\n",
    "print(\"l\", logL)\n",
    "logL_ = 0\n",
    "\n",
    "while np.abs(logL_ - logL) > eps:\n",
    "    print(np.abs(logL_ - logL))\n",
    "    # print(np.abs(logF_ - logF))\n",
    "    print(\"--------\")\n",
    "    if iter > 0:\n",
    "        logL = logL_\n",
    "        # logF = logF_\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## E-step ##################################################\n",
    "    ############################################################################################################\n",
    "    # Define the expected hidden data matrix\n",
    "    U = dict()\n",
    "    root_distr = estimated_root_distribution ########\n",
    "\n",
    "    for obs in states:\n",
    "        U[obs] = u_i[obs[n_int:]] # initialise by computing u_i of the expected hidden data matrix formula\n",
    "        f_ij = estimated_root_distribution[int(obs[0])] \n",
    "        for p in params.items():\n",
    "            u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "            f_ij *= p[1].transition_matrix[int(obs[u]), int(obs[v])]\n",
    "        # Now f_ij := numerator of the expected hidden data matrix formula \n",
    "\n",
    "        # We need to marginalise to obtain f_i := denominator of the expected hidden data matrix formula\n",
    "        f_i = 0 # we initialise the partial data model\n",
    "        observed_data = obs[n_int:]\n",
    "        for state in states:\n",
    "            if state[n_int:] == observed_data:\n",
    "                pi = estimated_root_distribution[int(state[0])]\n",
    "                for p in params.items():\n",
    "                    u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "                    pi *= p[1].transition_matrix[int(state[u]), int(state[v])]\n",
    "                f_i += pi\n",
    "\n",
    "        # observed_data = obs[n_int:]\n",
    "        # tree_leaves = dict()\n",
    "        # for i in range(len(observed_data)):\n",
    "        #     if observed_data[i] == str(0):\n",
    "        #         tree_leaves[n_int+i] = \"A\"\n",
    "        #     elif observed_data[i] == str(1):\n",
    "        #         tree_leaves[n_int+i] = \"G\"\n",
    "        #     elif observed_data[i] == str(2):\n",
    "        #         tree_leaves[n_int+i] = \"C\"\n",
    "        #     elif observed_data[i] == str(3):\n",
    "        #         tree_leaves[n_int+i] = \"T\"\n",
    "        # tree_inference = TreeInference(tree_df, tree_leaves, params, root_distr)\n",
    "        # tree_inference.calculateLikelihood(params, True)\n",
    "\n",
    "        U[obs] *= f_ij/f_i\n",
    "        # U[obs] *= f_ij/tree_inference.L_root\n",
    "        # del tree_inference\n",
    "        # del tree_leaves\n",
    "\n",
    "\n",
    "    assert(len(U) <= 4**(n_int+n_leaves))\n",
    "    sum = 0\n",
    "    sum2 = 0\n",
    "    for i in U.items():\n",
    "        sum += i[1]\n",
    "    for i in u_i.items():\n",
    "        sum2 += i[1]\n",
    "    assert(sum < sequence_length+0.001 and sum > sequence_length-0.001)\n",
    "    assert(sum2 < sequence_length+0.001 and sum2 > sequence_length-0.001)\n",
    "\n",
    "    ############################################################################################################\n",
    "    ################################################## M-step ##################################################\n",
    "    ############################################################################################################\n",
    "    estimated_parameters = dict()\n",
    "    root_distr = np.zeros(4)\n",
    "    iter_root = 0 # counter for the root distribution, we only want to count once\n",
    "\n",
    "    for p in params.items():\n",
    "        u, v = int(p[1].edge[0].name.split(\"_\")[1]), int(p[1].edge[1].name.split(\"_\")[1])\n",
    "        name = \"M_\" + str(u) + \"_to_\" + str(v)\n",
    "        estimated_parameters[name] = np.zeros((4,4))\n",
    "        for i in range(4):\n",
    "            for j in range(4):\n",
    "                ui = 0\n",
    "                for obs in U.items():\n",
    "                    # For the root distribution, we only need to consider the first internal node\n",
    "                    if iter_root == 0 and obs[0][0] == str(i):\n",
    "                        root_distr[i] += obs[1]\n",
    "                    # For the transition matrices:\n",
    "                    if obs[0][u] == str(i) and obs[0][v] == str(j):\n",
    "                        estimated_parameters[name][i][j] += obs[1]\n",
    "                    if obs[0][u] == str(i):\n",
    "                        ui += obs[1]\n",
    "                # Update root distribution\n",
    "                if iter_root == 0:\n",
    "                    root_distr[i] /= sequence_length\n",
    "                iter_root = 1\n",
    "                estimated_parameters[name][i][j] /= ui\n",
    "            iter_root = 0\n",
    "            \n",
    "        # Update parameters\n",
    "        params[name].transition_matrix = estimated_parameters[name]\n",
    "\n",
    "    ##########################################################################################################\n",
    "    ########################### Compute log-likelihood of the estimated parameters ###########################\n",
    "    ##########################################################################################################\n",
    "    # logF_ = Felsenstein_log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "    logL_ = log_likelihood(states, u_i, params, root_distr, n_int)\n",
    "\n",
    "    iter += 1\n",
    "\n",
    "print(\"Done!\")\n",
    "print(\"Number of iterations: \", iter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################### DEBUGGING ############################\n",
    "\n",
    "# # All possible (theoretical) occurrences of nucleotides in the leaves\n",
    "# alphabet = [0, 1, 2, 3]\n",
    "# combinations = itertools.product(alphabet, repeat=n_leaves)\n",
    "# sorted_combinations = sorted(combinations)\n",
    "# for i in range(len(sorted_combinations)):\n",
    "#     sorted_combinations[i] = ''.join([str(s) for s in sorted_combinations[i]])\n",
    "\n",
    "# # All possible (hidden) states of the internal nodes (i.e. indexes of columns of U)\n",
    "# hidden_combinations = itertools.product(alphabet, repeat=n_int)\n",
    "# sorted_hidden_combinations = sorted(hidden_combinations)\n",
    "# for i in range(len(sorted_hidden_combinations)):\n",
    "#     sorted_hidden_combinations[i] = ''.join([str(s) for s in sorted_hidden_combinations[i]])\n",
    "\n",
    "# # To check if an element is within a dictionary or not \n",
    "# if '0101' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')\n",
    "# if '0000' in u_i:\n",
    "#     print('ok')\n",
    "# else:\n",
    "#     print('not ok')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix: [[0.64642183 0.02803076 0.27483013 0.05071728]\n",
      " [0.03848122 0.7757021  0.10189982 0.08391686]\n",
      " [0.03184807 0.07934462 0.82109495 0.06771236]\n",
      " [0.04702478 0.01879095 0.04076404 0.89342023]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix: [[0.55850753 0.13612552 0.2499751  0.05539185]\n",
      " [0.01791204 0.82605787 0.05907209 0.096958  ]\n",
      " [0.04534628 0.008074   0.85149918 0.09508054]\n",
      " [0.00646156 0.06355415 0.02348564 0.90649866]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix: [[0.75152183 0.09827146 0.07355421 0.0766525 ]\n",
      " [0.04888819 0.86735324 0.02989229 0.05386628]\n",
      " [0.05446854 0.02450495 0.75018255 0.17084396]\n",
      " [0.12039713 0.03979299 0.05715314 0.78265675]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix: [[0.82835309 0.03976936 0.05146899 0.08040856]\n",
      " [0.02327998 0.80873654 0.06043788 0.10754561]\n",
      " [0.11544022 0.13895831 0.71027209 0.03532938]\n",
      " [0.09988558 0.08065349 0.01737091 0.80209001]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix: [[0.78296278 0.06942125 0.08171685 0.06589912]\n",
      " [0.02166725 0.83743552 0.01932756 0.12156967]\n",
      " [0.11182788 0.03378373 0.78484422 0.06954417]\n",
      " [0.11063295 0.0881492  0.0495941  0.75162375]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23195748 0.24737645 0.23524372 0.28542235]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.6464218290229772"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the theoretical parameters:\n",
    "theoretical_matrices = dict()\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    u = edges[i].edge[0].name.split(\"_\")[1]\n",
    "    v = edges[i].edge[1].name.split(\"_\")[1]\n",
    "    matrix_name = \"M_\" + u + \"_to_\" + v\n",
    "    print(\"Transition matrix:\", edges[i].transition_matrix)\n",
    "    theoretical_matrices[matrix_name] = np.array(edges[i].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", node_distr['Int_0'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'M_0_to_2': array([[0.64642183, 0.02803076, 0.27483013, 0.05071728],\n",
      "       [0.03848122, 0.7757021 , 0.10189982, 0.08391686],\n",
      "       [0.03184807, 0.07934462, 0.82109495, 0.06771236],\n",
      "       [0.04702478, 0.01879095, 0.04076404, 0.89342023]]), 'M_0_to_3': array([[0.55850753, 0.13612552, 0.2499751 , 0.05539185],\n",
      "       [0.01791204, 0.82605787, 0.05907209, 0.096958  ],\n",
      "       [0.04534628, 0.008074  , 0.85149918, 0.09508054],\n",
      "       [0.00646156, 0.06355415, 0.02348564, 0.90649866]]), 'M_0_to_1': array([[0.75152183, 0.09827146, 0.07355421, 0.0766525 ],\n",
      "       [0.04888819, 0.86735324, 0.02989229, 0.05386628],\n",
      "       [0.05446854, 0.02450495, 0.75018255, 0.17084396],\n",
      "       [0.12039713, 0.03979299, 0.05715314, 0.78265675]]), 'M_1_to_4': array([[0.82835309, 0.03976936, 0.05146899, 0.08040856],\n",
      "       [0.02327998, 0.80873654, 0.06043788, 0.10754561],\n",
      "       [0.11544022, 0.13895831, 0.71027209, 0.03532938],\n",
      "       [0.09988558, 0.08065349, 0.01737091, 0.80209001]]), 'M_1_to_5': array([[0.78296278, 0.06942125, 0.08171685, 0.06589912],\n",
      "       [0.02166725, 0.83743552, 0.01932756, 0.12156967],\n",
      "       [0.11182788, 0.03378373, 0.78484422, 0.06954417],\n",
      "       [0.11063295, 0.0881492 , 0.0495941 , 0.75162375]])}\n"
     ]
    }
   ],
   "source": [
    "# # save the dictionary to a file\n",
    "# np.save('data.npy', theoretical_matrices)\n",
    "# # load the dictionary from the file\n",
    "# loaded_dict = np.load('data.npy', allow_pickle=True).item()\n",
    "# # print the loaded dictionary\n",
    "# print(loaded_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M_0_to_2\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix:\n",
      " [[0.86753892 0.05213542 0.06946348 0.01086218]\n",
      " [0.08065343 0.85671968 0.04982893 0.01279796]\n",
      " [0.11219846 0.17464684 0.65461785 0.05853685]\n",
      " [0.02775028 0.08494961 0.04088151 0.8464186 ]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_3\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix:\n",
      " [[0.75527498 0.04249873 0.00019955 0.20202674]\n",
      " [0.10446698 0.73349645 0.10777451 0.05426206]\n",
      " [0.009178   0.09167603 0.88105832 0.01808765]\n",
      " [0.10672648 0.08298123 0.01949473 0.79079756]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_0_to_1\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix:\n",
      " [[0.69009493 0.15321495 0.12518281 0.03150731]\n",
      " [0.03390489 0.89612499 0.03748613 0.032484  ]\n",
      " [0.13612944 0.00355591 0.86031238 0.00000227]\n",
      " [0.13829849 0.07169377 0.04358193 0.74642581]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_4\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix:\n",
      " [[0.79787321 0.09123409 0.0748348  0.0360579 ]\n",
      " [0.1525167  0.62670853 0.12700414 0.09377063]\n",
      " [0.06182208 0.09423139 0.82703856 0.01690797]\n",
      " [0.09434814 0.05487633 0.01727208 0.83350345]]\n",
      "------------------------------------------------------------------------------------------\n",
      "M_1_to_5\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix:\n",
      " [[0.77670263 0.03884329 0.11981174 0.06464234]\n",
      " [0.07306013 0.6964252  0.14381485 0.08669983]\n",
      " [0.0001472  0.02174493 0.93060026 0.04750762]\n",
      " [0.01290462 0.10677068 0.06826479 0.81205991]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Root distribution: [0.23508015 0.26506847 0.26101933 0.23983305]\n"
     ]
    }
   ],
   "source": [
    "# Check the estimated parameters:\n",
    "for i in params.items():\n",
    "    print(i[0])\n",
    "    print(\"Edge:\", i[1].edge)\n",
    "    print(\"Transition matrix:\\n\", i[1].transition_matrix)\n",
    "    print(\"---\"*30)\n",
    "\n",
    "print(\"Root distribution:\", root_distr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_2'))\n",
      "Transition matrix difference:\n",
      " [[0.09960865 0.07093256 0.00498621 0.03366231]\n",
      " [0.01576357 0.02703813 0.00892483 0.03387688]\n",
      " [0.00533819 0.03058066 0.01301418 0.01222828]\n",
      " [0.00504287 0.04282721 0.00457214 0.05244223]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Leaf_3'))\n",
      "Transition matrix difference:\n",
      " [[0.00386991 0.01431933 0.0324696  0.04291902]\n",
      " [0.03240956 0.00438465 0.00578344 0.03101077]\n",
      " [0.00129948 0.00595081 0.00836875 0.00111846]\n",
      " [0.00368219 0.0201782  0.01427531 0.03077132]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(name='Int_0'), Clade(branch_length=1.0, name='Int_1'))\n",
      "Transition matrix difference:\n",
      " [[0.00043271 0.04586535 0.01971483 0.02571781]\n",
      " [0.04628465 0.03270525 0.01225697 0.00132243]\n",
      " [0.00405161 0.02596747 0.04979691 0.01977783]\n",
      " [0.05635943 0.0032355  0.00691523 0.06003916]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_4'))\n",
      "Transition matrix difference:\n",
      " [[0.02206907 0.00438435 0.02045318 0.00276846]\n",
      " [0.04114356 0.07985678 0.0152662  0.02344702]\n",
      " [0.01766089 0.02629588 0.04017144 0.00378534]\n",
      " [0.00126844 0.00706299 0.01874386 0.02707528]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Edge: (Clade(branch_length=1.0, name='Int_1'), Clade(branch_length=1.0, name='Leaf_5'))\n",
      "Transition matrix difference:\n",
      " [[0.04539034 0.01368587 0.03621623 0.00451176]\n",
      " [0.00302602 0.00613968 0.01219578 0.00908212]\n",
      " [0.00442834 0.01538392 0.00158094 0.0213932 ]\n",
      " [0.02699704 0.00888034 0.00551462 0.03036276]]\n",
      "------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Compute the matrix difference between the theoretical and estimated parameters\n",
    "for i in range(len(edges)):\n",
    "    print(\"Edge:\", edges[i].edge)\n",
    "    print(\"Transition matrix difference:\\n\", np.abs(edges[i].transition_matrix - params['M_' + str(edges[i].edge[0].name.split(\"_\")[1]) + '_to_' + str(edges[i].edge[1].name.split(\"_\")[1])].transition_matrix))\n",
    "    print(\"---\"*30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Real pi\n",
      "[0.21314739 0.28342213 0.23706711 0.26636337]\n",
      "Real M\n",
      "M_0_to_2 [[0.94678804 0.00873393 0.01465123 0.0298268 ]\n",
      " [0.00552733 0.92875971 0.04912965 0.01658331]\n",
      " [0.01496792 0.01831153 0.95164462 0.01507594]\n",
      " [0.03859277 0.0111101  0.01273422 0.93756291]]\n",
      "M_0_to_3 [[0.94568745 0.0129391  0.03351668 0.00785677]\n",
      " [0.06698824 0.92332335 0.00570654 0.00398187]\n",
      " [0.01583212 0.00697127 0.94776919 0.02942741]\n",
      " [0.01235995 0.00521254 0.03154398 0.95088353]]\n",
      "M_0_to_1 [[0.94978744 0.0344919  0.00367433 0.01204634]\n",
      " [0.02719184 0.92320545 0.01909197 0.03051075]\n",
      " [0.00280613 0.02422721 0.95695618 0.01601048]\n",
      " [0.02228667 0.03958845 0.00614081 0.93198407]]\n",
      "M_1_to_4 [[0.88809048 0.06280357 0.03662466 0.01248129]\n",
      " [0.00248335 0.98609982 0.00503081 0.00638602]\n",
      " [0.03592295 0.00837013 0.91485025 0.04085668]\n",
      " [0.00568673 0.02308479 0.00878014 0.96244835]]\n",
      "M_1_to_5 [[0.93248038 0.03340507 0.01881597 0.01529859]\n",
      " [0.01717432 0.94099276 0.03588036 0.00595256]\n",
      " [0.02359382 0.00613886 0.94583873 0.02442859]\n",
      " [0.02151848 0.00609908 0.02869287 0.94368957]]\n",
      "------------------------------------------------------------------------------------------\n",
      "Estimated M\n",
      "M_0_to_2 [[0.92040227 0.00865435 0.03355228 0.03739111]\n",
      " [0.         0.92051676 0.06744108 0.01204217]\n",
      " [0.00829572 0.01756156 0.96954098 0.00460173]\n",
      " [0.03169372 0.01504237 0.         0.95326391]]\n",
      "M_0_to_3 [[0.89636584 0.02404797 0.07444584 0.00514035]\n",
      " [0.09184197 0.89352723 0.         0.0146308 ]\n",
      " [0.00499033 0.         0.94975139 0.04525828]\n",
      " [0.01007904 0.01140701 0.03373476 0.94477919]]\n",
      "M_0_to_1 [[0.93217303 0.04941642 0.00449804 0.01391251]\n",
      " [0.03443008 0.92150621 0.01179612 0.03226759]\n",
      " [0.00065166 0.04359074 0.95150674 0.00425085]\n",
      " [0.01432634 0.06010134 0.         0.92557233]]\n",
      "M_1_to_4 [[0.88690074 0.04861923 0.05985962 0.00462041]\n",
      " [0.00367358 0.98405131 0.00798494 0.00429018]\n",
      " [0.02722976 0.01142609 0.91353956 0.04780459]\n",
      " [0.0072166  0.02323309 0.00364566 0.96590465]]\n",
      "M_1_to_5 [[0.96840051 0.00984577 0.02175373 0.        ]\n",
      " [0.01744847 0.9420556  0.04049594 0.        ]\n",
      " [0.033144   0.00409342 0.9397836  0.02297898]\n",
      " [0.016022   0.01492654 0.02156933 0.94748214]]\n",
      "Estimated l\n",
      "M_0_to_2 0.24695744287481536\n",
      "M_0_to_3 0.3383816598620235\n",
      "M_0_to_1 0.2818284009835473\n",
      "M_1_to_4 0.25196494961543703\n",
      "M_1_to_5 0.21688433829475823\n",
      "Estimated pi\n",
      "[0.21421083 0.26280666 0.2375805  0.28640302]\n",
      "Niter\n",
      "41\n",
      "TExec\n",
      "15.420313835144043\n"
     ]
    }
   ],
   "source": [
    "print(\"Real pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a25_b25_1000_random/real_root_distr.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Real M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a25_b25_1000_random/real_matrices.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"---\"*30)\n",
    "print(\"Estimated M\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a25_b25_1000_random/RESULTS/repetition_98/M_estimation.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated l\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a25_b25_1000_random/RESULTS/repetition_98/estimated_branch_lengths.npy', allow_pickle=True).item()\n",
    "for i in loaded_dict.items():\n",
    "    print(i[0], i[1])\n",
    "print(\"Estimated pi\")\n",
    "loaded_dict = np.load('./OUTPUTS/EM_a25_b25_1000_random/RESULTS/repetition_98/root_estimation.npy', allow_pickle=True)\n",
    "print(loaded_dict)\n",
    "print(\"Niter\")\n",
    "file = open(\"./OUTPUTS/EM_a25_b25_1000_random/RESULTS/repetition_98/Niter.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)\n",
    "print(\"TExec\")\n",
    "file = open(\"./OUTPUTS/EM_a25_b25_1000_random/RESULTS/repetition_98/TExec.txt\", \"r\")\n",
    "a = file.read()\n",
    "file.close()\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "debugger",
   "language": "python",
   "name": "debugger"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "03f790eb89590da6bbc9dab944ceea1669c4a69ff2bd32047ab9c0f81f755bd1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
